{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import mpltern\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.dissimilarity import dmatrix, load_dmat_as_pickle\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"paper\")\n",
    "plt.rcParams[\"figure.dpi\"] = 300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type_palette = {\n",
    "    \"SPGC\": \"tab:green\",\n",
    "    \"MAG\": \"tab:orange\",\n",
    "    \"Isolate\": \"tab:blue\",\n",
    "    \"Ref\": \"black\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[lambda x: x.species_group_id == \"hmp2\"]\n",
    "    .species_id.astype(str)\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "species_taxonomy = (\n",
    "    pd.read_table(species_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    "    .Lineage.apply(parse_taxonomy_string)\n",
    ")\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "phylum_order0 = sorted(species_taxonomy.p__.unique())\n",
    "\n",
    "phylum_palette0 = lib.plot.construct_ordered_palette(\n",
    "    phylum_order0,\n",
    "    cm=\"rainbow\",\n",
    "    desaturate_levels=[1.0, 0.5],\n",
    ")\n",
    "\n",
    "for p__ in phylum_order0:\n",
    "    plt.scatter([], [], color=phylum_palette0[p__], label=p__)\n",
    "plt.legend(ncols=4)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_genome(x):\n",
    "    if (x.genome_type == \"Isolate\") & x.passes_filter:\n",
    "        return \"isolate\"\n",
    "    elif (x.genome_type == \"Isolate\") & ~x.passes_filter:\n",
    "        return \"isolate_fails_qc\"\n",
    "    elif (x.genome_type == \"MAG\") & x.passes_filter:\n",
    "        return \"mag\"\n",
    "    elif (x.genome_type == \"MAG\") & ~x.passes_filter:\n",
    "        return \"mag_fails_qc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_filter:\n",
    "        return \"spgc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_geno_positions:\n",
    "        return \"sfacts_only\"\n",
    "    elif (x.genome_type == \"SPGC\") & ~(x.passes_geno_positions):\n",
    "        return \"sfacts_fails_qc\"\n",
    "    else:\n",
    "        raise ValueError(\"Genome did not match classification criteria:\", x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.strain_meta_spgc_and_ref.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath).assign(species=species, inpath=inpath)\n",
    "    filt_stats.append(data)\n",
    "filt_stats = (\n",
    "    pd.concat(filt_stats)\n",
    "    .assign(\n",
    "        genome_class=lambda x: x.apply(classify_genome, axis=1),\n",
    "        species_strain=lambda x: x.species + \"_\" + x.genome_id,\n",
    "    )\n",
    "    .set_index(\"species_strain\")\n",
    ")\n",
    "\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define different subsets of the species:\n",
    "\n",
    "# All species:\n",
    "# species_list\n",
    "spgc_strain_list = filt_stats[lambda x: x.genome_type.isin([\"SPGC\"])].index.values\n",
    "\n",
    "# All species with enough positions\n",
    "species_list0 = filt_stats[lambda x: x.passes_geno_positions].species.unique()\n",
    "spgc_strain_list0 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list0)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains\n",
    "species_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains to talk about distributions (>=10)\n",
    "species_list1b = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list1b = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1b)\n",
    "].index.values\n",
    "\n",
    "# All species with spgc strains\n",
    "species_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list2)\n",
    "].index.values\n",
    "\n",
    "# All species with enough spgc strains for pangenome analysis (>=10)\n",
    "species_list3 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list3 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list3)\n",
    "].index.values\n",
    "\n",
    "# Species with large numbers of strains (>=20)\n",
    "species_list4 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 20\n",
    ")\n",
    "spgc_strain_list4 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list4)\n",
    "].index.values\n",
    "\n",
    "_species_list_map = {\n",
    "    \"All considered species\": (species_list, spgc_strain_list),\n",
    "    \"0: Species with enough genotyped positions\": (species_list0, spgc_strain_list0),\n",
    "    \"1: With sfacts strains\": (species_list1, spgc_strain_list1),\n",
    "    \"1b: With (>=10) sfacts strains\": (species_list1b, spgc_strain_list1b),\n",
    "    \"2: With SPGC inferences\": (species_list2, spgc_strain_list2),\n",
    "    \"3: With >=10 inferences\": (species_list3, spgc_strain_list3),\n",
    "    \"4: With >=20 inferences\": (species_list4, spgc_strain_list4),\n",
    "}\n",
    "for _species_list_name, (_species_list, _strain_list) in _species_list_map.items():\n",
    "    print(_species_list_name, len(_species_list), len(_strain_list))\n",
    "    print(species_taxonomy.loc[_species_list].p__.value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats.loc[spgc_strain_list3].species.value_counts().quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Phylum palette"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy.loc[species_list1].p__.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_order = [\n",
    "    \"p__Euryarchaeota\",\n",
    "    \"p__Thermoplasmatota\",\n",
    "    \"p__Firmicutes\",\n",
    "    \"p__Firmicutes_A\",\n",
    "    \"p__Firmicutes_C\",\n",
    "    # \"p__Firmicutes_B\", # None in species_list1\n",
    "    # \"p__Firmicutes_G\", # B/G/I not sure how related to C or A\n",
    "    # \"p__Firmicutes_I\", #\n",
    "    # \"p__Cyanobacteria\", # None in species_list1\n",
    "    \"p__Actinobacteriota\",\n",
    "    \"p__Synergistota\",\n",
    "    \"p__Fusobacteriota\",\n",
    "    \"p__Campylobacterota\",\n",
    "    \"p__Proteobacteria\",\n",
    "    \"p__Desulfobacterota_A\",\n",
    "    \"p__Bacteroidota\",\n",
    "    \"p__Verrucomicrobiota\",\n",
    "    # \"dummy0\", # 18\n",
    "    # \"dummy1\", # 19\n",
    "    # \"dummy2\", # 20\n",
    "]\n",
    "\n",
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    phylum_order,\n",
    "    cm=\"rainbow\",\n",
    "    desaturate_levels=[1.0, 0.5],\n",
    ")\n",
    "\n",
    "for p__ in phylum_order:\n",
    "    print(p__, phylum_palette[p__])\n",
    "    plt.scatter([], [], color=phylum_palette[p__], label=p__)\n",
    "plt.legend(ncols=4)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prevalences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_prevalence = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list3\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.uhgg-strain_gene.prevalence.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(\n",
    "        inpath, names=[\"gene_id\", \"prevalence\"], index_col=\"gene_id\"\n",
    "    ).prevalence\n",
    "    spgc_gene_prevalence.append(data)\n",
    "spgc_gene_prevalence = pd.concat(spgc_gene_prevalence).sort_index()\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_prevalence = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list3\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/species/sp-{species}/midasdb.gene75_v20.uhgg-strain_gene.ref_prevalence.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(\n",
    "        inpath, names=[\"gene_id\", \"prevalence\"], index_col=\"gene_id\"\n",
    "    ).prevalence\n",
    "    ref_gene_prevalence.append(data)\n",
    "ref_gene_prevalence = pd.concat(ref_gene_prevalence).sort_index()\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "exclude_genes_never_greater_than = 0.01\n",
    "d = pd.DataFrame(dict(ref=ref_gene_prevalence, spgc=spgc_gene_prevalence)).fillna(0)[\n",
    "    lambda x: x.max(1) > exclude_genes_never_greater_than\n",
    "]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3.5, 3))\n",
    "bins = np.linspace(0, 1, num=51)\n",
    "*_, art = ax.hist2d(\n",
    "    \"ref\",\n",
    "    \"spgc\",\n",
    "    data=d,\n",
    "    bins=bins,\n",
    "    norm=mpl.colors.SymLogNorm(1, vmin=1, vmax=1e4),\n",
    "    cmap=\"Grays\",\n",
    ")\n",
    "fig.colorbar(art, ax=ax, label=\"Genes (count)\", shrink=0.94)\n",
    "ax.plot([0, 1], [0, 1], lw=2, linestyle=\":\", color=\"tab:red\")\n",
    "\n",
    "\n",
    "ax.set_aspect(1)\n",
    "ax.set_xlabel(\"Reference Prevalence\")\n",
    "ax.set_ylabel(\"SPGC Prevalence\")\n",
    "ax.set_xticks([0, 0.15, 0.9, 1.0])\n",
    "ax.set_yticks([0, 0.15, 0.9, 1.0])\n",
    "ax.xaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "ax.yaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "\n",
    "ax.annotate(\"core\", xy=(0.5, 0.901), ha=\"center\", va=\"bottom\", color=\"tab:red\")\n",
    "ax.axhline(0.9, lw=1, linestyle=\"--\", color=\"tab:red\")\n",
    "ax.annotate(\"shell\", xy=(0.5, 0.151), ha=\"center\", va=\"bottom\", color=\"tab:red\")\n",
    "ax.axhline(0.15, lw=1, linestyle=\"--\", color=\"tab:red\")\n",
    "ax.annotate(\"cloud\", xy=(0.5, 0.0), ha=\"center\", va=\"bottom\", color=\"tab:red\")\n",
    "lib.plot.rotate_xticklabels(ax=ax, rotation=90, ha=\"center\")\n",
    "# lib.plot.rotate_yticklabels(ax=ax, rotation=45, va='top')\n",
    "\n",
    "sp.stats.pearsonr(d.ref, d.spgc)\n",
    "\n",
    "fig.savefig(\"fig/fig4a_gene_prevalence.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genome Fractions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution of genome fractions in inferred strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Gather genome fractions for strains\n",
    "# Filter strains\n",
    "\n",
    "spgc_prevalence_class_counts = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list2\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.uhgg-strain_gene.prevalence_class_fraction.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "\n",
    "    strain_list = filt_stats[\n",
    "        lambda x: (x.species == species) & (x.passes_filter) & (x.genome_type == \"SPGC\")\n",
    "    ].genome_id.unique()\n",
    "    data = (\n",
    "        pd.read_table(inpath, index_col=\"strain\")\n",
    "        .rename(index=str)\n",
    "        .loc[strain_list]\n",
    "        .assign(species=species)\n",
    "        .reset_index()\n",
    "        .set_index([\"species\", \"strain\"])\n",
    "    )\n",
    "    spgc_prevalence_class_counts.append(data)\n",
    "spgc_prevalence_class_counts = pd.concat(spgc_prevalence_class_counts).sort_index()\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "median_prevalence_class_fraction = (\n",
    "    spgc_prevalence_class_counts.groupby(\"species\")\n",
    "    .median()\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    ")\n",
    "median_prevalence_class_fraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = median_prevalence_class_fraction.loc[species_list3].assign(\n",
    "    p__=lambda x: x.index.to_series().map(species_taxonomy.p__),\n",
    "    num_genomes=spgc_prevalence_class_counts.reset_index()[\"species\"].value_counts(),\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 3))\n",
    "_, ax_legend = plt.subplots()\n",
    "\n",
    "for p__ in phylum_order:\n",
    "    d1 = d0[lambda x: x.p__ == p__]\n",
    "    ax.scatter(\n",
    "        \"core\",\n",
    "        \"shell\",\n",
    "        data=d1,\n",
    "        color=phylum_palette[p__],\n",
    "        facecolor=\"none\",\n",
    "        s=15,\n",
    "        label=\"__nolegend__\",\n",
    "        # marker=\"o\",\n",
    "        # lw=2,\n",
    "        # facecolor=\"none\",\n",
    "        # alpha=0.85,\n",
    "    )\n",
    "    ax_legend.scatter(\n",
    "        [], [], color=phylum_palette[p__], facecolor=\"none\", label=p__, s=50, lw=3\n",
    "    )\n",
    "ax_legend.legend(bbox_to_anchor=(1, 0.5))\n",
    "lib.plot.hide_axes_and_spines(ax_legend)\n",
    "\n",
    "\n",
    "for cloud_frac, shift in [(0.0, 0.06), (0.1, 0.04), (0.2, 0.04), (0.3, 0.04)]:\n",
    "    if cloud_frac == 0.3:\n",
    "        annot = f\"Cloud {cloud_frac:.0%}\"\n",
    "    else:\n",
    "        annot = f\"{cloud_frac:.0%}\"\n",
    "    ax.plot(\n",
    "        [0, 1 - cloud_frac],\n",
    "        [1 - cloud_frac, 0],\n",
    "        lw=1,\n",
    "        linestyle=\":\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax.annotate(\n",
    "        annot,\n",
    "        xy=(1 - cloud_frac - shift, -0.04 + shift),\n",
    "        rotation=-45,\n",
    "        va=\"bottom\",\n",
    "        ha=\"right\",\n",
    "        fontsize=10,\n",
    "    )\n",
    "\n",
    "ax.set_xlim(0.35, 0.95)\n",
    "ax.set_ylim(0.0, 0.6)\n",
    "ax.xaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "ax.yaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "ax.set_xlabel(\"Core\")\n",
    "ax.set_ylabel(\"Shell\")\n",
    "ax.set_aspect(1)\n",
    "\n",
    "fig.savefig(\"fig/fig4b_genome_fraction.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "median_prevalence_class_fraction.loc[species_list3].quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Distribution of genome fractions in reference strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Gather genome fractions for strains\n",
    "# Filter strains\n",
    "\n",
    "ref_prevalence_class_counts = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list2\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/species/sp-{species}/midasdb.gene75_v20.uhgg-strain_gene.prevalence_class_fraction.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "\n",
    "    strain_list = filt_stats[\n",
    "        lambda x: (x.species == species)\n",
    "        & (x.passes_filter)\n",
    "        & (x.genome_type.isin([\"Isolate\", \"MAG\"]))\n",
    "    ].genome_id.unique()\n",
    "    data = (\n",
    "        pd.read_table(inpath, index_col=\"strain\")\n",
    "        .rename(index=str)\n",
    "        .loc[strain_list]\n",
    "        .assign(species=species)\n",
    "        .reset_index()\n",
    "        .set_index([\"species\", \"strain\"])\n",
    "    )\n",
    "    ref_prevalence_class_counts.append(data)\n",
    "ref_prevalence_class_counts = pd.concat(ref_prevalence_class_counts).sort_index()\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_median_prevalence_class_fraction = (\n",
    "    ref_prevalence_class_counts.groupby(\"species\")\n",
    "    .median()\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    ")\n",
    "ref_median_prevalence_class_fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = ref_median_prevalence_class_fraction.loc[species_list3].assign(\n",
    "    p__=lambda x: x.index.to_series().map(species_taxonomy.p__),\n",
    "    num_genomes=ref_prevalence_class_counts.reset_index()[\"species\"].value_counts(),\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 3))\n",
    "_, ax_legend = plt.subplots()\n",
    "\n",
    "for p__ in phylum_order:\n",
    "    d1 = d0[lambda x: x.p__ == p__]\n",
    "    ax.scatter(\n",
    "        \"core\",\n",
    "        \"shell\",\n",
    "        data=d1,\n",
    "        color=phylum_palette[p__],\n",
    "        facecolor=\"none\",\n",
    "        s=15,\n",
    "        label=\"__nolegend__\",\n",
    "        # marker=\"o\",\n",
    "        # lw=2,\n",
    "        # facecolor=\"none\",\n",
    "        # alpha=0.85,\n",
    "    )\n",
    "    ax_legend.scatter(\n",
    "        [], [], color=phylum_palette[p__], facecolor=\"none\", label=p__, s=50, lw=3\n",
    "    )\n",
    "ax_legend.legend(bbox_to_anchor=(1, 0.5))\n",
    "lib.plot.hide_axes_and_spines(ax_legend)\n",
    "\n",
    "\n",
    "for cloud_frac, shift in [(0.0, 0.06), (0.1, 0.04), (0.2, 0.04), (0.3, 0.04)]:\n",
    "    if cloud_frac == 0.3:\n",
    "        annot = f\"Cloud {cloud_frac:.0%}\"\n",
    "    else:\n",
    "        annot = f\"{cloud_frac:.0%}\"\n",
    "    ax.plot(\n",
    "        [0, 1 - cloud_frac],\n",
    "        [1 - cloud_frac, 0],\n",
    "        lw=1,\n",
    "        linestyle=\":\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax.annotate(\n",
    "        annot,\n",
    "        xy=(1 - cloud_frac - shift, -0.04 + shift),\n",
    "        rotation=-45,\n",
    "        va=\"bottom\",\n",
    "        ha=\"right\",\n",
    "        fontsize=10,\n",
    "    )\n",
    "\n",
    "ax.set_xlim(0.35, 0.95)\n",
    "ax.set_ylim(0.0, 0.6)\n",
    "ax.xaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "ax.yaxis.set_major_formatter(mpl.ticker.PercentFormatter(xmax=1))\n",
    "ax.set_xlabel(\"Core\")\n",
    "ax.set_ylabel(\"Shell\")\n",
    "ax.set_aspect(1)\n",
    "\n",
    "fig.savefig(\"fig/fig_s2.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_median_prevalence_class_fraction.loc[species_list3].quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## COG Cateogry Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"cog_category\", \"color\", \"description\"],\n",
    "    index_col=\"cog_category\",\n",
    ").description\n",
    "cog_category_description.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category = []\n",
    "for species in tqdm(species_list3):\n",
    "    cog_category_inpath = (\n",
    "        f\"data/species/sp-{species}/midasdb_v20.emapper.gene75_x_cog_category.tsv\"\n",
    "    )\n",
    "    cog_category.append(pd.read_table(cog_category_inpath))\n",
    "\n",
    "cog_category = pd.concat(cog_category)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category_matrix = (\n",
    "    cog_category.set_index([\"centroid_75\", \"cog_category\"])\n",
    "    .assign(annotation=True)\n",
    "    .unstack(\"cog_category\", fill_value=False)\n",
    "    .annotation\n",
    "    # .reindex(spgc_gene_prevalence.index)\n",
    "    # .fillna({'no_category': True})\n",
    "    # .fillna(False)\n",
    ")\n",
    "gene_x_cog_category_matrix.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Double check that only genes with no other cog category get assigned \"no_category\".\n",
    "assert not idxwhere(gene_x_cog_category_matrix[lambda x: x.no_category].sum(1) > 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_prevalence_class = spgc_gene_prevalence.map(\n",
    "    lambda x: np.where(\n",
    "        x > 0.9, \"core\", np.where(x > 0.15, \"shell\", np.where(x > 0, \"cloud\", \"absent\"))\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    spgc_prevalence_class.to_frame(\"prevalence_class\")\n",
    "    .join(gene_x_cog_category_matrix)\n",
    "    .assign(\n",
    "        cloud=lambda x: x.prevalence_class == \"cloud\",\n",
    "        shell=lambda x: x.prevalence_class == \"shell\",\n",
    "        core=lambda x: x.prevalence_class == \"core\",\n",
    "        absent=lambda x: x.prevalence_class == \"absent\",\n",
    "    )\n",
    ")[lambda x: ~x.absent]\n",
    "\n",
    "result = []\n",
    "for _prevalence_class, _cog_category in tqdm(\n",
    "    list(\n",
    "        product(\n",
    "            [\"core\", \"shell\", \"cloud\", \"absent\"], gene_x_cog_category_matrix.columns\n",
    "        )\n",
    "    )\n",
    "):\n",
    "    d1 = (\n",
    "        d0[[_prevalence_class, _cog_category]]\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[True, False], columns=[True, False])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    d1_pc = d1 + 1\n",
    "    log_oddsratio = np.log2(\n",
    "        (d1_pc.loc[True, True] / d1_pc.loc[True, False])\n",
    "        / (d1_pc.loc[False, True] / d1_pc.loc[False, False])\n",
    "    )\n",
    "    result.append(\n",
    "        (\n",
    "            _prevalence_class,\n",
    "            _cog_category,\n",
    "            d1.loc[True, True],\n",
    "            log_oddsratio,\n",
    "            sp.stats.fisher_exact(d1)[1],\n",
    "        )\n",
    "    )\n",
    "prevalence_class_cog_category_enrichment = pd.DataFrame(\n",
    "    result,\n",
    "    columns=[\n",
    "        \"prevalence_class\",\n",
    "        \"cog_category\",\n",
    "        \"num_genes\",\n",
    "        \"log2_oddsratio\",\n",
    "        \"pvalue\",\n",
    "    ],\n",
    ").set_index([\"prevalence_class\", \"cog_category\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_label = {\n",
    "    \"J\": \"Ribosomes / Translation - J\",\n",
    "    \"A\": \"RNA Processing - A\",\n",
    "    \"K\": \"Transcription - K\",\n",
    "    \"L\": \"DNA replication/recombination/repair - L\",\n",
    "    \"B\": \"Chromatin - B\",\n",
    "    \"D\": \"Cell cycle control - D\",\n",
    "    \"Y\": \"Nucleus - Y\",\n",
    "    \"V\": \"Defense - V\",\n",
    "    \"T\": \"Signal transduction - T\",\n",
    "    \"M\": \"Cell envelope - M\",\n",
    "    \"N\": \"Motility - N\",\n",
    "    \"Z\": \"Cytoskeleton - Z\",\n",
    "    \"W\": \"Extracellular structures - W\",\n",
    "    \"U\": \"Secretion / vesicular transport - U\",\n",
    "    \"O\": \"Protein processing - O\",\n",
    "    \"X\": '\"Mobilome\" - X',\n",
    "    \"C\": \"Energy - C\",\n",
    "    \"G\": \"Carbohydrates - G\",\n",
    "    \"E\": \"Amino acids - E\",\n",
    "    \"F\": \"Nucleotides - F\",\n",
    "    \"H\": \"Coenzymes - H\",\n",
    "    \"I\": \"Lipids - I\",\n",
    "    \"P\": \"Inorganic ions - P\",\n",
    "    \"Q\": \"Secondary metabolites - Q\",\n",
    "    \"R\": \"General only - R\",\n",
    "    \"S\": \"TODO: This shouldn't show up\",\n",
    "    \"no_category\": \"Unknown\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class_cog_category_enrichment.groupby(\n",
    "    \"cog_category\"\n",
    ").num_genes.sum().sort_values()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class_cog_category_enrichment.num_genes.unstack(\"prevalence_class\").loc[\n",
    "    [\"Y\", \"B\", \"A\"]\n",
    "].sum(\n",
    "    1\n",
    ")  # .sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 4C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = prevalence_class_cog_category_enrichment\n",
    "\n",
    "cog_category_drop_list = [\"Y\", \"B\", \"A\"]\n",
    "\n",
    "cog_category_order = [\n",
    "    c\n",
    "    for c in d.xs(\"core\").log2_oddsratio.sort_values(ascending=True).index\n",
    "    if c not in cog_category_drop_list\n",
    "]\n",
    "cog_category_idx = pd.Series(\n",
    "    np.arange(len(cog_category_order)), index=cog_category_order\n",
    ").rename_axis(\"cog_category\")\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]\n",
    "prevalence_class_idx = pd.Series(\n",
    "    np.arange(len(prevalence_class_order)), index=prevalence_class_order\n",
    ").rename_axis(\"prevalence_class\")\n",
    "\n",
    "num_genes_to_size = lambda x: 20 * np.log(x + 1)\n",
    "signif_size = 20\n",
    "\n",
    "d = (\n",
    "    d.join(prevalence_class_idx.rename(\"prevalence_class_idx\"))\n",
    "    .join(cog_category_idx.rename(\"cog_category_idx\"))\n",
    "    .assign(\n",
    "        num_genes_s=lambda x: num_genes_to_size(x.num_genes),\n",
    "        signif=lambda x: signif_size * (x.pvalue >= 0.05),\n",
    "    )\n",
    ")\n",
    "\n",
    "vmin, vmax = -np.log10(20), np.log10(20)\n",
    "\n",
    "cmap = \"coolwarm\"\n",
    "norm = mpl.colors.PowerNorm(1, vmin=vmin, vmax=vmax)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(1, 8), facecolor=\"none\")\n",
    "ax.scatter(\n",
    "    x=\"prevalence_class_idx\",\n",
    "    y=\"cog_category_idx\",\n",
    "    data=d,\n",
    "    c=\"log2_oddsratio\",\n",
    "    s=\"num_genes_s\",\n",
    "    cmap=cmap,\n",
    "    norm=norm,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "ax.scatter(\n",
    "    x=\"prevalence_class_idx\",\n",
    "    y=\"cog_category_idx\",\n",
    "    data=d,\n",
    "    s=\"signif\",\n",
    "    color=\"k\",\n",
    "    marker=\"x\",\n",
    "    label=\"__nolegend__\",\n",
    "    lw=1,\n",
    ")\n",
    "# for _, d1 in d.iterrows():\n",
    "#     ax.annotate(d1.signif, xy=(d1.prevalence_class_idx, d1.cog_category_idx), ha='center', va='center')\n",
    "\n",
    "\n",
    "ax.set_xlim(-0.5, len(prevalence_class_order) - 0.5)\n",
    "ax.set_ylim(-1.0, len(cog_category_order))\n",
    "ax.set_xticks(prevalence_class_idx)\n",
    "ax.set_xticklabels(prevalence_class_order)\n",
    "ax.set_yticks(cog_category_idx)\n",
    "ax.set_yticklabels([cog_category_label[c] for c in cog_category_order])\n",
    "lib.plot.rotate_yticklabels(ax=ax, rotation=30, va=\"top\")\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "# lib.plot.rotate_yticklabels(ax=ax, rotation=0, va='top')\n",
    "\n",
    "\n",
    "# Remove frame\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.spines[\"bottom\"].set_visible(False)\n",
    "ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "for num_genes in np.logspace(0, 4, num=5):\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        color=\"grey\",\n",
    "        edgecolor=\"black\",\n",
    "        label=num_genes,\n",
    "        s=num_genes_to_size(num_genes),\n",
    "    )\n",
    "ax.legend(\n",
    "    bbox_to_anchor=(1, 1),\n",
    "    frameon=False,\n",
    "    labelspacing=1,\n",
    ")\n",
    "fig.savefig(\"fig/fig4c_cogcat_enrichment.pdf\", bbox_inches=\"tight\")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(1.5, 0.1))\n",
    "fig.colorbar(\n",
    "    mpl.cm.ScalarMappable(norm=norm, cmap=cmap),\n",
    "    cax=ax,\n",
    "    extend=\"both\",\n",
    "    label=\"Enrichment\",\n",
    "    orientation=\"horizontal\",\n",
    ")\n",
    "ax.set_xticks([-np.log10(20), -np.log10(5), 0, np.log10(5), np.log10(20)])\n",
    "ax.set_xticklabels([\"\u22641:20\", \"1:5\", \"1:1\", \"5:1\", \"\u226520:1\"])\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "\n",
    "fig.savefig(\"fig/fig4c_cogcat_enrichment_cbar.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = prevalence_class_cog_category_enrichment\n",
    "\n",
    "cog_category_order = d.xs(\"core\").log2_oddsratio.sort_values(ascending=True).index\n",
    "cog_category_idx = pd.Series(\n",
    "    np.arange(len(cog_category_order)), index=cog_category_order\n",
    ").rename_axis(\"cog_category\")\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]\n",
    "prevalence_class_idx = pd.Series(\n",
    "    np.arange(len(prevalence_class_order)), index=prevalence_class_order\n",
    ").rename_axis(\"prevalence_class\")\n",
    "\n",
    "num_genes_to_size = lambda x: 55 * np.log(x + 1)\n",
    "signif_size = 20\n",
    "\n",
    "d = (\n",
    "    d.join(prevalence_class_idx.rename(\"prevalence_class_idx\"))\n",
    "    .join(cog_category_idx.rename(\"cog_category_idx\"))\n",
    "    .assign(\n",
    "        num_genes_s=lambda x: num_genes_to_size(x.num_genes),\n",
    "        signif=lambda x: signif_size * (x.pvalue >= 0.05),\n",
    "    )\n",
    ")\n",
    "\n",
    "d.xs(\"core\").sort_values(\"log2_oddsratio\", ascending=False).join(\n",
    "    cog_category_description\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.clustermap(1 - dmatrix(gene_x_cog_category_matrix.T, metric=\"correlation\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AMR Genes Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "amr_gene = []\n",
    "missing = []\n",
    "\n",
    "for species in tqdm(species_list):\n",
    "    amr_gene_inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.amr-strain_gene.tsv\"\n",
    "    if os.path.exists(amr_gene_inpath):\n",
    "        amr_gene.append(\n",
    "            pd.read_table(amr_gene_inpath, index_col=\"gene_id\").rename(\n",
    "                columns=lambda x: species + \"_\" + x\n",
    "            )\n",
    "        )\n",
    "        assert amr_gene[-1].index.is_unique\n",
    "    else:\n",
    "        missing.append(species)\n",
    "\n",
    "amr_gene = pd.concat(amr_gene, axis=1).fillna(0) > 0\n",
    "print(len(missing))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_strain_list = spgc_strain_list2\n",
    "\n",
    "d0 = (\n",
    "    amr_gene.reindex(columns=_strain_list)\n",
    "    .dropna(axis=1)\n",
    "    .sum()\n",
    "    .rename_axis(index=\"species_strain\")\n",
    "    .reset_index(name=\"num_amr_accessions\")\n",
    "    .assign(\n",
    "        species=lambda x: x.species_strain.str.split(\"_\").str[0],\n",
    "        strain=lambda x: x.species_strain.str.split(\"_\").str[1],\n",
    "    )\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    ")\n",
    "\n",
    "print((d0.num_amr_accessions > 0).mean())\n",
    "plt.hist(d0.num_amr_accessions, density=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print((d0.num_amr_accessions > 0).sum(), (d0.num_amr_accessions >= 0).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0.assign(has_amr=lambda x: x.num_amr_accessions > 0).groupby(\"p__\").has_amr.agg(\n",
    "    [\"sum\", \"count\", \"mean\"]\n",
    ").sort_values(\"mean\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1 = (\n",
    "    d0.assign(has_amr_accessions=lambda x: x.num_amr_accessions.gt(0))\n",
    "    .groupby(\"species\")\n",
    "    .has_amr_accessions.agg([\"sum\", \"mean\", \"count\"])\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"sum\": \"num_strains_with_amr\",\n",
    "            \"mean\": \"frac_strains_with_amr\",\n",
    "            \"count\": \"total_num_strains\",\n",
    "        }\n",
    "    )\n",
    "    # .to_frame(\"frac_strains_with_amr\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "d1\n",
    "plt.hist(d1.frac_strains_with_amr, bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.stripplot(data=d1, x=\"p__\", y=\"frac_strains_with_amr\")\n",
    "lib.plot.rotate_xticklabels()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_amr_gene = []\n",
    "for species in tqdm(species_list3):\n",
    "    amr_annot_inpath = f\"data/species/sp-{species}/midasdb_v20.gene75_x_amr.tsv\"\n",
    "    is_amr_gene.append(pd.read_table(amr_annot_inpath))\n",
    "\n",
    "amr_gene_list = list(\n",
    "    pd.concat(is_amr_gene).centroid_75.unique()\n",
    ")  # .drop_duplicates(subset='centroid_75').assign(is_amr_gene=True).set_index('centroid_75').is_amr_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    spgc_prevalence_class[lambda x: x != \"absent\"]\n",
    "    .to_frame()\n",
    "    .assign(is_amr_gene=lambda x: x.index.to_series().isin(amr_gene_list))\n",
    ")\n",
    "\n",
    "for prevalence_class in [\"core\", \"shell\", \"cloud\", \"absent\"]:\n",
    "    d1 = (\n",
    "        d0.assign(is_prevalence_class=lambda x: x.prevalence == prevalence_class)[\n",
    "            [\"is_amr_gene\", \"is_prevalence_class\"]\n",
    "        ]\n",
    "        .value_counts()\n",
    "        .unstack(\"is_prevalence_class\")\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    d1_pc = d1 + 1\n",
    "    log_odds_ratio = np.log2(\n",
    "        (d1_pc.loc[True, True] / d1_pc.loc[True, False])\n",
    "        / (d1_pc.loc[False, True] / d1_pc.loc[False, False])\n",
    "    )\n",
    "    print(prevalence_class, log_odds_ratio, sp.stats.fisher_exact(d1))\n",
    "    print(d1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0.is_amr_gene.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0[lambda x: (x.prevalence == \"absent\")]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz5",
   "language": "python",
   "name": "toolz5"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}