{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import mpltern\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.dissimilarity import load_dmat_as_pickle\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"paper\")\n",
    "plt.rcParams[\"figure.dpi\"] = 200"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type_palette = {\n",
    "    \"SPGC\": \"tab:green\",\n",
    "    \"MAG\": \"tab:orange\",\n",
    "    \"Isolate\": \"tab:blue\",\n",
    "    \"Ref\": \"black\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[lambda x: x.species_group_id == \"hmp2\"]\n",
    "    .species_id.astype(str)\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "species_taxonomy = (\n",
    "    pd.read_table(species_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    "    .Lineage.apply(parse_taxonomy_string)\n",
    ")\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_order = [\n",
    "    \"p__Euryarchaeota\",\n",
    "    \"p__Thermoplasmatota\",\n",
    "    \"p__Firmicutes\",\n",
    "    \"p__Firmicutes_A\",\n",
    "    \"p__Firmicutes_C\",\n",
    "    # \"p__Firmicutes_B\", # None in species_list1\n",
    "    # \"p__Firmicutes_G\", # B/G/I not sure how related to C or A\n",
    "    # \"p__Firmicutes_I\", #\n",
    "    # \"p__Cyanobacteria\", # None in species_list1\n",
    "    \"p__Actinobacteriota\",\n",
    "    \"p__Synergistota\",\n",
    "    \"p__Fusobacteriota\",\n",
    "    \"p__Campylobacterota\",\n",
    "    \"p__Proteobacteria\",\n",
    "    \"p__Desulfobacterota_A\",\n",
    "    \"p__Bacteroidota\",\n",
    "    \"p__Verrucomicrobiota\",\n",
    "    # \"dummy0\", # 18\n",
    "    # \"dummy1\", # 19\n",
    "    # \"dummy2\", # 20\n",
    "]\n",
    "\n",
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    phylum_order,\n",
    "    cm=\"rainbow\",\n",
    "    desaturate_levels=[1.0, 0.5],\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(2, 3), facecolor=\"none\")\n",
    "for p__ in phylum_order:\n",
    "    print(p__, phylum_palette[p__])\n",
    "    plt.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        color=phylum_palette[p__],\n",
    "        lw=0.5,\n",
    "        edgecolor=\"k\",\n",
    "        label=p__.replace(\"p__\", \"\"),\n",
    "    )\n",
    "plt.legend(ncols=1, markerscale=1.5, frameon=False)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "fig.savefig('fig/fig2_phylum_legend.pdf')\n",
    "\n",
    "fig = plt.figure(figsize=(2, 1.5), facecolor=\"none\")\n",
    "for p__ in phylum_order:\n",
    "    print(p__, phylum_palette[p__])\n",
    "    plt.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        color=phylum_palette[p__],\n",
    "        lw=0.5,\n",
    "        edgecolor=\"k\",\n",
    "        label=p__.replace(\"p__\", \"\"),\n",
    "    )\n",
    "plt.legend(ncols=2, markerscale=1.5, frameon=False)\n",
    "lib.plot.hide_axes_and_spines()\n",
    "fig.savefig('fig/fig4_phylum_legend_2col.pdf')\n",
    "\n",
    "\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen = pd.read_table(\"meta/hmp2/mgen.tsv\", index_col=\"library_id\")\n",
    "preparation = pd.read_table(\"meta/hmp2/preparation.tsv\", index_col=\"preparation_id\")\n",
    "stool = pd.read_table(\"meta/hmp2/stool.tsv\", index_col=\"stool_id\")\n",
    "visit = pd.read_table(\"meta/hmp2/visit.tsv\", index_col=\"visit_id\")\n",
    "subject = pd.read_table(\"meta/hmp2/subject.tsv\", index_col=\"subject_id\")\n",
    "\n",
    "meta_all = (\n",
    "    mgen.join(preparation.drop(columns=\"library_type\"), on=\"preparation_id\")\n",
    "    .join(stool, on=\"stool_id\")\n",
    "    .join(visit, on=\"visit_id\", rsuffix=\"_\")\n",
    "    .join(subject, on=\"subject_id\")\n",
    "    .assign(\n",
    "        new_name=lambda x: (\n",
    "            x[[\"subject_id\", \"week_number\"]]\n",
    "            .assign(library_id=x.index)\n",
    "            .assign(week_number=lambda x: x.week_number.fillna(999).astype(int))\n",
    "            .apply(lambda x: \"_\".join(x.astype(str)), axis=1)\n",
    "        )\n",
    "    )\n",
    "    # .reset_index()\n",
    "    # .set_index('new_name')\n",
    ")\n",
    "\n",
    "library_id_to_new_name = meta_all.new_name\n",
    "\n",
    "assert not any(meta_all.subject_id.isna())\n",
    "\n",
    "# TODO: Rename samples based on subject and visit number\n",
    "# TODO: Drop duplicate stools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Species Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = []\n",
    "_missing_species = []\n",
    "\n",
    "for species in tqdm(species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gene99_v20-v23-agg75.spgc_specgene-ref-filt-p95.species_depth.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        _missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, names=[\"sample\", \"depth\"]).assign(species=species)\n",
    "    species_depth.append(data)\n",
    "species_depth = (\n",
    "    pd.concat(species_depth)\n",
    "    .set_index([\"sample\", \"species\"])\n",
    "    .depth.unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "print(\n",
    "    len(_missing_species),\n",
    "    \"out of\",\n",
    "    len(species_list),\n",
    "    \"species are missing.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "depth_thresh = 0.2\n",
    "\n",
    "species_found = species_depth > depth_thresh\n",
    "species_prevalence = (\n",
    "    species_found.groupby(meta_all.subject_id).any().mean().sort_values(ascending=False)\n",
    ")\n",
    "\n",
    "species_rabund_when_found = species_depth.divide(species_depth.sum(1), axis=0).where(\n",
    "    species_found, np.nan\n",
    ")\n",
    "species_mean_rabund_when_found = (\n",
    "    species_rabund_when_found.groupby(meta_all.subject_id)\n",
    "    .mean()\n",
    "    .mean()\n",
    "    .sort_values(ascending=False)\n",
    ")\n",
    "species_median_rabund_when_found = (\n",
    "    species_rabund_when_found.groupby(meta_all.subject_id)\n",
    "    .median()\n",
    "    .median()\n",
    "    .sort_values(ascending=False)\n",
    ")\n",
    "\n",
    "species_prevalence.to_frame(\"prevalence\").assign(\n",
    "    mean_rabund=species_mean_rabund_when_found,\n",
    "    median_rabund=species_median_rabund_when_found,\n",
    ").join(species_taxonomy).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_genome(x):\n",
    "    if (x.genome_type == \"Isolate\") & x.passes_filter:\n",
    "        return \"isolate\"\n",
    "    elif (x.genome_type == \"Isolate\") & ~x.passes_filter:\n",
    "        return \"isolate_fails_qc\"\n",
    "    elif (x.genome_type == \"MAG\") & x.passes_filter:\n",
    "        return \"mag\"\n",
    "    elif (x.genome_type == \"MAG\") & ~x.passes_filter:\n",
    "        return \"mag_fails_qc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_filter:\n",
    "        return \"spgc\"\n",
    "    elif (x.genome_type == \"SPGC\") & x.passes_geno_positions:\n",
    "        return \"sfacts_only\"\n",
    "    elif (x.genome_type == \"SPGC\") & ~(x.passes_geno_positions):\n",
    "        return \"sfacts_fails_qc\"\n",
    "    else:\n",
    "        raise ValueError(\"Genome did not match classification criteria:\", x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.strain_meta_spgc_and_ref.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath).assign(species=species, inpath=inpath)\n",
    "    filt_stats.append(data)\n",
    "filt_stats = (\n",
    "    pd.concat(filt_stats)\n",
    "    .assign(\n",
    "        genome_class=lambda x: x.apply(classify_genome, axis=1),\n",
    "        species_strain=lambda x: x.species + \"_\" + x.genome_id,\n",
    "    )\n",
    "    .set_index(\"species_strain\")\n",
    ")\n",
    "\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many strains and species were detected at each filtering level."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define different subsets of the species:\n",
    "\n",
    "# All species:\n",
    "# species_list\n",
    "spgc_strain_list = filt_stats[lambda x: x.genome_type.isin([\"SPGC\"])].index.values\n",
    "\n",
    "# All species with enough positions\n",
    "species_list0 = filt_stats[lambda x: x.passes_geno_positions].species.unique()\n",
    "spgc_strain_list0 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list0)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains\n",
    "species_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list1 = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1)\n",
    "].index.values\n",
    "\n",
    "# All species with sf strains to talk about distributions (>=10)\n",
    "species_list1b = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_geno_positions & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list1b = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list1b)\n",
    "].index.values\n",
    "\n",
    "# All species with spgc strains\n",
    "species_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "].species.unique()\n",
    "spgc_strain_list2 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list2)\n",
    "].index.values\n",
    "\n",
    "# All species with enough spgc strains for pangenome analysis (>=10)\n",
    "species_list3 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 10\n",
    ")\n",
    "spgc_strain_list3 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list3)\n",
    "].index.values\n",
    "\n",
    "# Species with large numbers of strains (>=20)\n",
    "species_list4 = idxwhere(\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "    ].species.value_counts()\n",
    "    >= 20\n",
    ")\n",
    "spgc_strain_list4 = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & x.genome_type.isin([\"SPGC\"])\n",
    "    & x.species.isin(species_list4)\n",
    "].index.values\n",
    "\n",
    "_species_list_map = {\n",
    "    \"All considered species\": (species_list, spgc_strain_list),\n",
    "    \"0: Species with enough genotyped positions\": (species_list0, spgc_strain_list0),\n",
    "    \"1: With sfacts strains\": (species_list1, spgc_strain_list1),\n",
    "    \"1b: With (>=10) sfacts strains\": (species_list1b, spgc_strain_list1b),\n",
    "    \"2: With SPGC inferences\": (species_list2, spgc_strain_list2),\n",
    "    \"3: With >=10 inferences\": (species_list3, spgc_strain_list3),\n",
    "    \"4: With >=20 inferences\": (species_list4, spgc_strain_list4),\n",
    "}\n",
    "for _species_list_name, (_species_list, _strain_list) in _species_list_map.items():\n",
    "    print(_species_list_name, len(_species_list), len(_strain_list))\n",
    "    print(species_taxonomy.loc[_species_list].p__.value_counts())\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supplementary Table 1\n",
    "d = filt_stats.loc[spgc_strain_list2][\n",
    "    [\n",
    "        \"species\",\n",
    "        \"num_gene\",\n",
    "        \"num_strain_sample\",\n",
    "        \"sum_strain_depth\",\n",
    "        \"species_gene_frac\",\n",
    "        \"log_selected_gene_depth_ratio_std\",\n",
    "        \"num_geno_positions\",\n",
    "    ]\n",
    "].astype(dict(num_gene=int, num_strain_sample=int, num_geno_positions=int))\n",
    "d.to_csv(\"fig/hmp2_inferred_strains_supplementary_table1.tsv\", sep=\"\\t\")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = filt_stats.loc[spgc_strain_list2][\n",
    "    [\n",
    "        \"species\",\n",
    "        \"num_gene\",\n",
    "        \"num_strain_sample\",\n",
    "        \"sum_strain_depth\",\n",
    "        \"species_gene_frac\",\n",
    "        \"log_selected_gene_depth_ratio_std\",\n",
    "        \"num_geno_positions\",\n",
    "    ]\n",
    "].astype(dict(num_gene=int, num_strain_sample=int, num_geno_positions=int))\n",
    "\n",
    "bins = [0] + list(np.logspace(0, 8, base=2, num=9).astype(int))\n",
    "\n",
    "print(list(zip(*np.histogram(d.num_strain_sample, bins=bins))))\n",
    "print(d.num_strain_sample.quantile([0.25, 0.5, 0.75]))\n",
    "plt.hist(d.num_strain_sample, bins=bins, histtype='bar')\n",
    "plt.xlim(0, 256)\n",
    "plt.xscale('symlog', linthresh=1, linscale=0.2)\n",
    "plt.xticks(ticks=bins, labels=bins)\n",
    "plt.xticks(ticks=[], labels=[], minor=True)\n",
    "plt.xlabel('Num. Strain-Pure Samples')\n",
    "plt.ylabel('Strains (count)')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3E"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Annotations for Figure 3 species tree:\n",
    "filt_stats.loc[spgc_strain_list2].species.value_counts().to_csv(\n",
    "    \"fig/hmp2_spgc_strain_counts.tsv\", sep=\"\\t\", header=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diss_stats = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list2\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.strain_diss_spgc_and_ref.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath).assign(species=species, inpath=inpath)\n",
    "    diss_stats.append(data)\n",
    "diss_stats = (\n",
    "    pd.concat(diss_stats)\n",
    "    .set_index([\"species\", \"genome_id\"])\n",
    "    .join(filt_stats.set_index([\"species\", \"genome_id\"]), rsuffix=\"_b\")\n",
    "    .reset_index()\n",
    "    .assign(\n",
    "        species_strain=lambda x: x.species + \"_\" + x.genome_id,\n",
    "    )\n",
    "    .set_index(\"species_strain\")\n",
    ")\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many Archaea (+Bacteria)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats.loc[spgc_strain_list2].join(species_taxonomy, on=\"species\")[\n",
    "    lambda x: (x.d__ == \"d__Archaea\")\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats.loc[spgc_strain_list2].join(species_taxonomy, on=\"species\")[\n",
    "    lambda x: (x.d__ == \"d__Bacteria\")\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Novelty of SPGC/SFacts strains"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Novel Branch Length Fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_lengths2 = []\n",
    "missing_species2 = []\n",
    "\n",
    "for species in tqdm(species_list):\n",
    "    dmat_path = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.spgc_ss-all.geno_uhgg-v20_pdist-mask10-pseudo10.pkl\"\n",
    "    meta_path = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.strain_meta_spgc_and_ref.tsv\"\n",
    "    if not (os.path.exists(dmat_path) and os.path.exists(meta_path)):\n",
    "        missing_species2.append(species)\n",
    "        continue\n",
    "    _meta = pd.read_table(meta_path, index_col=\"genome_id\")\n",
    "    dmat_all = lib.dissimilarity.load_dmat_as_pickle(dmat_path)\n",
    "    without_spgc_list = idxwhere(\n",
    "        _meta.passes_geno_positions & _meta.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "    )\n",
    "    with_spgc_list = idxwhere(\n",
    "        _meta.passes_geno_positions & _meta.genome_type.isin([\"Isolate\", \"MAG\", \"SPGC\"])\n",
    "    )\n",
    "    dmat_without_spgc = dmat_all.loc[without_spgc_list, without_spgc_list]\n",
    "    dmat_with_spgc = dmat_all.loc[with_spgc_list, with_spgc_list]\n",
    "    if dmat_without_spgc.empty:\n",
    "        missing_species2.append(species)\n",
    "        continue\n",
    "    _tree_without_spgc = sp.cluster.hierarchy.linkage(\n",
    "        squareform(dmat_without_spgc), method=\"average\"\n",
    "    )\n",
    "    _tree_with_spgc = sp.cluster.hierarchy.linkage(\n",
    "        squareform(dmat_with_spgc), method=\"average\"\n",
    "    )\n",
    "    branch_lengths2.append(\n",
    "        dict(\n",
    "            species=species,\n",
    "            length_with_spgc=_tree_with_spgc[:, 2].sum(),\n",
    "            length_without_spgc=_tree_without_spgc[:, 2].sum(),\n",
    "            num_ref=len(without_spgc_list),\n",
    "            num_spgc=len(with_spgc_list) - len(without_spgc_list),\n",
    "            mean_diss_with_spgc=squareform(dmat_with_spgc).mean(),\n",
    "            mean_diss_without_spgc=squareform(dmat_without_spgc).mean(),\n",
    "        )\n",
    "    )\n",
    "\n",
    "branch_lengths2 = pd.DataFrame(branch_lengths2).set_index(\"species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(missing_species2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "branch_lengths2.assign(\n",
    "    mean_diss_diff=lambda x: x.mean_diss_with_spgc - x.mean_diss_without_spgc,\n",
    "    mean_diss_ratio=lambda x: x.mean_diss_diff / x.mean_diss_without_spgc,\n",
    ").sort_values(\"mean_diss_ratio\").head(50)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Figure 3D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    # Take all genomes that have enough geno positions for relatedness estimation\n",
    "    filt_stats[lambda x: x.passes_geno_positions]\n",
    "    # Count the number of genomes of each type.\n",
    "    [[\"species\", \"genome_id\", \"genome_type\"]]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    # Tag each cluster by it's \"best type\"\n",
    "    .assign(best_genome_type=lambda x: x[[\"Isolate\", \"MAG\", \"SPGC\"]].idxmax(1))\n",
    "    # Count for each species the number of clusters with each tag.\n",
    "    .groupby(\"species\")\n",
    "    .best_genome_type.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .join(\n",
    "        branch_lengths2.assign(\n",
    "            branch_length_diff=lambda x: x.length_with_spgc - x.length_without_spgc,\n",
    "            branch_length_ratio=lambda x: x.branch_length_diff / x.length_without_spgc,\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        species_prevalence.to_frame(\"prevalence\").assign(\n",
    "            mean_rabund=species_mean_rabund_when_found,\n",
    "            median_rabund=species_median_rabund_when_found,\n",
    "        )\n",
    "    )\n",
    "    .join(species_taxonomy)\n",
    "    .loc[species_list1]\n",
    ")\n",
    "\n",
    "p__meta = pd.DataFrame([], index=phylum_order[::-1]).assign(\n",
    "    pos=lambda x: np.arange(len(x))\n",
    ")\n",
    "\n",
    "s_func = lambda x: 4 * x**0.5\n",
    "\n",
    "np.random.seed(42)\n",
    "fig, ax = plt.subplots(figsize=(3.8, 5), facecolor=\"none\")\n",
    "ax.scatter(\n",
    "    \"branch_length_ratio\",\n",
    "    \"pos_jitter\",\n",
    "    data=d.join(p__meta, on=\"p__\").assign(\n",
    "        pos_jitter=lambda x: x.pos + np.random.uniform(-0.35, 0.35, size=len(x)),\n",
    "        c=lambda x: x.p__.map(phylum_palette),\n",
    "        s=lambda x: s_func(x[\"SPGC\"]),  # Total SPGC genomes\n",
    "    ),\n",
    "    # c='c',\n",
    "    marker=\"o\",\n",
    "    s=\"s\",\n",
    "    facecolors=\"c\",\n",
    "    edgecolors=\"k\",\n",
    "    lw=0.25,\n",
    "    alpha=1.0,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "ax.set_yticks(p__meta.pos.unique())\n",
    "ax.set_yticklabels(p__meta.index.to_series().str.replace(\"p__\", \"\"))\n",
    "# for ytick, p__, c in zip(ax.get_yticklabels(), p__meta.index.to_series(), p__meta.index.to_series().map(phylum_palette)):\n",
    "#     print(ytick, p__, c)\n",
    "#     ytick.set_color(c)\n",
    "\n",
    "\n",
    "ax.set_xlabel(\"Branch Length Increase\")\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-2, linscale=0.2)\n",
    "ax.set_xticks([0, 0.01, 0.1, 1.0, 3.0])\n",
    "ax.set_xticklabels([\"0%\", \"1%\", \"10%\", \"100%\", \"300%\"])\n",
    "ax.set_ylabel(\"Phylum\")\n",
    "ax.set_xlim(-0.01, 6.0)  # NOTE: This is expl\n",
    "\n",
    "# # sns.stripplot(x='p__', y='Ref_branch_frac', data=d)\n",
    "\n",
    "for count in [1, 10, 100]:\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        edgecolor=\"black\",\n",
    "        lw=0.25,\n",
    "        facecolor=\"grey\",\n",
    "        s=s_func(count),\n",
    "        label=count,\n",
    "    )\n",
    "plt.legend(title=\"Num.\\ninferred\\nstrains\", fontsize=8, bbox_to_anchor=(0.755, 0.55))\n",
    "\n",
    "plt.savefig(\"fig/fig3d_branch_length.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    # Take all genomes that have enough geno positions for relatedness estimation\n",
    "    filt_stats[lambda x: x.passes_geno_positions]\n",
    "    # Count the number of genomes of each type.\n",
    "    [[\"species\", \"genome_id\", \"genome_type\"]]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    # Tag each cluster by it's \"best type\"\n",
    "    .assign(best_genome_type=lambda x: x[[\"Isolate\", \"MAG\", \"SPGC\"]].idxmax(1))\n",
    "    # Count for each species the number of clusters with each tag.\n",
    "    .groupby(\"species\")\n",
    "    .best_genome_type.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .join(\n",
    "        branch_lengths2.assign(\n",
    "            branch_length_diff=lambda x: x.length_with_spgc - x.length_without_spgc,\n",
    "            branch_length_ratio=lambda x: x.branch_length_diff / x.length_without_spgc,\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        species_prevalence.to_frame(\"prevalence\").assign(\n",
    "            mean_rabund=species_mean_rabund_when_found,\n",
    "            median_rabund=species_median_rabund_when_found,\n",
    "        )\n",
    "    )\n",
    "    .join(species_taxonomy)\n",
    "    .loc[species_list1]\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    d.num_ref, d.branch_length_ratio, c=d.num_spgc, norm=mpl.colors.PowerNorm(1 / 3)\n",
    ")\n",
    "plt.colorbar(label=\"Num. inferred strains\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-2)\n",
    "plt.xscale(\"log\")\n",
    "plt.xlabel(\"Num. reference strains\")\n",
    "plt.ylabel(\"Branch Length Increase\")\n",
    "\n",
    "d[\n",
    "    [\n",
    "        \"Isolate\",\n",
    "        \"MAG\",\n",
    "        \"SPGC\",\n",
    "        \"branch_length_ratio\",\n",
    "        \"prevalence\",\n",
    "        \"mean_rabund\",\n",
    "        \"median_rabund\",\n",
    "        \"d__\",\n",
    "        \"p__\",\n",
    "        \"c__\",\n",
    "        \"o__\",\n",
    "        \"f__\",\n",
    "        \"g__\",\n",
    "        \"s__\",\n",
    "    ]\n",
    "].sort_values(\"branch_length_ratio\", ascending=False).head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = branch_lengths2.assign(\n",
    "    branch_length_diff=lambda x: x.length_with_spgc - x.length_without_spgc,\n",
    "    branch_length_ratio=lambda x: x.branch_length_diff / x.length_without_spgc,\n",
    ").branch_length_ratio\n",
    "for thresh in [0.1, 0.2, 0.5]:\n",
    "    print(thresh, (d > thresh).sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"101493\"\n",
    "\n",
    "spgc_meta = pd.read_table(\n",
    "    f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.strain_meta_spgc_and_ref.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ")\n",
    "spgc_data = xr.load_dataset(\n",
    "    f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v20-v23-agg75.spgc-fit.nc\"\n",
    ")\n",
    "mgtp_diss_all = lib.dissimilarity.load_dmat_as_pickle(\n",
    "    f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.spgc_ss-all.geno_uhgg-v20_pdist-mask10-pseudo10.pkl\"\n",
    ")\n",
    "\n",
    "mgtp_diss = mgtp_diss_all.loc[\n",
    "    spgc_meta.passes_geno_positions, spgc_meta.passes_geno_positions\n",
    "]\n",
    "mgtp_linkage = sp.cluster.hierarchy.linkage(\n",
    "    squareform(mgtp_diss), method=\"average\", optimal_ordering=False\n",
    ")\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        type=spgc_meta.genome_type.map(genome_type_palette),\n",
    "        filt=spgc_meta.passes_filter.map({True: \"black\", False: \"grey\"}),\n",
    "    )\n",
    ")\n",
    "sns.clustermap(\n",
    "    mgtp_diss,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=mgtp_linkage,\n",
    "    col_linkage=mgtp_linkage,\n",
    "    figsize=(10, 10),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "w = sf.data.World.load(\n",
    "    f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "position_ss = w.random_sample(position=1500).position\n",
    "sf.plot.plot_metagenotype(\n",
    "    w.sel(position=position_ss), col_linkage_func=lambda w: w.metagenotype.linkage()\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    w.sel(position=position_ss), col_linkage_func=lambda w: w.metagenotype.linkage()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genotype Dissimilarity to Closest Reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This inlined calculation of genotype dissimilarity is necessary because the diss_stats are incomplete where no strains pass filtering.\n",
    "min_geno_diss = {}\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list1  # NOT Species list #2, here we're looking at the overall diversity of genotypes identified.\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    _spgc_list = filt_stats[\n",
    "        lambda x: (x.species == species)\n",
    "        & x.genome_type.isin([\"SPGC\"])\n",
    "        & x.passes_geno_positions\n",
    "    ].genome_id\n",
    "    _ref_list = filt_stats[\n",
    "        lambda x: (x.species == species)\n",
    "        & x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "        & x.passes_geno_positions\n",
    "    ].genome_id\n",
    "    inpath = f\"data/group/hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.spgc_ss-all.geno_uhgg-v20_pdist-mask10-pseudo10.pkl\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    min_geno_diss[species] = (\n",
    "        load_dmat_as_pickle(inpath)\n",
    "        .loc[_ref_list, _spgc_list]\n",
    "        .min()\n",
    "        .to_frame()\n",
    "        .assign(species=species)\n",
    "    )\n",
    "\n",
    "min_geno_diss = (\n",
    "    pd.concat(min_geno_diss.values())\n",
    "    .rename(columns={0: \"min_geno_diss\"})\n",
    "    .rename_axis(index=\"strain\")\n",
    "    .reset_index()\n",
    "    .set_index([\"species\", \"strain\"])\n",
    "    .min_geno_diss\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-5, 0, num=50)\n",
    "plt.hist(min_geno_diss, bins=bins)\n",
    "plt.axvline(0.05, lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.axvline(min_geno_diss.median(), lw=1, linestyle=\"-\", color=\"k\")\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "\n",
    "print(min_geno_diss.median(), (min_geno_diss > 0.05).mean())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Content Dissimilarity to Closest Reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "diss_stats.ref_nn_uhgg_diss.reindex(spgc_strain_list2).dropna().median()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Relationship between Genotype and Gene Content Dissimilarity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    diss_stats.reindex(spgc_strain_list3)  # >= 10 strains\n",
    "    .dropna(subset=[\"min_ref_geno_diss\", \"ref_nn_uhgg_diss\"])\n",
    "    .assign(log_min_ref_geno_diss=lambda x: np.log10(x.min_ref_geno_diss))\n",
    ")\n",
    "\n",
    "diss_corr = []\n",
    "for species in d0.species.unique():\n",
    "    d1 = d0[lambda x: x.species == species]\n",
    "    diss_corr.append(\n",
    "        (\n",
    "            species,\n",
    "            *sp.stats.spearmanr(\n",
    "                d1.min_ref_geno_diss, d1.min_ref_eggnog_diss, alternative=\"greater\"\n",
    "            ),\n",
    "            d1.min_ref_geno_diss.std(),\n",
    "            len(d1),\n",
    "        )\n",
    "    )\n",
    "\n",
    "diss_corr = pd.DataFrame(\n",
    "    diss_corr,\n",
    "    columns=[\"species\", \"spearmanr\", \"pvalue\", \"geno_diss_stdev\", \"num_strains\"],\n",
    ")\n",
    "diss_corr.spearmanr.quantile([0.05, 0.25, 0.5, 0.75, 0.95])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"geno_diss_stdev\",\n",
    "    \"spearmanr\",\n",
    "    data=diss_corr,\n",
    "    c=\"pvalue\",\n",
    "    norm=mpl.colors.LogNorm(),\n",
    "    cmap=\"viridis_r\",\n",
    "    s=\"num_strains\",\n",
    ")\n",
    "plt.colorbar()\n",
    "\n",
    "sp.stats.spearmanr(diss_corr.geno_diss_stdev, diss_corr.spearmanr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(diss_corr.pvalue < 0.05).sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### E. coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    diss_stats.reindex(spgc_strain_list2)\n",
    "    .dropna(subset=[\"min_ref_geno_diss\", \"ref_nn_uhgg_diss\"])\n",
    "    .assign(log_min_ref_geno_diss=lambda x: np.log10(x.min_ref_geno_diss))\n",
    ")\n",
    "d1 = d0[lambda x: x.species == \"102506\"]\n",
    "print(d1.shape[0])\n",
    "plt.scatter(d1[\"log_min_ref_geno_diss\"], d1[\"ref_nn_eggnog_diss\"])\n",
    "sp.stats.spearmanr(d1[\"min_ref_geno_diss\"], d1[\"ref_nn_eggnog_diss\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Figure 3F"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.ticker as mtick\n",
    "\n",
    "d = diss_stats[lambda x: x.passes_filter].assign(\n",
    "    log_min_ref_geno_diss=lambda x: np.log10(x.min_ref_geno_diss)\n",
    ")\n",
    "d_ref = d[lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])]\n",
    "d_spgc = d[lambda x: x.genome_type.isin([\"SPGC\"])]\n",
    "x_bins = np.linspace(\n",
    "    -6,\n",
    "    0,\n",
    "    num=30,\n",
    ")\n",
    "y_bins = np.linspace(0, 1)\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    2,\n",
    "    2,\n",
    "    gridspec_kw=dict(\n",
    "        width_ratios=(3, 0.5), height_ratios=(0.5, 3), hspace=0.15, wspace=0.15\n",
    "    ),\n",
    "    sharex=\"col\",\n",
    "    sharey=\"row\",\n",
    "    figsize=(4, 4),\n",
    ")\n",
    "\n",
    "axs[1, 0].hist2d(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_ref,\n",
    "    bins=(x_bins, y_bins),\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmax=3000),\n",
    "    cmap=\"binary\",\n",
    ")\n",
    "# axs[1, 0].scatter('log_min_ref_geno_diss', 'ref_nn_eggnog_diss', data=d_spgc, s=0.5, alpha=1.0, color=genome_type_palette['SPGC'])\n",
    "# sns.kdeplot(x='log_min_ref_geno_diss', y='ref_nn_eggnog_diss', data=d_ref.sample(n=100), color=genome_type_palette['Ref'], ax=axs[1, 0], linewidths=1, levels=np.linspace(0, 1, num=11))\n",
    "sns.kdeplot(\n",
    "    x=\"log_min_ref_geno_diss\",\n",
    "    y=\"ref_nn_eggnog_diss\",\n",
    "    data=d_spgc,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    ax=axs[1, 0],\n",
    "    linewidths=1,\n",
    "    levels=np.linspace(0, 1, num=11),\n",
    "    alpha=0.7,\n",
    ")\n",
    "focal_species, focal_species_label = \"102506\", \"E. coli\"\n",
    "axs[1, 0].scatter(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_spgc[lambda x: x.species == focal_species],\n",
    "    color=\"darkgreen\",\n",
    "    s=20,\n",
    "    edgecolor=\"white\",\n",
    "    linewidth=0.5,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "axs[1, 0].set_xlabel(\"Genotype Dissimilarity\")\n",
    "axs[1, 0].set_ylabel(\"Gene Content Dissimilarity\")\n",
    "axs[1, 0].set_ylim(0, 0.39)\n",
    "\n",
    "# Legend\n",
    "with mpl.rc_context({\"path.sketch\": (5, 50, 1)}):\n",
    "    axs[1, 0].plot([], [], lw=1, label=\"All Novel\", color=genome_type_palette[\"SPGC\"])\n",
    "axs[1, 0].scatter(\n",
    "    [],\n",
    "    [],\n",
    "    color=\"darkgreen\",\n",
    "    s=20,\n",
    "    edgecolor=\"silver\",\n",
    "    linewidth=0.5,\n",
    "    label=focal_species_label,\n",
    ")\n",
    "axs[1, 0].legend(loc=\"upper left\")\n",
    "\n",
    "axs[1, 1].hist(\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_ref,\n",
    "    orientation=\"horizontal\",\n",
    "    bins=y_bins,\n",
    "    histtype=\"step\",\n",
    "    alpha=0.8,\n",
    "    lw=1,\n",
    "    color=genome_type_palette[\"Ref\"],\n",
    "    density=True,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "axs[1, 1].hist(\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_ref,\n",
    "    orientation=\"horizontal\",\n",
    "    bins=y_bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"Ref\"],\n",
    "    density=True,\n",
    "    label=\"Reference\",\n",
    ")\n",
    "axs[1, 1].hist(\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_spgc,\n",
    "    orientation=\"horizontal\",\n",
    "    bins=y_bins,\n",
    "    histtype=\"step\",\n",
    "    alpha=0.8,\n",
    "    lw=1,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    density=True,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "axs[1, 1].hist(\n",
    "    \"ref_nn_eggnog_diss\",\n",
    "    data=d_spgc,\n",
    "    orientation=\"horizontal\",\n",
    "    bins=y_bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    density=True,\n",
    "    label=\"Novel\",\n",
    ")\n",
    "# axs[1, 1].set_xticks([0])\n",
    "axs[1, 1].set_xlabel(\"Strains (density)\")\n",
    "# axs[1, 1].legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "\n",
    "axs[0, 0].hist(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    data=d_ref,\n",
    "    orientation=\"vertical\",\n",
    "    bins=x_bins,\n",
    "    histtype=\"step\",\n",
    "    alpha=0.8,\n",
    "    lw=1,\n",
    "    color=genome_type_palette[\"Ref\"],\n",
    "    density=True,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "axs[0, 0].hist(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    data=d_ref,\n",
    "    orientation=\"vertical\",\n",
    "    bins=x_bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"Ref\"],\n",
    "    density=True,\n",
    "    label=\"Reference\",\n",
    ")\n",
    "axs[0, 0].hist(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    data=d_spgc,\n",
    "    orientation=\"vertical\",\n",
    "    bins=x_bins,\n",
    "    histtype=\"step\",\n",
    "    alpha=0.8,\n",
    "    lw=1,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    density=True,\n",
    "    label=\"__nolegend__\",\n",
    ")\n",
    "axs[0, 0].hist(\n",
    "    \"log_min_ref_geno_diss\",\n",
    "    data=d_spgc,\n",
    "    orientation=\"vertical\",\n",
    "    bins=x_bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    density=True,\n",
    "    label=\"Novel\",\n",
    ")\n",
    "# axs[0, 0].set_yticks([0])\n",
    "# axs[0, 0].set_ylabel('Strains (density)')\n",
    "# axs[0, 0].legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "# axs[0, 0].set_ylabel('dens.')\n",
    "\n",
    "axs[1, 0].xaxis.set_major_formatter(mtick.FormatStrFormatter(\"$10^{%d}$\"))\n",
    "\n",
    "\n",
    "lib.plot.hide_axes_and_spines(axs[0, 1])\n",
    "\n",
    "# fig.savefig('fig/spgc_genotype_gene_content_joint_distributions.svg', bbox_inches='tight')\n",
    "sp.stats.spearmanr(\n",
    "    d_spgc[lambda x: x.species == focal_species].min_ref_geno_diss,\n",
    "    d_spgc[lambda x: x.species == focal_species].ref_nn_eggnog_diss,\n",
    ")\n",
    "\n",
    "plt.savefig(\"fig/fig3f_closest_reference.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax1 = plt.subplots(figsize=(3, 3), facecolor=\"none\")\n",
    "\n",
    "ax2 = ax1.twinx()\n",
    "\n",
    "# Legend\n",
    "with mpl.rc_context({\"path.sketch\": (5, 50, 1)}):\n",
    "    ax1.plot([], [], lw=2, label=\"All Novel\", color=genome_type_palette[\"SPGC\"])\n",
    "ax1.scatter(\n",
    "    [],\n",
    "    [],\n",
    "    color=\"darkgreen\",\n",
    "    s=150,\n",
    "    edgecolor=\"silver\",\n",
    "    linewidth=2,\n",
    "    label=focal_species_label,\n",
    ")\n",
    "ax1.legend(loc=\"upper left\", facecolor=\"none\")\n",
    "lib.plot.hide_axes_and_spines(ax=ax1)\n",
    "\n",
    "ax2.hist(\n",
    "    [],\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"Ref\"],\n",
    "    label=\"Reference\",\n",
    ")\n",
    "ax2.hist(\n",
    "    [],\n",
    "    histtype=\"stepfilled\",\n",
    "    alpha=0.5,\n",
    "    color=genome_type_palette[\"SPGC\"],\n",
    "    label=\"Novel\",\n",
    ")\n",
    "ax2.legend(loc=\"lower left\", facecolor=\"none\")\n",
    "lib.plot.hide_axes_and_spines(ax=ax2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(4, 2), facecolor=\"none\")\n",
    "x, _, _, _ = plt.hist2d(\n",
    "    [], [], norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=3000), cmap=\"binary\"\n",
    ")\n",
    "ax2 = plt.colorbar(label=\"Reference Strains (count)\", orientation=\"horizontal\")\n",
    "lib.plot.hide_axes_and_spines()\n",
    "lib.plot.rotate_xticklabels(ax2.ax, rotation=30)\n",
    "# plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig = plt.figure(figsize=(2, 4), facecolor=\"none\")\n",
    "x, _, _, _ = plt.hist2d(\n",
    "    [], [], norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=3000), cmap=\"binary\"\n",
    ")\n",
    "ax2 = plt.colorbar(label=\"Reference Strains (count)\", ticks=[0, 500, 2500])\n",
    "lib.plot.hide_axes_and_spines()\n",
    "lib.plot.rotate_xticklabels(ax2.ax, rotation=30)\n",
    "# plt.tight_layout()\n",
    "\n",
    "fig.savefig(\n",
    "    \"fig/spgc_genotype_gene_content_joint_distributions_colorbar.svg\",\n",
    "    bbox_inches=\"tight\",\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz5",
   "language": "python",
   "name": "toolz5"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}