{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.dissimilarity import load_dmat_as_pickle\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type_palette = {\"SPGC\": \"tab:green\", \"MAG\": \"tab:orange\", \"Isolate\": \"tab:blue\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Copy/pasted from https://www.genome.jp/kegg/docs/module_statistics.html\n",
    "_kegg_module_page = \"\"\"\n",
    "KEGG modules statistics\n",
    "As of 2023/10/3\n",
    "\n",
    "M00001\t884\t93.1\t4558\t54.3\t563\t95.7\t141\t95.9\t155\t97.5\t25\t44.6\t4442\t55.6\t116\t28.0\tGlycolysis (Embden-Meyerhof pathway), glucose => pyruvate\n",
    "M00002\t913\t96.1\t7334\t87.3\t569\t96.8\t142\t96.6\t155\t97.5\t47\t83.9\t7027\t88.0\t307\t74.0\tGlycolysis, core module involving three-carbon compounds\n",
    "M00003\t855\t90.0\t4916\t58.5\t556\t94.6\t138\t93.9\t145\t91.2\t16\t28.6\t4749\t59.5\t167\t40.2\tGluconeogenesis, oxaloacetate => fructose-6P\n",
    "M00004\t842\t88.6\t4183\t49.8\t545\t92.7\t141\t95.9\t135\t84.9\t21\t37.5\t4183\t52.4\t0\t0.0\tPentose phosphate pathway (Pentose phosphate cycle)\n",
    "M00005\t938\t98.7\t8142\t97.0\t587\t99.8\t146\t99.3\t152\t95.6\t53\t94.6\t7736\t96.9\t406\t97.8\tPRPP biosynthesis, ribose 5P => PRPP\n",
    "M00006\t908\t95.6\t4735\t56.4\t572\t97.3\t143\t97.3\t152\t95.6\t41\t73.2\t4723\t59.2\t12\t2.9\tPentose phosphate pathway, oxidative phase, glucose 6P => ribulose 5P\n",
    "M00007\t870\t91.6\t6686\t79.6\t559\t95.1\t145\t98.6\t140\t88.1\t26\t46.4\t6642\t83.2\t44\t10.6\tPentose phosphate pathway, non-oxidative phase, fructose 6P => ribose 5P\n",
    "M00008\t0\t0.0\t2188\t26.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2188\t27.4\t0\t0.0\tEntner-Doudoroff pathway, glucose-6P => glyceraldehyde-3P + pyruvate\n",
    "M00009\t674\t70.9\t5431\t64.7\t463\t78.7\t78\t53.1\t129\t81.1\t4\t7.1\t5265\t66.0\t166\t40.0\tCitrate cycle (TCA cycle, Krebs cycle)\n",
    "M00010\t880\t92.6\t6759\t80.5\t547\t93.0\t141\t95.9\t152\t95.6\t40\t71.4\t6514\t81.6\t245\t59.0\tCitrate cycle, first carbon oxidation, oxaloacetate => 2-oxoglutarate\n",
    "M00011\t709\t74.6\t5703\t67.9\t493\t83.8\t82\t55.8\t129\t81.1\t5\t8.9\t5525\t69.2\t178\t42.9\tCitrate cycle, second carbon oxidation, 2-oxoglutarate => oxaloacetate\n",
    "M00012\t292\t30.7\t3064\t36.5\t4\t0.7\t131\t89.1\t150\t94.3\t7\t12.5\t3008\t37.7\t56\t13.5\tGlyoxylate cycle\n",
    "M00013\t158\t16.6\t0\t0.0\t85\t14.5\t1\t0.7\t72\t45.3\t0\t0.0\t0\t0.0\t0\t0.0\tMalonate semialdehyde pathway, propanoyl-CoA => acetyl-CoA\n",
    "M00014\t498\t52.4\t0\t0.0\t498\t84.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlucuronate pathway (uronate pathway)\n",
    "M00015\t884\t93.1\t6293\t74.9\t573\t97.4\t144\t98.0\t147\t92.5\t20\t35.7\t6197\t77.6\t96\t23.1\tProline biosynthesis, glutamate => proline\n",
    "M00016\t0\t0.0\t4129\t49.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4127\t51.7\t2\t0.5\tLysine biosynthesis, succinyl-DAP pathway, aspartate => lysine\n",
    "M00017\t106\t11.2\t2635\t31.4\t0\t0.0\t3\t2.0\t102\t64.2\t1\t1.8\t2635\t33.0\t0\t0.0\tMethionine biosynthesis, aspartate => homoserine => methionine\n",
    "M00018\t282\t29.7\t6370\t75.9\t0\t0.0\t140\t95.2\t138\t86.8\t4\t7.1\t6117\t76.6\t253\t61.0\tThreonine biosynthesis, aspartate => homoserine => threonine\n",
    "M00019\t287\t30.2\t6329\t75.4\t0\t0.0\t141\t95.9\t146\t91.8\t0\t0.0\t6058\t75.9\t271\t65.3\tValine/isoleucine biosynthesis, pyruvate => valine / 2-oxobutanoate => isoleucine\n",
    "M00020\t804\t84.6\t4447\t53.0\t540\t91.8\t144\t98.0\t111\t69.8\t9\t16.1\t4428\t55.5\t19\t4.6\tSerine biosynthesis, glycerate-3P => serine\n",
    "M00021\t163\t17.2\t5851\t69.7\t0\t0.0\t147\t100.0\t1\t0.6\t15\t26.8\t5705\t71.5\t146\t35.2\tCysteine biosynthesis, serine => cysteine\n",
    "M00022\t293\t30.8\t4106\t48.9\t0\t0.0\t136\t92.5\t147\t92.5\t10\t17.9\t4098\t51.3\t8\t1.9\tShikimate pathway, phosphoenolpyruvate + erythrose-4P => chorismate\n",
    "M00023\t273\t28.7\t5306\t63.2\t0\t0.0\t137\t93.2\t136\t85.5\t0\t0.0\t4995\t62.6\t311\t74.9\tTryptophan biosynthesis, chorismate => tryptophan\n",
    "M00024\t100\t10.5\t1914\t22.8\t0\t0.0\t1\t0.7\t99\t62.3\t0\t0.0\t1914\t24.0\t0\t0.0\tPhenylalanine biosynthesis, chorismate => phenylpyruvate => phenylalanine\n",
    "M00025\t133\t14.0\t1348\t16.1\t0\t0.0\t0\t0.0\t133\t83.6\t0\t0.0\t1348\t16.9\t0\t0.0\tTyrosine biosynthesis, chorismate => HPP => tyrosine\n",
    "M00026\t261\t27.5\t3960\t47.2\t2\t0.3\t130\t88.4\t126\t79.2\t3\t5.4\t3886\t48.7\t74\t17.8\tHistidine biosynthesis, PRPP => histidine\n",
    "M00027\t836\t88.0\t836\t10.0\t562\t95.6\t132\t89.8\t142\t89.3\t0\t0.0\t821\t10.3\t15\t3.6\tGABA (gamma-Aminobutyrate) shunt\n",
    "M00028\t244\t25.7\t5434\t64.7\t0\t0.0\t144\t98.0\t97\t61.0\t3\t5.4\t5313\t66.6\t121\t29.2\tOrnithine biosynthesis, glutamate => ornithine\n",
    "M00029\t346\t36.4\t0\t0.0\t341\t58.0\t0\t0.0\t0\t0.0\t5\t8.9\t0\t0.0\t0\t0.0\tUrea cycle\n",
    "M00030\t126\t13.3\t0\t0.0\t0\t0.0\t0\t0.0\t126\t79.2\t0\t0.0\t0\t0.0\t0\t0.0\tLysine biosynthesis, AAA pathway, 2-oxoglutarate => 2-aminoadipate => lysine\n",
    "M00031\t0\t0.0\t153\t1.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t55\t0.7\t98\t23.6\tLysine biosynthesis, mediated by LysW, 2-aminoadipate => lysine\n",
    "M00032\t372\t39.2\t0\t0.0\t371\t63.1\t0\t0.0\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tLysine degradation, lysine => saccharopine => acetoacetyl-CoA\n",
    "M00033\t0\t0.0\t929\t11.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t914\t11.4\t15\t3.6\tEctoine biosynthesis, aspartate => ectoine\n",
    "M00034\t583\t61.4\t743\t8.8\t372\t63.3\t130\t88.4\t74\t46.5\t7\t12.5\t743\t9.3\t0\t0.0\tMethionine salvage pathway\n",
    "M00035\t669\t70.4\t1000\t11.9\t571\t97.1\t2\t1.4\t87\t54.7\t9\t16.1\t977\t12.2\t23\t5.5\tMethionine degradation\n",
    "M00036\t684\t72.0\t869\t10.3\t466\t79.3\t132\t89.8\t78\t49.1\t8\t14.3\t869\t10.9\t0\t0.0\tLeucine degradation, leucine => acetoacetate + acetyl-CoA\n",
    "M00037\t322\t33.9\t0\t0.0\t322\t54.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMelatonin biosynthesis, animals, tryptophan => serotonin => melatonin\n",
    "M00038\t379\t39.9\t57\t0.7\t376\t63.9\t0\t0.0\t3\t1.9\t0\t0.0\t57\t0.7\t0\t0.0\tTryptophan metabolism, tryptophan => kynurenine => 2-aminomuconate\n",
    "M00039\t124\t13.1\t0\t0.0\t0\t0.0\t124\t84.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMonolignol biosynthesis, phenylalanine/tyrosine => monolignol\n",
    "M00040\t145\t15.3\t474\t5.6\t0\t0.0\t143\t97.3\t0\t0.0\t2\t3.6\t474\t5.9\t0\t0.0\tTyrosine biosynthesis, chorismate => arogenate => tyrosine\n",
    "M00042\t312\t32.8\t0\t0.0\t312\t53.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCatecholamine biosynthesis, tyrosine => dopamine => noradrenaline => adrenaline\n",
    "M00043\t359\t37.8\t1\t0.0\t359\t61.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tThyroid hormone biosynthesis, tyrosine => triiodothyronine/thyroxine\n",
    "M00044\t728\t76.6\t493\t5.9\t536\t91.2\t131\t89.1\t50\t31.4\t11\t19.6\t493\t6.2\t0\t0.0\tTyrosine degradation, tyrosine => homogentisate\n",
    "M00045\t416\t43.8\t3010\t35.8\t409\t69.6\t0\t0.0\t0\t0.0\t7\t12.5\t2956\t37.0\t54\t13.0\tHistidine degradation, histidine => N-formiminoglutamate => glutamate\n",
    "M00046\t690\t72.6\t867\t10.3\t549\t93.4\t136\t92.5\t0\t0.0\t5\t8.9\t867\t10.9\t0\t0.0\tPyrimidine degradation, uracil => beta-alanine, thymine => 3-aminoisobutanoate\n",
    "M00047\t353\t37.2\t0\t0.0\t350\t59.5\t0\t0.0\t0\t0.0\t3\t5.4\t0\t0.0\t0\t0.0\tCreatine pathway\n",
    "M00048\t694\t73.1\t6780\t80.7\t456\t77.6\t125\t85.0\t112\t70.4\t1\t1.8\t6562\t82.2\t218\t52.5\tDe novo purine biosynthesis, PRPP + glutamine => IMP\n",
    "M00049\t893\t94.0\t6961\t82.9\t573\t97.4\t134\t91.2\t144\t90.6\t42\t75.0\t6578\t82.4\t383\t92.3\tAdenine ribonucleotide biosynthesis, IMP => ADP,ATP\n",
    "M00050\t888\t93.5\t6569\t78.2\t565\t96.1\t142\t96.6\t139\t87.4\t42\t75.0\t6569\t82.3\t0\t0.0\tGuanine ribonucleotide biosynthesis, IMP => GDP,GTP\n",
    "M00051\t638\t67.2\t1384\t16.5\t518\t88.1\t2\t1.4\t113\t71.1\t5\t8.9\t1103\t13.8\t281\t67.7\tDe novo pyrimidine biosynthesis, glutamine (+ PRPP) => UMP\n",
    "M00052\t915\t96.3\t7088\t84.4\t579\t98.5\t146\t99.3\t144\t90.6\t46\t82.1\t6699\t83.9\t389\t93.7\tPyrimidine ribonucleotide biosynthesis, UMP => UDP/UTP,CDP/CTP\n",
    "M00053\t916\t96.4\t5919\t70.5\t576\t98.0\t147\t100.0\t152\t95.6\t41\t73.2\t5808\t72.8\t111\t26.7\tDeoxyribonucleotide biosynthesis, ADP/GDP/CDP/UDP => dATP/dGTP/dCTP/dUTP\n",
    "M00055\t621\t65.4\t0\t0.0\t412\t70.1\t125\t85.0\t81\t50.9\t3\t5.4\t0\t0.0\t0\t0.0\tN-glycan precursor biosynthesis\n",
    "M00056\t112\t11.8\t0\t0.0\t112\t19.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tO-glycan biosynthesis, mucin type core\n",
    "M00057\t477\t50.2\t0\t0.0\t476\t81.0\t0\t0.0\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tGlycosaminoglycan biosynthesis, linkage tetrasaccharide\n",
    "M00058\t488\t51.4\t0\t0.0\t488\t83.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosaminoglycan biosynthesis, chondroitin sulfate backbone\n",
    "M00059\t562\t59.2\t0\t0.0\t562\t95.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosaminoglycan biosynthesis, heparan sulfate backbone\n",
    "M00060\t0\t0.0\t782\t9.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t782\t9.8\t0\t0.0\tKDO2-lipid A biosynthesis, Raetz pathway, LpxL-LpxM type\n",
    "M00061\t0\t0.0\t1274\t15.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1274\t16.0\t0\t0.0\tD-Glucuronate degradation, D-glucuronate => pyruvate + D-glyceraldehyde 3P\n",
    "M00063\t0\t0.0\t3255\t38.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3255\t40.8\t0\t0.0\tCMP-KDO biosynthesis\n",
    "M00064\t0\t0.0\t1822\t21.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1822\t22.8\t0\t0.0\tADP-L-glycero-D-manno-heptose biosynthesis\n",
    "M00065\t278\t29.3\t0\t0.0\t253\t43.0\t12\t8.2\t13\t8.2\t0\t0.0\t0\t0.0\t0\t0.0\tGPI-anchor biosynthesis, core oligosaccharide\n",
    "M00066\t367\t38.6\t0\t0.0\t367\t62.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tLactosylceramide biosynthesis\n",
    "M00067\t371\t39.1\t0\t0.0\t370\t62.9\t0\t0.0\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tSulfoglycolipids biosynthesis, ceramide/1-alkyl-2-acylglycerol => sulfatide/seminolipid\n",
    "M00068\t238\t25.1\t0\t0.0\t238\t40.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosphingolipid biosynthesis, globo-series, LacCer => Gb4Cer\n",
    "M00069\t352\t37.1\t0\t0.0\t352\t59.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosphingolipid biosynthesis, ganglio series, LacCer => GT3\n",
    "M00070\t376\t39.6\t0\t0.0\t376\t63.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosphingolipid biosynthesis, lacto-series, LacCer => Lc4Cer\n",
    "M00071\t373\t39.3\t0\t0.0\t373\t63.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlycosphingolipid biosynthesis, neolacto-series, LacCer => nLc4Cer\n",
    "M00072\t127\t13.4\t0\t0.0\t0\t0.0\t1\t0.7\t126\t79.2\t0\t0.0\t0\t0.0\t0\t0.0\tN-glycosylation by oligosaccharyltransferase\n",
    "M00073\t752\t79.2\t0\t0.0\t496\t84.4\t134\t91.2\t119\t74.8\t3\t5.4\t0\t0.0\t0\t0.0\tN-glycan precursor trimming\n",
    "M00074\t20\t2.1\t0\t0.0\t0\t0.0\t0\t0.0\t20\t12.6\t0\t0.0\t0\t0.0\t0\t0.0\tN-glycan biosynthesis, high-mannose type\n",
    "M00075\t329\t34.6\t0\t0.0\t329\t56.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tN-glycan biosynthesis, complex type\n",
    "M00076\t385\t40.5\t0\t0.0\t385\t65.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tDermatan sulfate degradation\n",
    "M00077\t375\t39.5\t0\t0.0\t375\t63.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tChondroitin sulfate degradation\n",
    "M00078\t355\t37.4\t0\t0.0\t355\t60.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tHeparan sulfate degradation\n",
    "M00079\t401\t42.2\t0\t0.0\t401\t68.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tKeratan sulfate degradation\n",
    "M00081\t194\t20.4\t0\t0.0\t0\t0.0\t133\t90.5\t61\t38.4\t0\t0.0\t0\t0.0\t0\t0.0\tPectin degradation\n",
    "M00082\t850\t89.5\t5577\t66.4\t542\t92.2\t143\t97.3\t146\t91.8\t19\t33.9\t5577\t69.9\t0\t0.0\tFatty acid biosynthesis, initiation\n",
    "M00083\t876\t92.2\t6598\t78.6\t559\t95.1\t145\t98.6\t149\t93.7\t23\t41.1\t6598\t82.7\t0\t0.0\tFatty acid biosynthesis, elongation\n",
    "M00085\t587\t61.8\t0\t0.0\t575\t97.8\t0\t0.0\t4\t2.5\t8\t14.3\t0\t0.0\t0\t0.0\tFatty acid elongation in mitochondria\n",
    "M00086\t949\t99.9\t6356\t75.7\t588\t100.0\t147\t100.0\t159\t100.0\t55\t98.2\t6140\t76.9\t216\t52.0\tbeta-Oxidation, acyl-CoA synthesis\n",
    "M00087\t781\t82.2\t2600\t31.0\t585\t99.5\t143\t97.3\t33\t20.8\t20\t35.7\t2600\t32.6\t0\t0.0\tbeta-Oxidation\n",
    "M00088\t0\t0.0\t10\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t10\t0.1\t0\t0.0\tKetone body biosynthesis, acetyl-CoA => acetoacetate/3-hydroxybutyrate/acetone\n",
    "M00089\t820\t86.3\t0\t0.0\t577\t98.1\t146\t99.3\t75\t47.2\t22\t39.3\t0\t0.0\t0\t0.0\tTriacylglycerol biosynthesis\n",
    "M00090\t541\t56.9\t0\t0.0\t517\t87.9\t0\t0.0\t24\t15.1\t0\t0.0\t0\t0.0\t0\t0.0\tPhosphatidylcholine (PC) biosynthesis, choline => PC\n",
    "M00091\t549\t57.8\t1140\t13.6\t380\t64.6\t3\t2.0\t149\t93.7\t17\t30.4\t1104\t13.8\t36\t8.7\tPhosphatidylcholine (PC) biosynthesis, PE => PC\n",
    "M00092\t852\t89.7\t0\t0.0\t575\t97.8\t140\t95.2\t94\t59.1\t43\t76.8\t0\t0.0\t0\t0.0\tPhosphatidylethanolamine (PE) biosynthesis, ethanolamine => PE\n",
    "M00093\t143\t15.1\t5356\t63.8\t0\t0.0\t0\t0.0\t143\t89.9\t0\t0.0\t5356\t67.1\t0\t0.0\tPhosphatidylethanolamine (PE) biosynthesis, PA => PS => PE\n",
    "M00094\t709\t74.6\t0\t0.0\t574\t97.6\t3\t2.0\t131\t82.4\t1\t1.8\t0\t0.0\t0\t0.0\tCeramide biosynthesis\n",
    "M00095\t729\t76.7\t100\t1.2\t487\t82.8\t129\t87.8\t105\t66.0\t8\t14.3\t58\t0.7\t42\t10.1\tC5 isoprenoid biosynthesis, mevalonate pathway\n",
    "M00096\t139\t14.6\t2185\t26.0\t0\t0.0\t134\t91.2\t0\t0.0\t5\t8.9\t2185\t27.4\t0\t0.0\tC5 isoprenoid biosynthesis, non-mevalonate pathway\n",
    "M00097\t131\t13.8\t35\t0.4\t0\t0.0\t131\t89.1\t0\t0.0\t0\t0.0\t35\t0.4\t0\t0.0\tbeta-Carotene biosynthesis, GGAP => beta-carotene\n",
    "M00098\t870\t91.6\t356\t4.2\t564\t95.9\t147\t100.0\t142\t89.3\t17\t30.4\t356\t4.5\t0\t0.0\tAcylglycerol degradation\n",
    "M00099\t665\t70.0\t0\t0.0\t574\t97.6\t0\t0.0\t90\t56.6\t1\t1.8\t0\t0.0\t0\t0.0\tSphingosine biosynthesis\n",
    "M00100\t830\t87.4\t0\t0.0\t528\t89.8\t138\t93.9\t144\t90.6\t20\t35.7\t0\t0.0\t0\t0.0\tSphingosine degradation\n",
    "M00101\t267\t28.1\t0\t0.0\t267\t45.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCholesterol biosynthesis, squalene 2,3-epoxide => cholesterol\n",
    "M00102\t107\t11.3\t0\t0.0\t0\t0.0\t0\t0.0\t107\t67.3\t0\t0.0\t0\t0.0\t0\t0.0\tErgocalciferol biosynthesis, squalene 2,3-epoxide => ergosterol/ergocalciferol\n",
    "M00103\t307\t32.3\t0\t0.0\t307\t52.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCholecalciferol biosynthesis\n",
    "M00104\t30\t3.2\t0\t0.0\t30\t5.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tBile acid biosynthesis, cholesterol => cholate/chenodeoxycholate\n",
    "M00106\t141\t14.8\t0\t0.0\t141\t24.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tConjugated bile acid biosynthesis, cholate => taurocholate/glycocholate\n",
    "M00107\t339\t35.7\t0\t0.0\t339\t57.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tSteroid hormone biosynthesis, cholesterol => pregnenolone => progesterone\n",
    "M00108\t28\t2.9\t0\t0.0\t28\t4.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tC21-Steroid hormone biosynthesis, progesterone => corticosterone/aldosterone\n",
    "M00109\t256\t26.9\t0\t0.0\t256\t43.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tC21-Steroid hormone biosynthesis, progesterone => cortisol/cortisone\n",
    "M00110\t351\t36.9\t0\t0.0\t351\t59.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tC19/C18-Steroid hormone biosynthesis, pregnenolone => androstenedione => estrone\n",
    "M00112\t147\t15.5\t67\t0.8\t0\t0.0\t143\t97.3\t0\t0.0\t4\t7.1\t67\t0.8\t0\t0.0\tTocopherol/tocotorienol biosynthesis, homogentisate + phytyl/geranylgeranyl-PP => tocopherol/tocotorienol\n",
    "M00113\t131\t13.8\t0\t0.0\t0\t0.0\t131\t89.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tJasmonic acid biosynthesis\n",
    "M00114\t135\t14.2\t0\t0.0\t0\t0.0\t135\t91.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tAscorbate biosynthesis, plants, fructose-6P => ascorbate\n",
    "M00115\t133\t14.0\t4713\t56.1\t0\t0.0\t133\t90.5\t0\t0.0\t0\t0.0\t4713\t59.0\t0\t0.0\tNAD biosynthesis, aspartate => quinolinate => NAD\n",
    "M00116\t0\t0.0\t1003\t11.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1003\t12.6\t0\t0.0\tMenaquinone biosynthesis, chorismate (+ polyprenyl-PP) => menaquinol\n",
    "M00117\t0\t0.0\t1505\t17.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1505\t18.9\t0\t0.0\tUbiquinone biosynthesis, prokaryotes, chorismate (+ polyprenyl-PP) => ubiquinol\n",
    "M00118\t780\t82.1\t3109\t37.0\t562\t95.6\t143\t97.3\t74\t46.5\t1\t1.8\t3108\t38.9\t1\t0.2\tGlutathione biosynthesis, glutamate => glutathione\n",
    "M00119\t0\t0.0\t2724\t32.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2724\t34.1\t0\t0.0\tPantothenate biosynthesis, valine/L-aspartate => pantothenate\n",
    "M00120\t651\t68.5\t6854\t81.6\t495\t84.2\t101\t68.7\t49\t30.8\t6\t10.7\t6854\t85.9\t0\t0.0\tCoenzyme A biosynthesis, pantothenate => CoA\n",
    "M00121\t140\t14.7\t3068\t36.5\t0\t0.0\t136\t92.5\t0\t0.0\t4\t7.1\t3068\t38.4\t0\t0.0\tHeme biosynthesis, plants and bacteria, glutamate => heme\n",
    "M00122\t0\t0.0\t2323\t27.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2323\t29.1\t0\t0.0\tCobalamin biosynthesis, cobyrinate a,c-diamide => cobalamin\n",
    "M00123\t241\t25.4\t4382\t52.2\t0\t0.0\t143\t97.3\t93\t58.5\t5\t8.9\t4358\t54.6\t24\t5.8\tBiotin biosynthesis, pimeloyl-ACP/CoA => biotin\n",
    "M00124\t0\t0.0\t1093\t13.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1093\t13.7\t0\t0.0\tPyridoxal-P biosynthesis, erythrose-4P => pyridoxal-P\n",
    "M00125\t134\t14.1\t2329\t27.7\t0\t0.0\t134\t91.2\t0\t0.0\t0\t0.0\t2329\t29.2\t0\t0.0\tRiboflavin biosynthesis, plants and bacteria, GTP => riboflavin/FMN/FAD\n",
    "M00126\t176\t18.5\t3272\t39.0\t0\t0.0\t116\t78.9\t60\t37.7\t0\t0.0\t3272\t41.0\t0\t0.0\tTetrahydrofolate biosynthesis, GTP => THF\n",
    "M00127\t0\t0.0\t640\t7.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t635\t8.0\t5\t1.2\tThiamine biosynthesis, prokaryotes, AIR (+ DXP/tyrosine) => TMP/TPP\n",
    "M00128\t668\t70.3\t0\t0.0\t534\t90.8\t0\t0.0\t132\t83.0\t2\t3.6\t0\t0.0\t0\t0.0\tUbiquinone biosynthesis, eukaryotes, 4-hydroxybenzoate + polyprenyl-PP => ubiquinol\n",
    "M00129\t212\t22.3\t0\t0.0\t212\t36.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tAscorbate biosynthesis, animals, glucose-1P => ascorbate\n",
    "M00130\t577\t60.7\t0\t0.0\t574\t97.6\t0\t0.0\t0\t0.0\t3\t5.4\t0\t0.0\t0\t0.0\tInositol phosphate metabolism, PI=> PIP2 => Ins(1,4,5)P3 => Ins(1,3,4,5)P4\n",
    "M00131\t373\t39.3\t0\t0.0\t373\t63.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tInositol phosphate metabolism, Ins(1,3,4,5)P4 => Ins(1,3,4)P3 => myo-inositol\n",
    "M00132\t604\t63.6\t0\t0.0\t467\t79.4\t131\t89.1\t0\t0.0\t6\t10.7\t0\t0.0\t0\t0.0\tInositol phosphate metabolism, Ins(1,3,4)P3 => phytate\n",
    "M00133\t224\t23.6\t1345\t16.0\t224\t38.1\t0\t0.0\t0\t0.0\t0\t0.0\t1256\t15.7\t89\t21.4\tPolyamine biosynthesis, arginine => agmatine => putrescine => spermidine\n",
    "M00134\t825\t86.8\t645\t7.7\t552\t93.9\t103\t70.1\t148\t93.1\t22\t39.3\t644\t8.1\t1\t0.2\tPolyamine biosynthesis, arginine => ornithine => putrescine\n",
    "M00135\t457\t48.1\t422\t5.0\t457\t77.7\t0\t0.0\t0\t0.0\t0\t0.0\t422\t5.3\t0\t0.0\tGABA biosynthesis, eukaryotes, putrescine => GABA\n",
    "M00136\t0\t0.0\t193\t2.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t193\t2.4\t0\t0.0\tGABA biosynthesis, prokaryotes, putrescine => GABA\n",
    "M00137\t132\t13.9\t0\t0.0\t0\t0.0\t132\t89.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tFlavanone biosynthesis, phenylalanine => naringenin\n",
    "M00138\t114\t12.0\t0\t0.0\t0\t0.0\t114\t77.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tFlavonoid biosynthesis, naringenin => pelargonidin\n",
    "M00140\t0\t0.0\t2785\t33.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2732\t34.2\t53\t12.8\tC1-unit interconversion, prokaryotes\n",
    "M00141\t737\t77.6\t37\t0.4\t575\t97.8\t1\t0.7\t151\t95.0\t10\t17.9\t37\t0.5\t0\t0.0\tC1-unit interconversion, eukaryotes\n",
    "M00142\t410\t43.2\t0\t0.0\t337\t57.3\t53\t36.1\t19\t11.9\t1\t1.8\t0\t0.0\t0\t0.0\tNADH:ubiquinone oxidoreductase, mitochondria\n",
    "M00143\t736\t77.5\t0\t0.0\t534\t90.8\t80\t54.4\t117\t73.6\t5\t8.9\t0\t0.0\t0\t0.0\tNADH dehydrogenase (ubiquinone) Fe-S protein/flavoprotein complex, mitochondria\n",
    "M00144\t0\t0.0\t3894\t46.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3894\t48.8\t0\t0.0\tNADH:quinone oxidoreductase, prokaryotes\n",
    "M00145\t98\t10.3\t150\t1.8\t0\t0.0\t98\t66.7\t0\t0.0\t0\t0.0\t150\t1.9\t0\t0.0\tNAD(P)H:quinone oxidoreductase, chloroplasts and cyanobacteria\n",
    "M00146\t383\t40.3\t0\t0.0\t383\t65.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tNADH dehydrogenase (ubiquinone) 1 alpha subcomplex\n",
    "M00147\t348\t36.6\t0\t0.0\t348\t59.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tNADH dehydrogenase (ubiquinone) 1 beta subcomplex\n",
    "M00148\t740\t77.9\t0\t0.0\t511\t86.9\t83\t56.5\t141\t88.7\t5\t8.9\t0\t0.0\t0\t0.0\tSuccinate dehydrogenase (ubiquinone)\n",
    "M00149\t0\t0.0\t4021\t47.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3805\t47.7\t216\t52.0\tSuccinate dehydrogenase, prokaryotes\n",
    "M00150\t0\t0.0\t891\t10.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t891\t11.2\t0\t0.0\tFumarate reductase, prokaryotes\n",
    "M00151\t451\t47.5\t4003\t47.7\t340\t57.8\t76\t51.7\t26\t16.4\t9\t16.1\t4003\t50.1\t0\t0.0\tCytochrome bc1 complex respiratory unit\n",
    "M00152\t374\t39.4\t0\t0.0\t340\t57.8\t0\t0.0\t34\t21.4\t0\t0.0\t0\t0.0\t0\t0.0\tCytochrome bc1 complex\n",
    "M00153\t0\t0.0\t1487\t17.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1487\t18.6\t0\t0.0\tCytochrome bd ubiquinol oxidase\n",
    "M00154\t320\t33.7\t0\t0.0\t304\t51.7\t0\t0.0\t16\t10.1\t0\t0.0\t0\t0.0\t0\t0.0\tCytochrome c oxidase\n",
    "M00155\t0\t0.0\t4345\t51.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4232\t53.0\t113\t27.2\tCytochrome c oxidase, prokaryotes\n",
    "M00156\t0\t0.0\t2013\t24.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2013\t25.2\t0\t0.0\tCytochrome c oxidase, cbb3-type\n",
    "M00157\t112\t11.8\t7438\t88.6\t0\t0.0\t110\t74.8\t0\t0.0\t2\t3.6\t7420\t92.9\t18\t4.3\tF-type ATPase, prokaryotes and chloroplasts\n",
    "M00158\t172\t18.1\t0\t0.0\t156\t26.5\t0\t0.0\t16\t10.1\t0\t0.0\t0\t0.0\t0\t0.0\tF-type ATPase, eukaryotes\n",
    "M00159\t0\t0.0\t728\t8.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t394\t4.9\t334\t80.5\tV/A-type ATPase, prokaryotes\n",
    "M00160\t462\t48.6\t0\t0.0\t462\t78.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tV-type ATPase, eukaryotes\n",
    "M00161\t110\t11.6\t146\t1.7\t0\t0.0\t109\t74.1\t0\t0.0\t1\t1.8\t146\t1.8\t0\t0.0\tPhotosystem II\n",
    "M00162\t0\t0.0\t103\t1.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t103\t1.3\t0\t0.0\tCytochrome b6f complex\n",
    "M00163\t114\t12.0\t147\t1.8\t0\t0.0\t112\t76.2\t0\t0.0\t2\t3.6\t147\t1.8\t0\t0.0\tPhotosystem I\n",
    "M00165\t120\t12.6\t378\t4.5\t0\t0.0\t120\t81.6\t0\t0.0\t0\t0.0\t378\t4.7\t0\t0.0\tReductive pentose phosphate cycle (Calvin cycle)\n",
    "M00168\t155\t16.3\t3630\t43.2\t0\t0.0\t146\t99.3\t0\t0.0\t9\t16.1\t3416\t42.8\t214\t51.6\tCAM (Crassulacean acid metabolism), dark\n",
    "M00169\t147\t15.5\t1175\t14.0\t0\t0.0\t142\t96.6\t0\t0.0\t5\t8.9\t1175\t14.7\t0\t0.0\tCAM (Crassulacean acid metabolism), light\n",
    "M00170\t138\t14.5\t0\t0.0\t0\t0.0\t134\t91.2\t0\t0.0\t4\t7.1\t0\t0.0\t0\t0.0\tC4-dicarboxylic acid cycle, phosphoenolpyruvate carboxykinase type\n",
    "M00171\t133\t14.0\t0\t0.0\t0\t0.0\t133\t90.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tC4-dicarboxylic acid cycle, NAD - malic enzyme type\n",
    "M00172\t114\t12.0\t0\t0.0\t0\t0.0\t114\t77.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tC4-dicarboxylic acid cycle, NADP - malic enzyme type\n",
    "M00173\t0\t0.0\t27\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t27\t0.3\t0\t0.0\tReductive citrate cycle (Arnon-Buchanan cycle)\n",
    "M00174\t0\t0.0\t37\t0.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t37\t0.5\t0\t0.0\tMethane oxidation, methanotroph, methane => formaldehyde\n",
    "M00175\t0\t0.0\t761\t9.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t695\t8.7\t66\t15.9\tNitrogen fixation, nitrogen => ammonia\n",
    "M00176\t276\t29.1\t2785\t33.2\t0\t0.0\t135\t91.8\t135\t84.9\t6\t10.7\t2764\t34.6\t21\t5.1\tAssimilatory sulfate reduction, sulfate => H2S\n",
    "M00307\t888\t93.5\t7850\t93.5\t570\t96.9\t146\t99.3\t135\t84.9\t37\t66.1\t7475\t93.6\t375\t90.4\tPyruvate oxidation, pyruvate => acetyl-CoA\n",
    "M00308\t0\t0.0\t73\t0.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t67\t16.1\tSemi-phosphorylative Entner-Doudoroff pathway, gluconate => glycerate-3P\n",
    "M00309\t0\t0.0\t51\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t51\t12.3\tNon-phosphorylative Entner-Doudoroff pathway, gluconate/galactonate => glycerate\n",
    "M00338\t734\t77.3\t1185\t14.1\t580\t98.6\t2\t1.4\t140\t88.1\t12\t21.4\t1166\t14.6\t19\t4.6\tCysteine biosynthesis, homocysteine + serine => cysteine\n",
    "M00344\t57\t6.0\t0\t0.0\t0\t0.0\t0\t0.0\t57\t35.8\t0\t0.0\t0\t0.0\t0\t0.0\tFormaldehyde assimilation, xylulose monophosphate pathway\n",
    "M00345\t0\t0.0\t552\t6.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t548\t6.9\t4\t1.0\tFormaldehyde assimilation, ribulose monophosphate pathway\n",
    "M00346\t0\t0.0\t51\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t51\t0.6\t0\t0.0\tFormaldehyde assimilation, serine pathway\n",
    "M00356\t0\t0.0\t47\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t47\t11.3\tMethanogenesis, methanol => methane\n",
    "M00357\t0\t0.0\t76\t0.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t76\t18.3\tMethanogenesis, acetate => methane\n",
    "M00358\t0\t0.0\t54\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t53\t12.8\tCoenzyme M biosynthesis\n",
    "M00363\t0\t0.0\t14\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t14\t0.2\t0\t0.0\tEHEC pathogenicity signature, Shiga toxin\n",
    "M00364\t150\t15.8\t3842\t45.7\t0\t0.0\t144\t98.0\t0\t0.0\t6\t10.7\t3434\t43.0\t408\t98.3\tC10-C20 isoprenoid biosynthesis, bacteria\n",
    "M00365\t0\t0.0\t1362\t16.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t954\t12.0\t408\t98.3\tC10-C20 isoprenoid biosynthesis, archaea\n",
    "M00366\t141\t14.8\t0\t0.0\t0\t0.0\t140\t95.2\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tC10-C20 isoprenoid biosynthesis, plants\n",
    "M00367\t690\t72.6\t0\t0.0\t534\t90.8\t0\t0.0\t141\t88.7\t15\t26.8\t0\t0.0\t0\t0.0\tC10-C20 isoprenoid biosynthesis, non-plant eukaryotes\n",
    "M00368\t131\t13.8\t0\t0.0\t0\t0.0\t131\t89.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tEthylene biosynthesis, methionine => ethylene\n",
    "M00369\t10\t1.1\t0\t0.0\t0\t0.0\t10\t6.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCyanogenic glycoside biosynthesis, tyrosine => dhurrin\n",
    "M00370\t9\t0.9\t0\t0.0\t0\t0.0\t9\t6.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGlucosinolate biosynthesis, tryptophan => glucobrassicin\n",
    "M00371\t99\t10.4\t0\t0.0\t0\t0.0\t99\t67.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCastasterone biosynthesis, campesterol => castasterone\n",
    "M00372\t129\t13.6\t0\t0.0\t0\t0.0\t129\t87.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tAbscisic acid biosynthesis, beta-carotene => abscisic acid\n",
    "M00373\t0\t0.0\t129\t1.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t129\t1.6\t0\t0.0\tEthylmalonyl pathway\n",
    "M00374\t0\t0.0\t11\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t11\t2.7\tDicarboxylate-hydroxybutyrate cycle\n",
    "M00375\t0\t0.0\t41\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t41\t9.9\tHydroxypropionate-hydroxybutylate cycle\n",
    "M00376\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\t3-Hydroxypropionate bi-cycle\n",
    "M00377\t0\t0.0\t51\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t51\t0.6\t0\t0.0\tReductive acetyl-CoA pathway (Wood-Ljungdahl pathway)\n",
    "M00378\t0\t0.0\t255\t3.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t255\t61.4\tF420 biosynthesis, archaea\n",
    "M00415\t846\t89.1\t0\t0.0\t570\t96.9\t136\t92.5\t113\t71.1\t27\t48.2\t0\t0.0\t0\t0.0\tFatty acid elongation in endoplasmic reticulum\n",
    "M00416\t0\t0.0\t468\t5.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t468\t5.9\t0\t0.0\tCytochrome aa3-600 menaquinol oxidase\n",
    "M00417\t0\t0.0\t2006\t23.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2006\t25.1\t0\t0.0\tCytochrome o ubiquinol oxidase\n",
    "M00418\t0\t0.0\t6\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t0\t0.0\tToluene degradation, anaerobic, toluene => benzoyl-CoA\n",
    "M00419\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tCymene degradation, p-cymene => p-cumate\n",
    "M00422\t0\t0.0\t96\t1.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t95\t22.9\tAcetyl-CoA pathway, CO2 => acetyl-CoA\n",
    "M00432\t287\t30.2\t6495\t77.3\t0\t0.0\t140\t95.2\t146\t91.8\t1\t1.8\t6219\t77.9\t276\t66.5\tLeucine biosynthesis, 2-oxoisovalerate => 2-oxoisocaproate\n",
    "M00433\t141\t14.8\t41\t0.5\t0\t0.0\t0\t0.0\t141\t88.7\t0\t0.0\t41\t0.5\t0\t0.0\tLysine biosynthesis, 2-oxoglutarate => 2-oxoadipate\n",
    "M00525\t0\t0.0\t459\t5.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t459\t5.7\t0\t0.0\tLysine biosynthesis, acetyl-DAP pathway, aspartate => lysine\n",
    "M00526\t0\t0.0\t539\t6.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t536\t6.7\t3\t0.7\tLysine biosynthesis, DAP dehydrogenase pathway, aspartate => lysine\n",
    "M00527\t139\t14.6\t1315\t15.7\t0\t0.0\t138\t93.9\t0\t0.0\t1\t1.8\t1220\t15.3\t95\t22.9\tLysine biosynthesis, DAP aminotransferase pathway, aspartate => lysine\n",
    "M00528\t0\t0.0\t36\t0.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t36\t0.5\t0\t0.0\tNitrification, ammonia => nitrite\n",
    "M00529\t0\t0.0\t283\t3.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t283\t3.5\t0\t0.0\tDenitrification, nitrate => nitrogen\n",
    "M00530\t0\t0.0\t1354\t16.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1354\t17.0\t0\t0.0\tDissimilatory nitrate reduction, nitrate => ammonia\n",
    "M00531\t224\t23.6\t1453\t17.3\t0\t0.0\t142\t96.6\t77\t48.4\t5\t8.9\t1397\t17.5\t56\t13.5\tAssimilatory nitrate reduction, nitrate => ammonia\n",
    "M00532\t113\t11.9\t0\t0.0\t0\t0.0\t113\t76.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPhotorespiration\n",
    "M00533\t0\t0.0\t532\t6.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t532\t6.7\t0\t0.0\tHomoprotocatechuate degradation, homoprotocatechuate => 2-oxohept-3-enedioate\n",
    "M00534\t0\t0.0\t8\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t8\t0.1\t0\t0.0\tNaphthalene degradation, naphthalene => salicylate\n",
    "M00535\t0\t0.0\t422\t5.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t259\t3.2\t163\t39.3\tIsoleucine biosynthesis, pyruvate => 2-oxobutanoate\n",
    "M00537\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tXylene degradation, xylene => methylbenzoate\n",
    "M00538\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tToluene degradation, toluene => benzoate\n",
    "M00539\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tCumate degradation, p-cumate => 2-oxopent-4-enoate + 2-methylpropanoate\n",
    "M00540\t0\t0.0\t19\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t19\t0.2\t0\t0.0\tBenzoate degradation, cyclohexanecarboxylic acid =>pimeloyl-CoA\n",
    "M00541\t0\t0.0\t27\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t27\t0.3\t0\t0.0\tBenzoyl-CoA degradation, benzoyl-CoA => 3-hydroxypimeloyl-CoA\n",
    "M00542\t0\t0.0\t9\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t9\t0.1\t0\t0.0\tEHEC/EPEC pathogenicity signature, T3SS and effectors\n",
    "M00543\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tBiphenyl degradation, biphenyl => 2-oxopent-4-enoate + benzoate\n",
    "M00544\t0\t0.0\t10\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t10\t0.1\t0\t0.0\tCarbazole degradation, carbazole => 2-oxopent-4-enoate + anthranilate\n",
    "M00545\t0\t0.0\t352\t4.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t352\t4.4\t0\t0.0\tTrans-cinnamate degradation, trans-cinnamate => acetyl-CoA\n",
    "M00546\t172\t18.1\t388\t4.6\t171\t29.1\t0\t0.0\t0\t0.0\t1\t1.8\t388\t4.9\t0\t0.0\tPurine degradation, xanthine => urea\n",
    "M00547\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tBenzene/toluene degradation, benzene => catechol / toluene => 3-methylcatechol\n",
    "M00548\t0\t0.0\t166\t2.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t166\t2.1\t0\t0.0\tBenzene degradation, benzene => catechol\n",
    "M00549\t930\t97.9\t4646\t55.3\t584\t99.3\t145\t98.6\t155\t97.5\t46\t82.1\t4634\t58.0\t12\t2.9\tNucleotide sugar biosynthesis, glucose => UDP-glucose\n",
    "M00550\t0\t0.0\t607\t7.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t607\t7.6\t0\t0.0\tAscorbate degradation, ascorbate => D-xylulose-5P\n",
    "M00551\t0\t0.0\t555\t6.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t555\t7.0\t0\t0.0\tBenzoate degradation, benzoate => catechol / methylbenzoate => methylcatechol\n",
    "M00552\t0\t0.0\t996\t11.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t996\t12.5\t0\t0.0\tD-galactonate degradation, De Ley-Doudoroff pathway, D-galactonate => glycerate-3P\n",
    "M00554\t648\t68.2\t3004\t35.8\t514\t87.4\t2\t1.4\t132\t83.0\t0\t0.0\t2957\t37.0\t47\t11.3\tNucleotide sugar biosynthesis, galactose => UDP-galactose\n",
    "M00555\t579\t60.9\t2138\t25.5\t438\t74.5\t134\t91.2\t1\t0.6\t6\t10.7\t2138\t26.8\t0\t0.0\tBetaine biosynthesis, choline => betaine\n",
    "M00563\t0\t0.0\t34\t0.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t34\t8.2\tMethanogenesis, methylamine/dimethylamine/trimethylamine => methane\n",
    "M00564\t0\t0.0\t45\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t45\t0.6\t0\t0.0\tHelicobacter pylori pathogenicity signature, cagA pathogenicity island\n",
    "M00565\t0\t0.0\t573\t6.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t573\t7.2\t0\t0.0\tTrehalose biosynthesis, D-glucose 1P => trehalose\n",
    "M00567\t0\t0.0\t104\t1.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t104\t25.1\tMethanogenesis, CO2 => methane\n",
    "M00568\t0\t0.0\t673\t8.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t673\t8.4\t0\t0.0\tCatechol ortho-cleavage, catechol => 3-oxoadipate\n",
    "M00569\t0\t0.0\t317\t3.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t317\t4.0\t0\t0.0\tCatechol meta-cleavage, catechol => acetyl-CoA / 4-methylcatechol => propanoyl-CoA\n",
    "M00570\t286\t30.1\t5709\t68.0\t0\t0.0\t141\t95.9\t145\t91.2\t0\t0.0\t5564\t69.7\t145\t34.9\tIsoleucine biosynthesis, threonine => 2-oxobutanoate => isoleucine\n",
    "M00572\t0\t0.0\t1481\t17.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1481\t18.6\t0\t0.0\tPimeloyl-ACP biosynthesis, BioC-BioH pathway, malonyl-ACP => pimeloyl-ACP\n",
    "M00573\t0\t0.0\t62\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t62\t0.8\t0\t0.0\tBiotin biosynthesis, BioI pathway, long-chain-acyl-ACP => pimeloyl-ACP => biotin\n",
    "M00574\t0\t0.0\t11\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t11\t0.1\t0\t0.0\tPertussis pathogenicity signature, pertussis toxin\n",
    "M00575\t0\t0.0\t8\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t8\t0.1\t0\t0.0\tPertussis pathogenicity signature, T1SS\n",
    "M00576\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tETEC pathogenicity signature, heat-labile and heat-stable enterotoxins\n",
    "M00577\t0\t0.0\t304\t3.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t281\t3.5\t23\t5.5\tBiotin biosynthesis, BioW pathway, pimelate => pimeloyl-CoA => biotin\n",
    "M00579\t0\t0.0\t5364\t63.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5341\t66.9\t23\t5.5\tPhosphate acetyltransferase-acetate kinase pathway, acetyl-CoA => acetate\n",
    "M00580\t0\t0.0\t654\t7.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t399\t5.0\t255\t61.4\tPentose phosphate pathway, archaea, fructose 6P => ribose 5P\n",
    "M00595\t0\t0.0\t465\t5.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t465\t5.8\t0\t0.0\tThiosulfate oxidation by SOX complex, thiosulfate => sulfate\n",
    "M00596\t0\t0.0\t154\t1.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t139\t1.7\t15\t3.6\tDissimilatory sulfate reduction, sulfate => H2S\n",
    "M00597\t0\t0.0\t169\t2.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t169\t2.1\t0\t0.0\tAnoxygenic photosystem II\n",
    "M00598\t0\t0.0\t13\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t13\t0.2\t0\t0.0\tAnoxygenic photosystem I\n",
    "M00608\t0\t0.0\t106\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t106\t25.5\t2-Oxocarboxylic acid chain extension, 2-oxoglutarate => 2-oxoadipate => 2-oxopimelate => 2-oxosuberate\n",
    "M00609\t0\t0.0\t267\t3.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t267\t3.3\t0\t0.0\tCysteine biosynthesis, methionine => cysteine\n",
    "M00611\t112\t11.8\t142\t1.7\t0\t0.0\t112\t76.2\t0\t0.0\t0\t0.0\t142\t1.8\t0\t0.0\tOxygenic photosynthesis in plants and cyanobacteria\n",
    "M00612\t0\t0.0\t48\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t48\t0.6\t0\t0.0\tAnoxygenic photosynthesis in purple bacteria\n",
    "M00613\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tAnoxygenic photosynthesis in green nonsulfur bacteria\n",
    "M00614\t0\t0.0\t9\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t9\t0.1\t0\t0.0\tAnoxygenic photosynthesis in green sulfur bacteria\n",
    "M00615\t220\t23.2\t1431\t17.0\t0\t0.0\t142\t96.6\t73\t45.9\t5\t8.9\t1376\t17.2\t55\t13.3\tNitrate assimilation\n",
    "M00616\t1\t0.1\t1732\t20.6\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t1732\t21.7\t0\t0.0\tSulfate-sulfur assimilation\n",
    "M00617\t0\t0.0\t113\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t113\t27.2\tMethanogen\n",
    "M00618\t0\t0.0\t43\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t43\t0.5\t0\t0.0\tAcetogen\n",
    "M00620\t0\t0.0\t61\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t61\t14.7\tIncomplete reductive citrate cycle, acetyl-CoA => oxoglutarate\n",
    "M00621\t857\t90.2\t5700\t67.9\t549\t93.4\t144\t98.0\t139\t87.4\t25\t44.6\t5530\t69.3\t170\t41.0\tGlycine cleavage system\n",
    "M00622\t0\t0.0\t181\t2.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t181\t2.3\t0\t0.0\tNicotinate degradation, nicotinate => fumarate\n",
    "M00623\t0\t0.0\t59\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t59\t0.7\t0\t0.0\tPhthalate degradation, phthalate => protocatechuate\n",
    "M00624\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tTerephthalate degradation, terephthalate => 3,4-dihydroxybenzoate\n",
    "M00625\t0\t0.0\t18\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t18\t0.2\t0\t0.0\tMethicillin resistance\n",
    "M00627\t0\t0.0\t441\t5.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t441\t5.5\t0\t0.0\tbeta-Lactam resistance, Bla system\n",
    "M00630\t41\t4.3\t0\t0.0\t0\t0.0\t0\t0.0\t41\t25.8\t0\t0.0\t0\t0.0\t0\t0.0\tD-Galacturonate degradation (fungi), D-galacturonate => glycerol\n",
    "M00631\t0\t0.0\t895\t10.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t895\t11.2\t0\t0.0\tD-Galacturonate degradation (bacteria), D-galacturonate => pyruvate + D-glyceraldehyde 3P\n",
    "M00632\t623\t65.6\t2660\t31.7\t499\t84.9\t1\t0.7\t123\t77.4\t0\t0.0\t2652\t33.2\t8\t1.9\tGalactose degradation, Leloir pathway, galactose => alpha-D-glucose-1P\n",
    "M00633\t0\t0.0\t42\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t42\t10.1\tSemi-phosphorylative Entner-Doudoroff pathway, gluconate/galactonate => glycerate-3P\n",
    "M00636\t0\t0.0\t9\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t9\t0.1\t0\t0.0\tPhthalate degradation, phthalate => protocatechuate\n",
    "M00637\t0\t0.0\t213\t2.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t213\t2.7\t0\t0.0\tAnthranilate degradation, anthranilate => catechol\n",
    "M00638\t0\t0.0\t62\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t62\t0.8\t0\t0.0\tSalicylate degradation, salicylate => gentisate\n",
    "M00639\t0\t0.0\t117\t1.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t117\t1.5\t0\t0.0\tMultidrug resistance, efflux pump MexCD-OprJ\n",
    "M00641\t0\t0.0\t187\t2.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t187\t2.3\t0\t0.0\tMultidrug resistance, efflux pump MexEF-OprN\n",
    "M00642\t0\t0.0\t415\t4.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t415\t5.2\t0\t0.0\tMultidrug resistance, efflux pump MexJK-OprM\n",
    "M00643\t0\t0.0\t107\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t107\t1.3\t0\t0.0\tMultidrug resistance, efflux pump MexXY-OprM\n",
    "M00649\t0\t0.0\t175\t2.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t175\t2.2\t0\t0.0\tMultidrug resistance, efflux pump AdeABC\n",
    "M00651\t0\t0.0\t8\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t8\t0.1\t0\t0.0\tVancomycin resistance, D-Ala-D-Lac type\n",
    "M00652\t0\t0.0\t33\t0.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t33\t0.4\t0\t0.0\tVancomycin resistance, D-Ala-D-Ser type\n",
    "M00660\t0\t0.0\t27\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t27\t0.3\t0\t0.0\tXanthomonas spp. pathogenicity signature, T3SS and effectors\n",
    "M00661\t7\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t7\t4.4\t0\t0.0\t0\t0.0\t0\t0.0\tPaspaline biosynthesis, geranylgeranyl-PP + indoleglycerol phosphate => paspaline\n",
    "M00664\t0\t0.0\t90\t1.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t90\t1.1\t0\t0.0\tNodulation\n",
    "M00672\t6\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t6\t3.8\t0\t0.0\t0\t0.0\t0\t0.0\tPenicillin biosynthesis, aminoadipate + cycteine + valine => penicillin\n",
    "M00673\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tCephamycin C biosynthesis, aminoadipate + cycteine + valine => cephamycin C\n",
    "M00674\t0\t0.0\t9\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t9\t0.1\t0\t0.0\tClavaminate biosynthesis, arginine + glyceraldehyde-3P => clavaminate\n",
    "M00675\t0\t0.0\t16\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t16\t0.2\t0\t0.0\tCarbapenem-3-carboxylate biosynthesis, pyrroline-5-carboxylate + malonyl-CoA => carbapenem-3-carboxylate\n",
    "M00696\t0\t0.0\t247\t2.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t247\t3.1\t0\t0.0\tMultidrug resistance, efflux pump AcrEF-TolC\n",
    "M00697\t0\t0.0\t81\t1.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t81\t1.0\t0\t0.0\tMultidrug resistance, efflux pump MdtEF-TolC\n",
    "M00698\t0\t0.0\t286\t3.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t286\t3.6\t0\t0.0\tMultidrug resistance, efflux pump BpeEF-OprC\n",
    "M00700\t0\t0.0\t235\t2.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t235\t2.9\t0\t0.0\tMultidrug resistance, efflux pump AbcA\n",
    "M00702\t0\t0.0\t106\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t106\t1.3\t0\t0.0\tMultidrug resistance, efflux pump NorB\n",
    "M00704\t0\t0.0\t67\t0.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t67\t0.8\t0\t0.0\tTetracycline resistance, efflux pump Tet38\n",
    "M00705\t0\t0.0\t200\t2.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t200\t2.5\t0\t0.0\tMultidrug resistance, efflux pump MepA\n",
    "M00714\t0\t0.0\t14\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t14\t0.2\t0\t0.0\tMultidrug resistance, efflux pump QacA\n",
    "M00718\t0\t0.0\t27\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t27\t0.3\t0\t0.0\tMultidrug resistance, efflux pump MexAB-OprM\n",
    "M00725\t0\t0.0\t91\t1.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t91\t1.1\t0\t0.0\tCationic antimicrobial peptide (CAMP) resistance, dltABCD operon\n",
    "M00726\t0\t0.0\t100\t1.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t100\t1.3\t0\t0.0\tCationic antimicrobial peptide (CAMP) resistance, lysyl-phosphatidylglycerol (L-PG) synthase MprF\n",
    "M00730\t0\t0.0\t76\t0.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t76\t1.0\t0\t0.0\tCationic antimicrobial peptide (CAMP) resistance, VraFG transporter\n",
    "M00736\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tNocardicin A biosynthesis, L-pHPG + arginine + serine => nocardicin A\n",
    "M00740\t0\t0.0\t14\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t14\t3.4\tMethylaspartate cycle\n",
    "M00741\t417\t43.9\t1299\t15.5\t403\t68.5\t3\t2.0\t0\t0.0\t11\t19.6\t1159\t14.5\t140\t33.7\tPropanoyl-CoA metabolism, propanoyl-CoA => succinyl-CoA\n",
    "M00744\t0\t0.0\t54\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t54\t0.7\t0\t0.0\tCationic antimicrobial peptide (CAMP) resistance, protease PgtE\n",
    "M00745\t0\t0.0\t203\t2.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t203\t2.5\t0\t0.0\tImipenem resistance, repression of porin OprD\n",
    "M00746\t0\t0.0\t43\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t43\t0.5\t0\t0.0\tMultidrug resistance, repression of porin OmpF\n",
    "M00761\t0\t0.0\t563\t6.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t563\t7.1\t0\t0.0\tUndecaprenylphosphate alpha-L-Ara4N biosynthesis, UDP-GlcA => undecaprenyl phosphate alpha-L-Ara4N\n",
    "M00763\t0\t0.0\t160\t1.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t160\t38.6\tOrnithine biosynthesis, mediated by LysW, glutamate => ornithine\n",
    "M00769\t0\t0.0\t20\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t20\t0.3\t0\t0.0\tMultidrug resistance, efflux pump MexPQ-OpmE\n",
    "M00773\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tTylosin biosynthesis, methylmalonyl-CoA + malonyl-CoA => tylactone => tylosin\n",
    "M00774\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tErythromycin biosynthesis, propanoyl-CoA + methylmalonyl-CoA => deoxyerythronolide B => erythromycin A/B\n",
    "M00775\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tOleandomycin biosynthesis, malonyl-CoA + methylmalonyl-CoA => 8,8a-deoxyoleandolide => oleandomycin\n",
    "M00776\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPikromycin/methymycin biosynthesis, methylmalonyl-CoA + malonyl-CoA => narbonolide/10-deoxymethynolide => pikromycin/methymycin\n",
    "M00777\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tAvermectin biosynthesis, 2-methylbutanoyl-CoA/isobutyryl-CoA => 6,8a-Seco-6,8a-deoxy-5-oxoavermectin 1a/1b aglycone => avermectin A1a/B1a/A1b/B1b\n",
    "M00778\t0\t0.0\t116\t1.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t116\t1.5\t0\t0.0\tType II polyketide backbone biosynthesis, acyl-CoA + malonyl-CoA => polyketide\n",
    "M00779\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tDihydrokalafungin biosynthesis, octaketide => dihydrokalafungin\n",
    "M00780\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tTetracycline/oxytetracycline biosynthesis, pretetramide => tetracycline/oxytetracycline\n",
    "M00781\t0\t0.0\t8\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t8\t0.1\t0\t0.0\tNogalavinone/aklavinone biosynthesis, deoxynogalonate/deoxyaklanonate => nogalavinone/aklavinone\n",
    "M00782\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMithramycin biosynthesis, 4-demethylpremithramycinone => mithramycin\n",
    "M00783\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tTetracenomycin C/8-demethyltetracenomycin C biosynthesis, tetracenomycin F2 => tetracenomycin C/8-demethyltetracenomycin C\n",
    "M00784\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tElloramycin biosynthesis, 8-demethyltetracenomycin C => elloramycin A\n",
    "M00785\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCycloserine biosynthesis, arginine/serine => cycloserine\n",
    "M00786\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tFumitremorgin alkaloid biosynthesis, tryptophan + proline => fumitremorgin C/A\n",
    "M00787\t0\t0.0\t62\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t62\t0.8\t0\t0.0\tBacilysin biosynthesis, prephenate => bacilysin\n",
    "M00788\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tTerpentecin biosynthesis, GGAP => terpentecin\n",
    "M00789\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tRebeccamycin biosynthesis, tryptophan => rebeccamycin\n",
    "M00790\t0\t0.0\t47\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t47\t0.6\t0\t0.0\tPyrrolnitrin biosynthesis, tryptophan => pyrrolnitrin\n",
    "M00793\t0\t0.0\t4939\t58.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4757\t59.6\t182\t43.9\tdTDP-L-rhamnose biosynthesis\n",
    "M00794\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tdTDP-6-deoxy-D-allose biosynthesis\n",
    "M00795\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tdTDP-beta-L-noviose biosynthesis\n",
    "M00796\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tdTDP-D-mycaminose biosynthesis\n",
    "M00797\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tdTDP-D-desosamine biosynthesis\n",
    "M00798\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tdTDP-L-mycarose biosynthesis\n",
    "M00799\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tdTDP-L-oleandrose biosynthesis\n",
    "M00800\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tdTDP-L-megosamine biosynthesis\n",
    "M00801\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tdTDP-L-olivose biosynthesis\n",
    "M00802\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tdTDP-D-forosamine biosynthesis\n",
    "M00803\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tdTDP-D-angolosamine biosynthesis\n",
    "M00804\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tComplete nitrification, comammox, ammonia => nitrite => nitrate\n",
    "M00805\t0\t0.0\t5\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5\t0.1\t0\t0.0\tStaurosporine biosynthesis, tryptophan => staurosporine\n",
    "M00808\t0\t0.0\t14\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t14\t0.2\t0\t0.0\tViolacein biosynthesis, tryptophan => violacein\n",
    "M00810\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tNicotine degradation, pyridine pathway, nicotine => 2,6-dihydroxypyridine/succinate semialdehyde\n",
    "M00811\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tNicotine degradation, pyrrolidine pathway, nicotine => succinate semialdehyde\n",
    "M00814\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tAcarbose biosynthesis, sedoheptulopyranose-7P => acarbose\n",
    "M00815\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tValidamycin A biosynthesis, sedoheptulopyranose-7P => validamycin A\n",
    "M00819\t0\t0.0\t6\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t0\t0.0\tPentalenolactone biosynthesis, farnesyl-PP => pentalenolactone\n",
    "M00823\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tChlortetracycline biosynthesis, pretetramide => chlortetracycline\n",
    "M00824\t0\t0.0\t39\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t39\t0.5\t0\t0.0\t9-membered enediyne core biosynthesis, malonyl-CoA => 3-hydroxyhexadeca-4,6,8,10,12,14-hexaenoyl-ACP => 9-membered enediyne core\n",
    "M00825\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t10-membered enediyne core biosynthesis, malonyl-CoA => 3-hydroxyhexadeca-4,6,8,10,12,14-hexaenoyl-ACP => 10-membered enediyne core\n",
    "M00826\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tC-1027 benzoxazolinate moiety biosynthesis, chorismate => benzoxazolinyl-CoA\n",
    "M00827\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tC-1027 beta-amino acid moiety biosynthesis, tyrosine => 3-chloro-4,5-dihydroxy-beta-phenylalanyl-PCP\n",
    "M00828\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tMaduropeptin beta-hydroxy acid moiety biosynthesis, tyrosine => 3-(4-hydroxyphenyl)-3-oxopropanoyl-PCP\n",
    "M00829\t0\t0.0\t6\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t0\t0.0\t3,6-Dimethylsalicylyl-CoA biosynthesis, malonyl-CoA => 6-methylsalicylate => 3,6-dimethylsalicylyl-CoA\n",
    "M00830\t0\t0.0\t3\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t3\t0.0\t0\t0.0\tNeocarzinostatin naphthoate moiety biosynthesis, malonyl-CoA => 2-hydroxy-5-methyl-1-naphthoate => 2-hydroxy-7-methoxy-5-methyl-1-naphthoyl-CoA\n",
    "M00831\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tKedarcidin 2-hydroxynaphthoate moiety biosynthesis, malonyl-CoA => 3,6,8-trihydroxy-2-naphthoate => 3-hydroxy-7,8-dimethoxy-6-isopropoxy-2-naphthoyl-CoA\n",
    "M00832\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tKedarcidin 2-aza-3-chloro-beta-tyrosine moiety biosynthesis, azatyrosine => 2-aza-3-chloro-beta-tyrosyl-PCP\n",
    "M00833\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tCalicheamicin biosynthesis, calicheamicinone => calicheamicin\n",
    "M00834\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tCalicheamicin orsellinate moiety biosynthesis, malonyl-CoA => orsellinate-ACP => 5-iodo-2,3-dimethoxyorsellinate-ACP\n",
    "M00835\t0\t0.0\t23\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t23\t0.3\t0\t0.0\tPyocyanine biosynthesis, chorismate => pyocyanine\n",
    "M00836\t0\t0.0\t40\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t40\t9.6\tCoenzyme F430 biosynthesis, sirohydrochlorin => coenzyme F430\n",
    "M00837\t0\t0.0\t15\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t15\t0.2\t0\t0.0\tProdigiosin biosynthesis, L-proline => prodigiosin\n",
    "M00838\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tUndecylprodigiosin biosynthesis, L-proline => undecylprodigiosin\n",
    "M00840\t0\t0.0\t82\t1.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t82\t1.0\t0\t0.0\tTetrahydrofolate biosynthesis, mediated by ribA and trpF, GTP => THF\n",
    "M00841\t13\t1.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t13\t23.2\t0\t0.0\t0\t0.0\tTetrahydrofolate biosynthesis, mediated by PTPS, GTP => THF\n",
    "M00842\t543\t57.2\t0\t0.0\t539\t91.7\t0\t0.0\t0\t0.0\t4\t7.1\t0\t0.0\t0\t0.0\tTetrahydrobiopterin biosynthesis, GTP => BH4\n",
    "M00843\t0\t0.0\t16\t0.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t16\t0.2\t0\t0.0\tL-threo-Tetrahydrobiopterin biosynthesis, GTP => L-threo-BH4\n",
    "M00844\t668\t70.3\t6233\t74.2\t377\t64.1\t144\t98.0\t141\t88.7\t6\t10.7\t5890\t73.8\t343\t82.7\tArginine biosynthesis, ornithine => arginine\n",
    "M00845\t0\t0.0\t130\t1.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t130\t1.6\t0\t0.0\tArginine biosynthesis, glutamate => acetylcitrulline => arginine\n",
    "M00846\t0\t0.0\t2419\t28.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2417\t30.3\t2\t0.5\tSiroheme biosynthesis, glutamyl-tRNA => siroheme\n",
    "M00847\t0\t0.0\t265\t3.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t158\t2.0\t107\t25.8\tHeme biosynthesis, archaea, siroheme => heme\n",
    "M00848\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tAurachin biosynthesis, anthranilate => aurachin A\n",
    "M00849\t0\t0.0\t341\t4.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t341\t82.2\tC5 isoprenoid biosynthesis, mevalonate pathway, archaea\n",
    "M00850\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tVibrio cholerae pathogenicity signature, cholera toxins\n",
    "M00851\t0\t0.0\t135\t1.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t135\t1.7\t0\t0.0\tCarbapenem resistance\n",
    "M00852\t0\t0.0\t10\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t10\t0.1\t0\t0.0\tVibrio cholerae pathogenicity signature, toxin coregulated pilus\n",
    "M00853\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tETEC pathogenicity signature, colonization factors\n",
    "M00854\t756\t79.6\t3581\t42.6\t478\t81.3\t144\t98.0\t132\t83.0\t2\t3.6\t3564\t44.6\t17\t4.1\tGlycogen biosynthesis, glucose-1P => glycogen/starch\n",
    "M00855\t722\t76.0\t1329\t15.8\t573\t97.4\t0\t0.0\t137\t86.2\t12\t21.4\t1328\t16.6\t1\t0.2\tGlycogen degradation, glycogen => glucose-6P\n",
    "M00856\t0\t0.0\t6\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t0\t0.0\tSalmonella enterica pathogenicity signature, typhoid toxin\n",
    "M00857\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tSalmonella enterica pathogenicity signature, Vi antigen\n",
    "M00859\t0\t0.0\t12\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t12\t0.2\t0\t0.0\tBacillus anthracis pathogenicity signature, anthrax toxin\n",
    "M00860\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tBacillus anthracis pathogenicity signature, polyglutamic acid capsule biosynthesis\n",
    "M00861\t538\t56.6\t0\t0.0\t535\t91.0\t0\t0.0\t0\t0.0\t3\t5.4\t0\t0.0\t0\t0.0\tbeta-Oxidation, peroxisome, VLCFA\n",
    "M00862\t239\t25.2\t0\t0.0\t239\t40.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tbeta-Oxidation, peroxisome, tri/dihydroxycholestanoyl-CoA => choloyl/chenodeoxycholoyl-CoA\n",
    "M00866\t0\t0.0\t68\t0.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t68\t0.9\t0\t0.0\tKDO2-lipid A biosynthesis, Raetz pathway, non-LpxL-LpxM type\n",
    "M00867\t0\t0.0\t59\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t59\t0.7\t0\t0.0\tKDO2-lipid A modification pathway\n",
    "M00868\t551\t58.0\t13\t0.2\t432\t73.5\t0\t0.0\t119\t74.8\t0\t0.0\t13\t0.2\t0\t0.0\tHeme biosynthesis, animals and fungi, glycine => heme\n",
    "M00872\t300\t31.6\t0\t0.0\t299\t50.9\t0\t0.0\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tO-glycan biosynthesis, mannose type (core M3)\n",
    "M00873\t216\t22.7\t0\t0.0\t216\t36.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tFatty acid biosynthesis in mitochondria, animals\n",
    "M00874\t38\t4.0\t0\t0.0\t0\t0.0\t0\t0.0\t38\t23.9\t0\t0.0\t0\t0.0\t0\t0.0\tFatty acid biosynthesis in mitochondria, fungi\n",
    "M00875\t0\t0.0\t72\t0.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t72\t0.9\t0\t0.0\tStaphyloferrin B biosynthesis, L-serine => staphyloferrin B\n",
    "M00876\t0\t0.0\t107\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t107\t1.3\t0\t0.0\tStaphyloferrin A biosynthesis, L-ornithine => staphyloferrin A\n",
    "M00877\t0\t0.0\t70\t0.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t70\t0.9\t0\t0.0\tKanosamine biosynthesis, glucose 6-phosphate => kanosamine\n",
    "M00878\t0\t0.0\t179\t2.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t179\t2.2\t0\t0.0\tPhenylacetate degradation, phenylaxetate => acetyl-CoA/succinyl-CoA\n",
    "M00879\t0\t0.0\t861\t10.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t861\t10.8\t0\t0.0\tArginine succinyltransferase pathway, arginine => glutamate\n",
    "M00880\t637\t67.1\t3992\t47.5\t498\t84.7\t127\t86.4\t12\t7.5\t0\t0.0\t3645\t45.7\t347\t83.6\tMolybdenum cofactor biosynthesis, GTP => molybdenum cofactor\n",
    "M00881\t157\t16.5\t5456\t65.0\t0\t0.0\t141\t95.9\t0\t0.0\t16\t28.6\t5445\t68.2\t11\t2.7\tLipoic acid biosynthesis, plants and bacteria, octanoyl-ACP => dihydrolipoyl-E2/H\n",
    "M00882\t773\t81.4\t0\t0.0\t490\t83.3\t139\t94.6\t142\t89.3\t2\t3.6\t0\t0.0\t0\t0.0\tLipoic acid biosynthesis, eukaryotes, octanoyl-ACP => dihydrolipoyl-H\n",
    "M00883\t472\t49.7\t463\t5.5\t472\t80.3\t0\t0.0\t0\t0.0\t0\t0.0\t463\t5.8\t0\t0.0\tLipoic acid biosynthesis, animals and bacteria, octanoyl-ACP => dihydrolipoyl-H => dihydrolipoyl-E2\n",
    "M00884\t40\t4.2\t0\t0.0\t0\t0.0\t0\t0.0\t40\t25.2\t0\t0.0\t0\t0.0\t0\t0.0\tLipoic acid biosynthesis, octanoyl-CoA => dihydrolipoyl-E2\n",
    "M00889\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPuromycin biosynthesis, ATP => puromycin\n",
    "M00890\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tRoseoflavin biosynthesis, FMN => roseoflavin\n",
    "M00891\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tDitryptophenaline biosynthesis, tryptophan + phenylalanine => ditryptophenaline\n",
    "M00892\t870\t91.6\t0\t0.0\t572\t97.3\t139\t94.6\t144\t90.6\t15\t26.8\t0\t0.0\t0\t0.0\tUDP-N-acetyl-D-glucosamine biosynthesis, eukaryotes, glucose => UDP-GlcNAc\n",
    "M00893\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tLovastatin biosynthesis, malonyl-CoA => lovastatin acid\n",
    "M00894\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCannabidiol biosynthesis, malonyl-CoA => cannabidiol/dronabinol\n",
    "M00895\t0\t0.0\t77\t0.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t77\t1.0\t0\t0.0\tThiamine biosynthesis, prokaryotes, AIR (+ DXP/glycine) => TMP/TPP\n",
    "M00896\t0\t0.0\t337\t4.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t58\t0.7\t279\t67.2\tThiamine biosynthesis, archaea, AIR (+ NAD+) => TMP/TPP\n",
    "M00897\t112\t11.8\t0\t0.0\t0\t0.0\t112\t76.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tThiamine biosynthesis, plants, AIR (+ NAD+) => TMP/thiamine/TPP\n",
    "M00898\t73\t7.7\t0\t0.0\t0\t0.0\t0\t0.0\t73\t45.9\t0\t0.0\t0\t0.0\t0\t0.0\tThiamine biosynthesis, pyridoxal-5P => TMP/thiamine/TPP\n",
    "M00899\t136\t14.3\t2730\t32.5\t0\t0.0\t135\t91.8\t0\t0.0\t1\t1.8\t2588\t32.4\t142\t34.2\tThiamine salvage pathway, HMP/HET => TMP\n",
    "M00900\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCrocin biosynthesis, crocetin => crocin\n",
    "M00901\t3\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t3\t1.9\t0\t0.0\t0\t0.0\t0\t0.0\tFumiquinazoline biosynthesis, tryptophan + alanine + anthranilate => fumiquinazoline\n",
    "M00902\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPodophyllotoxin biosynthesis, coniferyl alcohol => podophyllotoxin\n",
    "M00903\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tFosfomycin biosynthesis, phosphoenolpyruvate => fosfomycin\n",
    "M00904\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tDapdiamides biosynthesis, L-2,3-diaminopropanoate => dapdiamide A/B/C\n",
    "M00905\t0\t0.0\t4\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t4\t0.1\t0\t0.0\tGrixazone biosynthesis, aspartate 4-semialdehyde => grixazone B\n",
    "M00906\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tEthynylserine biosynthesis, lysine => ethynylserine\n",
    "M00909\t0\t0.0\t5170\t61.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t5170\t64.8\t0\t0.0\tUDP-N-acetyl-D-glucosamine biosynthesis, prokaryotes, glucose => UDP-GlcNAc\n",
    "M00910\t143\t15.1\t0\t0.0\t0\t0.0\t142\t96.6\t0\t0.0\t1\t1.8\t0\t0.0\t0\t0.0\tPhenylalanine biosynthesis, chorismate => arogenate => phenylalanine\n",
    "M00911\t37\t3.9\t0\t0.0\t0\t0.0\t0\t0.0\t37\t23.3\t0\t0.0\t0\t0.0\t0\t0.0\tRiboflavin biosynthesis, fungi, GTP => riboflavin/FMN/FAD\n",
    "M00912\t353\t37.2\t66\t0.8\t293\t49.8\t0\t0.0\t59\t37.1\t1\t1.8\t66\t0.8\t0\t0.0\tNAD biosynthesis, tryptophan => quinolinate => NAD\n",
    "M00913\t43\t4.5\t0\t0.0\t0\t0.0\t0\t0.0\t43\t27.0\t0\t0.0\t0\t0.0\t0\t0.0\tPantothenate biosynthesis, 2-oxoisovalerate/spermine => pantothenate\n",
    "M00914\t0\t0.0\t149\t1.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t149\t35.9\tCoenzyme A biosynthesis, archaea, 2-oxoisovalerate => 4-phosphopantoate => CoA\n",
    "M00915\t0\t0.0\t1\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t0\t0.0\tCaffeine degradation, caffeine => xanthine\n",
    "M00916\t333\t35.1\t2645\t31.5\t30\t5.1\t144\t98.0\t136\t85.5\t23\t41.1\t2278\t28.5\t367\t88.4\tPyridoxal-P biosynthesis, R5P + glyceraldehyde-3P + glutamine => pyridoxal-P\n",
    "M00917\t126\t13.3\t0\t0.0\t0\t0.0\t126\t85.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPhytosterol biosynthesis, squalene 2,3-epoxide => campesterol/sitosterol\n",
    "M00918\t0\t0.0\t190\t2.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t190\t2.4\t0\t0.0\tAerobactin biosynthesis, lysine => aerobactin\n",
    "M00919\t0\t0.0\t304\t3.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t304\t3.8\t0\t0.0\tEctoine degradation, ectoine => aspartate\n",
    "M00921\t0\t0.0\t9\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t9\t0.1\t0\t0.0\tCyclooctatin biosynthesis, dimethylallyl-PP + isopentenyl-PP => cyclooctatin\n",
    "M00922\t0\t0.0\t87\t1.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t87\t1.1\t0\t0.0\tCMP-Neu5Ac biosynthesis\n",
    "M00923\t0\t0.0\t7\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t7\t0.1\t0\t0.0\tUDP-L-FucNAm biosynthesis\n",
    "M00924\t0\t0.0\t634\t7.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t581\t7.3\t53\t12.8\tCobalamin biosynthesis, anaerobic, uroporphyrinogen III => sirohydrochlorin => cobyrinate a,c-diamide\n",
    "M00925\t0\t0.0\t258\t3.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t258\t3.2\t0\t0.0\tCobalamin biosynthesis, aerobic, uroporphyrinogen III => precorrin 2 => cobyrinate a,c-diamide\n",
    "M00926\t0\t0.0\t1383\t16.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1383\t17.3\t0\t0.0\tHeme biosynthesis, bacteria, glutamyl-tRNA => coproporphyrin III => heme\n",
    "M00927\t127\t13.4\t0\t0.0\t0\t0.0\t127\t86.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGibberellin A12 biosynthesis, GAPP => GA12\n",
    "M00928\t132\t13.9\t0\t0.0\t0\t0.0\t132\t89.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGibberellin A4/A1 biosynthesis, GA12/GA53 => GA4/GA1\n",
    "M00929\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tGibberellin A1 biosynthesis, GGPP => GA1\n",
    "M00930\t0\t0.0\t421\t5.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t421\t5.3\t0\t0.0\tMenaquinone biosynthesis, futalosine pathway\n",
    "M00931\t0\t0.0\t221\t2.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t221\t2.8\t0\t0.0\tMenaquinone biosynthesis, modified futalosine pathway\n",
    "M00932\t126\t13.3\t54\t0.6\t0\t0.0\t126\t85.7\t0\t0.0\t0\t0.0\t54\t0.7\t0\t0.0\tPhylloquinone biosynthesis, chorismate (+ phytyl-PP) => phylloquinol\n",
    "M00933\t140\t14.7\t0\t0.0\t0\t0.0\t138\t93.9\t0\t0.0\t2\t3.6\t0\t0.0\t0\t0.0\tPlastoquinone biosynthesis, homogentisate + solanesyl-PP => plastoquinol\n",
    "M00934\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMycinamicin biosynthesis, malonyl-CoA + methylmalonyl-CoA => protomycinolide IV => mycinamicin II\n",
    "M00935\t0\t0.0\t40\t0.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t40\t9.6\tMethanofuran biosynthesis\n",
    "M00936\t62\t6.5\t0\t0.0\t0\t0.0\t62\t42.2\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMelatonin biosynthesis, plants, tryptophan => serotonin => melatonin\n",
    "M00937\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tAflatoxin biosynthesis, malonyl-CoA => aflatoxin B1\n",
    "M00938\t811\t85.4\t4336\t51.6\t505\t85.9\t145\t98.6\t123\t77.4\t38\t67.9\t4265\t53.4\t71\t17.1\tPyrimidine deoxyribonucleotide biosynthesis, UDP => dTTP\n",
    "M00939\t0\t0.0\t307\t3.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t307\t3.8\t0\t0.0\tPyrimidine degradation, uracil => 3-hydroxypropanoate\n",
    "M00940\t39\t4.1\t0\t0.0\t0\t0.0\t39\t26.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tFlavanone biosynthesis, p-coumaroyl-CoA => liquiritigenin\n",
    "M00941\t17\t1.8\t0\t0.0\t0\t0.0\t17\t11.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tIsoflavone biosynthesis, liquiritigenin/naringenin => daidzein/genistein\n",
    "M00942\t16\t1.7\t0\t0.0\t0\t0.0\t16\t10.9\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tPterocarpan biosynthesis, daidzein => medicarpin\n",
    "M00943\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tReticuline biosynthesis, dopamine + 4HPAA => (S)-reticuline\n",
    "M00944\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMorphine biosynthesis, (S)-reticuline => morphine\n",
    "M00945\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tSanguinarine biosynthesis, (S)-reticuline => sanguinarine\n",
    "M00946\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tNoscapine biosynthesis, (S)-reticuline => noscapine\n",
    "M00947\t0\t0.0\t108\t1.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t108\t1.4\t0\t0.0\tD-Arginine racemization, D-arginine => L-arginine\n",
    "M00948\t0\t0.0\t619\t7.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t619\t7.8\t0\t0.0\tHydroxyproline degradation, trans-4-hydroxy-L-proline => 2-oxoglutarate\n",
    "M00949\t0\t0.0\t116\t1.4\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t116\t1.5\t0\t0.0\tStaphylopine biosynthesis, L-histidine => staphylopine\n",
    "M00950\t0\t0.0\t131\t1.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t92\t1.2\t39\t9.4\tBiotin biosynthesis, BioU pathway, pimeloyl-ACP/CoA => biotin\n",
    "M00951\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tCremeomycin biosynthesis, aspartate/3,4-AHBA => cremeomycin\n",
    "M00952\t1\t0.1\t0\t0.0\t0\t0.0\t1\t0.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tBenzoxazinoid biosynthesis, indoleglycerol phosphate => DIMBOA-glucoside\n",
    "M00953\t17\t1.8\t0\t0.0\t0\t0.0\t17\t11.6\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tMugineic acid biosynthesis, methionine => 3-epihydroxymugineic acid\n",
    "M00956\t0\t0.0\t180\t2.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t180\t2.3\t0\t0.0\tLysine degradation, bacteria, L-lysine => succinate\n",
    "M00957\t0\t0.0\t151\t1.8\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t151\t1.9\t0\t0.0\tLysine degradation, bacteria, L-lysine => glutarate => succinate/acetyl-CoA\n",
    "M00958\t593\t62.4\t2109\t25.1\t548\t93.2\t0\t0.0\t36\t22.6\t9\t16.1\t2109\t26.4\t0\t0.0\tAdenine ribonucleotide degradation, AMP => Urate\n",
    "M00959\t544\t57.3\t1705\t20.3\t526\t89.5\t0\t0.0\t14\t8.8\t4\t7.1\t1705\t21.4\t0\t0.0\tGuanine ribonucleotide degradation, GMP => Urate\n",
    "M00960\t0\t0.0\t23\t0.3\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t23\t0.3\t0\t0.0\tLysine degradation, bacteria, L-lysine => D-lysine => succinate\n",
    "M00961\t4\t0.4\t0\t0.0\t0\t0.0\t4\t2.7\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tBetacyanin biosynthesis, L-tyrosine => amaranthin\n",
    "M00962\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tPsilocybin biosynthesis, tryptophan => psilocybin\n",
    "M00963\t6\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\t6\t3.8\t0\t0.0\t0\t0.0\t0\t0.0\tChanoclavine aldehyde biosynthesis, tryptophan => chanoclavine-I aldehyde\n",
    "M00964\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tFumigaclavine biosynthesis, chanoclavine-I aldehyde => fumigaclavine C\n",
    "M00965\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tVindoline biosynthesis, tabersonine => vindoline\n",
    "M00966\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tEquol biosynthesis, daidzein => equol\n",
    "M00967\t0\t0.0\t6\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t6\t0.1\t0\t0.0\tFlavone degradation, luteolin/apigenin => DHCA/phloretate\n",
    "M00968\t0\t0.0\t210\t2.5\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.0\t209\t50.4\tPentose bisphosphate pathway (nucleoside degradation), archaea, nucleoside/NMP => 3-PGA/glycerone phosphate\n",
    "M00969\t1\t0.1\t0\t0.0\t0\t0.0\t0\t0.0\t1\t0.6\t0\t0.0\t0\t0.0\t0\t0.0\tFumagillin biosynthesis, farnesyl-PP => fumagillin\n",
    "M00970\t824\t86.7\t5301\t63.1\t552\t93.9\t142\t96.6\t102\t64.2\t28\t50.0\t5282\t66.2\t19\t4.6\tProline degradation, proline => glutamate\n",
    "M00971\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\tQS-7 biosynthesis, 2,3-epoxysqualene => QS-7\n",
    "M00972\t821\t86.4\t1354\t16.1\t567\t96.4\t138\t93.9\t99\t62.3\t17\t30.4\t1354\t17.0\t0\t0.0\tProline metabolism\n",
    "M00973\t0\t0.0\t2\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t0\t0.0\t2\t0.0\t0\t0.0\tAnammox, nitrite + ammonia => nitrogen\"\"\"\n",
    "\n",
    "kegg_module_description = pd.Series({line.split('\\t')[0]: line.split('\\t')[-1] for line in _kegg_module_page.split('\\n')[4:]}, name='description')\n",
    "assert kegg_module_description.index.str.startswith('M0').all()\n",
    "len(kegg_module_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[\n",
    "        lambda x: x.species_group_id == \"xjin_ucfmt_hmp2\"\n",
    "    ]\n",
    "    .species_id.astype(str)\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "species_taxonomy = (\n",
    "    pd.read_table(species_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    "    .Lineage.apply(parse_taxonomy_string)\n",
    ")\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    sorted(species_taxonomy.p__.unique()),\n",
    "    cm=\"tab10\",\n",
    ")\n",
    "\n",
    "# assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_genome(x):\n",
    "    if (x.genome_type == 'Isolate') & x.passes_filter:\n",
    "        return 'isolate'\n",
    "    elif (x.genome_type == 'Isolate') & ~x.passes_filter:\n",
    "        return 'isolate_fails_qc'\n",
    "    elif (x.genome_type == 'MAG') & x.passes_filter:\n",
    "        return 'mag'\n",
    "    elif (x.genome_type == 'MAG') & ~x.passes_filter:\n",
    "        return 'mag_fails_qc'\n",
    "    elif (x.genome_type == 'SPGC') & x.passes_filter:\n",
    "        return 'spgc'\n",
    "    elif (x.genome_type == 'SPGC') & x.passes_geno_positions & x.passes_in_sample_list:\n",
    "        return 'sfacts_only'\n",
    "    elif (x.genome_type == 'SPGC') & ~(x.passes_geno_positions & x.passes_in_sample_list):\n",
    "        return 'sfacts_fails_qc'\n",
    "    else:\n",
    "        raise ValueError(\"Genome did not match classification criteria:\", x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats = []\n",
    "missing_species = []\n",
    "\n",
    "_species_list = species_list\n",
    "# _species_list = [\"100003\"]\n",
    "\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.eggnog-strain_gene.strain_meta_for_analysis.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, index_col=\"genome_id\").assign(species=species)\n",
    "    filt_stats.append(data)\n",
    "filt_stats = pd.concat(filt_stats).assign(genome_class=lambda x: x.apply(classify_genome, axis=1))\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = filt_stats.groupby(['species', 'clust']).genome_class.value_counts().unstack('genome_class', fill_value=0).assign(highest_genome_class=lambda x: (x[['isolate', 'mag', 'spgc', 'sfacts_only', 'isolate_fails_qc', 'mag_fails_qc', 'sfacts_fails_qc']] > 0).idxmax(1))\n",
    "d1 = d0.groupby('species').highest_genome_class.value_counts().unstack(fill_value=0)[['isolate', 'mag', 'spgc', 'sfacts_only', 'isolate_fails_qc', 'mag_fails_qc', 'sfacts_fails_qc']]\n",
    "d2 = d1.join(species_taxonomy).sort_values(list(species_taxonomy.columns.values))\n",
    "\n",
    "row_colors = d2.p__.map(phylum_palette)\n",
    "sns.clustermap(d2[['isolate', 'mag', 'spgc', 'sfacts_only']], norm=mpl.colors.SymLogNorm(1), col_cluster=False, row_cluster=False, row_colors=row_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist2d(\n",
    "    \"min_ref_gene_raw_diss\",\n",
    "    \"min_ref_gene_filt_diss\",\n",
    "    data=filt_stats[lambda x: x.passes_filter].dropna(subset=[\"ref_nn_genome_id\"]),\n",
    "    bins=50,\n",
    "    norm=mpl.colors.PowerNorm(1 / 3),\n",
    ")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What fraction of StrainFacts genotypes (with enough positions == 100) have a genotype dissimilarity of less than 1%?\n",
    "filt_stats[\n",
    "    lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "    & x.passes_geno_positions\n",
    "    & x.passes_in_sample_list\n",
    "].min_ref_geno_diss.lt(0.01).mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set of species for dereplication analysis\n",
    "d = filt_stats[\n",
    "    lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "    & x.passes_geno_positions\n",
    "    & x.passes_in_sample_list\n",
    "]\n",
    "species_list1 = list(d.species.unique())\n",
    "len(species_list1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dereplication analysis: how many clusters are solely represented by an SPGC (or MAG/Isolate)?\n",
    "\n",
    "d = filt_stats[\n",
    "    lambda x: x.passes_geno_positions\n",
    "    & (\n",
    "        (x.genome_type.isin([\"SPGC\"]) & x.passes_in_sample_list)\n",
    "        | (x.genome_type.isin([\"Isolate\", \"MAG\"]))\n",
    "    )\n",
    "]\n",
    "d.genome_type.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    # Take all genomes with enough genotyped positions, and drop any SPGC\n",
    "    # genomes that are only found in UCFMT or XJIN samples.\n",
    "    filt_stats[\n",
    "        lambda x: x.passes_geno_positions\n",
    "        & (\n",
    "            (x.genome_type.isin([\"SPGC\"]) & x.passes_in_sample_list)\n",
    "            | (x.genome_type.isin([\"Isolate\", \"MAG\"]))\n",
    "        )\n",
    "    ]\n",
    "    # Count the number of genomes of each type in each cluster.\n",
    "    [[\"species\", \"clust\", \"genome_type\"]]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    # Tag each cluster by it's \"best type\"\n",
    "    .assign(best_genome_type=lambda x: x[[\"Isolate\", \"MAG\", \"SPGC\"]].idxmax(1))\n",
    "    # Count for each species the number of clusters with each tag.\n",
    "    .groupby(\"species\")\n",
    "    .best_genome_type.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "(\n",
    "    d.join(d.divide(d.sum(1), axis=0), rsuffix=\"_frac\")\n",
    "    .join(species_taxonomy[[\"p__\", \"f__\", \"g__\", \"s__\"]])\n",
    "    .sort_values(\"SPGC\", ascending=False)\n",
    "    .head(20)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    # Take all genomes that pass the full filter.\n",
    "    filt_stats[lambda x: x.passes_filter]\n",
    "    # Count the number of genomes of each type in each cluster.\n",
    "    [[\"species\", \"clust\", \"genome_type\"]]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    # Tag each cluster by it's \"best type\"\n",
    "    .assign(best_genome_type=lambda x: x[[\"Isolate\", \"MAG\", \"SPGC\"]].idxmax(1))\n",
    "    # Count for each species the number of clusters with each tag.\n",
    "    .groupby(\"species\")\n",
    "    .best_genome_type.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    ")\n",
    "\n",
    "(\n",
    "    d.join(d.divide(d.sum(1), axis=0), rsuffix=\"_frac\")\n",
    "    .join(species_taxonomy[[\"f__\", \"g__\", \"s__\"]])\n",
    "    .sort_values(\"SPGC\", ascending=False)\n",
    "    .head(40)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\n",
    "    filt_stats[\n",
    "        lambda x: x.min_ref_geno_diss.isna() & x.passes_geno_positions\n",
    "    ].species.value_counts()\n",
    ")\n",
    "\n",
    "assert ~filt_stats[lambda x: x.passes_geno_positions].min_ref_geno_diss.isna().any()\n",
    "assert ~filt_stats[lambda x: x.passes_geno_positions].ref_nn_gene_raw_diss.isna().any()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = filt_stats[lambda x: x.genome_type.isin([\"SPGC\"]) & x.passes_filter]\n",
    "species_list2 = idxwhere((d.species.value_counts() >= 10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats.groupby(\"species\").min_ref_gene_raw_diss.apply(\n",
    "    lambda x: x.isna().mean()\n",
    ").sort_values().tail(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats[lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])].dropna(\n",
    "    subset=[\"ref_nn_genome_id\"]\n",
    ")[\n",
    "    [\n",
    "        \"min_ref_gene_raw_diss\",\n",
    "        \"min_ref_gene_filt_diss\",\n",
    "        \"ref_nn_gene_raw_diss\",\n",
    "        \"ref_nn_gene_filt_diss\",\n",
    "    ]\n",
    "].quantile(\n",
    "    [0.25, 0.5, 0.75]\n",
    ").T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Relationship between genotype and gene distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = \"min_ref_geno_diss\"\n",
    "\n",
    "d0 = filt_stats[lambda x: x.passes_filter].dropna(subset=[\"ref_nn_genome_id\"])\n",
    "\n",
    "for y in [\n",
    "    \"min_ref_gene_raw_diss\",\n",
    "    \"min_ref_gene_filt_diss\",\n",
    "    \"ref_nn_gene_raw_diss\",\n",
    "    \"ref_nn_gene_filt_diss\",\n",
    "]:\n",
    "    for genome_set in [[\"SPGC\"], [\"Isolate\", \"MAG\"]]:\n",
    "        d1 = d0[lambda x: x.genome_type.isin(genome_set)]\n",
    "        print(y, genome_set, sp.stats.spearmanr(d1[x], d1[y]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = filt_stats[lambda x: x.passes_filter].dropna(subset=[\"ref_nn_genome_id\"])\n",
    "\n",
    "fit_raw_ref = smf.ols(\n",
    "    (\n",
    "        \"ref_nn_gene_raw_diss ~ \"\n",
    "        \"0 + C(species, Sum) + np.log10(min_ref_geno_diss)\"\n",
    "        # \"+ C(species, Sum):genome_type \"\n",
    "        # \"+ C(species, Sum):np.log10(min_ref_geno_diss) \"\n",
    "        # \"+ genome_type:np.log10(min_ref_geno_diss)\"\n",
    "        # \"+ C(species, Sum):genome_type:np.log10(min_ref_geno_diss)\"\n",
    "    ),\n",
    "    data=d0[lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])],\n",
    ").fit()\n",
    "print(fit_raw_ref.aic)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(2, 2, sharex=True, sharey=True, figsize=(7, 9))\n",
    "\n",
    "for (genome_type, d1), ax in zip(\n",
    "    d0.assign(predict=lambda x: fit_raw_ref.predict(x)).groupby(\"genome_type\"),\n",
    "    axs.flatten(),\n",
    "):\n",
    "    ax.set_title(genome_type)\n",
    "    bins = np.linspace(0, 0.5, num=50)\n",
    "    ax.hist2d(\n",
    "        \"predict\",\n",
    "        \"ref_nn_gene_raw_diss\",\n",
    "        data=d1,\n",
    "        bins=bins,\n",
    "        norm=mpl.colors.PowerNorm(1 / 2),\n",
    "    )\n",
    "    ax.set_aspect(1)\n",
    "    ax.plot([0, 0.6], [0, 0.6], c=\"w\", lw=1)\n",
    "fig.tight_layout()\n",
    "ax.set_xlabel(\"predicted\")\n",
    "ax.set_ylabel(\"observed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1 = d0[lambda x: x.genome_type == \"SPGC\"].assign(\n",
    "    predict=lambda x: fit_raw_ref.predict(x)\n",
    ")\n",
    "print(d1[\"ref_nn_gene_raw_diss\"].quantile([0.25, 0.5, 0.75]))\n",
    "print(d1[\"predict\"].quantile([0.25, 0.5, 0.75]))\n",
    "print(sp.stats.pearsonr(d1[\"ref_nn_gene_raw_diss\"], d1[\"predict\"]))\n",
    "print((d1.ref_nn_gene_raw_diss - d1.predict).quantile([0.25, 0.5, 0.75]))\n",
    "plt.hist(d1.ref_nn_gene_raw_diss - d1.predict, bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(d1.ref_nn_gene_raw_diss - d1.predict).quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = filt_stats[lambda x: x.passes_filter].dropna(subset=[\"ref_nn_genome_id\"])\n",
    "\n",
    "fit_raw_ref2 = smf.ols(\n",
    "    (\n",
    "        \"ref_nn_gene_raw_diss ~ \"\n",
    "        \"0 + C(species, Sum) + np.log10(min_ref_geno_diss) + genome_type\"\n",
    "        # \"+ C(species, Sum):genome_type \"\n",
    "        # \"+ C(species, Sum):np.log10(min_ref_geno_diss) \"\n",
    "        # \"+ genome_type:np.log10(min_ref_geno_diss)\"\n",
    "        # \"+ C(species, Sum):genome_type:np.log10(min_ref_geno_diss)\"\n",
    "    ),\n",
    "    data=d,\n",
    ").fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit_raw_ref2.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# species = '102506'\n",
    "d = filt_stats[lambda x: x.passes_filter]\n",
    "plt.scatter('min_ref_geno_diss', 'ref_nn_gene_raw_diss', data=d[lambda x: x.genome_type.isin(['Isolate'])], s=1)\n",
    "# plt.scatter('min_ref_geno_diss', 'ref_nn_gene_raw_diss', data=d[lambda x: x.genome_type.isin(['MAG'])], s=1)\n",
    "plt.scatter('min_ref_geno_diss', 'ref_nn_gene_raw_diss', data=d[lambda x: x.genome_type.isin(['SPGC'])], s=1)\n",
    "\n",
    "# plt.scatter('min_ref_geno_diss', 'ref_nn_gene_raw_diss', data=d, s=1)\n",
    "\n",
    "plt.xscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Taxonomic diversity of strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "filt_stats[\n",
    "    lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])\n",
    "].species.value_counts().to_frame(\"num_spgc_strains\").join(species_taxonomy)[\n",
    "    lambda x: x.p__ == \"p__Euryarchaeota\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    filt_stats[lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])]\n",
    "    .species.value_counts()\n",
    "    .to_frame(\"num_spgc_strains\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 20))\n",
    "ax.invert_yaxis()\n",
    "\n",
    "_phylum_list = (\n",
    "    d0.groupby(\"p__\").num_spgc_strains.sum().sort_values(ascending=False).index\n",
    ")\n",
    "_phylum_palette = lib.plot.construct_ordered_palette(_phylum_list, cm=\"rainbow\")\n",
    "\n",
    "y_start = 0\n",
    "for p__ in _phylum_list:\n",
    "    d1 = d0[lambda x: x.p__ == p__]\n",
    "    num_species = len(d1)\n",
    "    yy = np.arange(y_start, y_start + num_species)\n",
    "    plt.scatter(\n",
    "        d1.num_spgc_strains,\n",
    "        yy,\n",
    "        color=_phylum_palette[p__],\n",
    "        s=50,\n",
    "        marker=\"x\",\n",
    "        lw=2,\n",
    "        label=(p__, num_species, d1.num_spgc_strains.sum()),\n",
    "    )\n",
    "    y_start += num_species\n",
    "ax.set_xscale(\"log\")\n",
    "ax.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    filt_stats[lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])]\n",
    "    .species.value_counts()\n",
    "    .to_frame(\"num_spgc_strains\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "\n",
    "d0.groupby(\"p__\").num_spgc_strains.sum().sort_values(\n",
    "    ascending=False\n",
    ")  # sort_values('num_spgc_strains', ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    filt_stats[lambda x: x.passes_filter & x.genome_type.isin([\"SPGC\"])]\n",
    "    .groupby(\"species\")\n",
    "    .apply(\n",
    "        lambda x: pd.Series(\n",
    "            dict(\n",
    "                num_spgc_strains=len(x),\n",
    "                num_geno_positions=x.num_geno_positions.median(),\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "    # .to_frame(\"num_spgc_strains\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "\n",
    "\n",
    "for p__ in [\n",
    "    \"p__Firmicutes_A\",\n",
    "    \"p__Bacteroidota\",\n",
    "    \"p__Proteobacteria\",\n",
    "    \"p__Euryarchaeota\",\n",
    "]:\n",
    "    d1 = d0[lambda x: x.p__ == p__].sort_values(\"num_spgc_strains\", ascending=False)\n",
    "    print(p__)\n",
    "    print(d1[[\"num_spgc_strains\", \"s__\"]].head(8))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(14, 8))\n",
    "\n",
    "ax.scatter(\n",
    "    \"min_ref_geno_diss\",\n",
    "    \"ref_nn_gene_raw_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"lightgrey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "focal_species = \"102506\"\n",
    "genome_type_s_palette = {\"Isolate\": 2, \"MAG\": 2, \"SPGC\": 40}\n",
    "for genome_type in [\"SPGC\", \"MAG\", \"Isolate\"]:\n",
    "    ax.scatter(\n",
    "        \"min_ref_geno_diss\",\n",
    "        \"min_ref_gene_raw_diss\",\n",
    "        data=d0[\n",
    "            lambda x: x.genome_type.isin([genome_type])\n",
    "            & x.species.isin([focal_species])\n",
    "        ],\n",
    "        label=\"__nolegend__\",\n",
    "        color=genome_type_palette[genome_type],\n",
    "        s=genome_type_s_palette[genome_type],\n",
    "        alpha=0.5,\n",
    "    )\n",
    "\n",
    "for genome_type in [\"SPGC\", \"MAG\", \"Isolate\"]:\n",
    "    xx = np.logspace(-4, -1)\n",
    "    yy = (\n",
    "        fit_raw.predict(\n",
    "            pd.DataFrame(dict(min_ref_geno_diss=xx)).assign(\n",
    "                species=focal_species, genome_type=genome_type\n",
    "            )\n",
    "        ).values\n",
    "        # - fit.params[_arbitrary_species_term]\n",
    "    )\n",
    "    ax.plot(\n",
    "        xx,\n",
    "        yy,\n",
    "        color=genome_type_palette[genome_type],\n",
    "    )\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-6)\n",
    "\n",
    "ax.set_xlabel(\"Minimum SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Minimum Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "\n",
    "ax.set_ylim(0.0, 0.45)\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "name_to_contrast = dict(\n",
    "    # Slope of the relationship between log10(min_ref_geno_diss) and ref_nn_gene_raw_diss\n",
    "    # Isolate at\n",
    "    slope=\"np.log10(min_ref_geno_diss)\",\n",
    "    # MAG\n",
    "    slope_mag=\"np.log10(min_ref_geno_diss) + genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # SPGC\n",
    "    slope_spgc=\"np.log10(min_ref_geno_diss) + genome_type[T.SPGC]:np.log10(min_ref_geno_diss)\",\n",
    "    # Difference in slope of the relationship between log10(min_ref_geno_diss) and ref_nn_gene_raw_diss\n",
    "    # SPGC vs. Isolate\n",
    "    slope_spgc_vs_iso=\"genome_type[T.SPGC]:np.log10(min_ref_geno_diss)\",\n",
    "    # SPGC vs. MAG\n",
    "    slope_spgc_vs_mag=\"genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # MAG vs. Isolate\n",
    "    slope_mag_vs_iso=\"genome_type[T.SPGC]:np.log10(min_ref_geno_diss) = genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # Difference in means a genotype dissimilarity of 10**(-4)\n",
    "    # SPGC vs. Isolate at\n",
    "    mean_spgc_vs_iso=\"genome_type[T.SPGC] + (-4 * genome_type[T.SPGC]:np.log10(min_ref_geno_diss))\",\n",
    "    # SPGC vs. MAG at a dissimilarity of 10**(-4)\n",
    "    mean_spgc_vs_mag=\"genome_type[T.SPGC] + (-4 * genome_type[T.SPGC]:np.log10(min_ref_geno_diss)) = genome_type[T.MAG] + (-4 * genome_type[T.MAG]:np.log10(min_ref_geno_diss))\",\n",
    "    # MAG vs. Isolate at a dissimilarity of 10**(-4)\n",
    "    mean_mag_vs_iso=\"genome_type[T.MAG] + (-4 * genome_type[T.MAG]:np.log10(min_ref_geno_diss))\",\n",
    ")\n",
    "\n",
    "fit_raw.t_test(list(name_to_contrast.values())).summary_frame().rename(\n",
    "    {f\"c{i}\": name for i, name in enumerate(name_to_contrast.keys())}\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "#### Filtered"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = filt_stats[lambda x: x.passes_filter].dropna(subset=[\"ref_nn_genome_id\"])\n",
    "\n",
    "fit_filt = smf.ols(\n",
    "    (\n",
    "        \"ref_nn_gene_filt_diss ~ \"\n",
    "        \"0 + C(species, Sum) + genome_type + np.log10(min_ref_geno_diss)\"\n",
    "        \"+ C(species, Sum):genome_type \"\n",
    "        \"+ C(species, Sum):np.log10(min_ref_geno_diss) \"\n",
    "        \"+ genome_type:np.log10(min_ref_geno_diss)\"\n",
    "        \"+ C(species, Sum):genome_type:np.log10(min_ref_geno_diss)\"\n",
    "    ),\n",
    "    data=d0,\n",
    ").fit()\n",
    "print(fit_filt.aic)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fit_filt.summary()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "fig, ax = plt.subplots(figsize=(14, 8))\n",
    "\n",
    "ax.scatter(\n",
    "    \"min_ref_geno_diss\",\n",
    "    \"ref_nn_gene_filt_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"lightgrey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "focal_species = \"102492\"\n",
    "genome_type_s_palette = {\"Isolate\": 2, \"MAG\": 2, \"SPGC\": 40}\n",
    "for genome_type in [\"SPGC\", \"MAG\", \"Isolate\"]:\n",
    "    ax.scatter(\n",
    "        \"min_ref_geno_diss\",\n",
    "        \"min_ref_gene_filt_diss\",\n",
    "        data=d0[\n",
    "            lambda x: x.genome_type.isin([genome_type])\n",
    "            & x.species.isin([focal_species])\n",
    "        ],\n",
    "        label=\"__nolegend__\",\n",
    "        color=genome_type_palette[genome_type],\n",
    "        s=genome_type_s_palette[genome_type],\n",
    "        alpha=0.5,\n",
    "    )\n",
    "\n",
    "for genome_type in [\"SPGC\", \"MAG\", \"Isolate\"]:\n",
    "    xx = np.logspace(-4, -1)\n",
    "    yy = (\n",
    "        fit_filt.predict(\n",
    "            pd.DataFrame(dict(min_ref_geno_diss=xx)).assign(\n",
    "                species=focal_species, genome_type=genome_type\n",
    "            )\n",
    "        ).values\n",
    "        # - fit.params[_arbitrary_species_term]\n",
    "    )\n",
    "    ax.plot(\n",
    "        xx,\n",
    "        yy,\n",
    "        color=genome_type_palette[genome_type],\n",
    "    )\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-6)\n",
    "\n",
    "ax.set_xlabel(\"Minimum SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Minimum Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "\n",
    "ax.set_ylim(0.0, 0.45)\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "name_to_contrast = dict(\n",
    "    # Slope of the relationship between log10(min_ref_geno_diss) and ref_nn_gene_raw_diss\n",
    "    # Isolate at\n",
    "    slope=\"np.log10(min_ref_geno_diss)\",\n",
    "    # MAG\n",
    "    slope_mag=\"np.log10(min_ref_geno_diss) + genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # SPGC\n",
    "    slope_spgc=\"np.log10(min_ref_geno_diss) + genome_type[T.SPGC]:np.log10(min_ref_geno_diss)\",\n",
    "    # Difference in slope of the relationship between log10(min_ref_geno_diss) and ref_nn_gene_raw_diss\n",
    "    # SPGC vs. Isolate\n",
    "    slope_spgc_vs_iso=\"genome_type[T.SPGC]:np.log10(min_ref_geno_diss)\",\n",
    "    # SPGC vs. MAG\n",
    "    slope_spgc_vs_mag=\"genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # MAG vs. Isolate\n",
    "    slope_mag_vs_iso=\"genome_type[T.SPGC]:np.log10(min_ref_geno_diss) = genome_type[T.MAG]:np.log10(min_ref_geno_diss)\",\n",
    "    # Difference in means a genotype dissimilarity of 10**(-4)\n",
    "    # SPGC vs. Isolate at\n",
    "    mean_spgc_vs_iso=\"genome_type[T.SPGC] + (-4 * genome_type[T.SPGC]:np.log10(min_ref_geno_diss))\",\n",
    "    # SPGC vs. MAG at a dissimilarity of 10**(-4)\n",
    "    mean_spgc_vs_mag=\"genome_type[T.SPGC] + (-4 * genome_type[T.SPGC]:np.log10(min_ref_geno_diss)) = genome_type[T.MAG] + (-4 * genome_type[T.MAG]:np.log10(min_ref_geno_diss))\",\n",
    "    # MAG vs. Isolate at a dissimilarity of 10**(-4)\n",
    "    mean_mag_vs_iso=\"genome_type[T.MAG] + (-4 * genome_type[T.MAG]:np.log10(min_ref_geno_diss))\",\n",
    ")\n",
    "\n",
    "fit_filt.t_test(list(name_to_contrast.values())).summary_frame().rename(\n",
    "    {f\"c{i}\": name for i, name in enumerate(name_to_contrast.keys())}\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How many species/genomes for pangenomics?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "gene_meta = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/species/sp-{species}/midasdb.gene_meta.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, index_col=\"gene_id\").assign(species=species)\n",
    "    gene_meta.append(data)\n",
    "gene_meta = pd.concat(gene_meta)\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "_result = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/species/sp-{species}/midasdb.gene_x_cog_category.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, index_col=\"gene_id\")\n",
    "    _result.append(data)\n",
    "gene_x_cog_category = pd.concat(_result)\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")\n",
    "\n",
    "gene_x_cog_category_matrix = gene_x_cog_category.reset_index().set_index(['gene_id', 'cog_category']).assign(present=True).present.unstack('cog_category', fill_value=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table('ref/cog-20.categories.tsv', names=['cog_category', 'color', 'description'], index_col='cog_category').description\n",
    "cog_category_description.sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dereplication analysis: how many clusters are solely represented by an SPGC (or MAG/Isolate)?\n",
    "\n",
    "d = filt_stats[\n",
    "    lambda x: x.passes_filter\n",
    "    & (\n",
    "        (x.genome_type.isin([\"SPGC\"]) & x.passes_in_sample_list)\n",
    "        | (x.genome_type.isin([\"Isolate\", \"MAG\"]))\n",
    "    )\n",
    "]\n",
    "print(\"Num strains of each type:\", d.genome_type.value_counts())\n",
    "\n",
    "species_with_gt10_spgc_strains = idxwhere(\n",
    "    d[lambda x: x.genome_type.isin([\"SPGC\"])].species.value_counts() >= 10\n",
    ")\n",
    "print(\"Num species with >=10 genomes:\", len(species_with_gt10_spgc_strains))\n",
    "\n",
    "print(\n",
    "    \"Num strains of each type in pangenomics species:\",\n",
    "    d[\n",
    "        lambda x: x.species.isin(species_with_gt10_spgc_strains)\n",
    "    ].genome_type.value_counts(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d[lambda x: x.species.isin(species_with_gt10_spgc_strains)][\n",
    "    [\"species\", \"genome_type\"]\n",
    "].value_counts().unstack(\"genome_type\", fill_value=0).SPGC.quantile(\n",
    "    [0.25, 0.5, 0.75], interpolation=\"nearest\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d[lambda x: x.species.isin(species_with_gt10_spgc_strains)][\n",
    "    [\"species\", \"genome_type\"]\n",
    "].value_counts().unstack(\"genome_type\", fill_value=0).assign(\n",
    "    spgc_ratio=lambda x: x.SPGC / (x.Isolate + x.MAG)\n",
    ").sort_values(\n",
    "    \"spgc_ratio\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pangenomics (TODO: this doesn't really belong in this notebook)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence = []\n",
    "\n",
    "for species in tqdm(species_with_gt10_spgc_strains):\n",
    "    spgc_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene.prevalence-hmp2.tsv\"\n",
    "    ref_inpath = (\n",
    "        f\"data/species/sp-{species}/midasdb.gene75_new.uhgg-strain_gene.prevalence.tsv\"\n",
    "    )\n",
    "    spgc_prev = pd.read_table(\n",
    "        spgc_inpath, names=[\"gene_id\", \"prevalence\"], index_col=\"gene_id\"\n",
    "    ).prevalence\n",
    "    ref_prev = pd.read_table(\n",
    "        ref_inpath, names=[\"gene_id\", \"prevalence\"], index_col=\"gene_id\"\n",
    "    ).prevalence\n",
    "    data = pd.DataFrame(dict(spgc=spgc_prev, ref=ref_prev)).assign(species=species)\n",
    "    prevalence.append(data)\n",
    "prevalence = pd.concat(prevalence).fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(sp.stats.pearsonr(prevalence.ref, prevalence.spgc))\n",
    "plt.hist2d(\"ref\", \"spgc\", data=prevalence, bins=20, norm=mpl.colors.LogNorm())\n",
    "plt.colorbar()\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(prevalence.spgc[lambda x: x > 0], bins=np.linspace(0, 1, num=20))\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class = prevalence.spgc.map(lambda x: np.where(x > 0.9, \"core\", np.where(x > 0.15, \"shell\", \"cloud\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    prevalence_class.to_frame(\"prevalence_class\")\n",
    "    .join(gene_x_cog_category_matrix)\n",
    "    .assign(\n",
    "        cloud=lambda x: x.prevalence_class == \"cloud\",\n",
    "        shell=lambda x: x.prevalence_class == \"shell\",\n",
    "        core=lambda x: x.prevalence_class == \"core\",\n",
    "    )\n",
    ")\n",
    "\n",
    "result = []\n",
    "for _prevalence_class, _cog_category in tqdm(list(product(['core', 'shell', 'cloud'], gene_x_cog_category_matrix.columns))):\n",
    "    d1 = d0[[_prevalence_class, _cog_category]].value_counts().unstack().reindex(index=[True, False], columns=[True, False]).fillna(0)\n",
    "    d1_pc = d1 + 1\n",
    "    log_oddsratio = np.log2((d1_pc.loc[True, True] / d1_pc.loc[True, False]) / (d1_pc.loc[False, True] / d1_pc.loc[False, False]))\n",
    "    result.append((_prevalence_class, _cog_category, log_oddsratio, sp.stats.fisher_exact(d1)[1]))\n",
    "prevalence_class_cog_category_enrichment = pd.DataFrame(result, columns=['prevalence_class', 'cog_category', 'log2_oddsratio', 'pvalue']).set_index(['prevalence_class', 'cog_category'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = prevalence_class_cog_category_enrichment\n",
    "\n",
    "d_oddsr = d.log2_oddsratio.unstack('prevalence_class')\n",
    "d_signf = d.pvalue.map(lambda x: np.where(x < 0.05, '\u00b7', '')).unstack('prevalence_class')\n",
    "\n",
    "prevalence_class_order = ['core', 'shell', 'cloud']\n",
    "cog_category_order = d_oddsr['core'].sort_values(ascending=False).index\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "sns.heatmap(d_oddsr.loc[cog_category_order, prevalence_class_order], norm=mpl.colors.PowerNorm(1, vmin=-4, vmax=+4), cmap='coolwarm', ax=ax)\n",
    "# Annotations (because seaborn annotations are failing)\n",
    "for (i, _cog_category), (j, _prevalence_class) in product(enumerate(cog_category_order), enumerate(prevalence_class_order)):\n",
    "    ax.annotate(d_signf.loc[_cog_category, _prevalence_class], xy=(j + 0.5, i + 0.5), ha='center', va='center')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class_cog_category_enrichment.loc['core'].sort_values(['log2_oddsratio'], ascending=False).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class_cog_category_enrichment.loc['shell'].sort_values(['log2_oddsratio'], ascending=False).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prevalence_class_cog_category_enrichment.loc['cloud'].sort_values(['log2_oddsratio'], ascending=False).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_prevalence_class_tally = []\n",
    "ref_prevalence_class_tally = []\n",
    "\n",
    "for species in tqdm(species_with_gt10_spgc_strains):\n",
    "    spgc_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene.prevalence_class_fraction-hmp2.tsv\"\n",
    "    spgc_data = (\n",
    "        pd.read_table(spgc_inpath)\n",
    "        .assign(species=species)\n",
    "        .rename(columns={\"strain\": \"genome_id\"})\n",
    "        .astype({\"genome_id\": str})\n",
    "        .set_index([\"species\", \"genome_id\"])\n",
    "    )\n",
    "    spgc_prevalence_class_tally.append(spgc_data)\n",
    "\n",
    "    # TODO: ref_prevalence_class_tally\n",
    "    ref_inpath = f\"data/species/sp-{species}/midasdb.gene75_new.uhgg-strain_gene.prevalence_class_fraction.tsv\"\n",
    "    ref_data = (\n",
    "        pd.read_table(ref_inpath)\n",
    "        .assign(species=species)\n",
    "        .rename(columns={\"strain\": \"genome_id\"})\n",
    "        .astype({\"genome_id\": str})\n",
    "        .set_index([\"species\", \"genome_id\"])\n",
    "    )\n",
    "    ref_prevalence_class_tally.append(ref_data)\n",
    "\n",
    "# Filter SPGC\n",
    "spgc_prevalence_class_tally = (\n",
    "    pd.concat(spgc_prevalence_class_tally)\n",
    "    .fillna(0)\n",
    "    .loc[\n",
    "        filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "            & x.species.isin(species_with_gt10_spgc_strains)\n",
    "        ]\n",
    "        .reset_index()\n",
    "        .set_index([\"species\", \"genome_id\"])\n",
    "        .passes_filter\n",
    "    ]\n",
    ")\n",
    "\n",
    "# Filter Ref\n",
    "ref_prevalence_class_tally = (\n",
    "    pd.concat(ref_prevalence_class_tally)\n",
    "    .fillna(0)\n",
    "    .loc[\n",
    "        filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"MAG\", \"Isolate\"])\n",
    "            & x.species.isin(species_with_gt10_spgc_strains)\n",
    "        ]\n",
    "        .reset_index()\n",
    "        .set_index([\"species\", \"genome_id\"])\n",
    "        .passes_filter\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_prevalence_class_frac = spgc_prevalence_class_tally.divide(\n",
    "    spgc_prevalence_class_tally.sum(1), axis=0\n",
    ")  # .groupby(\"species\").median()\n",
    "ref_prevalence_class_frac = ref_prevalence_class_tally.divide(\n",
    "    ref_prevalence_class_tally.sum(1), axis=0\n",
    ")  # .groupby(\"species\").median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1 = spgc_prevalence_class_frac.groupby(\"species\").median()\n",
    "\n",
    "for frac in [\"core\", \"shell\", \"cloud\"]:\n",
    "    plt.hist(d1[frac], label=frac, alpha=0.7)\n",
    "plt.legend()\n",
    "\n",
    "d1.groupby(\"species\").median().quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d1 = ref_prevalence_class_frac.groupby(\"species\").median()\n",
    "\n",
    "for frac in [\"core\", \"shell\", \"cloud\"]:\n",
    "    plt.hist(d1[frac], label=frac, alpha=0.7)\n",
    "plt.legend()\n",
    "\n",
    "d1.groupby(\"species\").median().quantile([0.25, 0.5, 0.75])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        ref=ref_prevalence_class_frac.groupby(\"species\").median().stack(),\n",
    "        spgc=spgc_prevalence_class_frac.groupby(\"species\").median().stack(),\n",
    "    )\n",
    ").rename_axis(index=[\"species\", \"pangenome_fraction\"])\n",
    "for pangenome_fraction, d1 in d0.groupby(\"pangenome_fraction\"):\n",
    "    plt.scatter(\"ref\", \"spgc\", data=d1, s=10, alpha=0.7)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = spgc_prevalence_class_tally\n",
    "\n",
    "d1 = (\n",
    "    d0.divide(d0.sum(1), axis=0)\n",
    "    .groupby(\"species\")\n",
    "    .median()\n",
    "    .assign(num_strains=d0.index.to_frame().species.value_counts())\n",
    ")\n",
    "plt.scatter(\"core\", \"cloud\", c=\"num_strains\", data=d1, norm=mpl.colors.PowerNorm(1 / 2))\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "plt.scatter(\n",
    "    \"shell\", \"cloud\", c=\"num_strains\", data=d1, norm=mpl.colors.PowerNorm(1 / 2)\n",
    ")\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .join(ref_prevalence_class_frac.groupby(\"species\").median(), rsuffix=\"_ref\")\n",
    "    .assign(\n",
    "        total_num_ref_genomes=filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"MAG\", \"Isolate\"])\n",
    "        ][\"species\"].value_counts()\n",
    "    )\n",
    "    .join(species_taxonomy[[\"f__\", \"g__\", \"s__\"]])\n",
    "    .sort_values(\"cloud\", ascending=False)\n",
    "    .head(20)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .join(ref_prevalence_class_frac.groupby(\"species\").median(), rsuffix=\"_ref\")\n",
    "    .assign(\n",
    "        total_num_ref_genomes=filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"MAG\", \"Isolate\"])\n",
    "        ][\"species\"].value_counts()\n",
    "    )\n",
    "    .join(species_taxonomy[[\"f__\", \"g__\", \"s__\"]])\n",
    "    .sort_values(\"cloud\", ascending=True)\n",
    "    .head(20)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = (\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .rename_axis(columns=\"pangenome_fraction\")\n",
    "    .stack()\n",
    "    .to_frame(\"frac\")\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .reset_index()\n",
    ")\n",
    "\n",
    "sns.stripplot(x=\"pangenome_fraction\", hue=\"p__\", y=\"frac\", data=d2, dodge=True)\n",
    "lib.plot.rotate_xticklabels()\n",
    "plt.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .join(ref_prevalence_class_frac.groupby(\"species\").median(), rsuffix=\"_ref\")\n",
    "    .assign(\n",
    "        total_num_ref_genomes=filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"MAG\", \"Isolate\"])\n",
    "        ][\"species\"].value_counts(),\n",
    "        total_num_spgc_genomes=filt_stats[lambda x: x.genome_type.isin([\"SPGC\"])][\n",
    "            \"species\"\n",
    "        ].value_counts(),\n",
    "    )\n",
    "    .join(species_taxonomy[[\"p__\", \"f__\", \"g__\", \"s__\"]])\n",
    ")\n",
    "\n",
    "_phylum_palette = lib.plot.construct_ordered_palette(d.p__.unique(), cm=\"tab10\")\n",
    "\n",
    "\n",
    "import mpltern\n",
    "\n",
    "fig = plt.figure(figsize=(15, 15))\n",
    "ax = fig.add_subplot(projection=\"ternary\", ternary_sum=100.0)\n",
    "ax.grid()\n",
    "\n",
    "\n",
    "ax.set_tlabel(\"Core (%)\")\n",
    "ax.set_llabel(\"Shell (%)\")\n",
    "ax.set_rlabel(\"Cloud (%)\")\n",
    "\n",
    "ax.set_tlim(20, 100)\n",
    "ax.set_llim(10, 62)\n",
    "ax.set_rlim(0, 35)\n",
    "\n",
    "for p__, d1 in d.groupby(\"p__\"):\n",
    "    ax.scatter(\n",
    "        \"core\",\n",
    "        \"shell\",\n",
    "        \"cloud\",\n",
    "        data=d1,\n",
    "        s=d1[\"total_num_spgc_genomes\"],\n",
    "        marker=\"o\",\n",
    "        lw=2,\n",
    "        edgecolor=_phylum_palette[p__],\n",
    "        facecolor=\"none\",\n",
    "        alpha=0.85,\n",
    "    )\n",
    "    ax.scatter(\n",
    "        [],\n",
    "        [],\n",
    "        [],\n",
    "        edgecolor=_phylum_palette[p__],\n",
    "        label=p__,\n",
    "        lw=2,\n",
    "        facecolor=\"none\",\n",
    "    )\n",
    "\n",
    "ax.legend(loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phylum_palette = lib.plot.construct_ordered_palette(species_taxonomy.p__.unique(), cm='tab10')\n",
    "\n",
    "d = (\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .join(ref_prevalence_class_frac.groupby(\"species\").median(), rsuffix=\"_ref\")\n",
    "    .assign(\n",
    "        total_num_ref_genomes=filt_stats[\n",
    "            lambda x: x.genome_type.isin([\"MAG\", \"Isolate\"])\n",
    "        ][\"species\"].value_counts()\n",
    "    )\n",
    "    .join(species_taxonomy[[\"f__\", \"g__\", \"s__\"]])\n",
    ")\n",
    "\n",
    "import mpltern\n",
    "\n",
    "fig = plt.figure(figsize=(5, 5))\n",
    "ax = fig.add_subplot(projection=\"ternary\", ternary_sum=100.0)\n",
    "\n",
    "ax.set_tlabel(\"Core (%)\")\n",
    "ax.set_llabel(\"Shell (%)\")\n",
    "ax.set_rlabel(\"Cloud (%)\")\n",
    "\n",
    "# ax.grid()\n",
    "\n",
    "ax.scatter(\"core_ref\", \"shell_ref\", \"cloud_ref\", data=d, s=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# phylum_palette = lib.plot.construct_ordered_palette(species_taxonomy.p__.unique(), cm='tab10')\n",
    "\n",
    "d = (\n",
    "    spgc_prevalence_class_frac.groupby(\"species\")\n",
    "    .median()\n",
    "    .rename_axis(columns=\"pangenome_fraction\")\n",
    "    # .stack()\n",
    "    # .to_frame(\"frac\")\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    # .assign(phylum_color=lambda x: x.p__.map(phylum_palette))\n",
    "    # .reset_index()\n",
    ")\n",
    "\n",
    "import mpltern\n",
    "\n",
    "fig = plt.figure(figsize=(5, 5))\n",
    "ax = fig.add_subplot(projection=\"ternary\", ternary_sum=100.0)\n",
    "\n",
    "ax.set_tlabel(\"Core (%)\")\n",
    "ax.set_llabel(\"Shell (%)\")\n",
    "ax.set_rlabel(\"Cloud (%)\")\n",
    "\n",
    "# ax.grid()\n",
    "\n",
    "ax.scatter(\"core\", \"shell\", \"cloud\", data=d, s=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "morans_i = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene.morans_i.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    data = pd.read_table(inpath, index_col=\"gene_id\")\n",
    "    morans_i.append(data)\n",
    "morans_i = pd.concat(morans_i).assign(spgc_species_percentile=lambda x: x.spgc.groupby(gene_meta.species).rank(pct=True))\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inpath"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = filt_stats[lambda x: x.genome_type.isin([\"SPGC\"]) & x.passes_filter]\n",
    "species_list3 = idxwhere((d.species.value_counts() >= 20))\n",
    "len(species_list3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = morans_i.join(gene_meta.species)[lambda x: x.species.isin(species_list2)]\n",
    "\n",
    "plt.hist2d('ref', 'spgc', data=d.dropna(), bins=20, norm=mpl.colors.PowerNorm(1/3))\n",
    "sp.stats.pearsonr(d.dropna()['ref'], d.dropna()['spgc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = morans_i.join(gene_meta.species)[lambda x: x.species.isin(species_list3)]\n",
    "\n",
    "plt.hist2d('ref', 'spgc', data=d.dropna(), bins=20, norm=mpl.colors.PowerNorm(1/5))\n",
    "sp.stats.pearsonr(d.dropna()['ref'], d.dropna()['spgc'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(d.spgc.dropna(), bins=100, alpha=0.5)\n",
    "plt.hist(d.ref.dropna(), bins=100, alpha=0.5)\n",
    "\n",
    "plt.yscale('log')\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_meta.join(gene_x_cog_category_matrix).assign(cmi=morans_i.ref, prevalence=prevalence.ref).fillna({'cmi': 0})\n",
    "d"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for in_cat, d2 in d1.groupby(category):\n",
    "    plt.scatter('prevalence', 'cmi_rank', data=d2)\n",
    "    plt.plot('prevalence', 'predict', data=d2.assign(predict=lambda x: fit.predict(x), residual=lambda x: x.cmi_rank - x.predict).sort_values('prevalence'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Can take 15 or more minutes to run.\n",
    "\n",
    "d0 = pd.DataFrame(dict(cmi=morans_i.spgc, prevalence=prevalence.spgc)).assign(species=gene_meta.species).join(gene_x_cog_category_matrix).dropna()\n",
    "\n",
    "results = {}\n",
    "for species in tqdm(species_list3):\n",
    "    d1 = d0[lambda x: (x.species == species) & (x.prevalence > 0.15) & (x.prevalence < 0.9)]\n",
    "    for category in gene_x_cog_category_matrix.columns:\n",
    "        if (d1[category].mean() == 0) or (d1[category].mean() == 1):\n",
    "            continue\n",
    "        fit = smf.rlm(f'cmi ~ cr(prevalence, 4) + {category}', data=d1).fit()\n",
    "        results[(species, category)] = [fit.params[f'{category}[T.True]'], fit.pvalues[f'{category}[T.True]']]\n",
    "\n",
    "for category in tqdm(gene_x_cog_category_matrix.columns):\n",
    "    d1 = d0[lambda x: (x.species.isin(species_list3)) & (x.prevalence > 0.15) & (x.prevalence < 0.9)]\n",
    "    fit = smf.rlm(f'cmi ~ cr(prevalence, 4) + {category} + species', data=d1).fit()\n",
    "    results[('overall', category)] = [fit.params[f'{category}[T.True]'], fit.pvalues[f'{category}[T.True]']]\n",
    "\n",
    "spgc_species_by_cog_category_cdi_bias_test = pd.DataFrame(results.values(), index=results.keys(), columns=['coef', 'pvalue'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = pd.DataFrame(dict(cmi=morans_i.spgc, prevalence=prevalence.spgc)).assign(species=gene_meta.species).join(gene_x_cog_category_matrix).dropna()\n",
    "d1 = d0[lambda x: (x.species.isin(species_list3))]\n",
    "_fit = smf.rlm(f'cmi ~ cr(prevalence, 10) + species', data=d1).fit()\n",
    "# _fit.params[f'{category}[T.True]'], _fit.pvalues[f'{category}[T.True]']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(predict=_fit.predict(), resid=_fit.resid)\n",
    "\n",
    "bins = np.linspace(d1.cmi.min(), d1.cmi.max(), num=10)\n",
    "\n",
    "fig, axs = plt.subplots(2, 2, sharex=True, sharey=True)\n",
    "for cat, ax in zip(['G', 'W', 'X', 'J'], axs.flatten()):\n",
    "    ax.hist2d('predict', 'resid', data=d2[d2[cat]], bins=bins, norm=mpl.colors.PowerNorm(1/2), density=True)\n",
    "    ax.set_title(cat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(np.random.rand(d2['W'].sum()) + 1, d2[d2['W']].prevalence, s=1)\n",
    "plt.scatter(np.random.rand(d2['G'].sum()) + 2, d2[d2['G']].prevalence, s=1, alpha=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter('prevalence', 'resid', data=d2[d2['X']], s=3, alpha=0.2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d3 = d2[lambda x: (x.species == '101337')]\n",
    "\n",
    "plt.scatter('prevalence', 'cmi', data=d3, s=3, alpha=0.2)\n",
    "pp = np.linspace(0, 1)\n",
    "plt.plot(pp, _fit.predict(pd.DataFrame(dict(prevalence=pp, species='100022'))), color='tab:orange')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2[lambda x: (x.species == '101337')].assign(cmi_rank=lambda x: x.cmi.rank(pct=True)).cmi_rank.median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2[lambda x: (x.species == '101337')].assign(cmi_rank=lambda x: x.cmi.rank(pct=True))[lambda x: x.G].cmi_rank.median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "qq = np.linspace(0, 1, num=100)\n",
    "plt.scatter(d2[d2['W']].resid.quantile(qq), d2[d2['G']].resid.quantile(qq))\n",
    "plt.plot([0, 0.15], [0, 0.15])\n",
    "# plt.yscale('log')\n",
    "# plt.xscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Can take 15 or more minutes to run.\n",
    "\n",
    "d0 = pd.DataFrame(dict(cmi=morans_i.ref, prevalence=prevalence.ref)).assign(species=gene_meta.species).join(gene_x_cog_category_matrix).dropna()\n",
    "\n",
    "results = {}\n",
    "for species in tqdm(species_list3):\n",
    "    d1 = d0[lambda x: (x.species == species) & (x.prevalence > 0.15) & (x.prevalence < 0.9)]\n",
    "    for category in gene_x_cog_category_matrix.columns:\n",
    "        if (d1[category].mean() == 0) or (d1[category].mean() == 1):\n",
    "            continue\n",
    "        fit = smf.rlm(f'cmi ~ cr(prevalence, 4) + {category}', data=d1).fit()\n",
    "        results[(species, category)] = [fit.params[f'{category}[T.True]'], fit.pvalues[f'{category}[T.True]']]\n",
    "\n",
    "for category in tqdm(gene_x_cog_category_matrix.columns):\n",
    "    d1 = d0[lambda x: (x.species.isin(species_list3)) & (x.prevalence > 0.15) & (x.prevalence < 0.9)]\n",
    "    fit = smf.rlm(f'cmi ~ cr(prevalence, 4) + {category} + species', data=d1).fit()\n",
    "    results[('overall', category)] = [fit.params[f'{category}[T.True]'], fit.pvalues[f'{category}[T.True]']]\n",
    "\n",
    "ref_species_by_cog_category_cdi_bias_test = pd.DataFrame(results.values(), index=results.keys(), columns=['coef', 'pvalue'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = spgc_species_by_cog_category_cdi_bias_test\n",
    "d1_coef = d0.coef.unstack(fill_value=0)\n",
    "d1_signf = d0.pvalue.unstack(fill_value=1.0).map(lambda p: np.where(p < 0.05, '\u2022', ''))\n",
    "\n",
    "cmi_category_order = list(d1_coef.mean().sort_values().index)\n",
    "cmi_species_order = list(np.square(d1_coef).mean(1).sort_values(ascending=False).drop('overall').index) + ['overall']\n",
    "\n",
    "d2_coef = d1_coef.loc[cmi_species_order, cmi_category_order]\n",
    "d2_signf = d1_signf.loc[cmi_species_order, cmi_category_order]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "# sns.heatmap(d2_coef)\n",
    "# sns.heatmap(d1_coef.loc[cmi_species_order, cmi_category_order], annot=d1_signf.loc[cmi_species_order, cmi_category_order], fmt='', norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-1e5, vmax=1e5), xticklabels=1, yticklabels=1, cmap='coolwarm', ax=ax)\n",
    "sns.heatmap(d2_coef, norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-.1, vmax=.1), xticklabels=1, yticklabels=1, cmap='PuOr', ax=ax)\n",
    "\n",
    "# Annotations (because seaborn annotations are failing)\n",
    "for (i, species), (j, category) in product(enumerate(cmi_species_order), enumerate(cmi_category_order)):\n",
    "    ax.annotate(d2_signf.loc[species, category], xy=(j + 0.5, i + 0.5), ha='center', va='center')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_species_by_cog_category_cdi_bias_test.loc['overall'].reindex(cmi_category_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description.reindex(cmi_category_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 20\n",
    "r = 0.1\n",
    "offset = 1\n",
    "\n",
    "p = int(round(n * r))\n",
    "q = int(round(n * (1 - r)))\n",
    "print(p, q)\n",
    "x = [0]*p + [1]*q\n",
    "y = [0]*(p - offset) + [1]*(q + offset)\n",
    "print(x)\n",
    "print(y)\n",
    "sp.stats.pearsonr(x, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "','.join(species_list3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = ref_species_by_cog_category_cdi_bias_test\n",
    "d1_coef = d0.coef.unstack(fill_value=0)\n",
    "d1_signf = d0.pvalue.unstack(fill_value=1.0).map(lambda p: np.where(p < 0.05, '\u2022', ''))\n",
    "\n",
    "d2_coef = d1_coef.loc[cmi_species_order, cmi_category_order]\n",
    "d2_signf = d1_signf.loc[cmi_species_order, cmi_category_order]\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "# sns.heatmap(d2_coef)\n",
    "# sns.heatmap(d1_coef.loc[cmi_species_order, cmi_category_order], annot=d1_signf.loc[cmi_species_order, cmi_category_order], fmt='', norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-1e5, vmax=1e5), xticklabels=1, yticklabels=1, cmap='coolwarm', ax=ax)\n",
    "sns.heatmap(d2_coef, norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-.1, vmax=.1), xticklabels=1, yticklabels=1, cmap='PuOr', ax=ax)\n",
    "\n",
    "# Annotations (because seaborn annotations are failing)\n",
    "for (i, species), (j, category) in product(enumerate(cmi_species_order), enumerate(cmi_category_order)):\n",
    "    ax.annotate(d2_signf.loc[species, category], xy=(j + 0.5, i + 0.5), ha='center', va='center')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "co_clust = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    ref_inpath = f\"data/species/sp-{species}/midasdb.gene75_new.uhgg-strain_gene.gene_clust-t10.tsv\"\n",
    "    spgc_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene.gene_clust-t10.tsv\"\n",
    "    if not os.path.exists(spgc_inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    ref_data = pd.read_table(ref_inpath, names=['gene_id', 'cluster'], index_col=\"gene_id\").cluster\n",
    "    spgc_data = pd.read_table(spgc_inpath, names=['gene_id', 'cluster'], index_col=\"gene_id\").cluster\n",
    "    co_clust.append(pd.DataFrame(dict(ref_clust=ref_data, spgc_clust=spgc_data)).assign(species=species))\n",
    "co_clust = pd.concat(co_clust).fillna({'spgc_clust': -5, 'ref_clust': -5}).astype({'spgc_clust': int, 'ref_clust': int})\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (co_clust[lambda x: (x.spgc_clust >= 0) & x.species.isin(species_list3)][['species', 'spgc_clust']].value_counts()[lambda x: (x > 1)])\n",
    "clust_list = d.index\n",
    "len(d), d.sum(), d.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d.groupby('species').apply(len).median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_module = gene_meta.KEGG_Module.dropna().str.split(',').explode()#[lambda x: x.str.startswith('map')]\n",
    "gene_x_module.value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_pathway = gene_meta.KEGG_inpathway.dropna().str.split(',').explode()[lambda x: x.str.startswith('map')]\n",
    "gene_x_pathway.value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    co_clust[lambda x: (x.spgc_clust >= 0) & (x.index.isin(gene_x_module.index))]\n",
    "    .join(gene_x_module)\n",
    "    .assign(\n",
    "        # ref_clust=lambda x: x.species + \"-\" + x.ref_clust.astype(str),\n",
    "        spgc_clust=lambda x: x.species + \"-\" + x.spgc_clust.astype(str),\n",
    "    )\n",
    ")\n",
    "d0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh = 1\n",
    "\n",
    "obs_count_module_multihit = (\n",
    "    # Count the total number of clusters with multiple hits to the same KEGG_Module\n",
    "    d0[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    ")\n",
    "perm_count_module_multihit = []\n",
    "for i in tqdm(range(999)):\n",
    "    perm_count_module_multihit.append(\n",
    "        d0.assign(\n",
    "            # Permute clusters within species.\n",
    "            spgc_clust=lambda x: x.groupby(\"species\").spgc_clust.sample(frac=1).values\n",
    "        )[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    "    )\n",
    "perm_count_module_multihit = np.array(perm_count_module_multihit)\n",
    "plt.hist(perm_count_module_multihit)\n",
    "plt.axvline(obs_count_module_multihit)\n",
    "\n",
    "print(perm_count_module_multihit.mean(), perm_count_module_multihit.std(), obs_count_module_multihit, obs_count_module_multihit / perm_count_module_multihit.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh = 2\n",
    "\n",
    "obs_count_module_multihit = (\n",
    "    # Count the total number of clusters with multiple hits to the same KEGG_Module\n",
    "    d0[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    ")\n",
    "perm_count_module_multihit = []\n",
    "for i in tqdm(range(999)):\n",
    "    perm_count_module_multihit.append(\n",
    "        d0.assign(\n",
    "            # Permute clusters within species.\n",
    "            spgc_clust=lambda x: x.groupby(\"species\").spgc_clust.sample(frac=1).values\n",
    "        )[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    "    )\n",
    "perm_count_module_multihit = np.array(perm_count_module_multihit)\n",
    "plt.hist(perm_count_module_multihit)\n",
    "plt.axvline(obs_count_module_multihit)\n",
    "\n",
    "print(perm_count_module_multihit.mean(), perm_count_module_multihit.std(), obs_count_module_multihit, obs_count_module_multihit / perm_count_module_multihit.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh = 3\n",
    "\n",
    "obs_count_module_multihit = (\n",
    "    # Count the total number of clusters with multiple hits to the same KEGG_Module\n",
    "    d0[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    ")\n",
    "perm_count_module_multihit = []\n",
    "for i in tqdm(range(999)):\n",
    "    perm_count_module_multihit.append(\n",
    "        d0.assign(\n",
    "            # Permute clusters within species.\n",
    "            spgc_clust=lambda x: x.groupby(\"species\").spgc_clust.sample(frac=1).values\n",
    "        )[[\"spgc_clust\", \"KEGG_Module\"]]\n",
    "    .value_counts()\n",
    "    .gt(thresh)\n",
    "    .groupby('spgc_clust')\n",
    "    .any()\n",
    "    .sum()\n",
    "    )\n",
    "perm_count_module_multihit = np.array(perm_count_module_multihit)\n",
    "plt.hist(perm_count_module_multihit)\n",
    "plt.axvline(obs_count_module_multihit)\n",
    "\n",
    "print(perm_count_module_multihit.mean(), perm_count_module_multihit.std(), obs_count_module_multihit, obs_count_module_multihit / perm_count_module_multihit.mean())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject = pd.read_table('meta/ucfmt/subject.tsv', index_col='subject_id')\n",
    "subject[['donor_subject_id', 'remission']].value_counts(dropna=False).unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject[lambda x: (~x.withdrawal_due_to_failure.fillna(True))][['donor_subject_id', 'remission']].value_counts().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Unfortunately, no evidence that donors performed differently. :-/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "donor_comparison = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene-ucfmt.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    donor_comparison.append(pd.read_table(inpath, index_col=\"gene_id\"))\n",
    "\n",
    "donor_comparison = (\n",
    "    pd.concat(donor_comparison)\n",
    "    [[\"D0044\", \"D0097\"]].dropna()\n",
    "    .assign(\n",
    "        label=lambda x: (\n",
    "            # Encode in binary\n",
    "            x.D0097 * 2**0 + x.D0044 * 2**1\n",
    "        ).map(  \n",
    "            {\n",
    "                # Lookup\n",
    "                0: \"neither\",\n",
    "                1: \"d97\",\n",
    "                2: \"d44\",\n",
    "                3: \"both\",\n",
    "            }\n",
    "        )\n",
    "    )\n",
    ")\n",
    "# mwas = pd.concat(mwas).assign(\n",
    "#     log2_oddsratio_pc_ibd=lambda x: np.log2(x.oddsratio_pc_ibd),\n",
    "#     neg_log10_pvalue=lambda x: -np.log10(x.fisher_exact_pvalue_ibd),\n",
    "#     num_subjects_present=lambda x: x[\"present-nonIBD\"] + x[\"present-IBD\"],\n",
    "#     num_subjects_absent=lambda x: x[\"absent-nonIBD\"] + x[\"absent-IBD\"],\n",
    "#     num_subjects_total=lambda x: x.num_subjects_present + x.num_subjects_absent,\n",
    "#     prevalence=lambda x: x.num_subjects_present\n",
    "#     / x.num_subjects_total,\n",
    "# )\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = donor_comparison.groupby(gene_meta.species).label.value_counts().unstack()\n",
    "d.quantile([0.25, 0.5, 0.75]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = donor_comparison.groupby(gene_meta.species).label.value_counts().unstack()\n",
    "d.apply(lambda x: x / x.sum(), axis=1).quantile([0.25, 0.5, 0.75]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = species_list\n",
    "\n",
    "mwas = []\n",
    "missing_species = []\n",
    "for species in tqdm(_species_list):\n",
    "    inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.uhgg-strain_gene.hmp2_mwas-f30-n1.tsv\"\n",
    "    if not os.path.exists(inpath):\n",
    "        missing_species.append(species)\n",
    "        continue\n",
    "    mwas.append(pd.read_table(inpath, index_col=\"gene_id\").assign(species=species))\n",
    "mwas = pd.concat(mwas).assign(\n",
    "    log2_oddsratio_pc_ibd=lambda x: np.log2(x.oddsratio_pc_ibd),\n",
    "    neg_log10_pvalue=lambda x: -np.log10(x.fisher_exact_pvalue_ibd),\n",
    "    num_subjects_present=lambda x: x[\"present-nonIBD\"] + x[\"present-IBD\"],\n",
    "    num_subjects_absent=lambda x: x[\"absent-nonIBD\"] + x[\"absent-IBD\"],\n",
    "    num_subjects_total=lambda x: x.num_subjects_present + x.num_subjects_absent,\n",
    "    prevalence=lambda x: x.num_subjects_present\n",
    "    / x.num_subjects_total,\n",
    ")\n",
    "\n",
    "print(\n",
    "    len(missing_species),\n",
    "    \"out of\",\n",
    "    len(_species_list),\n",
    "    \"species are missing stats.\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_comparison.join(gene_meta.species)[['species', 'label']].value_counts().unstack().sort_values('d44').head(10)\n",
    "# No species has 0 d44/d97 genes, meaning they're always different strains across the two donors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust_size = co_clust.groupby(['species', 'spgc_clust']).apply(len)\n",
    "clust_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust, _donor = align_indexes(co_clust[lambda x: x.species.isin(species_list3)], donor_comparison)\n",
    "# _clust_size = _donor.join(_clust).groupby(['species', 'spgc_clust']).apply(len)\n",
    "_clust_size = co_clust[lambda x: x.spgc_clust >= 0][['species', 'spgc_clust']].value_counts()\n",
    "\n",
    "d = _donor.join(_clust)\n",
    "# d1 = d.groupby(['species', 'spgc_clust'])[['D0044', 'D0097']].mean()[(_clust_size > 1) & (_clust_size.index.to_frame().spgc_clust >= 0)]\n",
    "# d2 = d.groupby(['species', 'spgc_clust']).label.value_counts().unstack('label', fill_value=0).divide(_clust_size, axis=0)[(_clust_size > 1) & (_clust_size.index.to_frame().spgc_clust >= 0)]\n",
    "donor_strain_cluster_membership = d.groupby(['species', 'spgc_clust'])[['D0044', 'D0097']].sum().astype(int).assign(clust_size=_clust_size)[lambda x: x.clust_size > 1].join(d.groupby(['species', 'spgc_clust']).label.value_counts().unstack(fill_value=0)).assign(D0044_frac=lambda x: x.D0044 / x.clust_size, D0097_frac=lambda x: x.D0097 / x.clust_size)\n",
    "\n",
    "donor_strain_cluster_membership.sort_values('clust_size', ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist2d('D0097_frac', 'D0044_frac', data=donor_strain_cluster_membership[lambda x: x.clust_size > 1], bins=np.linspace(0, 1, num=21))\n",
    "plt.colorbar()\n",
    "None\n",
    "# plt.hist(donor_strain_cluster_membership[lambda x: x.clust_size > 5]['D0097_frac'], bins=bins, alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_cluster_membership.assign(\n",
    "    d44_present=lambda x: x.D0044_frac > 0.9, d97_present=lambda x: x.D0097_frac > 0.9\n",
    ").assign(both_present=lambda x: x.d44_present & x.d97_present)[\n",
    "    lambda x: x.clust_size > 1\n",
    "][['both', 'd44', 'd97', 'd44_present', 'd97_present', 'both_present']].groupby('species').sum().join(species_taxonomy.s__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of species being considered\n",
    "donor_strain_cluster_membership.index.to_frame().species.value_counts().shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "differential_cluster_membership = donor_strain_cluster_membership[lambda x: x.clust_size > 1][['D0044_frac', 'D0097_frac']][lambda x: ((x.D0044_frac > 0.9) & (x.D0097_frac < 0.1)) | ((x.D0097_frac > 0.9) & (x.D0044_frac < 0.1))]\n",
    "differential_cluster_membership.round().groupby('species').sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = differential_cluster_membership.round().groupby('species').sum()\n",
    "d.shape, d.values.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "differential_cluster_membership.assign(num_genes=donor_strain_cluster_membership.clust_size).sort_values('num_genes', ascending=False).xs('102506', level='species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species = \"102506\"\n",
    "\n",
    "for _clust in (\n",
    "    differential_cluster_membership.assign(\n",
    "        num_genes=donor_strain_cluster_membership.clust_size\n",
    "    )\n",
    "    .sort_values(\"num_genes\", ascending=False)\n",
    "    .xs(_species, level=\"species\")\n",
    "    .head(50)\n",
    "    .index\n",
    "):\n",
    "    print(\"-------------\")\n",
    "    print(\"Cluster\", _clust)\n",
    "    print()\n",
    "    print(\n",
    "        donor_strain_cluster_membership.loc[[(_species, _clust)]][\n",
    "            [\"D0044\", \"D0097\", \"clust_size\"]\n",
    "        ]\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        gene_x_cog_category_matrix.loc[\n",
    "            co_clust[lambda x: (x.species == _species) & (x.spgc_clust == _clust)].index\n",
    "        ]\n",
    "        .sum()[lambda x: x > 0]\n",
    "        .sort_values(ascending=False)\n",
    "        .to_frame()\n",
    "        .T\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        \"\\n\".join(\n",
    "            gene_meta.loc[\n",
    "                co_clust[\n",
    "                    lambda x: (x.species == _species) & (x.spgc_clust == _clust)\n",
    "                ].index\n",
    "            ][lambda x: ~(x.Preferred_name.isna() & x.Description.isna())]\n",
    "            .astype(str)\n",
    "            .apply(\n",
    "                lambda x: f\"({x.Preferred_name}) {x.Description} ({x.KEGG_ko})\", axis=1\n",
    "            )\n",
    "            .values\n",
    "        )\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        mwas.loc[\n",
    "            co_clust[lambda x: (x.species == _species) & (x.spgc_clust == _clust)].index\n",
    "        ]\n",
    "        .rename(\n",
    "            columns={\n",
    "                \"num_subjects_total\": \"nsubj\",\n",
    "                \"log2_oddsratio_pc_ibd\": \"lor\",\n",
    "                \"fisher_exact_pvalue_ibd\": \"p\",\n",
    "            }\n",
    "        )[[\"nsubj\", \"prevalence\", \"lor\", \"p\"]]\n",
    "        .round(2)\n",
    "        .assign(morans_i=morans_i.spgc_species_percentile)\n",
    "        .value_counts()\n",
    "    )\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species = \"101337\"\n",
    "\n",
    "for _clust in (\n",
    "    differential_cluster_membership.assign(\n",
    "        num_genes=donor_strain_cluster_membership.clust_size\n",
    "    )\n",
    "    .sort_values(\"num_genes\", ascending=False)\n",
    "    .xs(_species, level=\"species\")\n",
    "    .head(50)\n",
    "    .index\n",
    "):\n",
    "    print(\"-------------\")\n",
    "    print(\"Cluster\", _clust)\n",
    "    print()\n",
    "    print(\n",
    "        donor_strain_cluster_membership.loc[[(_species, _clust)]][\n",
    "            [\"D0044\", \"D0097\", \"clust_size\"]\n",
    "        ]\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        gene_x_cog_category_matrix.loc[\n",
    "            co_clust[lambda x: (x.species == _species) & (x.spgc_clust == _clust)].index\n",
    "        ]\n",
    "        .sum()[lambda x: x > 0]\n",
    "        .sort_values(ascending=False)\n",
    "        .to_frame()\n",
    "        .T\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        \"\\n\".join(\n",
    "            gene_meta.loc[\n",
    "                co_clust[\n",
    "                    lambda x: (x.species == _species) & (x.spgc_clust == _clust)\n",
    "                ].index\n",
    "            ][lambda x: ~(x.Preferred_name.isna() & x.Description.isna())]\n",
    "            .astype(str)\n",
    "            .apply(\n",
    "                lambda x: f\"({x.Preferred_name}) {x.Description} ({x.KEGG_ko})\", axis=1\n",
    "            )\n",
    "            .values\n",
    "        )\n",
    "    )\n",
    "    print()\n",
    "    print(\n",
    "        mwas.loc[\n",
    "            co_clust[lambda x: (x.species == _species) & (x.spgc_clust == _clust)].index\n",
    "        ]\n",
    "        .rename(\n",
    "            columns={\n",
    "                \"num_subjects_total\": \"nsubj\",\n",
    "                \"log2_oddsratio_pc_ibd\": \"lor\",\n",
    "                \"fisher_exact_pvalue_ibd\": \"p\",\n",
    "            }\n",
    "        )[[\"nsubj\", \"prevalence\", \"lor\", \"p\"]]\n",
    "        .round(2)\n",
    "        .assign(morans_i=morans_i.spgc_species_percentile)\n",
    "        .value_counts()\n",
    "    )\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    mwas[lambda x:\n",
    "        # (x.species == \"102506\")\n",
    "    True\n",
    "        & (x.num_subjects_present >= 10)\n",
    "        & (x.num_subjects_absent >= 10)\n",
    "    ].assign(\n",
    "        donor=donor_comparison.label,\n",
    "        species=gene_meta.species,\n",
    "        prevalence_class=prevalence_class,\n",
    "        nlength=gene_meta.nlength,\n",
    "        clust=co_clust.spgc_clust,\n",
    "    )[\n",
    "        lambda x: x.donor.isin([\"d97\", \"d44\"])\n",
    "        & (x.prevalence_class == \"shell\")\n",
    "        & (x.nlength > 300)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1])\n",
    ").sort_values(['fisher_exact_pvalue_ibd', 'clust'])\n",
    "\n",
    "d[lambda x: x.fisher_exact_pvalue_ibd < 1e-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    mwas[lambda x:\n",
    "        # (x.species == \"102506\") &\n",
    "    True\n",
    "        & (x.num_subjects_present >= 10)\n",
    "        & (x.num_subjects_absent >= 10)\n",
    "    ].assign(\n",
    "        donor=donor_comparison.label,\n",
    "        species=gene_meta.species,\n",
    "        prevalence_class=prevalence_class,\n",
    "        nlength=gene_meta.nlength,\n",
    "        clust=co_clust.spgc_clust[lambda x: x >= 0],\n",
    "    )\n",
    "        .join(co_clust[['species', 'spgc_clust']].value_counts().rename(\"clust_size\"), on=['species', 'clust'])\n",
    "    [\n",
    "        lambda x: True\n",
    "        & x.donor.isin([\"d97\", \"d44\", \"both\"])\n",
    "        & (x.prevalence_class == \"shell\")\n",
    "        & (x.nlength > 300)\n",
    "        & (x.clust_size > 1)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1])\n",
    "    .sort_values('fisher_exact_pvalue_ibd')\n",
    ")\n",
    "\n",
    "plt.scatter('log2_oddsratio_pc_ibd', 'fisher_exact_pvalue_ibd', c='fdr', data=d, lw=1, facecolor='none', marker='o', cmap='viridis_r')\n",
    "plt.yscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.colorbar()\n",
    "\n",
    "d[lambda x: x.fisher_exact_pvalue_ibd < 1e-2][['fisher_exact_pvalue_ibd', 'oddsratio_pc_ibd', 'species',\n",
    "       'log2_oddsratio_pc_ibd', 'neg_log10_pvalue', 'num_subjects_present',\n",
    "       'num_subjects_absent', 'prevalence', 'donor',\n",
    "       'prevalence_class', 'nlength', 'clust', 'clust_size', 'fdr']].head(30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d[lambda x: x.fisher_exact_pvalue_ibd < 1e-3][\n",
    "    [\n",
    "        \"fisher_exact_pvalue_ibd\",\n",
    "        \"log2_oddsratio_pc_ibd\",\n",
    "        \"num_subjects_total\",\n",
    "        \"prevalence\",\n",
    "        \"donor\",\n",
    "        'fdr',\n",
    "        'clust',\n",
    "        'clust_size',\n",
    "    ]\n",
    "].join(\n",
    "    gene_meta[\n",
    "        [\n",
    "            \"species\",\n",
    "            \"PFAMs\",\n",
    "            \"eggNOG_OGs\",\n",
    "            \"COG_category\",\n",
    "            \"Description\",\n",
    "            \"Preferred_name\",\n",
    "        ]\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statsmodels.graphics.gofplots import qqplot\n",
    "\n",
    "d0 = (\n",
    "    mwas[lambda x:\n",
    "        # (x.species == \"102506\") &\n",
    "    True\n",
    "        & (x.num_subjects_present >= 10)\n",
    "        & (x.num_subjects_absent >= 10)\n",
    "    ].assign(\n",
    "        donor=donor_comparison.label,\n",
    "        species=gene_meta.species,\n",
    "        prevalence_class=prevalence_class,\n",
    "        nlength=gene_meta.nlength,\n",
    "        clust=co_clust.spgc_clust,\n",
    "    )[\n",
    "        lambda x: x.donor.isin([\"d97\", \"d44\"])\n",
    "        & (x.prevalence_class == \"shell\")\n",
    "        & (x.nlength > 300)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1])\n",
    "    .sort_values('fisher_exact_pvalue_ibd')\n",
    ")\n",
    "\n",
    "d1 = d0.join(gene_x_cog_category_matrix)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "cog_category_list = gene_x_cog_category_matrix.loc[d0.index].sum().sort_values(ascending=False).head(10).index\n",
    "\n",
    "for cat in cog_category_list:\n",
    "    d2 = d1[d1[cat]]\n",
    "    ax.scatter(np.linspace(0, 1, num=d2.shape[0] + 1)[1:], d2.fisher_exact_pvalue_ibd, label=cat, s=40)\n",
    "    # ax.scatter(np.linspace(0, 1, num=d2.shape[0] + 1)[1:], sp.stats.uniform.ppf(d2.fisher_exact_pvalue_ibd), label=cat, s=40)  # Equivalently, and generalizable to other distributions.\n",
    "ax.plot([0, 1], [0, 1], color='k')\n",
    "ax.legend(bbox_to_anchor=(1, 1), markerscale=1)\n",
    "ax.set_yscale('log')\n",
    "ax.set_xscale('log')\n",
    "# plt.xlim(right=1e-1)\n",
    "# plt.ylim(top=1e-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    mwas[lambda x:\n",
    "        (x.species == \"102506\") &\n",
    "    True\n",
    "        & (x.num_subjects_present >= 10)\n",
    "        & (x.num_subjects_absent >= 10)\n",
    "    ].assign(\n",
    "        donor=donor_comparison.label,\n",
    "        species=gene_meta.species,\n",
    "        prevalence_class=prevalence_class,\n",
    "        nlength=gene_meta.nlength,\n",
    "        clust=co_clust.spgc_clust,\n",
    "    )\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1])\n",
    "    .sort_values('fisher_exact_pvalue_ibd')\n",
    ")\n",
    "\n",
    "bins = np.linspace(-5, 5)\n",
    "plt.hist(mwas.loc[idxwhere(donor_comparison.label == 'd97')][lambda x: x.fisher_exact_pvalue_ibd < 1e-2].log2_oddsratio_pc_ibd, bins=bins, alpha=0.5)\n",
    "plt.hist(mwas.loc[idxwhere(donor_comparison.label == 'd44')][lambda x: x.fisher_exact_pvalue_ibd < 1e-2].log2_oddsratio_pc_ibd, bins=bins, alpha=0.5)\n",
    "# plt.hist(mwas.loc[idxwhere(donor_comparison.label == 'both')][lambda x: x.fisher_exact_pvalue_ibd < 1e-2].log2_oddsratio_pc_ibd, bins=bins, alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_enrichment_contingency = mwas.assign(donor=donor_comparison.label, enriched_in_ibd=lambda x: x.log2_oddsratio_pc_ibd > 0).dropna(subset=['donor'])[lambda x: x.fisher_exact_pvalue_ibd < 1e-2][['donor', 'enriched_in_ibd']].value_counts().unstack()\n",
    "donor_enrichment_contingency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sp.stats.chi2_contingency(donor_enrichment_contingency)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d3 = d1[lambda x: (x[['G', 'L', 'M', 'R']].any(axis=1)) & (x.fisher_exact_pvalue_ibd < 1e-2)][['oddsratio_pc_ibd', 'fisher_exact_pvalue_ibd', 'fdr', 'donor', 'species']].join(gene_meta[['eggNOG_OGs', 'nlength', 'COG_category', 'Description', 'Preferred_name']]).sort_values('oddsratio_pc_ibd')\n",
    "d3.sort_values('fdr')#gene_x_cog_category_matrix.loc[d3.index][['G', 'L', 'M', 'R', 'no_category']].assign(cat=d3.COG_category, cog=d3.eggNOG_OGs, oddrat=d3.oddsratio_pc_ibd, species=d3.species, )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = mwas.assign(donor=donor_comparison.label).assign(prevalence_class=prevalence_class)[\n",
    "    lambda x: (x.num_subjects_present >= 10)\n",
    "    & (x.num_subjects_absent >= 10)\n",
    "    & x.donor.isin([\"d97\", \"d44\"])\n",
    "    & (x.prevalence_class == 'shell')\n",
    "].assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1], enriched_in_ibd=lambda x: x.log2_oddsratio_pc_ibd > 0, low_pvalue_hit=lambda x: x.fisher_exact_pvalue_ibd < 5e-3).sort_values(\n",
    "    \"fisher_exact_pvalue_ibd\"\n",
    ")\n",
    "d0[lambda x: x.low_pvalue_hit][['donor', 'enriched_in_ibd']].value_counts().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = mwas.assign(donor=donor_comparison.label).assign(prevalence_class=prevalence_class)[\n",
    "    lambda x: (x.num_subjects_present >= 10)\n",
    "    & (x.num_subjects_absent >= 10)\n",
    "    & x.donor.isin([\"d97\", \"d44\"])\n",
    "    & (x.prevalence_class == 'shell')\n",
    "].assign(fdr=lambda x: fdrcorrection(x.fisher_exact_pvalue_ibd)[1], enriched_in_ibd=lambda x: x.log2_oddsratio_pc_ibd > 0, low_pvalue_hit=lambda x: x.fisher_exact_pvalue_ibd < 1e-3).sort_values(\n",
    "    \"fisher_exact_pvalue_ibd\"\n",
    ")\n",
    "d0[lambda x: x.low_pvalue_hit][['donor', 'enriched_in_ibd']].value_counts().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist2d('log2_oddsratio_pc_ibd', 'neg_log10_pvalue', data=mwas, norm=mpl.colors.PowerNorm(1/10), bins=100)\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = mwas[\n",
    "    lambda x: (x.num_subjects_present > 20)\n",
    "    & (x.num_subjects_absent > 20)\n",
    "].sort_values(\"fisher_exact_pvalue_ibd\")\n",
    "d.assign(lambda x: x.value_counts()\n",
    "# fdrcorrection(d.fisher_exact_pvalue_ibd)[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fdr_by_species.assign(signif=lambda x: x.fdr < 0.05)[['species', 'signif']].value_counts().unstack(fill_value=0).sort_values(True, ascending=False).join(species_taxonomy).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas = mwas.assign(fdr=lambda x: x[test_filter].groupby('species').fisher_exact_pvalue_ibd.pipe()\n",
    "plt.hist2d('log2_oddsratio_pc_ibd', 'neg_log10_pvalue', data=mwas[test_filter], norm=mpl.colors.PowerNorm(1/10), bins=100)\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-5, 0, num=40)\n",
    "plt.hist(mwas[test_filter].fisher_exact_pvalue_ibd, bins=bins)\n",
    "plt.plot(bins[1:], (bins[1:] - bins[:-1]) * mwas[test_filter].shape[0])\n",
    "plt.xscale('log')\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(np.sort(fdrcorrection(mwas[test_filter].fisher_exact_pvalue_ibd)[1]))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_perm_coclustering = d.groupby('species', group_keys=False).apply(lambda x: x.assign(permute_clust=lambda y: y.spgc_clust.sample(frac=1).values)).drop(columns=['spgc_clust'])\n",
    "_perm_coclustering"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "perm_num_coclustering = []\n",
    "for i in tqdm(range(99)):\n",
    "    _perm_coclustering = d.groupby('species', group_keys=False).apply(lambda x: x.assign(permute_clust=lambda y: y.spgc_clust.sample(frac=1).values)).drop(columns=['spgc_clust']).groupby(['species', 'permute_clust']).sum().gt(1)\n",
    "    _perm_num_coclustering = _perm_coclustering.groupby('species').sum()\n",
    "    _perm_num_coclustering.loc['overall'] = _perm_coclustering.sum()\n",
    "    _perm_num_coclustering.loc[:, 'overall'] = _perm_coclustering.drop(columns=['no_category', 'S', 'R']).any(axis=1).groupby('species').sum()\n",
    "    _perm_num_coclustering.loc['overall', 'overall'] = _perm_coclustering.drop(columns=['no_category', 'S', 'R']).any(axis=1).sum()\n",
    "    perm_num_coclustering.append(_perm_num_coclustering.stack())\n",
    "perm_num_coclustering = pd.DataFrame(perm_num_coclustering)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def _calculate_2tailed_pvalue_from_perm(obs, perms, return_zscore=False):\n",
    "    hypoth_left = perms >= obs\n",
    "    hypoth_right = perms <= obs\n",
    "    null_p_left = (hypoth_left.sum() + 1) / (len(hypoth_left) + 1)\n",
    "    null_p_right = (hypoth_right.sum() + 1) / (len(hypoth_right) + 1)\n",
    "    if return_zscore:\n",
    "        null_mean = perms.mean()\n",
    "        null_std = perms.std()\n",
    "        zscore = (obs - null_mean) / null_std\n",
    "        return zscore, np.minimum(null_p_left, null_p_right)\n",
    "    return np.minimum(null_p_left, null_p_right)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_perm, _obs = align_indexes(perm_num_coclustering.T, obs_num_coclustering.stack())\n",
    "\n",
    "result = []\n",
    "for idx in _obs.index:\n",
    "    result.append(_calculate_2tailed_pvalue_from_perm(_obs.loc[idx], _perm.loc[idx], return_zscore=True))\n",
    "spgc_coclustering_enrichment_test = pd.DataFrame(result, index=_obs.index, columns=['zscore', 'pvalue'])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = spgc_coclustering_enrichment_test\n",
    "d1_zscore = d0.zscore.replace({np.inf: np.nan}).unstack(fill_value=0)\n",
    "d1_signf = d0.pvalue.unstack(fill_value=1.0).map(lambda p: np.where(p < 0.05, '\u2022', ''))\n",
    "_species_order = d1_zscore.loc[:, 'overall'].sort_values().index\n",
    "_category_order = d1_zscore.loc['overall'].sort_values().index\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 20))\n",
    "sns.heatmap(d1_zscore.loc[_species_order, _category_order], xticklabels=1, yticklabels=1, norm=mpl.colors.SymLogNorm(linthresh=1.0, vmin=-50, vmax=50), cmap='coolwarm', ax=ax)\n",
    "\n",
    "# Annotations (because seaborn annotations are failing)\n",
    "for (i, species), (j, category) in product(enumerate(_species_order), enumerate(_category_order)):\n",
    "    ax.annotate(d1_signf.loc[species, category], xy=(j + 0.5, i + 0.5), ha='center', va='center')\n",
    "\n",
    "# d0 = ref_species_by_cog_category_cdi_bias_test\n",
    "# d1_coef = d0.coef.unstack(fill_value=0)\n",
    "# d1_signf = d0.pvalue.unstack(fill_value=1.0).map(lambda p: np.where(p < 0.05, '\u2022', ''))\n",
    "\n",
    "# d2_coef = d1_coef.loc[cmi_species_order, cmi_category_order]\n",
    "# d2_signf = d1_signf.loc[cmi_species_order, cmi_category_order]\n",
    "\n",
    "# fig, ax = plt.subplots(figsize=(20, 20))\n",
    "# # sns.heatmap(d2_coef)\n",
    "# # sns.heatmap(d1_coef.loc[cmi_species_order, cmi_category_order], annot=d1_signf.loc[cmi_species_order, cmi_category_order], fmt='', norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-1e5, vmax=1e5), xticklabels=1, yticklabels=1, cmap='coolwarm', ax=ax)\n",
    "# sns.heatmap(d2_coef, norm=mpl.colors.SymLogNorm(linthresh=0.001, vmin=-.1, vmax=.1), xticklabels=1, yticklabels=1, cmap='coolwarm', ax=ax)\n",
    "\n",
    "# # Annotations (because seaborn annotations are failing)\n",
    "# for (i, species), (j, category) in product(enumerate(cmi_species_order), enumerate(cmi_category_order)):\n",
    "#     ax.annotate(d2_signf.loc[species, category], xy=(j + 0.5, i + 0.5), ha='center', va='center')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "spgc_coclustering_enrichment_test.loc['overall', 'overall']"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "obs_num_coclustering.loc['overall', 'overall'] / perm_num_coclustering.T.loc['overall', 'overall'].mean()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "(obs_num_coclustering.loc['overall', :] / perm_num_coclustering.T.loc['overall', :].mean(1)).sort_values(ascending=False).to_frame('excess_clumpiness').join(spgc_coclustering_enrichment_test.loc['overall', :]).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "frac_gt = (pd.DataFrame(perm_num_coclustering) > obs_num_coclustering.stack()).mean()\n",
    "frac_lt = (pd.DataFrame(perm_num_coclustering) < obs_num_coclustering.stack()).mean()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pvalue = np.minimum(frac_gt, frac_lt).unstack()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zscore = pd.DataFrame(perm_num_coclustering).agg(['mean', 'std']).T.assign(obs=obs_num_coclustering.stack()).apply(lambda x: (x['obs'] - x['mean']) / x['std'], axis=1).unstack().replace({np.inf: np.nan})"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "zscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz3",
   "language": "python",
   "name": "toolz3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}