{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mpl.scale.register_scale(mpl.scale.FuncScale)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_list = idxwhere(pd.read_table(\"meta/genome_group.tsv\", index_col=\"genome_id\").genome_group_id == \"xjin\")\n",
    "len(genome_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = pd.read_table(\"meta/genome.tsv\", index_col=\"genome_id\").loc[genome_list]\n",
    "assert genome.index.is_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = list(\n",
    "    genome.loc[genome_list][\n",
    "        lambda x: (x.species_id != \"TODO\") & (x.genome_path != \"\")\n",
    "    ].species_id.unique()\n",
    ")\n",
    "len(species_list)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "xjin_sample_list = pd.read_table(\n",
    "    \"meta/XJIN_BENCHMARK/mgen.tsv\", index_col=\"mgen_id\"\n",
    ").index.to_list()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_depth = pd.read_table(depth_path, names=['sample_id', 'depth'], index_col='sample_id').depth.loc[xjin_sample_list]\n",
    "_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_group = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[\n",
    "        lambda x: x.species_group_id == \"xjin\"\n",
    "    ]\n",
    "    .species_id.astype(str)\n",
    "    .to_list()\n",
    ")\n",
    "species_group[:5]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "with open('test.args', 'w') as f:\n",
    "    for genome_id, d in genome.loc[genome_list].iterrows():\n",
    "        species = d.species_id\n",
    "        if species in species_group:\n",
    "            print(f\"data/group/XJIN_BENCHMARK/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.{genome_id}.geno_matching_stats.tsv\", file=f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome[lambda x: x.species_id == '100003']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.read_table(strain_meta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data/group/xjin/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.STRAIN_MATCH_BENCHMARK_GRID.flag\n",
    "\n",
    "strain_match = {}\n",
    "for genome_id, d in genome.loc[genome_list].iterrows():\n",
    "    species = d.species_id\n",
    "    if species == \"TODO\":\n",
    "        continue\n",
    "    strain_match_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.{genome_id}.geno_matching_stats.tsv\"\n",
    "    strain_meta_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta-s90-d100-a1-pos100.tsv\"\n",
    "    if os.path.exists(strain_match_path):\n",
    "        strain_meta = pd.read_table(strain_meta_path, index_col='genome_id').rename_axis('strain')\n",
    "        strain_match[genome_id] = pd.read_table(\n",
    "            strain_match_path, index_col=[\"genome_id\", \"strain\"]\n",
    "        ).assign(\n",
    "            genotype_matching_positions=lambda x: (1 - x.genotype_dissimilarity)\n",
    "            * x.num_geno_positions_compared,\n",
    "            genotype_dissimilarity_pc=lambda x: x.genotype_dissimilarity\n",
    "            + (1 / x.num_geno_positions_compared),\n",
    "            strain_match_path=strain_match_path,\n",
    "        ).join(strain_meta, on='strain')\n",
    "strain_match = (\n",
    "    pd.concat(strain_match.values()).reset_index().set_index([\"genome_id\", \"strain\"])\n",
    ")\n",
    "strain_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: We match strains only from among those with \"strain-pure-samples\" in the xjin set\n",
    "# and then picking the lowest genotype_dissimilarity.\n",
    "# TODO: Consider also allowing strains to have strain pure samples\n",
    "# outside the xjin set, but also be detected at appreciable relative abundances\n",
    "# inside the xjin samples?\n",
    "genome_to_spgc_strain = (\n",
    "    strain_match[lambda x: x.strain_depth_max > 0]\n",
    "    .reset_index()\n",
    "    .sort_values(\n",
    "        [\"genotype_dissimilarity_pc\", \"strain_depth_sum\"], ascending=(True, False)\n",
    "    )\n",
    "    .groupby(\"genome_id\")\n",
    "    .head(1)\n",
    "    .set_index(\"genome_id\")\n",
    "    .reindex(genome.index)\n",
    "    .fillna(\n",
    "        {\n",
    "            \"strain\": -1,\n",
    "            \"genotype_dissimilarity\": 1.0,\n",
    "            \"strain_depth_sum\": 0,\n",
    "            \"strain_depth_max\": 0,\n",
    "        }\n",
    "    )\n",
    "    .assign(strain=lambda x: x.strain.astype(int))\n",
    ")\n",
    "\n",
    "plt.hist(\n",
    "    genome_to_spgc_strain.genotype_dissimilarity_pc, bins=[0] + list(np.logspace(-5, 0))\n",
    ")\n",
    "plt.xscale(\"symlog\", linthresh=1e-5, linscale=0.1)\n",
    "genome_to_spgc_strain.sort_values(\"genotype_matching_positions\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity_pc\",\n",
    "    data=genome_to_spgc_strain,\n",
    "    alpha=0.4,\n",
    ")\n",
    "\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# data/group/xjin/r.proc.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv.SELECT_SPECIES.flag\n",
    "# data/group/xjin/r.proc.gene99_v15-v22-agg75.ACCURACY_BENCHMARK_GRID.flag\n",
    "# data/group/xjin/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta-hmp2-s90-d100-a1-pos100.tsv.SELECT_SPECIES.flag\n",
    "\n",
    "\n",
    "benchmark = {}\n",
    "missing = []\n",
    "unmatched = []\n",
    "depth_meta = {}\n",
    "spgc_qc = {}\n",
    "for genome_id, d in genome.iterrows():\n",
    "    species = d.species_id\n",
    "    if species == \"TODO\":\n",
    "        continue\n",
    "    depth_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv\"\n",
    "    _depth = pd.read_table(\n",
    "        depth_path, names=[\"sample_id\", \"depth\"], index_col=\"sample_id\"\n",
    "    ).depth\n",
    "    depth_meta[species] = pd.Series(\n",
    "        dict(species_depth_max=_depth.max(), species_depth_sum=_depth.sum())\n",
    "    )\n",
    "\n",
    "    # SPGC\n",
    "    for unit in [\"uhggtop\", \"eggnog\", \"cog\"]:\n",
    "        for tool in [\n",
    "            \"spgc-fit\",\n",
    "            # \"spgc2-fit\",\n",
    "            # \"nnmatched-m50\",\n",
    "            # \"nnmatched-m10\",\n",
    "            # \"nnmatched-m1\",\n",
    "            # \"nnmatched-m0\",\n",
    "            \"spgc-depth200\",\n",
    "        ]:\n",
    "            accuracy_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.{tool}.{genome_id}.{unit}-reconstruction_accuracy.tsv\"\n",
    "            if not os.path.exists(accuracy_path):\n",
    "                missing.append(accuracy_path)\n",
    "                continue\n",
    "            matched_strain = genome_to_spgc_strain.strain[genome_id]\n",
    "            data = (\n",
    "                pd.read_table(accuracy_path, index_col=\"strain\")\n",
    "                .assign(species=species, accuracy_path=accuracy_path, strain=lambda x: x.index)\n",
    "                .sort_values(\"f1\", ascending=False)\n",
    "            )\n",
    "\n",
    "            if matched_strain in data.index:\n",
    "                benchmark[(genome_id, tool, unit, \"match\")] = data.loc[matched_strain]\n",
    "            if data.shape[0] >= 1:\n",
    "                benchmark[(genome_id, tool, unit, \"top\")] = data.iloc[0]\n",
    "            if data.shape[0] >= 2:\n",
    "                benchmark[(genome_id, tool, unit, \"second\")] = data.iloc[1]\n",
    "        for tool in [\n",
    "            \"panphlan\",\n",
    "            # \"spanda-s2\",\n",
    "            # \"spanda-s3\",\n",
    "            \"spanda-s4\",\n",
    "        ]:\n",
    "            accuracy_path = f\"data/group/xjin/species/sp-{species}/r.proc.gene99_v15-v22-agg75.{tool}.{genome_id}.{unit}-reconstruction_accuracy.tsv\"\n",
    "            if not os.path.exists(accuracy_path):\n",
    "                missing.append(accuracy_path)\n",
    "                continue\n",
    "            data = (\n",
    "                pd.read_table(accuracy_path, index_col=\"strain\")\n",
    "                .assign(species=species, accuracy_path=accuracy_path, strain=lambda x: x.index)\n",
    "                .sort_values(\"f1\", ascending=False)\n",
    "            )\n",
    "            if data.shape[0] >= 1:\n",
    "                benchmark[(genome_id, tool, unit, \"top\")] = data.iloc[0]\n",
    "                benchmark[(genome_id, tool, unit, \"match\")] = data.iloc[0]\n",
    "            if data.shape[0] >= 2:\n",
    "                benchmark[(genome_id, tool, unit, \"second\")] = data.iloc[1]\n",
    "            if data.shape[0] >= 3:\n",
    "                benchmark[(genome_id, tool, unit, \"third\")] = data.iloc[2]\n",
    "\n",
    "    spgc_qc_path = f\"data/group/xjin/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_v15-v22-agg75.spgc-fit.strain_meta-hmp2-s90-d100-a1-pos100.tsv\"\n",
    "    if not os.path.exists(spgc_qc_path):\n",
    "        continue\n",
    "    _qc = pd.read_table(spgc_qc_path, index_col=\"genome_id\")\n",
    "    if matched_strain in _qc.index:\n",
    "        spgc_qc[genome_id] = _qc.loc[matched_strain]\n",
    "\n",
    "benchmark = pd.DataFrame(benchmark.values(), index=benchmark.keys()).rename_axis(\n",
    "    [\"genome_id\", \"tool\", \"unit\", \"match\"]\n",
    ")\n",
    "depth_meta = pd.DataFrame(depth_meta.values(), index=depth_meta.keys()).rename_axis(\n",
    "    \"species\"\n",
    ")\n",
    "spgc_qc = pd.DataFrame(spgc_qc.values(), index=spgc_qc.keys())\n",
    "\n",
    "benchmark.sort_values([\"species\", \"tool\"])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs('panphlan', level='tool').accuracy_path.iloc[0]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs(('panphlan', 'eggnog', 'match'), level=('tool', 'unit', 'match')).sort_values('f1').accuracy_path.values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=50)\n",
    "d = benchmark.xs((\"eggnog\", \"top\"), level=(\"unit\", \"match\")).f1.unstack().fillna(0)\n",
    "for tool in [\n",
    "    \"spgc-fit\",\n",
    "    # \"nnmatched-m0\", \"nnmatched-m50\",\n",
    "    \"spgc-depth200\",\n",
    "]:\n",
    "    plt.hist(d[tool], bins=bins, alpha=0.5, label=tool)\n",
    "# plt.hist(d[\"nnmatched-m0\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"nnmatched-m50\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spgc-depth200\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"panphlan\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s2\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s3\"], bins=bins, alpha=0.5)\n",
    "# plt.hist(d[\"spanda-s4\"], bins=bins, alpha=0.5)\n",
    "plt.legend()\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    ").fillna(0)#.loc[genome_list]\n",
    "\n",
    "plt.scatter(\n",
    "    \"spanda-s4\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark.xs((\"uhggtop\", \"match\", \"spanda-s4\"), level=(\"unit\", \"match\", \"tool\")).sort_values('f1', ascending=False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.xs(('eggnog', 'spgc-fit', 'match'), level=('unit', 'tool', 'match')).accuracy_path.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"spgc-fit\", \"eggnog\"), level=(\"tool\", \"unit\")).f1.unstack()\n",
    ").fillna(0)\n",
    "\n",
    "\n",
    "plt.scatter(\n",
    "    \"match\",\n",
    "    \"top\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    # c=\"strain_depth_max\",\n",
    ")\n",
    "# plt.colorbar()\n",
    "# # plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = genome_to_spgc_strain.join(\n",
    "    benchmark.xs((\"match\", \"eggnog\"), level=(\"match\", \"unit\")).f1.unstack()\n",
    ").fillna(0)\n",
    "\n",
    "\n",
    "plt.scatter(\n",
    "    \"spgc-fit\",\n",
    "    \"spgc2-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    # c=\"strain_depth_max\",\n",
    ")\n",
    "# plt.colorbar()\n",
    "# # plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter('precision', 'recall', data=benchmark.xs(('spgc-fit', 'eggnog', 'match'), level=('tool', 'unit', 'match')))\n",
    "plt.plot([0, 1], [0, 1])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0[lambda x: x['spgc-fit'] != x['spgc2-fit']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Which genomes are failing and why?\n",
    "(\n",
    "    genome_to_spgc_strain.join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack())\n",
    "    .fillna(0)[lambda x: x[\"spgc-fit\"] == 0]\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    .join(\n",
    "        genome.species_id.value_counts().rename(\"num_strains_in_species\"),\n",
    "        on=\"species_id\",\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What's the point of benchmarking strains that have no or very low depth?\n",
    "# that just even more heavily favors the tools that I'm taking \"best hit\" for.\n",
    "d0 = (\n",
    "    genome_to_spgc_strain.join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack())\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    [lambda x: x.species_depth_sum >= 0.5]\n",
    ")\n",
    "print(len(d0))\n",
    "\n",
    "plt.scatter(\n",
    "    \"panphlan\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    genome_to_spgc_strain.join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack())\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # .join(depth_meta, on=\"species_id\")\n",
    "    [lambda x: x.species_depth_sum >= 0.5]\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"panphlan\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.2),\n",
    "    c=\"strain_depth_max\",\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.ylim(0.55, 1.05)\n",
    "plt.xlim(0.55, 1.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1\n",
    "    .unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"genotype_dissimilarity\",\n",
    "    \"spgc-fit\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"strain_depth_max\",\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.1),\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.xscale(\"symlog\", linthresh=1e-5, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"spgc-fit\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack()\n",
    "    .fillna(0)\n",
    "    .join(\n",
    "        genome_to_spgc_strain[\n",
    "            [\n",
    "                \"genotype_dissimilarity\",\n",
    "                \"strain_depth_sum\",\n",
    "                \"strain_depth_max\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .fillna({\"genotype_dissimilarity\": 1.0})\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    \"strain_depth_sum\",\n",
    "    \"genotype_dissimilarity\",\n",
    "    data=d0,\n",
    "    alpha=0.4,\n",
    "    c=\"spgc-fit\",\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "plt.colorbar()\n",
    "# plt.scatter(d1['spanda-s2'], d1['spgc-fit'])\n",
    "# plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-4)\n",
    "plt.xscale(\"symlog\", linthresh=1e-2, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "multi_genome_species = idxwhere(\n",
    "    strain_match.join(genome.species_id)\n",
    "    .reset_index()[[\"species_id\", \"genome_id\"]]\n",
    "    .drop_duplicates()\n",
    "    .species_id.value_counts()\n",
    "    > 1\n",
    ")\n",
    "multi_genome_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(benchmark.xs((\"spgc-fit\", \"eggnog\", \"match\"), level=(\"tool\", \"unit\", \"match\")), rsuffix=\"_\")\n",
    "    .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    ")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")), rsuffix=\"_\")\n",
    "    # .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    "    [['precision', 'recall', 'f1', 'jaccard']]\n",
    "    .unstack('tool')\n",
    "    .xs('spgc-fit', level='tool', axis='columns')\n",
    "    .fillna(0)\n",
    "    .join(genome_to_spgc_strain.species_gene_frac)\n",
    ")\n",
    "\n",
    "plt.scatter('precision', \"recall\", data=d, c='species_gene_frac')\n",
    "plt.colorbar()\n",
    "plt.plot([0, 1], [0, 1])\n",
    "print(d[[\"precision\", \"recall\", \"f1\", \"jaccard\"]].quantile([0.25, 0.5, 0.75]))\n",
    "print(d[d.species_gene_frac > 0.9][[\"precision\", \"recall\", \"f1\", \"jaccard\"]].quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")), rsuffix=\"_\")\n",
    "    # .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    "    .f1.unstack(\"tool\")\n",
    "    .fillna(0)\n",
    "    .join(genome_to_spgc_strain.species_gene_frac)\n",
    ")\n",
    "\n",
    "d_filt = d[d.species_gene_frac > 0.9]\n",
    "\n",
    "print((-(d.subtract(d[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))\n",
    "print((-(d_filt.subtract(d_filt[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")), rsuffix=\"_\")\n",
    "    # .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    "    .precision.unstack(\"tool\")\n",
    "    .fillna(0)\n",
    "    .join(genome_to_spgc_strain.species_gene_frac)\n",
    ")\n",
    "\n",
    "d_filt = d[d.species_gene_frac > 0.9]\n",
    "\n",
    "print((-(d.subtract(d[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))\n",
    "print((-(d_filt.subtract(d_filt[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    genome.join(genome_to_spgc_strain)[\n",
    "        lambda x: (x.species_id != \"TODO\")\n",
    "        & (x.species_depth_sum > 0)\n",
    "        & (x.species_depth_max > 0)\n",
    "        # & ~x.species_id.isin(multi_genome_species)\n",
    "    ]\n",
    "    .join(benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")), rsuffix=\"_\")\n",
    "    # .fillna({\"precision\": 0, \"recall\": 0, \"f1\": 0, \"jaccard\": 0})\n",
    "    .recall.unstack(\"tool\")\n",
    "    .fillna(0)\n",
    "    .join(genome_to_spgc_strain.species_gene_frac)\n",
    ")\n",
    "\n",
    "d_filt = d[d.species_gene_frac > 0.9]\n",
    "\n",
    "print((-(d.subtract(d[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))\n",
    "print((-(d_filt.subtract(d_filt[\"spgc-fit\"], axis=0))).quantile([0.25, 0.5, 0.75]))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import statsmodels.formula.api as smf\n",
    "\n",
    "for tool in ['panphlan', 'spanda-s2', 'spgc-fit']:\n",
    "    d = genome_accuracy_table.assign(x=lambda x: x['max_depth'], y=lambda x: x[tool]).fillna(0)[lambda x: x.max_depth > 0]\n",
    "    fit = smf.ols('y ~ cr(x, 5)', data=d).fit()\n",
    "    d = d.assign(predict_y=lambda x: fit.predict(d))\n",
    "    plt.scatter('max_depth', tool, data=d, label=tool, alpha=0.2)\n",
    "    plt.plot('max_depth', 'predict_y', data=d.sort_values('max_depth'), label='__nolegend__')\n",
    "plt.xscale('symlog', linthresh=1e-2, linscale=0.1)\n",
    "plt.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "excluded_species = [\"100878\", \"100022\"]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "considered_species = list(\n",
    "    # set(species_list) - set(multi_genome_species) - set(excluded_species)\n",
    "    set(species_list) - set(excluded_species)\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark[lambda x: x.species.isin(considered_species)][\n",
    "    [\"species\", \"tool\"]\n",
    "].value_counts().sort_values(ascending=False)\n",
    "# benchmark[lambda x: x.species == '100196']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum > 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda-s4\"]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order) + 0.5, 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc2-fit\"\n",
    "nbins = 20\n",
    "left_bound = 0.0\n",
    "bins = [0] + list(np.linspace(left_bound, 1, num=nbins + 1)[1:])\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\")\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        *_, cbar_artist = ax.hist2d(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            bins=bins,\n",
    "            cmin=1,\n",
    "            norm=mpl.colors.PowerNorm(1 / 1, vmin=0, vmax=21),\n",
    "            cmap=\"magma_r\",\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "ax.set_xlim(left_bound, 1)\n",
    "ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "\n",
    "fig.subplots_adjust(right=0.8)\n",
    "cbar_ax = fig.add_axes([0.85, 0.15, 0.025, 0.67])\n",
    "fig.colorbar(\n",
    "    cbar_artist, cax=cbar_ax, ticks=[1, 10, 20, 30, 40, 50], label=\"count strains\"\n",
    ")\n",
    "# fig.tight_layout(rect=(0, 0, 0.85, 0.67))\n",
    "# ax.set_xlabel(x)\n",
    "# ax.set_ylabel(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda-s4\", \"spgc-depth200\"]\n",
    "\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = (d1.xs(score, level=\"score\", axis=\"columns\").join(\n",
    "        genome.join(genome_to_spgc_strain).join(\n",
    "            depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "        )\n",
    "    )\n",
    "          # .join(spgc_qc.species_gene_frac)\n",
    "         )\n",
    "\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=5,\n",
    "            c=\"species_gene_frac\",\n",
    "            norm=mpl.colors.PowerNorm(1/2),\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.01, 1.01)\n",
    "\n",
    "# ax.set_xscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_xticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_xlim(-0.01, 1.01)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "    lib.plot.rotate_xticklabels(ax=ax_col[-1], rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    # \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = (d1.xs(score, level=\"score\", axis=\"columns\").join(\n",
    "        genome.join(genome_to_spgc_strain).join(\n",
    "            depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "        )\n",
    "    )\n",
    "          # .join(genome_to_spgc_strain.species_gene_frac)\n",
    "          .sort_values('species_gene_frac')\n",
    "         )\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=10,\n",
    "            c=\"species_gene_frac\",\n",
    "            norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper',\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        # ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        # ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.8, 0.9, 0.95, 0.99]))\n",
    "ax.set_yticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_ylim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "ax.set_xticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_xlim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "    lib.plot.rotate_xticklabels(ax=ax_col[-1], rotation=45)\n",
    "\n",
    "fig = plt.figure(figsize=(5, 2))\n",
    "plt.scatter([], [], c=[], norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper')\n",
    "plt.colorbar(ticks=[0.5, 0.6, 0.7, 0.8, 0.9, 1.0], orientation='horizontal')\n",
    "ax = plt.gca()\n",
    "ax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"uhggtop\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda-s4\", \"spgc-depth200\"]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = (d1.xs(score, level=\"score\", axis=\"columns\").join(\n",
    "        genome.join(genome_to_spgc_strain).join(\n",
    "            depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "        )\n",
    "    )\n",
    "          # .join(spgc_qc.species_gene_frac)\n",
    "          # .sort_values('species_gene_frac')\n",
    "         )\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=10,\n",
    "            # c=\"species_gene_frac\",\n",
    "            norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper',\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        # ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        # ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.8, 0.9, 0.95, 0.99]))\n",
    "ax.set_yticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_ylim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "ax.set_xticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_xlim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "    lib.plot.rotate_xticklabels(ax=ax_col[-1], rotation=45)\n",
    "\n",
    "fig = plt.figure(figsize=(5, 2))\n",
    "plt.scatter([], [], c=[], norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper')\n",
    "plt.colorbar(ticks=[0.5, 0.6, 0.7, 0.8, 0.9, 1.0], orientation='horizontal')\n",
    "ax = plt.gca()\n",
    "ax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"cog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda-s4\", \"spgc-depth200\"]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = (d1.xs(score, level=\"score\", axis=\"columns\").join(\n",
    "        genome.join(genome_to_spgc_strain).join(\n",
    "            depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "        )\n",
    "    )\n",
    "          # .join(spgc_qc.species_gene_frac)\n",
    "          # .sort_values('species_gene_frac')\n",
    "         )\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=10,\n",
    "            # c=\"species_gene_frac\",\n",
    "            norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper',\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        # ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        # ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.8, 0.9, 0.95, 0.99]))\n",
    "ax.set_yticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_ylim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "ax.set_xticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "ax.set_xlim(-0.0, 1.0)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "    lib.plot.rotate_xticklabels(ax=ax_col[-1], rotation=45)\n",
    "\n",
    "fig = plt.figure(figsize=(5, 2))\n",
    "plt.scatter([], [], c=[], norm=mpl.colors.PowerNorm(3, vmin=0, vmax=1),\n",
    "            cmap='copper')\n",
    "plt.colorbar(ticks=[0.5, 0.6, 0.7, 0.8, 0.9, 1.0], orientation='horizontal')\n",
    "ax = plt.gca()\n",
    "ax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\", \"jaccard\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "d2 = d1.xs('spgc-fit', level=\"tool\", axis=\"columns\").join(\n",
    "    genome.join(genome_to_spgc_strain).join(\n",
    "        depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "    )\n",
    ").join(spgc_qc.species_gene_frac)[lambda x: x.strain != -1].sort_values('species_gene_frac')\n",
    "\n",
    "for score in ['precision', 'recall', 'f1']:\n",
    "    sns.regplot(x='species_gene_frac', y=score, data=d2)\n",
    "    print(sp.stats.pearsonr(d2['species_gene_frac'], d2[score]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"cog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum >= 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda-s4\", \"spgc-depth200\"]\n",
    "\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\").join(\n",
    "        genome.join(genome_to_spgc_strain).join(\n",
    "            depth_meta, on=\"species_id\", rsuffix=\"_\"\n",
    "        )\n",
    "    )\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=5,\n",
    "            c=\"strain_depth_max\",\n",
    "            norm=mpl.colors.PowerNorm(1/2),\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.1, 1.00)\n",
    "\n",
    "# ax.set_xscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "ax.set_xticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_xlim(-0.1, 1.00)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "    lib.plot.rotate_xticklabels(ax=ax_col[-1], rotation=45)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark#.iloc[0].accuracy_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum > 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    # \"nnmatched-m50\",\n",
    "    # \"nnmatched-m10\",\n",
    "    # \"nnmatched-m1\",\n",
    "    # \"nnmatched-m0\",\n",
    "    \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order) + 0.5, 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "nbins = 20\n",
    "left_bound = 0.0\n",
    "bins = [0] + list(np.linspace(left_bound, 1, num=nbins + 1)[1:])\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\")\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        *_, cbar_artist = ax.hist2d(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            bins=bins,\n",
    "            cmin=1,\n",
    "            norm=mpl.colors.PowerNorm(1 / 1, vmin=0, vmax=21),\n",
    "            cmap=\"magma_r\",\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "ax.set_xlim(left_bound, 1)\n",
    "ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "\n",
    "fig.subplots_adjust(right=0.8)\n",
    "cbar_ax = fig.add_axes([0.85, 0.15, 0.025, 0.67])\n",
    "fig.colorbar(\n",
    "    cbar_artist, cax=cbar_ax, ticks=[1, 10, 20, 30, 40, 50], label=\"count strains\"\n",
    ")\n",
    "# fig.tight_layout(rect=(0, 0, 0.85, 0.67))\n",
    "# ax.set_xlabel(x)\n",
    "# ax.set_ylabel(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_depth_sum > 0.1\n",
    "    ][[\"f1\", \"precision\", \"recall\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "d1.f1[[\"spgc-fit\", \"spgc-depth200\"]].sort_values(\"spgc-fit\").join(genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark.xs(('eggnog', 'match', 'spgc-fit'), level=('unit', 'match', 'tool')).sort_values('f1', ascending=False).join(genome.join(depth_meta, on=\"species_id\")).loc[lambda x: x.species_id.isin(multi_genome_species)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark.xs(('eggnog', 'match', 'Veillonella-sp-3-1-44_MAF-2'), level=('unit', 'match', 'genome_id')).sort_values('f1', ascending=False).head(20).join(genome.join(depth_meta, on=\"species_id\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark.xs(('eggnog', 'match', 'Veillonella-sp-6-1-27_MAF-2'), level=('unit', 'match', 'genome_id')).sort_values('f1', ascending=False).head(20).join(genome.join(depth_meta, on=\"species_id\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf.plot.plot_community(w1.drop_low_abundance_strains(0.05).community)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sfacts as sf\n",
    "\n",
    "w0 = sf.data.World.load('data/group/xjin/species/sp-100003/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc')\n",
    "w1 = sf.data.World.load('data/group/xjin/species/sp-100003/r.proc.gtpro.filt-poly00-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc')\n",
    "\n",
    "print(w0.sizes, w1.sizes)\n",
    "# sf.plot.plot_metagenotype(w0)\n",
    "sf.plot.plot_metagenotype(w1.sel(sample=list(set(w1.sample.values) & set(w0.sample.values))).random_sample(position=634))\n",
    "sf.plot.plot_community(w1.sel(sample=list(set(w1.sample.values) & set(w0.sample.values))).random_sample(position=634))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for species_id in ['100035', '100214', '101349', '102544']:\n",
    "    d = xr.load_dataarray(f'include/StrainPGC/example_data/{species_id}.gene_depth.nc')\n",
    "    d.to_pandas().T.rename_axis(index='gene').to_csv(f'include/StrainPGC/example_data/{species_id}.gene_depth.tsv.gz', sep='\\t')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genome_to_spgc_strain.loc[\"Bacteroides-finegoldii-DSM-17565_MAF-2\"]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "benchmark.loc[\n",
    "    (\"Bacteroides-finegoldii-DSM-17565_MAF-2\", \"spgc-fit\", \"eggnog\")\n",
    "].accuracy_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "_genome_list = idxwhere(genome_to_spgc_strain.species_depth_sum > 0)\n",
    "\n",
    "_tool_list = [\n",
    "    \"spgc-fit\",\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    \"spgc-depth200\",\n",
    "    # \"nnmatched-m50\"\n",
    "]\n",
    "_score_list = [\"precision\", \"recall\", \"f1\"]\n",
    "xvar = \"strain_depth_max\"\n",
    "\n",
    "# _tool_palette = lib.plot.construct_ordered_palette(_tool_list)\n",
    "_tool_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    _tool_list, colors=[\"tab:blue\", \"tab:green\", \"tab:orange\", \"tab:purple\"]\n",
    ")\n",
    "\n",
    "\n",
    "d0 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))[_score_list]\n",
    "    .reindex(product(_genome_list, _tool_list))\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # [lambda x: ~x.species_id.isin(multi_genome_species)]\n",
    "    .join(depth_meta, on=\"species_id\")\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_list),\n",
    "    figsize=(5, 3 * len(_score_list)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "for _tool in _tool_list:\n",
    "    d1 = d0.xs(_tool, level=\"tool\")\n",
    "    for _score, ax in zip(_score_list, axs.flatten()):\n",
    "        fit = smf.ols(f'{_score} ~ cr({xvar}, 5, constraints=\"center\")', data=d1).fit()\n",
    "        d2 = d1.assign(\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d[_score], d[xvar], it=10, frac=1 / 3, return_sorted=False\n",
    "            ),\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            xvar_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[xvar]\n",
    "            .mean(),\n",
    "            score_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[_score]\n",
    "            .mean(),\n",
    "        )\n",
    "        smoothed = pd.DataFrame({xvar: np.logspace(-0.5, 4, num=100)}).assign(\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d2[_score],\n",
    "                d2[xvar],\n",
    "                xvals=d[xvar],\n",
    "                it=10,\n",
    "                frac=1 / 3,\n",
    "                return_sorted=False,\n",
    "            ),\n",
    "        )\n",
    "        ax.scatter(\n",
    "            xvar,\n",
    "            _score,\n",
    "            data=d2,\n",
    "            label=_tool,\n",
    "            s=8,\n",
    "            alpha=0.5,\n",
    "            facecolor=\"none\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "        ax.plot(\n",
    "            \"xvar_rolling_average\",\n",
    "            \"score_rolling_average\",\n",
    "            data=d2.sort_values(xvar),\n",
    "            label=\"__nolegend__\",\n",
    "            lw=1,\n",
    "            linestyle=\"-\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "\n",
    "for _score, ax in zip(_score_list, axs.flatten()):\n",
    "    ax.set_ylabel(_score)\n",
    "\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-1, linscale=0.1)\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"uhggtop\"\n",
    "match = \"match\"\n",
    "\n",
    "_genome_list = idxwhere(genome_to_spgc_strain.species_depth_sum > 0)\n",
    "\n",
    "_tool_list = [\n",
    "    \"spgc-fit\",\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    \"spgc-depth200\",\n",
    "    # \"nnmatched-m50\"\n",
    "]\n",
    "_score_list = [\"precision\", \"recall\", \"f1\"]\n",
    "xvar = \"strain_depth_max\"\n",
    "\n",
    "# _tool_palette = lib.plot.construct_ordered_palette(_tool_list)\n",
    "_tool_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    _tool_list, colors=[\"tab:blue\", \"tab:green\", \"tab:orange\", \"tab:purple\"]\n",
    ")\n",
    "\n",
    "\n",
    "d0 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))[_score_list]\n",
    "    .reindex(product(_genome_list, _tool_list))\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # [lambda x: ~x.species_id.isin(multi_genome_species)]\n",
    "    .join(depth_meta, on=\"species_id\")\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_list),\n",
    "    figsize=(5, 3 * len(_score_list)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "for _tool in _tool_list:\n",
    "    d1 = d0.xs(_tool, level=\"tool\")\n",
    "    for _score, ax in zip(_score_list, axs.flatten()):\n",
    "        fit = smf.ols(f'{_score} ~ cr({xvar}, 5, constraints=\"center\")', data=d1).fit()\n",
    "        d2 = d1.assign(\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d[_score], d[xvar], it=10, frac=1 / 3, return_sorted=False\n",
    "            ),\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            xvar_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[xvar]\n",
    "            .mean(),\n",
    "            score_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[_score]\n",
    "            .mean(),\n",
    "        )\n",
    "        smoothed = pd.DataFrame({xvar: np.logspace(-0.5, 4, num=100)}).assign(\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d2[_score],\n",
    "                d2[xvar],\n",
    "                xvals=d[xvar],\n",
    "                it=10,\n",
    "                frac=1 / 3,\n",
    "                return_sorted=False,\n",
    "            ),\n",
    "        )\n",
    "        ax.scatter(\n",
    "            xvar,\n",
    "            _score,\n",
    "            data=d2,\n",
    "            label=_tool,\n",
    "            s=8,\n",
    "            alpha=0.5,\n",
    "            facecolor=\"none\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "        ax.plot(\n",
    "            \"xvar_rolling_average\",\n",
    "            \"score_rolling_average\",\n",
    "            data=d2.sort_values(xvar),\n",
    "            label=\"__nolegend__\",\n",
    "            lw=1,\n",
    "            linestyle=\"-\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "\n",
    "for _score, ax in zip(_score_list, axs.flatten()):\n",
    "    ax.set_ylabel(_score)\n",
    "\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-1, linscale=0.1)\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"cog\"\n",
    "match = \"match\"\n",
    "\n",
    "_genome_list = idxwhere(genome_to_spgc_strain.species_depth_sum > 0)\n",
    "\n",
    "_tool_list = [\n",
    "    \"spgc-fit\",\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    \"spgc-depth200\",\n",
    "    # \"nnmatched-m50\"\n",
    "]\n",
    "_score_list = [\"precision\", \"recall\", \"f1\"]\n",
    "xvar = \"strain_depth_max\"\n",
    "\n",
    "# _tool_palette = lib.plot.construct_ordered_palette(_tool_list)\n",
    "_tool_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    _tool_list, colors=[\"tab:blue\", \"tab:green\", \"tab:orange\", \"tab:purple\"]\n",
    ")\n",
    "\n",
    "\n",
    "d0 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))[_score_list]\n",
    "    .reindex(product(_genome_list, _tool_list))\n",
    "    .fillna(0)\n",
    "    .join(genome)\n",
    "    # [lambda x: ~x.species_id.isin(multi_genome_species)]\n",
    "    .join(depth_meta, on=\"species_id\")\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_list),\n",
    "    figsize=(5, 3 * len(_score_list)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "for _tool in _tool_list:\n",
    "    d1 = d0.xs(_tool, level=\"tool\")\n",
    "    for _score, ax in zip(_score_list, axs.flatten()):\n",
    "        fit = smf.ols(f'{_score} ~ cr({xvar}, 5, constraints=\"center\")', data=d1).fit()\n",
    "        d2 = d1.assign(\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d[_score], d[xvar], it=10, frac=1 / 3, return_sorted=False\n",
    "            ),\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            xvar_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[xvar]\n",
    "            .mean(),\n",
    "            score_rolling_average=lambda d: d.sort_values(xvar)\n",
    "            .rolling(window=10)[_score]\n",
    "            .mean(),\n",
    "        )\n",
    "        smoothed = pd.DataFrame({xvar: np.logspace(-0.5, 4, num=100)}).assign(\n",
    "            spline_y=lambda d: fit.predict(pd.DataFrame({xvar: d[xvar]})),\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d2[_score],\n",
    "                d2[xvar],\n",
    "                xvals=d[xvar],\n",
    "                it=10,\n",
    "                frac=1 / 3,\n",
    "                return_sorted=False,\n",
    "            ),\n",
    "        )\n",
    "        ax.scatter(\n",
    "            xvar,\n",
    "            _score,\n",
    "            data=d2,\n",
    "            label=_tool,\n",
    "            s=8,\n",
    "            alpha=0.5,\n",
    "            facecolor=\"none\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "        ax.plot(\n",
    "            \"xvar_rolling_average\",\n",
    "            \"score_rolling_average\",\n",
    "            data=d2.sort_values(xvar),\n",
    "            label=\"__nolegend__\",\n",
    "            lw=1,\n",
    "            linestyle=\"-\",\n",
    "            color=_tool_palette[_tool],\n",
    "        )\n",
    "\n",
    "for _score, ax in zip(_score_list, axs.flatten()):\n",
    "    ax.set_ylabel(_score)\n",
    "\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-1, linscale=0.1)\n",
    "\n",
    "pad = 1e-2\n",
    "# ax.set_yscale(\n",
    "#     \"function\",\n",
    "#     functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    "# )\n",
    "# ax.set_yticks(np.unique([0.0, 0.5, 0.75, 0.9, 0.95, 0.99]))\n",
    "ax.set_ylim(-0.05, 1.05)\n",
    "\n",
    "ax.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    benchmark.xs((unit, match, \"spgc-fit\"), level=(\"unit\", \"match\", \"tool\"))[_score_list]\n",
    "    .fillna(0)\n",
    "    .reindex(genome_to_spgc_strain.index, fill_value=0)\n",
    "    .join(genome_to_spgc_strain, rsuffix=\"_\")\n",
    "    .join(spgc_qc.species_gene_frac)\n",
    "\n",
    ")\n",
    "\n",
    "d1 = d0[lambda x: x.species_depth_sum > 0]\n",
    "d1_bad_qc =  d1[lambda x: x.species_gene_frac < 0.9]\n",
    "d1_good_qc = d1[lambda x: x.species_gene_frac >= 0.9]\n",
    "\n",
    "bins = np.linspace(0, 1, num=21)\n",
    "plt.hist(d1_bad_qc.f1, alpha=0.5, bins=bins, color='tab:orange')\n",
    "plt.hist(d1_good_qc.f1, alpha=0.5, bins=bins, color='tab:blue')\n",
    "\n",
    "# plt.axvline(d1_bad_qc.f1.median(), color='tab:orange')\n",
    "# plt.axvline(d1_good_qc.f1.median(), color='tab:blue')"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = benchmark.xs((\"eggnog\", \"match\"), level=(\"unit\", \"match\")).f1.unstack().fillna(0)\n",
    "plt.scatter(\"spanda-s2\", \"spanda-s4\", data=d)\n",
    "plt.plot([0, 1], [0, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "idxwhere(d[\"spanda-s4\"] - d[\"spanda-s2\"] > 0.03)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_id.isin(multi_genome_species)\n",
    "    ][[\"f1\", \"precision\", \"recall\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    # \"nnmatched-m50\",\n",
    "    # \"nnmatched-m10\",\n",
    "    # \"nnmatched-m1\",\n",
    "    # \"nnmatched-m0\",\n",
    "    \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order) + 0.5, 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "nbins = 20\n",
    "left_bound = 0.0\n",
    "bins = [0] + list(np.linspace(left_bound, 1, num=nbins + 1)[1:])\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\")\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        *_, cbar_artist = ax.hist2d(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            bins=bins,\n",
    "            cmin=1,\n",
    "            norm=mpl.colors.PowerNorm(1 / 1, vmin=0, vmax=21),\n",
    "            cmap=\"magma_r\",\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "ax.set_xlim(left_bound, 1)\n",
    "ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "\n",
    "fig.subplots_adjust(right=0.8)\n",
    "cbar_ax = fig.add_axes([0.85, 0.15, 0.025, 0.67])\n",
    "fig.colorbar(\n",
    "    cbar_artist, cax=cbar_ax, ticks=[1, 10, 20, 30, 40, 50], label=\"count strains\"\n",
    ")\n",
    "# fig.tight_layout(rect=(0, 0, 0.85, 0.67))\n",
    "# ax.set_xlabel(x)\n",
    "# ax.set_ylabel(y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "unit = \"eggnog\"\n",
    "match = \"match\"\n",
    "\n",
    "d1 = (\n",
    "    benchmark.xs((unit, match), level=(\"unit\", \"match\"))\n",
    "    .join(genome.join(depth_meta, on=\"species_id\"))[\n",
    "        lambda x: x.species_id.isin(multi_genome_species)\n",
    "    ][[\"f1\", \"precision\", \"recall\"]]\n",
    "    .unstack()\n",
    "    .rename_axis(columns=[\"score\", \"tool\"])\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\n",
    "    \"panphlan\",\n",
    "    \"spanda-s4\",\n",
    "    # \"nnmatched-m50\",\n",
    "    # \"nnmatched-m10\",\n",
    "    # \"nnmatched-m1\",\n",
    "    # \"nnmatched-m0\",\n",
    "    \"spgc-depth200\",\n",
    "]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order), 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc-fit\"\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\")\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        ax.scatter(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            s=5,\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.05, 0.25, 0.5, 0.75, 0.95]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "        ax.set_xticks([0, 0.6, 0.8, 1.0])\n",
    "        ax.set_yticks([0, 0.6, 0.8, 1.0])\n",
    "\n",
    "\n",
    "pad = 1e-2\n",
    "ax.set_yscale(\n",
    "    \"function\",\n",
    "    functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    ")\n",
    "ax.set_yticks(np.unique([0.0, 0.5, 0.9, 0.99, 1.00]))\n",
    "ax.set_ylim(-0.1, 1.00)\n",
    "\n",
    "ax.set_xscale(\n",
    "    \"function\",\n",
    "    functions=(lambda x: -np.log(1 + pad - x), lambda y: (1 + pad - np.exp(-y))),\n",
    ")\n",
    "ax.set_xticks(np.unique([0.0, 0.5, 0.9, 0.99, 1.00]))\n",
    "ax.set_xlim(-0.1, 1.00)\n",
    "\n",
    "# ax.set_xlim(left_bound, 1)\n",
    "# ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "unit = \"eggnog\"\n",
    "\n",
    "d0 = xjin_benchmarking[lambda x: ~x.to_drop & x.unit.isin([unit])]\n",
    "\n",
    "\n",
    "selector_panphlan = lambda x: ((x.tool == \"panphlan\") & True)\n",
    "selector_spanda = lambda x: ((x.tool == \"spanda\") & True)\n",
    "selector_depth = lambda x: (\n",
    "    (x.tool == \"spgc\")\n",
    "    & (x.specgene_params == \"ref-t25-p95\")\n",
    "    & (x.depth_thresh == 0.25)\n",
    "    & (x.correlation_thresh == 0)\n",
    "    & (x.max_strain_samples == 999)\n",
    "    & (x.trnsfm_exponent == 1)\n",
    "    & True\n",
    ")\n",
    "selector_spgc = lambda x: (\n",
    "    (x.tool == \"spgc\")\n",
    "    & (x.specgene_params == \"ref-t25-p95\")\n",
    "    & (x.depth_thresh == 0.25)\n",
    "    & (x.correlation_thresh == 0.1)\n",
    "    & (x.max_strain_samples == 999)\n",
    "    & (x.trnsfm_exponent == 3.0)\n",
    "    & True\n",
    ")\n",
    "selector_spgc_alt = lambda x: (\n",
    "    (x.tool == \"spgc\")\n",
    "    & (x.specgene_params == \"ref-t25-p95\")\n",
    "    & (x.depth_thresh == 0.25)\n",
    "    & (x.correlation_thresh == 0.5)\n",
    "    & (x.max_strain_samples == 999)\n",
    "    & (x.trnsfm_exponent == 1.0)\n",
    "    & True\n",
    ")\n",
    "\n",
    "\n",
    "# TODO: Plot SPGC against each of the three competitors.\n",
    "# Show comparison to all three: precision/recall/f1\n",
    "\n",
    "_tool_order = [\"panphlan\", \"spanda\", \"depth_only\", \"spgc\"]\n",
    "tool_palette = lib.plot.construct_ordered_palette(_tool_order, cm=\"rainbow\")\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "\n",
    "d1 = []\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d1.append(\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                panphlan=d0[selector_panphlan].set_index(\"genome_id\")[score],\n",
    "                spanda=d0[selector_spanda].set_index(\"genome_id\")[score],\n",
    "                depth_only=d0[selector_depth].set_index(\"genome_id\")[score],\n",
    "                spgc=d0[selector_spgc].set_index(\"genome_id\")[score],\n",
    "                spgc_alt=d0[selector_spgc_alt].set_index(\"genome_id\")[score],\n",
    "                # species_id=ref_strains.species_id,\n",
    "                # strain_depth_max=ref_strains.species_id.map(max_species_depth),\n",
    "                # strain_depth_sum=ref_strains.species_id.map(sum_species_depth),\n",
    "                score=score,\n",
    "            )\n",
    "        )\n",
    "    )\n",
    "d1 = (\n",
    "    pd.concat(d1)\n",
    "    .set_index(\"score\", append=True)\n",
    "    .unstack(\"score\")\n",
    "    .assign(\n",
    "        species_id=ref_strains.species_id,\n",
    "        strain_depth_max=ref_strains.species_id.map(max_species_depth),\n",
    "        strain_depth_sum=ref_strains.species_id.map(sum_species_depth),\n",
    "    )\n",
    "    .fillna(0)\n",
    "    .sort_values(\"strain_depth_max\")\n",
    ")  # FIXME\n",
    "\n",
    "### Figure A ###\n",
    "_tool_comparison_order = [\"panphlan\", \"spanda\", \"depth_only\"]\n",
    "_score_order = [\"precision\", \"recall\", \"f1\"]\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_comparison_order),\n",
    "    figsize=(2.5 * len(_tool_comparison_order) + 2.5, 2.5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "y = \"spgc\"\n",
    "nbins = 15\n",
    "left_bound = 0.0\n",
    "bins = [0] + list(np.linspace(left_bound, 1, num=nbins + 1)[1:])\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    d2 = d1.xs(score, level=\"score\", axis=\"columns\")\n",
    "    print(score, \"SPGC IQR:\", (d2[y]).quantile([0.25, 0.5, 0.75]).round(3).tolist())\n",
    "    for x, ax in zip(_tool_comparison_order, ax_row):\n",
    "        *_, cbar_artist = ax.hist2d(\n",
    "            x,\n",
    "            y,\n",
    "            data=d2,\n",
    "            bins=bins,\n",
    "            cmin=1,\n",
    "            norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=64),\n",
    "            cmap=\"magma_r\",\n",
    "        )\n",
    "        ax.plot([left_bound, 1], [left_bound, 1], lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.set_aspect(1)\n",
    "        print(\n",
    "            \"compared to:\",\n",
    "            x,\n",
    "            (d2[y] - d2[x]).quantile([0.25, 0.5, 0.75]).round(3).tolist(),\n",
    "            sp.stats.wilcoxon(d2[x], d2[y]).pvalue,\n",
    "        )\n",
    "\n",
    "ax.set_xlim(left_bound, 1)\n",
    "ax.set_ylim(left_bound, 1)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "    # *_, artist = ax_row[-1].hist2d(x, y, data=d1.head(0), bins=np.linspace(0, 1, num=21), norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=30), cmap='Blues')\n",
    "    # fig.colorbar(artist, cax=ax_row[-1])\n",
    "\n",
    "for tool_comparison, ax_col in zip(_tool_comparison_order, axs.T):\n",
    "    ax_col[0].set_title(tool_comparison)\n",
    "\n",
    "# fig.tight_layout()\n",
    "fig.subplots_adjust(right=0.8)\n",
    "cbar_ax = fig.add_axes([0.85, 0.15, 0.025, 0.67])\n",
    "fig.colorbar(\n",
    "    cbar_artist, cax=cbar_ax, ticks=[0, 1, 2, 4, 8, 16, 32, 64], label=\"count strains\"\n",
    ")\n",
    "# ax.set_xlabel(x)\n",
    "# ax.set_ylabel(y)\n",
    "\n",
    "\n",
    "### Figure B ###\n",
    "fig, axs = plt.subplots(\n",
    "    len(_score_order),\n",
    "    len(_tool_order) + 1,\n",
    "    figsize=(7 * len(_tool_order) + 1, 5 * len(_score_order)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    trend_ax = ax_row[-1]\n",
    "    for tool, ax in zip(_tool_order, ax_row):\n",
    "        d2 = pd.DataFrame(\n",
    "            dict(\n",
    "                x=d1[\"strain_depth_max\"],\n",
    "                y=d1[(tool, score)],\n",
    "            )\n",
    "        ).assign(\n",
    "            rolling_average_y=lambda d: d.sort_values(\"x\")\n",
    "            .rolling(window=30, min_periods=2, center=True)\n",
    "            .y.mean(),\n",
    "            lowess_y=lambda d: sm.nonparametric.lowess(\n",
    "                d.y, d.x, it=5, frac=1 / 4, return_sorted=False\n",
    "            ),\n",
    "        )\n",
    "        ax.scatter(\"x\", \"y\", data=d2, s=20)\n",
    "        # smoothed_y = d1.sort_values('strain_depth_max').rolling(window=20, min_periods=1, center=True).mean()\n",
    "        # smoothed_x = np.logspace(-1, 2, num=200)\n",
    "        # smoothed_y = sm.nonparametric.lowess(d1[tool].values, d1['strain_depth_max'].values, xvals=smoothed_x, it=5, frac=1/4, return_sorted=False)\n",
    "        ax.plot(\n",
    "            \"x\",\n",
    "            \"lowess_y\",\n",
    "            data=d2[(d2.x > 0.1) & (d2.x < 100)],\n",
    "            color=tool_palette[tool],\n",
    "            lw=4,\n",
    "            zorder=0,\n",
    "        )\n",
    "        trend_ax.plot(\n",
    "            \"x\",\n",
    "            \"lowess_y\",\n",
    "            data=d2[(d2.x > 0.1) & (d2.x < 100)],\n",
    "            color=tool_palette[tool],\n",
    "            lw=3,\n",
    "            zorder=0,\n",
    "            alpha=0.9,\n",
    "            label=tool,\n",
    "        )\n",
    "# ax.set_yscale('symlog', linthresh=1e-2)\n",
    "# ax.invert_yaxis()\n",
    "\n",
    "for ax_row in axs:\n",
    "    for ax in ax_row:\n",
    "        ax.axvline(1.0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "        ax.axhline(0.9, lw=1, linestyle=\":\", color=\"k\")\n",
    "        # ax.axhline(1 - 0.5, lw=1, linestyle='--', color='k')\n",
    "# trend_ax.legend()\n",
    "\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-1)\n",
    "ax.set_xlim(-1e-2)\n",
    "# ax.set_yscale('logit')\n",
    "# ax.set_ylim(0.5)\n",
    "\n",
    "for score, ax_row in zip(_score_order, axs):\n",
    "    ax_row[0].set_ylabel(score)\n",
    "ax_row[0].set_xlabel(\"strain_depth_max\")\n",
    "\n",
    "for tool, ax_col in zip(_tool_order, axs.T):\n",
    "    ax_col[0].set_title(tool)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = benchmark.join(depth_meta, on=\"species\")\n",
    "\n",
    "tool = \"spgc-fit\"\n",
    "d1 = d0[lambda x: x.tool == tool]\n",
    "plt.scatter(\"max_depth\", \"f1\", data=d1)\n",
    "\n",
    "tool = \"panphlan\"\n",
    "d1 = d0[lambda x: x.tool == tool]\n",
    "plt.scatter(\"max_depth\", \"f1\", data=d1)\n",
    "\n",
    "tool = \"spanda-s2\"\n",
    "d1 = d0[lambda x: x.tool == tool]\n",
    "plt.scatter(\"max_depth\", \"f1\", data=d1)\n",
    "\n",
    "\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_depth = (\n",
    "    pd.read_table(\n",
    "        \"data/group/xjin/r.proc.gtpro.species_depth.tsv\",\n",
    "        index_col=[\"sample\", \"species_id\"],\n",
    "    )\n",
    "    .depth.unstack(fill_value=0)\n",
    "    # .rename(columns=str, index=lambda x: \"CF_\" + str(int(x.split(\"_\")[1])))\n",
    "    # .rename({'CF_15': 'CF_11', 'CF_11': 'CF_15'})  # Sample swap\n",
    ")\n",
    "motu_rabund = motu_depth.divide(motu_depth.sum(1), axis=0)\n",
    "\n",
    "motu_rabund"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sotu_depth = []\n",
    "missing_files = []\n",
    "for species_id in motu_depth.columns:\n",
    "    path = f\"data/group/xjin/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.comm.tsv\"\n",
    "    try:\n",
    "        d = (\n",
    "            pd.read_table(path, index_col=[\"sample\", \"strain\"])\n",
    "            .squeeze()\n",
    "            .unstack()\n",
    "            # .rename(columns=str, index=lambda x: \"CF_\" + str(int(x.split(\"_\")[1])))\n",
    "            # .rename({'CF_11': 'CF_15', 'CF_15': 'CF_11'})  # Sample swap.\n",
    "        )\n",
    "    except FileNotFoundError:\n",
    "        missing_files.append(path)\n",
    "        d = pd.DataFrame([])\n",
    "    _keep_strains = idxwhere(d.sum() > 0.05)\n",
    "    assert d.index.isin(motu_depth.index).all()\n",
    "    d = d.reindex(index=motu_depth.index, columns=_keep_strains, fill_value=0)\n",
    "    d = d.assign(__other=lambda x: 1 - x.sum(1)).rename(columns={\"__other\": -1})\n",
    "    d[d < 0] = 0\n",
    "    d = d.divide(d.sum(1), axis=0)\n",
    "    d = d.multiply(motu_depth[species_id], axis=0)\n",
    "    d = d.rename(columns=lambda s: f\"{species_id}_{s}\")\n",
    "    sotu_depth.append(d)\n",
    "sotu_depth = pd.concat(sotu_depth, axis=1)\n",
    "sotu_rabund = sotu_depth.divide(sotu_depth.sum(1), axis=0)\n",
    "len(motu_depth.columns), len(missing_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_sotu_depth = sotu_depth[lambda x: x.index.str.startswith('xjin_')]\n",
    "xjin_sotu_depth = xjin_sotu_depth[idxwhere((xjin_sotu_depth.sum() > 0.01))]\n",
    "plt.hist(np.log10(xjin_sotu_depth.sum()), bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_sotu_strain_list = idxwhere(xjin_sotu_depth.sum() > 1)\n",
    "other_strain_depth = xjin_sotu_depth.drop(columns=xjin_sotu_strain_list).sum(1)\n",
    "\n",
    "d = xjin_sotu_depth[xjin_sotu_strain_list].assign(other=other_strain_depth)\n",
    "sns.clustermap(d.T, norm=mpl.colors.SymLogNorm(linthresh=0.1), metric='cosine', figsize=(5, 8), xticklabels=0, yticklabels=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_motu_depth = motu_depth[lambda x: x.index.str.startswith('xjin_')].rename(columns=str)\n",
    "xjin_motu_depth = xjin_motu_depth[idxwhere((xjin_motu_depth.sum() > 0.01))]\n",
    "plt.hist(np.log10(xjin_motu_depth.sum()), bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_has_genome_list = genome_to_spgc_strain.assign(species=genome.species_id)[lambda x: x.species != 'TODO'].species.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "xjin_motu_strain_list = idxwhere(xjin_motu_depth.sum() > 1)\n",
    "d = xjin_motu_depth[xjin_motu_strain_list]\n",
    "_num_genomes_per_species = genome_to_spgc_strain.assign(species=genome.species_id).species.value_counts().reindex(d.columns, fill_value=0)\n",
    "_num_genomes_palette = lib.plot.construct_ordered_palette(np.arange(_num_genomes_per_species.max()) + 1, cm='viridis', other='white')\n",
    "\n",
    "g = sns.clustermap(\n",
    "    d.T.rename_axis(index=''),\n",
    "    cmap=\"pink\",\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.1, vmin=0, vmax=1e4),\n",
    "    metric=\"cosine\",\n",
    "    figsize=(5, 8),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    dendrogram_ratio=(0.1, 0.1),\n",
    "    row_colors=_num_genomes_per_species.map(_num_genomes_palette),\n",
    ")\n",
    "g.ax_cbar.set_visible(False)\n",
    "# See https://stackoverflow.com/questions/65677031/move-seaborn-clustermap-row-colors-bar-to-the-other-side-of-the-plot\n",
    "ax_row_colors = g.ax_row_colors\n",
    "box = ax_row_colors.get_position()\n",
    "box_heatmap = g.ax_heatmap.get_position()\n",
    "ax_row_colors.set_position([box_heatmap.max[0] + 0.01, box.y0, box.width*1.5, box.height])\n",
    "\n",
    "\n",
    "fig = plt.figure(figsize=(5, 2))\n",
    "plt.scatter(\n",
    "    [],\n",
    "    [],\n",
    "    c=[],\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=0.1, vmin=0, vmax=1e3),\n",
    "    cmap=\"pink\",\n",
    ")\n",
    "plt.colorbar(orientation=\"horizontal\")\n",
    "ax = plt.gca()\n",
    "ax.set_visible(False)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 1))\n",
    "for n in np.arange(_num_genomes_per_species.max()) + 1:\n",
    "    ax.scatter([], [], color=_num_genomes_palette[n], label=n, lw=0.5, edgecolor='k', marker='s')\n",
    "ax.legend(ncols=2)\n",
    "lib.plot.hide_axes_and_spines(ax)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}