{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/gtpro/species_taxonomy_ext.tsv\"\n",
    "\n",
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(species_taxonomy_inpath)\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"cog_category\", \"description\"],\n",
    "    index_col=\"cog_category\",\n",
    ").assign(description=lambda x: x.index + \": \" + x.description)\n",
    "cog_category_description.loc[\"no_category\", \"description\"] = \"-: No Annotation\"\n",
    "cog_category_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_order = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"H\",\n",
    "    \"I\",\n",
    "    \"J\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "    \"V\",\n",
    "    \"W\",\n",
    "    \"X\",\n",
    "    \"Y\",\n",
    "    \"Z\",\n",
    "    \"no_category\",\n",
    "]\n",
    "\n",
    "cog_category_raw_columns = [f\"cog_category_{s}\" for s in cog_category_order]\n",
    "\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list = (\n",
    "    pd.read_table(\"meta/species_group.tsv\")[\n",
    "        lambda x: x.species_group_id == \"xjin_ucfmt_hmp2\"\n",
    "    ]\n",
    "    .species_id.astype(str)\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_stats = []\n",
    "ref_strain_stats = []\n",
    "missing_strain_stats = []\n",
    "for species in species_list:\n",
    "    spgc_path = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.spgc_strain_stats.tsv\"\n",
    "    ref_path = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.ref_strain_stats.tsv\"\n",
    "    if not os.path.exists(spgc_path):\n",
    "        missing_strain_stats.append(species)\n",
    "        continue\n",
    "    d_spgc = pd.read_table(spgc_path)\n",
    "    d_ref = pd.read_table(ref_path)\n",
    "    spgc_strain_stats.append(d_spgc.assign(species=species))\n",
    "    ref_strain_stats.append(d_ref.assign(species=species))\n",
    "spgc_strain_stats = pd.concat(spgc_strain_stats).join(species_taxonomy, on=\"species\")\n",
    "ref_strain_stats = pd.concat(ref_strain_stats).join(species_taxonomy, on=\"species\")\n",
    "\n",
    "\n",
    "print(\n",
    "    len(missing_strain_stats),\n",
    "    \"out of\",\n",
    "    len(species_list),\n",
    "    \"species are missing spgc strain stats.\",\n",
    ")\n",
    "print(\n",
    "    len(spgc_strain_stats),\n",
    "    \"spgc strains found across\",\n",
    "    len(spgc_strain_stats.species.unique()),\n",
    "    \"species\",\n",
    ")\n",
    "print(\n",
    "    len(ref_strain_stats),\n",
    "    \"ref strains found across\",\n",
    "    len(ref_strain_stats.species.unique()),\n",
    "    \"species\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_palette = lib.plot.construct_ordered_palette(\n",
    "    sorted(ref_strain_stats.sort_values(\"taxonomy_string\").p__.unique()),\n",
    "    cm=\"tab10\",\n",
    ")\n",
    "\n",
    "assert len(set(phylum_palette.values())) == len((phylum_palette.values()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_stats = []\n",
    "missing_mwas_stats = []\n",
    "for species in tqdm(species_list):\n",
    "    mwas_path = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.strain_gene.hmp2_mwas-f30-n2.tsv\"\n",
    "    if not os.path.exists(mwas_path):\n",
    "        missing_mwas_stats.append(species)\n",
    "        continue\n",
    "    d = pd.read_table(\n",
    "        mwas_path\n",
    "    )  # TODO: Drop the species mapping. We'll get that from the metadata table.\n",
    "    mwas_stats.append(d.assign(species_id=species))\n",
    "mwas_stats = (\n",
    "    pd.concat(mwas_stats)\n",
    "    .set_index('gene_id')\n",
    "    .assign(\n",
    "        total_present=lambda x: x[[\"present-CD\", \"present-UC\", \"present-nonIBD\"]].sum(\n",
    "            1\n",
    "        ),\n",
    "        total_absent=lambda x: x[[\"absent-CD\", \"absent-UC\", \"absent-nonIBD\"]].sum(1),\n",
    "        total_subjects=lambda x: x[[\"total_present\", \"total_absent\"]].sum(1),\n",
    "        subject_prevalence=lambda x: x['total_present'] / (x['total_present'] + x['total_absent']),\n",
    "        log10_fisher_exact_pvalue_ibd=lambda x: np.log10(x.fisher_exact_pvalue_ibd),\n",
    "    )\n",
    "    .sort_values(\"fisher_exact_pvalue_ibd\")\n",
    ")\n",
    "print(len(missing_mwas_stats), \"species are missing mwas stats\")\n",
    "mwas_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats = []\n",
    "_gene_meta = []\n",
    "missing_gene_stats = []\n",
    "for species in tqdm(species_list):\n",
    "    stats_path = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.gene_stats.tsv\"\n",
    "    meta_path = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc-fit.gene_meta.tsv\"\n",
    "    if not os.path.exists(stats_path):\n",
    "        missing_gene_stats.append(species)\n",
    "        continue\n",
    "    d_stats = pd.read_table(stats_path)\n",
    "    d_meta = pd.read_table(meta_path)\n",
    "    gene_stats.append(d_stats.assign(species=species))\n",
    "    _gene_meta.append(d_meta.assign(species=species))\n",
    "gene_stats = pd.concat(gene_stats).set_index(\"gene_id\")\n",
    "_gene_meta = pd.concat(_gene_meta).set_index(\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category_matrix = (\n",
    "    _gene_meta.loc[:, cog_category_raw_columns]\n",
    "    .rename(columns=lambda s: s[len(\"cog_category_\") :])\n",
    "    .fillna(False)\n",
    ")\n",
    "\n",
    "gene_meta = _gene_meta.drop(columns=cog_category_raw_columns, errors='ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Novelty / Believability / Variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist2d(\n",
    "    spgc_strain_stats.nearest_ref_geno_diss,\n",
    "    spgc_strain_stats.nearest_ref_gene_diss,\n",
    "    norm=mpl.colors.PowerNorm(1 / 3),\n",
    "    bins=(\n",
    "        [0] + list(np.logspace(-4, 0, num=40)),\n",
    "        np.linspace(0, 1, num=40),\n",
    "    ),\n",
    ")\n",
    "plt.xscale(\"symlog\", linthresh=1e-4, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist2d(\n",
    "    ref_strain_stats.nearest_ref_geno_diss,\n",
    "    ref_strain_stats.nearest_ref_gene_diss,\n",
    "    norm=mpl.colors.PowerNorm(1 / 3),\n",
    "    bins=(\n",
    "        [0] + list(np.logspace(-4, 0, num=40)),\n",
    "        np.linspace(0, 1, num=40),\n",
    "    ),\n",
    ")\n",
    "plt.xscale(\"symlog\", linthresh=1e-4, linscale=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Number of strains in each group after de-replication.\n",
    "pd.concat(\n",
    "    [\n",
    "        spgc_strain_stats[[\"species\", \"derep_clust\"]].assign(genome_type=\"SPGC\"),\n",
    "        ref_strain_stats[[\"species\", \"derep_clust\", \"genome_type\"]],\n",
    "    ]\n",
    ").groupby(\"species\").value_counts().unstack(\"genome_type\", fill_value=0).gt(0).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Entirely new strains \"discovered\" in my pipeline.\n",
    "\n",
    "d = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            spgc_strain_stats[[\"species\", \"derep_clust\"]].assign(genome_type=\"SPGC\"),\n",
    "            ref_strain_stats[[\"species\", \"derep_clust\", \"genome_type\"]],\n",
    "        ]\n",
    "    )\n",
    "    .groupby(\"species\")\n",
    "    .value_counts()\n",
    "    .unstack(\"genome_type\", fill_value=0)\n",
    "    .gt(0)\n",
    "    .groupby(level=\"species\")\n",
    "    .value_counts()\n",
    "    .unstack(\"species\", fill_value=0)\n",
    "    .T\n",
    ")\n",
    "\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "num_novel_strains = d[(False, False, True)]\n",
    "num_existing_strains = d.sum(1) - num_novel_strains\n",
    "plt.scatter(num_existing_strains, num_novel_strains)\n",
    "plt.yscale(\"symlog\")\n",
    "plt.xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(sp.stats.pearsonr(gene_stats.prevalence_ref, gene_stats.prevalence_spgc))\n",
    "plt.hist2d(\n",
    "    gene_stats.prevalence_ref,\n",
    "    gene_stats.prevalence_spgc,\n",
    "    norm=mpl.colors.PowerNorm(1 / 3),\n",
    ")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = spgc_strain_stats.sort_values(\"taxonomy_string\")\n",
    "focal_species = \"102438\"\n",
    "\n",
    "fit = smf.ols(\n",
    "    f\"nearest_ref_gene_diss ~ np.log10(nearest_ref_geno_diss) + C(species, Sum)\",\n",
    "    data=d0,\n",
    ").fit()\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "_color_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    [\"102506\"], colors=[\"red\"], other=\"grey\"\n",
    ")\n",
    "_size_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    [\"102506\"], colors=[20], other=5\n",
    ")\n",
    "\n",
    "ax = axs[0]\n",
    "# for species, d1 in d0.groupby(\"species\"):\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"grey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0[d0.species == focal_species],\n",
    "    label=focal_species,\n",
    "    color=\"red\",\n",
    "    s=20,\n",
    "    alpha=0.5,\n",
    ")\n",
    "ax.legend(loc=\"upper right\")\n",
    "# ax.scatter([], [], label=p__, color=phylum_palette[p__], s=10)\n",
    "\n",
    "xx = np.logspace(-3, -1)\n",
    "ax.plot(\n",
    "    xx,\n",
    "    np.log10(xx) * fit.params[\"np.log10(nearest_ref_geno_diss)\"]\n",
    "    + fit.params[\"Intercept\"],\n",
    "    color=\"k\",\n",
    "    label=\"Overall Slope\",\n",
    ")\n",
    "p_slope = fit.pvalues[\"np.log10(nearest_ref_geno_diss)\"]\n",
    "slope = fit.params[\"np.log10(nearest_ref_geno_diss)\"]  # TODO\n",
    "r2_adj = fit.rsquared_adj\n",
    "ax.annotate(\n",
    "    f\"$\\\\beta_{{\\\\mathrm{{slope}}}}$={slope:0.3f}\\n$P_{{\\\\mathrm{{slope}}}}$={p_slope:0.0e}\\n$R^2_{{\\\\mathrm{{adj}}}}$={r2_adj:0.2%}\",\n",
    "    xy=(0.1, 0.9),\n",
    "    xycoords=\"axes fraction\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax.annotate(f\"\", xy=(0.1, 0.9), xycoords=\"axes fraction\")\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "ax = axs[1]\n",
    "sns.boxplot(x=\"p__\", y=\"nearest_ref_gene_diss\", data=d0, ax=ax, palette=phylum_palette)\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel(\"\")\n",
    "\n",
    "ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = ref_strain_stats.sort_values(\"taxonomy_string\")\n",
    "focal_species = \"102438\"\n",
    "\n",
    "fit = smf.ols(\n",
    "    f\"nearest_ref_gene_diss ~ np.log10(nearest_ref_geno_diss) + C(species, Sum)\",\n",
    "    data=d0,\n",
    ").fit()\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "_color_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    [\"102506\"], colors=[\"red\"], other=\"grey\"\n",
    ")\n",
    "_size_palette = lib.plot.construct_ordered_palette_from_list(\n",
    "    [\"102506\"], colors=[20], other=5\n",
    ")\n",
    "\n",
    "ax = axs[0]\n",
    "# for species, d1 in d0.groupby(\"species\"):\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"grey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0[d0.species == focal_species],\n",
    "    label=focal_species,\n",
    "    color=\"red\",\n",
    "    s=20,\n",
    "    alpha=0.5,\n",
    ")\n",
    "ax.legend(loc=\"upper right\")\n",
    "# ax.scatter([], [], label=p__, color=phylum_palette[p__], s=10)\n",
    "\n",
    "xx = np.logspace(-3, -1)\n",
    "ax.plot(\n",
    "    xx,\n",
    "    np.log10(xx) * fit.params[\"np.log10(nearest_ref_geno_diss)\"]\n",
    "    + fit.params[\"Intercept\"],\n",
    "    color=\"k\",\n",
    "    label=\"Overall Slope\",\n",
    ")\n",
    "p_slope = fit.pvalues[\"np.log10(nearest_ref_geno_diss)\"]\n",
    "slope = fit.params[\"np.log10(nearest_ref_geno_diss)\"]  # TODO\n",
    "r2_adj = fit.rsquared_adj\n",
    "ax.annotate(\n",
    "    f\"$\\\\beta_{{\\\\mathrm{{slope}}}}$={slope:0.3f}\\n$P_{{\\\\mathrm{{slope}}}}$={p_slope:0.0e}\\n$R^2_{{\\\\mathrm{{adj}}}}$={r2_adj:0.2%}\",\n",
    "    xy=(0.1, 0.9),\n",
    "    xycoords=\"axes fraction\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax.annotate(f\"\", xy=(0.1, 0.9), xycoords=\"axes fraction\")\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "ax = axs[1]\n",
    "sns.boxplot(x=\"p__\", y=\"nearest_ref_gene_diss\", data=d0, ax=ax, palette=phylum_palette)\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel(\"\")\n",
    "\n",
    "ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = spgc_strain_stats.sort_values(\"taxonomy_string\")\n",
    "fit = smf.ols(\n",
    "    f\"nearest_ref_gene_diss ~ cr(nearest_ref_geno_diss, 20, constraints='center') + C(species, Sum)\",\n",
    "    data=d0,\n",
    ").fit()\n",
    "print(fit.aic)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "ax = axs[0]\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"grey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "_arbitrary_species = \"100003\"\n",
    "_arbitrary_species_term = f\"C(species, Sum)[S.{_arbitrary_species}]\"\n",
    "xx = np.logspace(-4, -1)\n",
    "yy = (\n",
    "    fit.predict(\n",
    "        pd.DataFrame(dict(nearest_ref_geno_diss=xx)).assign(species=_arbitrary_species)\n",
    "    ).values\n",
    "    - fit.params[_arbitrary_species_term]\n",
    ")\n",
    "ax.plot(\n",
    "    xx,\n",
    "    yy,\n",
    "    color=\"k\",\n",
    ")\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "\n",
    "ax = axs[1]\n",
    "_species_list = d0.species.unique()\n",
    "d1 = (\n",
    "    pd.Series(\n",
    "        fit.predict(\n",
    "            pd.DataFrame(dict(species=_species_list)).assign(nearest_ref_geno_diss=1e-3)\n",
    "        ).values,\n",
    "        index=_species_list,\n",
    "    )\n",
    "    .to_frame(name=\"species_min_predicted_minimum_gene_diss\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "sns.boxplot(\n",
    "    x=\"p__\",\n",
    "    y=\"species_min_predicted_minimum_gene_diss\",\n",
    "    data=d1,\n",
    "    ax=ax,\n",
    "    palette=phylum_palette,\n",
    ")\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_ylim(-0.005, 0.605)\n",
    "\n",
    "spgc_near_neighbor_predicted_gene_diss = d1.species_min_predicted_minimum_gene_diss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = ref_strain_stats.sort_values(\"taxonomy_string\")\n",
    "fit = smf.ols(\n",
    "    f\"nearest_ref_gene_diss ~ cr(nearest_ref_geno_diss, 20, constraints='center') + C(species, Sum)\",\n",
    "    data=d0,\n",
    ").fit()\n",
    "print(fit.aic)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "ax = axs[0]\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"grey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "_arbitrary_species = \"100003\"\n",
    "_arbitrary_species_term = f\"C(species, Sum)[S.{_arbitrary_species}]\"\n",
    "xx = np.logspace(-4, -1)\n",
    "yy = (\n",
    "    fit.predict(\n",
    "        pd.DataFrame(dict(nearest_ref_geno_diss=xx)).assign(species=_arbitrary_species)\n",
    "    ).values\n",
    "    - fit.params[_arbitrary_species_term]\n",
    ")\n",
    "ax.plot(\n",
    "    xx,\n",
    "    yy,\n",
    "    color=\"k\",\n",
    ")\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "\n",
    "ax = axs[1]\n",
    "_species_list = d0.species.unique()\n",
    "d1 = (\n",
    "    pd.Series(\n",
    "        fit.predict(\n",
    "            pd.DataFrame(dict(species=_species_list)).assign(nearest_ref_geno_diss=1e-3)\n",
    "        ).values,\n",
    "        index=_species_list,\n",
    "    )\n",
    "    .to_frame(name=\"species_min_predicted_minimum_gene_diss\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "sns.boxplot(\n",
    "    x=\"p__\",\n",
    "    y=\"species_min_predicted_minimum_gene_diss\",\n",
    "    data=d1,\n",
    "    ax=ax,\n",
    "    palette=phylum_palette,\n",
    ")\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_ylim(-0.005, 0.605)\n",
    "\n",
    "ref_near_neighbor_predicted_gene_diss = d1.species_min_predicted_minimum_gene_diss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# {ref,spgc}_near_neighbor_predicted_gene_diss is the predicted\n",
    "# gene dissimilarity at a genotype distance of 1e-3 (very closely related strains)\n",
    "# in just about any species.\n",
    "# Here we see both the relatively large mean of these values (suggesting that\n",
    "# going with the DBs nearest neighbor is fraught and gene content dissimilarity\n",
    "# really matters), the high degree of correlation (suggesting that SPGC\n",
    "# estimates are a truthfull view on reality), and the\n",
    "# high variance (suggesting that different taxa behave differently in this\n",
    "# regard.\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        ref=ref_near_neighbor_predicted_gene_diss,\n",
    "        spgc=spgc_near_neighbor_predicted_gene_diss,\n",
    "    )\n",
    ").join(species_taxonomy)\n",
    "for p__, d1 in d0.groupby(\"p__\"):\n",
    "    ax.scatter(\n",
    "        \"ref\", \"spgc\", data=d1, color=phylum_palette[p__], s=50, alpha=0.8, label=p__\n",
    "    )\n",
    "ax.legend()\n",
    "ax.plot([0, 0.2], [0, 0.2], lw=1, linestyle=\"--\", color=\"k\")\n",
    "ax.set_xlim(-0.04, 0.42)\n",
    "ax.set_ylim(-0.04, 0.42)\n",
    "ax.set_aspect(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = pd.concat(\n",
    "    [ref_strain_stats.assign(is_spgc=False), spgc_strain_stats.assign(is_spgc=True)]\n",
    ").sort_values(\"taxonomy_string\")\n",
    "fit = smf.ols(\n",
    "    f\"nearest_ref_gene_diss ~ cr(nearest_ref_geno_diss, 20, constraints='center') + C(species, Sum) + is_spgc\",\n",
    "    data=d0,\n",
    ").fit()\n",
    "print(fit.aic)\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "ax = axs[0]\n",
    "ax.scatter(\n",
    "    \"nearest_ref_geno_diss\",\n",
    "    \"nearest_ref_gene_diss\",\n",
    "    data=d0,\n",
    "    label=\"__nolegend__\",\n",
    "    color=\"grey\",\n",
    "    s=2,\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "_arbitrary_species = \"100003\"\n",
    "_arbitrary_species_term = f\"C(species, Sum)[S.{_arbitrary_species}]\"\n",
    "xx = np.logspace(-4, -1)\n",
    "yy_spgc = (\n",
    "    fit.predict(\n",
    "        pd.DataFrame(dict(nearest_ref_geno_diss=xx)).assign(\n",
    "            species=_arbitrary_species, is_spgc=True\n",
    "        )\n",
    "    ).values\n",
    "    - fit.params[_arbitrary_species_term]\n",
    ")\n",
    "yy_ref = (\n",
    "    fit.predict(\n",
    "        pd.DataFrame(dict(nearest_ref_geno_diss=xx)).assign(\n",
    "            species=_arbitrary_species, is_spgc=False\n",
    "        )\n",
    "    ).values\n",
    "    - fit.params[_arbitrary_species_term]\n",
    ")\n",
    "ax.plot(\n",
    "    xx,\n",
    "    yy_spgc,\n",
    "    color=\"darkred\",\n",
    ")\n",
    "ax.plot(\n",
    "    xx,\n",
    "    yy_ref,\n",
    "    color=\"k\",\n",
    ")\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\\n(filtered, Jaccard)\")\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "\n",
    "ax = axs[1]\n",
    "_species_list = d0.species.unique()\n",
    "d1 = (\n",
    "    pd.Series(\n",
    "        fit.predict(\n",
    "            pd.DataFrame(dict(species=_species_list)).assign(\n",
    "                nearest_ref_geno_diss=1e-4, is_spgc=False\n",
    "            )\n",
    "        ).values,\n",
    "        index=_species_list,\n",
    "    )\n",
    "    .to_frame(name=\"species_min_predicted_minimum_gene_diss\")\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "sns.boxplot(\n",
    "    x=\"p__\",\n",
    "    y=\"species_min_predicted_minimum_gene_diss\",\n",
    "    data=d1,\n",
    "    ax=ax,\n",
    "    palette=phylum_palette,\n",
    ")\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel(\"\")\n",
    "ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    spgc_strain_stats.assign(\n",
    "        nearest_ref_gene_diss_adj=lambda x: x.nearest_ref_gene_diss\n",
    "        - fit.params[\"is_spgc[T.True]\"]\n",
    "    )\n",
    "    .groupby(\"species\")[\n",
    "        [\"nearest_ref_geno_diss\", \"nearest_ref_gene_diss\", \"nearest_ref_gene_diss_adj\"]\n",
    "    ]\n",
    "    .median()\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(2, sharex=True, figsize=(7, 10))\n",
    "sns.stripplot(\n",
    "    x=\"p__\",\n",
    "    y=\"nearest_ref_geno_diss\",\n",
    "    data=d0,\n",
    "    size=10,\n",
    "    alpha=0.7,\n",
    "    palette=phylum_palette,\n",
    "    ax=axs[0],\n",
    ")\n",
    "sns.stripplot(\n",
    "    x=\"p__\",\n",
    "    y=\"nearest_ref_gene_diss_adj\",\n",
    "    data=d0,\n",
    "    size=10,\n",
    "    alpha=0.7,\n",
    "    palette=phylum_palette,\n",
    "    ax=axs[1],\n",
    ")\n",
    "lib.plot.rotate_xticklabels(ax=axs[1])\n",
    "# ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### strain-gene-MWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_filt_func = lambda x: (x.subject_prevalence < 0.8) & (x.subject_prevalence > 0.2) & (x.total_subjects > 40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_stats[mwas_filt_func].sort_values('fisher_exact_pvalue_ibd')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-6, 0, num=20)\n",
    "plt.hist(\n",
    "    mwas_stats[mwas_filt_func].fisher_exact_pvalue_ibd,\n",
    "    alpha=0.5,\n",
    "    bins=bins,\n",
    ")\n",
    "print(mwas_stats[mwas_filt_func].shape[0])\n",
    "plt.plot(bins[1:], (bins[1:] - bins[:-1]) * mwas_stats[mwas_filt_func].shape[0])\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# mwas_stats[mwas_filt_func].sort_values(\"fisher_exact_pvalue_ibd\").head(50)#.join(species_taxonomy, on='species_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "ax.scatter('oddsratio_pc_ibd', 'log10_fisher_exact_pvalue_ibd', data=mwas_stats[mwas_filt_func], s=5)\n",
    "ax.set_xscale('log')\n",
    "ax.invert_yaxis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(mwas_stats[mwas_filt_func].log10_fisher_exact_pvalue_ibd.sort_values().values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_stats[mwas_filt_func].join(gene_meta)[\n",
    "lambda x: (x.fisher_exact_pvalue_ibd < 1e-3)\n",
    "# & (x.species_id == '101345')\n",
    "].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_stats[mwas_filt_func].join(gene_meta)[lambda x: x.fisher_exact_pvalue_ibd < 1e-3].species.value_counts().to_frame().join(species_taxonomy[['f__', 'g__', 's__']]).assign(total=mwas_stats[mwas_filt_func].join(gene_meta).species.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = '101338'\n",
    "d = pd.DataFrame(dict(\n",
    "    in_G=mwas_stats[mwas_filt_func][lambda x: (x.species_id == species_id)].join(gene_meta).cog_categories.fillna('').str.contains('G'),\n",
    "    is_signif=mwas_stats[mwas_filt_func][lambda x: (x.species_id == species_id)].fisher_exact_pvalue_ibd < 1e-3,\n",
    "))\n",
    "contingency = d.value_counts().unstack()\n",
    "print(sp.stats.fisher_exact(contingency))\n",
    "contingency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(dict(\n",
    "    in_G=mwas_stats[mwas_filt_func].join(gene_meta).cog_categories.fillna('').str.contains('G'),\n",
    "    is_signif=mwas_stats[mwas_filt_func].fisher_exact_pvalue_ibd < 1e-3,\n",
    "))\n",
    "contingency = d.value_counts().unstack()\n",
    "print(sp.stats.fisher_exact(contingency))\n",
    "contingency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_stats[mwas_filt_func].join(gene_meta).iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-7, 0, num=50)\n",
    "# bins = np.linspace(0, 1, num=51)\n",
    "\n",
    "x = gene_stats[\n",
    "    lambda x: (x.prevalence_mwas_subject > 0.2)\n",
    "    & (x.prevalence_mwas_subject < 0.8)\n",
    "    & (x.num_mwas_subject > 20)\n",
    "].ibd_mwas_pvalue.sort_values()\n",
    "print(len(x))\n",
    "plt.hist(\n",
    "    x,\n",
    "    alpha=0.5,\n",
    "    bins=bins,\n",
    ")\n",
    "\n",
    "x = gene_stats[\n",
    "    lambda x: (x.prevalence_mwas_subject > 0.2)\n",
    "    & (x.prevalence_mwas_subject < 0.8)\n",
    "    & (x.num_mwas_subject > 40)\n",
    "].ibd_mwas_pvalue.sort_values()\n",
    "print(len(x))\n",
    "plt.hist(\n",
    "    x,\n",
    "    alpha=0.5,\n",
    "    bins=bins,\n",
    ")\n",
    "\n",
    "plt.plot(bins[1:], (bins[1:] - bins[:-1]) * x.shape)\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=gene_stats[\n",
    "        lambda x: (x.prevalence_mwas_subject > 0.2)\n",
    "        & (x.prevalence_mwas_subject < 0.8)\n",
    "        & (x.num_mwas_subject > 20)\n",
    "    ],\n",
    "    s=4,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "ax = plt.gca()\n",
    "ax.invert_yaxis()\n",
    "plt.xlabel(\"Odds-Ratio (w/ pseudo-counts)\")\n",
    "plt.ylabel(\"P-value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    gene_stats[\n",
    "        lambda x: (x.prevalence_mwas_subject > 0.2)\n",
    "        & (x.prevalence_mwas_subject < 0.8)\n",
    "        & (x.num_mwas_subject > 20)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.ibd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.2)],\n",
    "    s=20,\n",
    "    marker=\"o\",\n",
    "    edgecolor=\"k\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1\n",
    "\n",
    "high_confidence_mwas_hit = idxwhere(d0.fdr < 0.2)\n",
    "lowish_confidence_mwas_hit = idxwhere(d0.fdr < 0.3)\n",
    "lower_confidence_mwas_hit = idxwhere(d0.fdr < 0.4)\n",
    "low_confidence_mwas_hit = idxwhere(d0.ibd_mwas_pvalue < 0.01)\n",
    "\n",
    "print(len(high_confidence_mwas_hit), len(lower_confidence_mwas_hit), len(\n",
    "    low_confidence_mwas_hit))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    gene_meta.loc[high_confidence_mwas_hit]\n",
    "    .assign(mwas_ibd_pos_assoc=(gene_stats.ibd_mwas_oddsratio_pc > 1).replace({True: \"pos\", False: \"neg\"}))[\n",
    "        [\"species\", \"mwas_ibd_pos_assoc\"]\n",
    "    ]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .assign(total_hits=lambda x: x.neg + x.pos)\n",
    "    .join(species_taxonomy[[\"g__\", \"s__\", \"taxonomy_string\"]])\n",
    "    .sort_values(\"total_hits\", ascending=False)\n",
    "    .head(10)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    gene_meta.loc[low_confidence_mwas_hit]\n",
    "    .assign(mwas_ibd_pos_assoc=(gene_stats.ibd_mwas_oddsratio_pc > 1).replace({True: \"pos\", False: \"neg\"}))[\n",
    "        [\"species\", \"mwas_ibd_pos_assoc\"]\n",
    "    ]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .assign(total_hits=lambda x: x.neg + x.pos)\n",
    "    .join(species_taxonomy[[\"g__\", \"s__\", \"taxonomy_string\"]])\n",
    "    .sort_values(\"total_hits\", ascending=False)\n",
    "    .head(10)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    gene_meta.loc[lower_confidence_mwas_hit]\n",
    "    .assign(mwas_ibd_pos_assoc=(gene_stats.ibd_mwas_oddsratio_pc > 1).replace({True: \"pos\", False: \"neg\"}))[\n",
    "        [\"species\", \"mwas_ibd_pos_assoc\"]\n",
    "    ]\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .assign(total_hits=lambda x: x.neg + x.pos)\n",
    "    .join(species_taxonomy[[\"g__\", \"s__\", \"taxonomy_string\"]])\n",
    "    .sort_values(\"total_hits\", ascending=False)\n",
    "    .head(10)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species = '103681'\n",
    "_species_hit = gene_stats[lambda x: (x.species == _species) & x.index.isin(lower_confidence_mwas_hit)].index\n",
    "print(gene_x_cog_category_matrix.loc[_species_hit].sum().sort_values(ascending=False).to_frame('tally').join(cog_category_description).head(5))\n",
    "print(gene_stats.loc[_species_hit].coclust_label_spgc.value_counts().head(5))\n",
    "gene_meta.loc[_species_hit].join(gene_stats, rsuffix='_').sort_values('ibd_mwas_pvalue').head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats.loc[lower_confidence_mwas_hit][['species', 'coclust_label_spgc']].value_counts().head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.metrics import adjusted_mutual_info_score, adjusted_rand_score\n",
    "many_strains_species_list = idxwhere(spgc_strain_stats.species.value_counts() > 25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "result = {}\n",
    "for species in tqdm(many_strains_species_list):\n",
    "    _gene_list = idxwhere((gene_stats.species == species) & (gene_stats.coclust_label_spgc >= 0))\n",
    "    result[species] = adjusted_mutual_info_score(gene_stats.loc[_gene_list].coclust_label_spgc, gene_stats.loc[_gene_list].coclust_label_ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.Series(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    gene_x_cog_category_matrix.loc[lowish_confidence_mwas_hit]\n",
    "    .sum()\n",
    "    .sort_values()\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    "    .assign(frac=lambda x: x.tally / x.tally.sum())\n",
    "    .assign(\n",
    "        frac_overall=gene_x_cog_category_matrix.sum()\n",
    "        .sort_values()\n",
    "        .to_frame(\"tally\")\n",
    "        .join(cog_category_description)\n",
    "        .assign(frac=lambda x: x.tally / x.tally.sum())\n",
    "        .frac\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _gene_stats = gene_stats.loc[gene_x_cog_category_matrix.index]\n",
    "\n",
    "\n",
    "x = gene_x_cog_category_matrix\n",
    "y = gene_x_cog_category_matrix.index.to_series().isin(low_confidence_mwas_hit)\n",
    "# z_pos = gene_x_cog_category_matrix.index.to_series().isin(low_confidence_mwas_hit) & (_gene_stats.ibd_mwas_oddsratio_pc > 1)\n",
    "# z_neg = gene_x_cog_category_matrix.index.to_series().isin(low_confidence_mwas_hit) & (_gene_stats.ibd_mwas_oddsratio_pc < 1)\n",
    "\n",
    "print(\"done\")\n",
    "\n",
    "for category in x.columns:\n",
    "    # category = 'G'\n",
    "    dh = (\n",
    "        pd.DataFrame({\"hit\": y, \"in_category\": x[category]})\n",
    "        .value_counts()\n",
    "        .unstack(fill_value=0)\n",
    "    )\n",
    "    # dp = pd.DataFrame({'hit': z_pos, 'in_category': x[category]}).value_counts().unstack(fill_value=0)\n",
    "    # dn = pd.DataFrame({'hit': z_neg, 'in_category': x[category]}).value_counts().unstack(fill_value=0)\n",
    "    # odds_ratio_pc = (dp1.loc[True, True] / dp1.loc[True, False]) / (dp1.loc[False, True] / dp1.loc[False, False])\n",
    "    print(category)\n",
    "    print(sp.stats.fisher_exact(dh))\n",
    "    # print(sp.stats.fisher_exact(dp))\n",
    "    # print(sp.stats.fisher_exact(dn))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    (gene_x_cog_category_matrix)\n",
    "    .loc[low_confidence_mwas_hit]\n",
    "    .join(\n",
    "        (gene_stats.ibd_mwas_oddsratio_pc > 1)\n",
    "        .to_frame(\"mwas_ibd_pos_assoc\")\n",
    "        .replace({True: \"pos\", False: \"neg\"})\n",
    "    )\n",
    "    .groupby(\"mwas_ibd_pos_assoc\")\n",
    "    .sum()\n",
    "    .T.join(cog_category_description)\n",
    "    .assign(total_genes_in_category=gene_x_cog_category_matrix.sum())\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = (\n",
    "    gene_stats[\n",
    "        lambda x: (x.prevalence_mwas_subject > 0.1)\n",
    "        & (x.prevalence_mwas_subject < 0.9)\n",
    "        & (x.num_mwas_subject > 20)\n",
    "        & (x.ibd_mwas_pvalue < 1e-4)\n",
    "    ]\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .join(gene_meta, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "d[\n",
    "    [\n",
    "        \"prevalence_spgc\",\n",
    "        \"prevalence_ref\",\n",
    "        \"prevalence_mwas_subject\",\n",
    "        \"num_mwas_subject\",\n",
    "        \"ibd_mwas_oddsratio_pc\",\n",
    "        \"ibd_mwas_pvalue\",\n",
    "        \"s__\",\n",
    "        \"cog_categories\",\n",
    "        \"description\",\n",
    "    ]\n",
    "].sort_values(\"ibd_mwas_oddsratio_pc\")\n",
    "\n",
    "# d = (\n",
    "#     gene_stats\n",
    "#     .join(species_taxonomy, on=\"species\")\n",
    "#     .sort_values(\"ibd_mwas_pvalue\")\n",
    "# )\n",
    "\n",
    "# d[d.ibd_mwas_pvalue < 1e-4].assign(\n",
    "#     log_oddsratio_sign=lambda x: (x.ibd_mwas_oddsratio_pc > 1).map(\n",
    "#         {True: \"+\", False: \"-\"}\n",
    "#     )\n",
    "# )[[\"species\", \"log_oddsratio_sign\"]].value_counts().unstack(\n",
    "#     \"log_oddsratio_sign\", fill_value=0\n",
    "# ).reindex(\n",
    "#     columns=[\"-\", \"+\"], fill_value=0\n",
    "# ).assign(\n",
    "#     total=lambda x: x[\"+\"] + x[\"-\"]\n",
    "# ).sort_values(\n",
    "#     [\"total\", \"+\"], ascending=False\n",
    "# ).rename(\n",
    "#     species_taxonomy.taxonomy_string\n",
    "# )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = gene_stats[\n",
    "    lambda x: (x.prevalence_mwas_subject > 0.1)\n",
    "    & (x.prevalence_mwas_subject < 0.9)\n",
    "    & (x.num_mwas_subject > 30)\n",
    "].join(gene_meta, rsuffix=\"_\")\n",
    "d = (\n",
    "    d.assign(fdr=fdrcorrection(d.ibd_mwas_pvalue)[1])\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    "    .assign(\n",
    "        log_oddsratio_sign=lambda x: (x.ibd_mwas_oddsratio_pc > 1).map(\n",
    "            {True: \"+\", False: \"-\"}\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "_results = {}\n",
    "for _col in cog_category_raw_columns:\n",
    "    _results[_col] = (\n",
    "        d[d.fdr < 0.2][[_col, \"log_oddsratio_sign\"]]\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex([True], fill_value=0)\n",
    "        .squeeze()\n",
    "    )\n",
    "\n",
    "_results = (\n",
    "    pd.DataFrame(_results)\n",
    "    .fillna(0)\n",
    "    .T.assign(total=lambda x: x[\"+\"] + x[\"-\"])\n",
    "    .sort_values([\"total\", \"+\"], ascending=False)\n",
    ")\n",
    "_results[_results.total > 0].astype(int)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = (\n",
    "    gene_stats[\n",
    "        lambda x: (x.prevalence_spgc > 0.1)\n",
    "        & (x.prevalence_spgc < 0.9)\n",
    "        & (x.num_mwas_subject > 40)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.ibd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    "    .join(gene_meta, rsuffix=\"_\")\n",
    ")\n",
    "\n",
    "d0[\n",
    "    [\n",
    "        \"nlength\",\n",
    "        \"description\",\n",
    "        \"species\",\n",
    "        \"f__\",\n",
    "        \"g__\",\n",
    "        \"s__\",\n",
    "        \"ibd_mwas_pvalue\",\n",
    "        \"ibd_mwas_oddsratio_pc\",\n",
    "        \"prevalence_spgc\",\n",
    "        \"prevalence_ref\",\n",
    "        # \"phylogenetic_i_spgc\",\n",
    "    ]\n",
    "][d0.fdr < 0.2]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = (\n",
    "    gene_stats[\n",
    "        lambda x: (x.prevalence_spgc > 0.1)\n",
    "        & (x.prevalence_spgc < 0.9)\n",
    "        & (x.num_mwas_subject > 20)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.cd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"cd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"cd_mwas_oddsratio_pc\",\n",
    "    \"cd_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"cd_mwas_oddsratio_pc\",\n",
    "    \"cd_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.1)],\n",
    "    s=20,\n",
    "    marker=\"o\",\n",
    "    edgecolor=\"k\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = (\n",
    "    gene_stats[\n",
    "        lambda x: (x.prevalence_spgc > 0.1)\n",
    "        & (x.prevalence_spgc < 0.9)\n",
    "        & (x.num_mwas_subject > 20)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.uc_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"uc_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"uc_mwas_oddsratio_pc\",\n",
    "    \"uc_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"uc_mwas_oddsratio_pc\",\n",
    "    \"uc_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.1)],\n",
    "    s=20,\n",
    "    marker=\"o\",\n",
    "    edgecolor=\"k\",\n",
    "    facecolor=\"none\",\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prevalence Class Enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _assign_prevalence_class(p):\n",
    "    if p > 0.95:\n",
    "        return \"core\"\n",
    "    elif p > 0.1:\n",
    "        return \"shell\"\n",
    "    elif p < 0.1:\n",
    "        return \"cloud\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_gene_class_enrichment_test.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(cog_category_gene_class_enrichment_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    gene_stats.dropna(subset=[\"prevalence_spgc\"])\n",
    "    .assign(\n",
    "        prevalence_class=lambda x: x.prevalence_spgc.dropna().map(\n",
    "            _assign_prevalence_class\n",
    "        )\n",
    "    )\n",
    "    .join(gene_meta, rsuffix=\"_\")\n",
    "    .join(gene_x_cog_category_matrix)\n",
    ")\n",
    "\n",
    "x = d[\"prevalence_class\"]\n",
    "y = (\n",
    "    d.loc[:, cog_category_order]\n",
    "    .fillna(False)\n",
    ")\n",
    "\n",
    "cog_category_gene_class_enrichment_test = {}\n",
    "\n",
    "for _prevalence_class, _cog_category in tqdm(\n",
    "    list(product(prevalence_class_order, cog_category_order))\n",
    "):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_prev_class=(x == _prevalence_class),\n",
    "                is_cog_category=y[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    contingency_table_p1 = contingency_table + 1\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_gene_class_enrichment_test[(_cog_category, _prevalence_class)] = dict(\n",
    "        pvalue=_test[1],\n",
    "        gene_count=contingency_table.loc[True, True],\n",
    "        odds_ratio_pc=(\n",
    "            (\n",
    "                contingency_table_p1.loc[True, True]\n",
    "                / contingency_table_p1.loc[True, False]\n",
    "            )\n",
    "            / (\n",
    "                contingency_table_p1.loc[False, True]\n",
    "                / contingency_table_p1.loc[False, False]\n",
    "            )\n",
    "        ),\n",
    "    )\n",
    "\n",
    "cog_category_gene_class_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_gene_class_enrichment_test,\n",
    "    )\n",
    "    .T.rename_axis(index=[\"cog_category\", \"prevalence_class\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio_pc=lambda x: np.log2(x.odds_ratio_pc),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_gene_class_enrichment_test.log2_odds_ratio_pc.unstack(\n",
    "    \"prevalence_class\"\n",
    ").sort_values(\"core\", ascending=False).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    cog_category_gene_class_enrichment_test.log2_odds_ratio_pc.unstack(\"prevalence_class\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "\n",
    "def _assign_significance_marker(pvalue):\n",
    "    if pvalue < 0.0001:\n",
    "        return \"*\"\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"prevalence_class\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = x[\"core\"].sort_values(ascending=False).index\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-3,\n",
    "    vmax=3,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.assign(\n",
    "    prevalence_ref=lambda x: x.prevalence_ref.fillna(0),\n",
    "    prevalence_spgc=lambda x: x.prevalence_spgc.fillna(0),\n",
    ")[lambda x: (x.prevalence_ref > 0) | (x.prevalence_spgc > 0)]\n",
    "\n",
    "plt.hist2d(\n",
    "    \"prevalence_ref\",\n",
    "    \"prevalence_spgc\",\n",
    "    data=d,\n",
    "    bins=np.linspace(0, 1, num=26),\n",
    "    norm=mpl.colors.PowerNorm(1 / 5),\n",
    ")\n",
    "print(sp.stats.pearsonr(d[\"prevalence_ref\"], d[\"prevalence_spgc\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.assign(\n",
    "    prevalence_ref=lambda x: x.prevalence_ref.fillna(0),\n",
    "    prevalence_spgc=lambda x: x.prevalence_spgc.fillna(0),\n",
    ")[lambda x: (x.prevalence_ref > 0.1) & (x.prevalence_spgc > 0.1) & (x.prevalence_ref < 0.9) & (x.prevalence_spgc < 0.9)]\n",
    "\n",
    "plt.hist2d(\n",
    "    \"prevalence_ref\",\n",
    "    \"prevalence_spgc\",\n",
    "    data=d,\n",
    "    bins=np.linspace(0, 1, num=26),\n",
    "    # norm=mpl.colors.PowerNorm(1 / 5),\n",
    ")\n",
    "print(sp.stats.pearsonr(d[\"prevalence_ref\"], d[\"prevalence_spgc\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    spgc_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    ")\n",
    "plt.hist(d.spgc_core_gene_tally, label=\"core\", alpha=0.7)\n",
    "plt.hist(d.spgc_shell_gene_tally, label=\"shell\", alpha=0.7)\n",
    "plt.hist(d.spgc_cloud_gene_tally, label=\"cloud\", alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    ref_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    ")\n",
    "plt.hist(d.ref_core_gene_tally, label=\"core\", alpha=0.7)\n",
    "plt.hist(d.ref_shell_gene_tally, label=\"shell\", alpha=0.7)\n",
    "plt.hist(d.ref_cloud_gene_tally, label=\"cloud\", alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    spgc_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level=\"species\")\n",
    "    .median()\n",
    ")\n",
    "\n",
    "\n",
    "plt.hist(d.spgc_core_gene_tally, label=\"core\", alpha=0.7)\n",
    "plt.hist(d.spgc_shell_gene_tally, label=\"shell\", alpha=0.7)\n",
    "plt.hist(d.spgc_cloud_gene_tally, label=\"cloud\", alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    ref_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level=\"species\")\n",
    "    .median()\n",
    ")\n",
    "plt.hist(d.ref_core_gene_tally, label=\"core\", alpha=0.7)\n",
    "plt.hist(d.ref_shell_gene_tally, label=\"shell\", alpha=0.7)\n",
    "plt.hist(d.ref_cloud_gene_tally, label=\"cloud\", alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    spgc_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level=\"species\")\n",
    "    .median()\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "_phylum_list = d0.p__.unique()\n",
    "\n",
    "fig, axs = plt.subplots(3, sharex=True)\n",
    "bins = np.linspace(0, 1, num=21)\n",
    "for partition, ax in zip(\n",
    "    [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"], axs\n",
    "):\n",
    "    for p__, d1 in d0.groupby(\"p__\"):\n",
    "        sns.kdeplot(\n",
    "            d1[partition], color=phylum_palette[p__], ax=ax, label=\"__nolegend__\"\n",
    "        )\n",
    "    ax.set_ylabel(partition[len(\"spgc_\") : -len(\"_gene_tally\")])\n",
    "\n",
    "for p__, _ in d0.groupby(\"p__\"):\n",
    "    axs[0].plot([], [], color=phylum_palette[p__], label=p__)\n",
    "axs[0].legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "\n",
    "# plt.hist(d.spgc_core_gene_tally, label='core', alpha=0.7)\n",
    "# plt.hist(d.spgc_shell_gene_tally, label='shell', alpha=0.7)\n",
    "# plt.hist(d.spgc_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    ref_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level=\"species\")\n",
    "    .median()\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "_phylum_list = d0.p__.unique()\n",
    "\n",
    "fig, axs = plt.subplots(3, sharex=True)\n",
    "bins = np.linspace(0, 1, num=21)\n",
    "for partition, ax in zip(\n",
    "    [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"], axs\n",
    "):\n",
    "    for p__, d1 in d0.groupby(\"p__\"):\n",
    "        sns.kdeplot(\n",
    "            d1[partition], color=phylum_palette[p__], ax=ax, label=\"__nolegend__\"\n",
    "        )\n",
    "    ax.set_ylabel(partition[len(\"ref_\") : -len(\"_gene_tally\")])\n",
    "\n",
    "for p__, _ in d0.groupby(\"p__\"):\n",
    "    axs[0].plot([], [], color=phylum_palette[p__], label=p__)\n",
    "axs[0].legend(bbox_to_anchor=(1, 1))\n",
    "\n",
    "\n",
    "# plt.hist(d.ref_core_gene_tally, label='core', alpha=0.7)\n",
    "# plt.hist(d.ref_shell_gene_tally, label='shell', alpha=0.7)\n",
    "# plt.hist(d.ref_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### \"Phylogenetic\" Signal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = idxwhere(ref_strain_stats.species.value_counts() > 100)\n",
    "print(len(_species_list))\n",
    "_gene_list = idxwhere(gene_stats.species.isin(_species_list) & (gene_stats.prevalence_ref > 0.2) & (gene_stats.prevalence_ref < 0.8))\n",
    "print(len(_gene_list))\n",
    "\n",
    "d = gene_stats.loc[_gene_list]\n",
    "plt.hist(d.phylogenetic_i_ref, bins=100)\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.loc[idxwhere(d.phylogenetic_i_ref < -0.01)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.dropna(subset=[\"phylogenetic_i_ref\", \"phylogenetic_i_spgc\"])\n",
    "\n",
    "plt.hist2d(\n",
    "    \"phylogenetic_i_ref\",\n",
    "    \"phylogenetic_i_spgc\",\n",
    "    data=d,\n",
    "    bins=50,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=1),\n",
    ")\n",
    "print(sp.stats.pearsonr(d[\"phylogenetic_i_ref\"], d[\"phylogenetic_i_spgc\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(\n",
    "    gene_stats[lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9)]\n",
    "    .phylogenetic_i_spgc.sort_values()\n",
    "    .values\n",
    ")\n",
    "plt.plot(\n",
    "    gene_stats[lambda x: (x.prevalence_spgc > 0.2) & (x.prevalence_spgc < 0.8)]\n",
    "    .phylogenetic_i_spgc.sort_values()\n",
    "    .values\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_median_phylogenetic_i_ref = {}\n",
    "cog_category_median_phylogenetic_i_spgc = {}\n",
    "\n",
    "d = gene_stats.join(gene_x_cog_category_matrix)\n",
    "\n",
    "\n",
    "for _col in cog_category_order:\n",
    "    cog_category_median_phylogenetic_i_ref[_col] = d[\n",
    "        lambda x: (x.prevalence_ref > 0.1)\n",
    "        & (x.prevalence_ref < 0.9)\n",
    "        & (x[_col])\n",
    "        & x.species.isin([\"102506\"])\n",
    "    ].phylogenetic_i_ref.median()\n",
    "    cog_category_median_phylogenetic_i_spgc[_col] = d[\n",
    "        lambda x: (x.prevalence_spgc > 0.1)\n",
    "        & (x.prevalence_spgc < 0.9)\n",
    "        & (x[_col])\n",
    "        & x.species.isin([\"102506\"])\n",
    "    ].phylogenetic_i_spgc.median()\n",
    "\n",
    "cog_category_median_phylogenetic_i_ref = pd.Series(\n",
    "    cog_category_median_phylogenetic_i_ref\n",
    ")\n",
    "cog_category_median_phylogenetic_i_spgc = pd.Series(\n",
    "    cog_category_median_phylogenetic_i_spgc\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(\n",
    "    [cog_category_median_phylogenetic_i_spgc, cog_category_median_phylogenetic_i_ref]\n",
    ").T.sort_values(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.violinplot(\n",
    "    data=gene_stats.join(gene_x_cog_category_matrix)[lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9)],\n",
    "    x=\"no_category\",\n",
    "    y=\"phylogenetic_i_ref\",\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz2",
   "language": "python",
   "name": "toolz2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}