{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Project Template"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "from datetime import datetime\n",
    "from glob import glob\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import lib.thisproject.data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy_inpath = \"ref/gtpro/species_taxonomy_ext.tsv\"\n",
    "\n",
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(species_taxonomy_inpath)\n",
    "species_taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"cog_category\", \"description\"],\n",
    "    index_col=\"cog_category\",\n",
    ").assign(description=lambda x: x.index + \": \" + x.description)\n",
    "cog_category_description.loc[\"no_category\", \"description\"] = \"-: No Annotation\"\n",
    "cog_category_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_columns = [\n",
    "    \"cog_category_A\",\n",
    "    \"cog_category_B\",\n",
    "    \"cog_category_C\",\n",
    "    \"cog_category_D\",\n",
    "    \"cog_category_E\",\n",
    "    \"cog_category_F\",\n",
    "    \"cog_category_G\",\n",
    "    \"cog_category_H\",\n",
    "    \"cog_category_I\",\n",
    "    \"cog_category_J\",\n",
    "    \"cog_category_K\",\n",
    "    \"cog_category_L\",\n",
    "    \"cog_category_M\",\n",
    "    \"cog_category_N\",\n",
    "    \"cog_category_O\",\n",
    "    \"cog_category_P\",\n",
    "    \"cog_category_Q\",\n",
    "    \"cog_category_R\",\n",
    "    \"cog_category_S\",\n",
    "    \"cog_category_T\",\n",
    "    \"cog_category_U\",\n",
    "    \"cog_category_V\",\n",
    "    \"cog_category_W\",\n",
    "    \"cog_category_X\",\n",
    "    \"cog_category_Y\",\n",
    "    \"cog_category_Z\",\n",
    "    \"cog_category_no_category\",\n",
    "]\n",
    "\n",
    "cog_category_order = [x[len('cog_category_'):] for x in cog_category_columns]\n",
    "prevalence_class_order = ['core', 'shell', 'cloud']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats = []\n",
    "\n",
    "for path in tqdm(\n",
    "    glob(\n",
    "        \"data/group/xjin_ucfmt_hmp2/species/sp-*/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.gene_stats.tsv\"\n",
    "    )\n",
    "):\n",
    "    species = path[\n",
    "        len(\"data/group/xjin_ucfmt_hmp2/species/sp-\") : -len(\n",
    "            \"/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.gene_stats.tsv\"\n",
    "        )\n",
    "    ]\n",
    "    d = pd.read_table(path)\n",
    "    gene_stats.append(d.assign(species=species))\n",
    "gene_stats = pd.concat(gene_stats).set_index(\"gene_id\")\n",
    "gene_stats[cog_category_columns] = gene_stats[cog_category_columns].fillna(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_stats = []\n",
    "\n",
    "for path in tqdm(\n",
    "    glob(\n",
    "        \"data/group/xjin_ucfmt_hmp2/species/sp-*/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_strain_stats.tsv\"\n",
    "    )\n",
    "):\n",
    "    species = path[\n",
    "        len(\"data/group/xjin_ucfmt_hmp2/species/sp-\") : -len(\n",
    "            \"/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_strain_stats.tsv\"\n",
    "        )\n",
    "    ]\n",
    "    d = pd.read_table(path)\n",
    "    spgc_strain_stats.append(d.assign(species=species))\n",
    "spgc_strain_stats = pd.concat(spgc_strain_stats).join(species_taxonomy, on='species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_strain_stats = []\n",
    "\n",
    "for path in tqdm(\n",
    "    glob(\n",
    "        \"data/group/xjin_ucfmt_hmp2/species/sp-*/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.ref_strain_stats.tsv\"\n",
    "    )\n",
    "):\n",
    "    species = path[\n",
    "        len(\"data/group/xjin_ucfmt_hmp2/species/sp-\") : -len(\n",
    "            \"/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.ref_strain_stats.tsv\"\n",
    "        )\n",
    "    ]\n",
    "    d = pd.read_table(path)\n",
    "    ref_strain_stats.append(d.assign(species=species))\n",
    "ref_strain_stats = pd.concat(ref_strain_stats).join(species_taxonomy, on='species')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylum_palette = lib.plot.construct_ordered_palette(sorted(ref_strain_stats.p__.unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-7, 0, num=50)\n",
    "# bins = np.linspace(0, 1, num=51)\n",
    "\n",
    "x = gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 20)\n",
    "].ibd_mwas_pvalue.sort_values()\n",
    "\n",
    "plt.hist(\n",
    "    x,\n",
    "    alpha=0.5,\n",
    "    bins=bins,\n",
    ")\n",
    "\n",
    "plt.plot(bins[1:], (bins[1:] - bins[:-1]) * x.shape)\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=gene_stats[\n",
    "        lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 20)\n",
    "    ],\n",
    "    s=4,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")\n",
    "ax = plt.gca()\n",
    "ax.invert_yaxis()\n",
    "plt.xlabel(\"Odds-Ratio (w/ pseudo-counts)\")\n",
    "plt.ylabel(\"P-value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (gene_stats[\n",
    "        lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 40)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.ibd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"ibd_mwas_oddsratio_pc\",\n",
    "    \"ibd_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.2)],\n",
    "    s=20,\n",
    "    marker='o',\n",
    "    edgecolor='k',\n",
    "    facecolor='none',\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 40)\n",
    "]\n",
    "d = (\n",
    "    d.assign(fdr=fdrcorrection(d.ibd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "d[d.fdr<0.2].assign(log_oddsratio_sign=lambda x: (x.ibd_mwas_oddsratio_pc > 1).map({True: '+', False: '-'}))[['species', 'log_oddsratio_sign']].value_counts().unstack('log_oddsratio_sign', fill_value=0).reindex(columns=['-', '+'], fill_value=0).assign(total=lambda x: x['+'] + x['-']).sort_values(['total', '+'], ascending=False).rename(species_taxonomy.taxonomy_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 40)\n",
    "]\n",
    "d = (\n",
    "    d.assign(fdr=fdrcorrection(d.ibd_mwas_pvalue)[1])\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    "    .assign(log_oddsratio_sign=lambda x: (x.ibd_mwas_oddsratio_pc > 1).map({True: '+', False: '-'}))\n",
    ")\n",
    "\n",
    "_results = {}\n",
    "for _col in cog_category_columns:\n",
    "    _results[_col]  = d[d.fdr<0.2][[_col, 'log_oddsratio_sign']].value_counts().unstack().reindex([True], fill_value=0).squeeze()\n",
    "\n",
    "_results = pd.DataFrame(_results).fillna(0).T.assign(total=lambda x: x['+'] + x['-']).sort_values(['total', '+'], ascending=False)\n",
    "_results[_results.total > 0].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 40)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.ibd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"ibd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "d0[\n",
    "    [\n",
    "        \"nlength\",\n",
    "        \"description\",\n",
    "        \"species\",\n",
    "        \"f__\",\n",
    "        \"g__\",\n",
    "        \"s__\",\n",
    "        \"ibd_mwas_pvalue\",\n",
    "        \"ibd_mwas_oddsratio_pc\",\n",
    "        \"prevalence_spgc\",\n",
    "        \"prevalence_ref\",\n",
    "        \"phylogenetic_r_spgc\",\n",
    "    ]\n",
    "][d0.fdr<0.2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 20)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.cd_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"cd_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"cd_mwas_oddsratio_pc\",\n",
    "    \"cd_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"cd_mwas_oddsratio_pc\",\n",
    "    \"cd_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.1)],\n",
    "    s=20,\n",
    "    marker='o',\n",
    "    edgecolor='k',\n",
    "    facecolor='none',\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (gene_stats[\n",
    "    lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x.num_mwas_subject > 20)\n",
    "    ]\n",
    "    .assign(fdr=lambda x: fdrcorrection(x.uc_mwas_pvalue)[1])\n",
    "    .join(species_taxonomy, on=\"species\")\n",
    "    .sort_values(\"uc_mwas_pvalue\")\n",
    ")\n",
    "\n",
    "plt.hist2d(\n",
    "    \"uc_mwas_oddsratio_pc\",\n",
    "    \"uc_mwas_pvalue\",\n",
    "    data=d0,\n",
    "    bins=(np.logspace(-3, 3, num=100), np.logspace(-7, 0, num=100)),\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=2, linscale=3, vmin=1, vmax=1e5),\n",
    "    cmin=1,\n",
    "    cmap=\"copper_r\",\n",
    ")\n",
    "\n",
    "plt.colorbar()\n",
    "\n",
    "plt.scatter(\n",
    "    \"uc_mwas_oddsratio_pc\",\n",
    "    \"uc_mwas_pvalue\",\n",
    "    data=d0[(d0.fdr < 0.1)],\n",
    "    s=20,\n",
    "    marker='o',\n",
    "    edgecolor='k',\n",
    "    facecolor='none',\n",
    "    lw=1,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.xscale('log')\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Log(Odds Ratio)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "# d1 = d0[lambda x: (x.gene == 'UHGG000638_01150') & (x.pvalue < 1e-3)]\n",
    "# for _, d2 in d1.iterrows():\n",
    "#     plt.annotate('', xy=(d2.log_logratio_pc, d2.pvalue), ha='left', va='bottom', xytext=(10, 10), textcoords=\"offset points\", arrowprops=dict())\n",
    "# d1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _assign_prevalence_class(p):\n",
    "    if p > 0.95:\n",
    "        return \"core\"\n",
    "    elif p > 0.1:\n",
    "        return \"shell\"\n",
    "    elif p < 0.1:\n",
    "        return \"cloud\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.dropna(subset=['prevalence_spgc']).assign(prevalence_class=lambda x: x.prevalence_spgc.dropna().map(_assign_prevalence_class))\n",
    "\n",
    "x = d['prevalence_class']\n",
    "y = d.loc[:, cog_category_columns].rename(columns=lambda x: x[len('cog_category_'):]).fillna(False)\n",
    "\n",
    "cog_category_gene_class_enrichment_test = []\n",
    "\n",
    "for _prevalence_class, _cog_category in tqdm(list(product(\n",
    "    prevalence_class_order, cog_category_order\n",
    "))):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_prev_class=(x == _prevalence_class),\n",
    "                is_cog_category=y[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_gene_class_enrichment_test.append(\n",
    "        (\n",
    "            _prevalence_class,\n",
    "            _cog_category,\n",
    "            _test[0],\n",
    "            _test[1],\n",
    "            contingency_table.loc[True, True],\n",
    "        )\n",
    "    )\n",
    "\n",
    "cog_category_gene_class_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_gene_class_enrichment_test,\n",
    "        columns=[\n",
    "            \"prevalence_class\",\n",
    "            \"cog_category\",\n",
    "            \"statistic\",\n",
    "            \"pvalue\",\n",
    "            \"gene_count\",\n",
    "        ],\n",
    "    )\n",
    "    .set_index([\"prevalence_class\", \"cog_category\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio=lambda x: np.log2(x.statistic),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    cog_category_gene_class_enrichment_test.log2_odds_ratio.unstack(\"prevalence_class\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "\n",
    "def _assign_significance_marker(pvalue):\n",
    "    if pvalue < 0.0001:\n",
    "        return \"*\"\n",
    "    else:\n",
    "        return \"\"\n",
    "\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"prevalence_class\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = x[\"core\"].sort_values(ascending=False).index\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-3,\n",
    "    vmax=3,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats.columns"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "[\n",
    "    lambda x: ((x.prevalence_ref > 0.2) | (x.prevalence_spgc > 0.2))\n",
    "    & ((x.prevalence_ref < 0.8) | (x.prevalence_spgc < 0.8))\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.dropna(subset=[\"phylogenetic_r_ref\", \"phylogenetic_r_spgc\"])\n",
    "\n",
    "plt.hist2d(\n",
    "    \"phylogenetic_r_ref\",\n",
    "    \"phylogenetic_r_spgc\",\n",
    "    data=d,\n",
    "    bins=50,\n",
    "    norm=mpl.colors.SymLogNorm(linthresh=1),\n",
    ")\n",
    "print(sp.stats.pearsonr(d[\"phylogenetic_r_ref\"], d[\"phylogenetic_r_spgc\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.plot(gene_stats[lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9)].phylogenetic_r_spgc.sort_values().values)\n",
    "plt.plot(gene_stats[lambda x: (x.prevalence_spgc > 0.2) & (x.prevalence_spgc < 0.8)].phylogenetic_r_spgc.sort_values().values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats[\n",
    "        lambda x: (x.prevalence_ref > 0.1)\n",
    "        & (x.prevalence_ref < 0.9)\n",
    "        & (x[_col])\n",
    "    ].phylogenetic_r_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_median_phylogenetic_r_ref = {}\n",
    "cog_category_median_phylogenetic_r_spgc = {}\n",
    "\n",
    "\n",
    "for _col in cog_category_columns:\n",
    "    cog_category_median_phylogenetic_r_ref[_col] = gene_stats[lambda x: (x.prevalence_ref > 0.1) & (x.prevalence_ref < 0.9) & (x[_col]) & x.species.isin(['102506'])].phylogenetic_r_ref.median()\n",
    "    cog_category_median_phylogenetic_r_spgc[_col] = gene_stats[lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9) & (x[_col]) & x.species.isin(['102506'])].phylogenetic_r_spgc.median()\n",
    "\n",
    "cog_category_median_phylogenetic_r_ref = pd.Series(cog_category_median_phylogenetic_r_ref)\n",
    "cog_category_median_phylogenetic_r_spgc = pd.Series(cog_category_median_phylogenetic_r_spgc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame([cog_category_median_phylogenetic_r_spgc, cog_category_median_phylogenetic_r_ref]).T.sort_values(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.violinplot(data=gene_stats[lambda x: (x.prevalence_spgc > 0.1) & (x.prevalence_spgc < 0.9)], x='cog_category_no_category', y='phylogenetic_r_spgc')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = gene_stats.assign(\n",
    "        prevalence_ref=lambda x: x.prevalence_ref.fillna(0),\n",
    "        prevalence_spgc=lambda x: x.prevalence_spgc.fillna(0),\n",
    "    )[lambda x: (x.prevalence_ref > 0) | (x.prevalence_spgc > 0)]\n",
    "\n",
    "plt.hist2d(\n",
    "    'prevalence_ref',\n",
    "    'prevalence_spgc',\n",
    "    data=d,\n",
    "    bins=np.linspace(0, 1, num=26),\n",
    "    norm=mpl.colors.PowerNorm(1/5),\n",
    ")\n",
    "print(sp.stats.pearsonr(d['prevalence_ref'], d['prevalence_spgc']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = spgc_strain_stats.sort_values('taxonomy_string')\n",
    "\n",
    "fit = smf.ols(f\"nearest_ref_gene_raw_diss ~ np.log10(nearest_ref_geno_diss) + C(species, Sum)\", data=d0).fit()\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "ax = axs[0]\n",
    "for p__, d1 in d0.groupby(\"p__\"):\n",
    "    ax.scatter(\"nearest_ref_geno_diss\", \"nearest_ref_gene_raw_diss\", data=d1, label='__nolegend__', color=phylum_palette[p__], s=5, alpha=0.5)\n",
    "    ax.scatter([], [], label=p__, color=phylum_palette[p__], s=10)\n",
    "\n",
    "xx = np.logspace(-3, 0)\n",
    "ax.plot(xx, np.log10(xx) * fit.params['np.log10(nearest_ref_geno_diss)'] + fit.params['Intercept'], color='k')\n",
    "p_slope = fit.pvalues['np.log10(nearest_ref_geno_diss)']\n",
    "slope = fit.params['np.log10(nearest_ref_geno_diss)']  # TODO\n",
    "r2_adj = fit.rsquared_adj\n",
    "ax.annotate(f'$\\\\beta_{{\\\\mathrm{{slope}}}}$={slope:0.3f}\\n$P_{{\\\\mathrm{{slope}}}}$={p_slope:0.0e}\\n$R^2_{{\\\\mathrm{{adj}}}}$={r2_adj:0.2%}', xy=(0.1, 0.9), xycoords='axes fraction', va='top')\n",
    "ax.annotate(f'', xy=(0.1, 0.9), xycoords='axes fraction')\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel('SNP Profile Dissimilarity')\n",
    "ax.set_ylabel('Gene Content Dissimilarity\\n(filtered, Jaccard)')\n",
    "# plt.scatter(spgc_strain_stats.nearest_ref_geno_diss, spgc_strain_stats.nearest_ref_gene_raw_diss, s=1)\n",
    "\n",
    "ax = axs[1]\n",
    "sns.boxplot(x='p__', y='nearest_ref_gene_raw_diss', data=d0, ax=ax, palette=phylum_palette)\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel('')\n",
    "\n",
    "ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = ref_strain_stats.sort_values('taxonomy_string')\n",
    "\n",
    "fit = smf.ols(f\"nearest_ref_gene_raw_diss ~ np.log10(nearest_ref_geno_diss) + C(species, Sum)\", data=d0).fit()\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "\n",
    "ax = axs[0]\n",
    "for p__, d1 in d0.groupby(\"p__\"):\n",
    "    ax.scatter(\"nearest_ref_geno_diss\", \"nearest_ref_gene_raw_diss\", data=d1, label='__nolegend__', color=phylum_palette[p__], s=5, alpha=0.5)\n",
    "    ax.scatter([], [], label=p__, color=phylum_palette[p__], s=10)\n",
    "\n",
    "xx = np.logspace(-3, 0)\n",
    "ax.plot(xx, np.log10(xx) * fit.params['np.log10(nearest_ref_geno_diss)'] + fit.params['Intercept'], color='k')\n",
    "p_slope = fit.pvalues['np.log10(nearest_ref_geno_diss)']\n",
    "slope = fit.params['np.log10(nearest_ref_geno_diss)']  # TODO\n",
    "r2_adj = fit.rsquared_adj\n",
    "ax.annotate(f'$\\\\beta_{{\\\\mathrm{{slope}}}}$={slope:0.3f}\\n$P_{{\\\\mathrm{{slope}}}}$={p_slope:0.0e}\\n$R^2_{{\\\\mathrm{{adj}}}}$={r2_adj:0.2%}', xy=(0.1, 0.9), xycoords='axes fraction', va='top')\n",
    "ax.annotate(f'', xy=(0.1, 0.9), xycoords='axes fraction')\n",
    "\n",
    "# ax.plot(\"nearest_ref_geno_diss\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "# ax.legend(bbox_to_anchor=(1, 1), markerscale=4)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "\n",
    "ax.set_xlabel('SNP Profile Dissimilarity')\n",
    "ax.set_ylabel('Gene Content Dissimilarity\\n(filtered, Jaccard)')\n",
    "\n",
    "ax = axs[1]\n",
    "sns.violinplot(x='p__', y='nearest_ref_gene_raw_diss', data=d0, ax=ax, palette=phylum_palette)\n",
    "lib.plot.rotate_xticklabels(ax=ax)\n",
    "ax.set_ylabel('')\n",
    "\n",
    "ax.set_ylim(-0.005, 0.605)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1)\n",
    "\n",
    "fig, axs = plt.subplots(2, sharex=True)\n",
    "ax = axs[0]\n",
    "ax.hist(spgc_strain_stats.nearest_ref_geno_diss, bins=bins, alpha=0.5, density=True, label='SPGC', color='tab:purple')\n",
    "ax.hist(ref_strain_stats.nearest_ref_geno_diss, bins=bins, alpha=0.5, density=True, label='Reference', color='tab:green')\n",
    "ax.set_xlabel('Closest Ref. SNP Dissimilarity')\n",
    "ax.set_ylabel('Density')\n",
    "ax.legend()\n",
    "\n",
    "ax = axs[1]\n",
    "ax.hist(spgc_strain_stats.nearest_ref_gene_raw_diss, bins=bins, alpha=0.5, density=True, label='SPGC', color='tab:purple')\n",
    "ax.hist(ref_strain_stats.nearest_ref_gene_raw_diss, bins=bins, alpha=0.5, density=True, label='Reference', color='tab:green')\n",
    "ax.set_xlabel('Closest Ref. Gene Dissimilarity (Jaccard)')\n",
    "ax.set_ylabel('Density')\n",
    "\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1)\n",
    "plt.hist(spgc_strain_stats.nearest_ref_gene_raw_diss, bins=bins, alpha=0.5, density=True, label='SPGC', color='tab:purple')\n",
    "plt.hist(ref_strain_stats.nearest_ref_gene_raw_diss, bins=bins, alpha=0.5, density=True, label='Reference', color='tab:green')\n",
    "plt.xlabel('Closest Ref. Gene Dissimilarity (Jaccard)')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(spgc_strain_stats.nearest_ref_geno_diss, bins=np.logspace(-4, 0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = spgc_strain_stats.reset_index().set_index(['species', 'strain'])[['spgc_core_gene_tally', 'spgc_shell_gene_tally', 'spgc_cloud_gene_tally']].apply(lambda x: x / x.sum(), axis=1)\n",
    "plt.hist(d.spgc_core_gene_tally, label='core', alpha=0.7)\n",
    "plt.hist(d.spgc_shell_gene_tally, label='shell', alpha=0.7)\n",
    "plt.hist(d.spgc_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = ref_strain_stats.reset_index().set_index(['species', 'strain'])[['ref_core_gene_tally', 'ref_shell_gene_tally', 'ref_cloud_gene_tally']].apply(lambda x: x / x.sum(), axis=1)\n",
    "plt.hist(d.ref_core_gene_tally, label='core', alpha=0.7)\n",
    "plt.hist(d.ref_shell_gene_tally, label='shell', alpha=0.7)\n",
    "plt.hist(d.ref_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    spgc_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level='species').median()\n",
    ")\n",
    "\n",
    "\n",
    "plt.hist(d.spgc_core_gene_tally, label='core', alpha=0.7)\n",
    "plt.hist(d.spgc_shell_gene_tally, label='shell', alpha=0.7)\n",
    "plt.hist(d.spgc_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    ref_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level='species').median()\n",
    ")\n",
    "plt.hist(d.ref_core_gene_tally, label='core', alpha=0.7)\n",
    "plt.hist(d.ref_shell_gene_tally, label='shell', alpha=0.7)\n",
    "plt.hist(d.ref_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    spgc_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"spgc_core_gene_tally\", \"spgc_shell_gene_tally\", \"spgc_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level='species').median()\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "_phylum_list = d0.p__.unique()\n",
    "\n",
    "fig, axs = plt.subplots(3, sharex=True)\n",
    "bins = np.linspace(0, 1, num=21)\n",
    "for partition, ax in zip(['spgc_core_gene_tally', 'spgc_shell_gene_tally', 'spgc_cloud_gene_tally'], axs):\n",
    "    for p__, d1 in d0.groupby('p__'):\n",
    "        sns.kdeplot(d1[partition], color=phylum_palette[p__], ax=ax, label='__nolegend__')\n",
    "    ax.set_ylabel(partition[len('spgc_'):-len('_gene_tally')])\n",
    "\n",
    "for p__, _ in d0.groupby('p__'):\n",
    "    axs[0].plot([], [], color=phylum_palette[p__], label=p__)\n",
    "axs[0].legend(bbox_to_anchor=(1, 1))\n",
    "    \n",
    "\n",
    "\n",
    "# plt.hist(d.spgc_core_gene_tally, label='core', alpha=0.7)\n",
    "# plt.hist(d.spgc_shell_gene_tally, label='shell', alpha=0.7)\n",
    "# plt.hist(d.spgc_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    ref_strain_stats.reset_index()\n",
    "    .set_index([\"species\", \"strain\"])[\n",
    "        [\"ref_core_gene_tally\", \"ref_shell_gene_tally\", \"ref_cloud_gene_tally\"]\n",
    "    ]\n",
    "    .apply(lambda x: x / x.sum(), axis=1)\n",
    "    .groupby(level='species').median()\n",
    "    .join(species_taxonomy)\n",
    ")\n",
    "_phylum_list = d0.p__.unique()\n",
    "\n",
    "fig, axs = plt.subplots(3, sharex=True)\n",
    "bins = np.linspace(0, 1, num=21)\n",
    "for partition, ax in zip(['ref_core_gene_tally', 'ref_shell_gene_tally', 'ref_cloud_gene_tally'], axs):\n",
    "    for p__, d1 in d0.groupby('p__'):\n",
    "        sns.kdeplot(d1[partition], color=phylum_palette[p__], ax=ax, label='__nolegend__')\n",
    "    ax.set_ylabel(partition[len('ref_'):-len('_gene_tally')])\n",
    "\n",
    "for p__, _ in d0.groupby('p__'):\n",
    "    axs[0].plot([], [], color=phylum_palette[p__], label=p__)\n",
    "axs[0].legend(bbox_to_anchor=(1, 1))\n",
    "    \n",
    "\n",
    "\n",
    "# plt.hist(d.ref_core_gene_tally, label='core', alpha=0.7)\n",
    "# plt.hist(d.ref_shell_gene_tally, label='shell', alpha=0.7)\n",
    "# plt.hist(d.ref_cloud_gene_tally, label='cloud', alpha=0.7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "default",
   "language": "python",
   "name": "default"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}