{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "from warnings import filterwarnings\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "\n",
    "# from fastcluster import linkage\n",
    "from scipy.cluster.hierarchy import linkage\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from statsmodels.graphics.regressionplots import influence_plot\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "import lib.thisproject.data\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _calculate_2tailed_pvalue_from_perm(obs, perms):\n",
    "    hypoth_left = perms > obs\n",
    "    hypoth_right = perms < obs\n",
    "    null_p_left = (hypoth_left.sum() + 1) / (len(hypoth_left) + 1)\n",
    "    null_p_right = (hypoth_right.sum() + 1) / (len(hypoth_right) + 1)\n",
    "    return np.minimum(null_p_left, null_p_right) * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def linkage_order(linkage, labels):\n",
    "    return labels[sp.cluster.hierarchy.to_tree(linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "\n",
    "def is_prime(n):\n",
    "    if n <= 1:\n",
    "        return False\n",
    "    for i in range(2, int(n**0.5) + 1):\n",
    "        if n % i == 0:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "def iterate_primes_up_to(n, return_index=False):\n",
    "    n = int(np.ceil(n))\n",
    "    idx = 0\n",
    "    for i in range(n):\n",
    "        if is_prime(i):\n",
    "            if return_index:\n",
    "                yield (idx, i)\n",
    "            else:\n",
    "                yield i\n",
    "            idx += 1\n",
    "\n",
    "\n",
    "def maximally_shuffled_order(sorted_order):\n",
    "    n = len(sorted_order)\n",
    "    primes_list = list(iterate_primes_up_to(np.sqrt(n)))\n",
    "    table = pd.DataFrame(np.arange(n), index=sorted_order, columns=[\"original_order\"])\n",
    "    for prime in primes_list:\n",
    "        table[prime] = table.original_order % prime\n",
    "    table.sort_values(primes_list).original_order.values\n",
    "    table = table.assign(new_order=table.sort_values(primes_list).original_order.values)\n",
    "    z = table.sort_values(\"new_order\").original_order.values\n",
    "    table[\"delta\"] = [np.nan] + list(z[1:] - z[:-1])\n",
    "    return table.sort_values(\"new_order\").index.to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_meta = pd.read_table(\n",
    "    \"ref/cog-20.meta.tsv\",\n",
    "    encoding=\"latin10\",\n",
    "    names=[\"cog\", \"categories\", \"description\", \"gene_name\", \"pathway\", \"_5\", \"color\"],\n",
    "    index_col=\"cog\",\n",
    ")\n",
    "cog_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_meta = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"category\", \"color\", \"description\"],\n",
    "    index_col=\"category\",\n",
    ")\n",
    "cog_category_meta"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# NOTE: If I decide that it's the filenames that are wrong, then I need to swap labels in the metagenomic results (including sfacts worlds and\n",
    "# species relative abundances).\n",
    "\n",
    "SAMPLE_SWAP_CORRECTION = {\n",
    "    # CF_667 through CF_672 should actually be labeled CF_531 - CF_537\n",
    "    'CF_667': 'CF_531',\n",
    "    'CF_668': 'CF_532',\n",
    "    'CF_669': 'CF_533',\n",
    "    'CF_670': 'CF_534',\n",
    "    'CF_671': 'CF_535',\n",
    "    'CF_672': 'CF_536',\n",
    "\n",
    "    # CF_11 <-> CF_15\n",
    "    'CF_11': 'CF_15',\n",
    "    'CF_15': 'CF_11',\n",
    "\n",
    "    # Some/half of the newly shared samples\n",
    "    CF_379\n",
    "    CF_380\n",
    "    CF_381\n",
    "    CF_384\n",
    "    CF_385\n",
    "    CF_386\n",
    "    CF_426\n",
    "    CF_427\n",
    "    CF_428\n",
    "    CF_429\n",
    "    CF_430\n",
    "    CF_431\n",
    "    CF_395\n",
    "    CF_397\n",
    "    CF_402\n",
    "    CF_406\n",
    "    CF_409\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pair_type_palette = {\n",
    "    \"EEN\": \"teal\",\n",
    "    \"PostEEN\": \"mediumblue\",\n",
    "    \"Transition\": \"blueviolet\",\n",
    "}\n",
    "\n",
    "diet_palette = {\n",
    "    \"EEN\": \"lightgreen\",\n",
    "    \"PostEEN\": \"lightblue\",\n",
    "    \"InVitro\": \"plum\",\n",
    "    \"PreEEN\": \"lightpink\",\n",
    "}\n",
    "\n",
    "subject_order = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "\n",
    "# NOTE: Requires a dummy value because I want exactly 20 items.\n",
    "subject_palette = lib.plot.construct_ordered_palette(\n",
    "    subject_order + [f\"dummy{i}\" for i in range(20 - len(subject_order))], cm=\"tab20\"\n",
    ")\n",
    "subject_palette[\"X\"] = \"black\"\n",
    "pair_type_order = [\"EEN\", \"Transition\", \"PostEEN\"]\n",
    "pair_type_marker_palette = {\"EEN\": \"s\", \"Transition\": \">\", \"PostEEN\": \"o\"}\n",
    "pair_type_linestyle_palette = {\"EEN\": \":\", \"Transition\": \"-.\", \"PostEEN\": \"-\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _label_experiment_sample(x):\n",
    "    if x.sample_type == \"human\":\n",
    "        label = f\"{x.subject_id} [{x.sample_id}] {x.collection_date_relative_een_end} {x.diet_or_media}\"\n",
    "    elif x.sample_type in [\"Fermenter_inoculum\"]:\n",
    "        label = (\n",
    "            f\"{x.subject_id} [{x.sample_id}] {x.source_samples} inoc {x.diet_or_media}\"\n",
    "        )\n",
    "    elif x.sample_type in [\"Fermenter\"]:\n",
    "        label = (\n",
    "            f\"{x.subject_id} [{x.sample_id}] {x.source_samples} frmnt {x.diet_or_media}\"\n",
    "        )\n",
    "    elif x.sample_type in [\"mouse\"]:\n",
    "        if x.status_mouse_inflamed == \"Inflamed\":\n",
    "            label = f\"{x.subject_id} [{x.sample_id}] {x.source_samples} \ud83d\udc2d {x.mouse_genotype} {x.diet_or_media} inflam\"\n",
    "        elif x.status_mouse_inflamed == \"not_Inflamed\":\n",
    "            label = f\"{x.subject_id} [{x.sample_id}] {x.source_samples} \ud83d\udc2d {x.mouse_genotype} {x.diet_or_media} not_inf\"\n",
    "        else:\n",
    "            raise ValueError(f\"sample type {x.status_mouse_inflamed} not understood\")\n",
    "    else:\n",
    "        raise ValueError(f\"sample type {x.sample_type} not understood\")\n",
    "    return label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = (\n",
    "    pd.read_table(\"meta/een-mgen/sample.tsv\")\n",
    "    .assign(\n",
    "        label=lambda x: x[\n",
    "            [\n",
    "                \"subject_id\",\n",
    "                \"collection_date_relative_een_end\",\n",
    "                \"diet_or_media\",\n",
    "                \"sample_id\",\n",
    "            ]\n",
    "        ].apply(tuple, axis=1),\n",
    "        fuller_label=lambda d: d.apply(_label_experiment_sample, axis=1),\n",
    "    )\n",
    "    .set_index(\"sample_id\")\n",
    ")\n",
    "subject = pd.read_table(\"meta/een-mgen/subject.tsv\", index_col=\"subject_id\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Species Depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_depth = (\n",
    "    pd.read_table(\n",
    "        \"data/group/een/r.proc.gtpro.species_depth.tsv\",\n",
    "        index_col=[\"sample\", \"species_id\"],\n",
    "    )\n",
    "    .depth.unstack(fill_value=0)\n",
    "    .rename(\n",
    "        columns=str, index=lambda x: \"CF_\" + str(int(x.split(\"_\")[1]))\n",
    "    )  # Normalize names\n",
    "    .rename({\"CF_15\": \"CF_11\", \"CF_11\": \"CF_15\"})  # Sample swap\n",
    ")\n",
    "motu_rabund = motu_depth.divide(motu_depth.sum(1), axis=0)\n",
    "\n",
    "motu_rabund"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cogs\n",
    "# take up to 7 minutes to compile everything\n",
    "cog_depth = {}\n",
    "\n",
    "for species in tqdm(motu_rabund.columns):\n",
    "    gene_x_cog_inpath = (\n",
    "        f\"data/species/sp-{species}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    gene_depth_inpath = (\n",
    "        f\"data/group/een/species/sp-{species}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    _gene_x_cog = (\n",
    "        pd.read_table(gene_x_cog_inpath)\n",
    "        .drop_duplicates()\n",
    "        .set_index(\"gene_id\")\n",
    "        .squeeze()\n",
    "    )\n",
    "    _cog_depth = (\n",
    "        xr.load_dataarray(gene_depth_inpath)\n",
    "        .to_pandas()\n",
    "        .T.join(_gene_x_cog)\n",
    "        .groupby(\"cog\")\n",
    "        .sum()\n",
    "    )\n",
    "    cog_depth[species] = _cog_depth.stack()\n",
    "\n",
    "cog_depth = (\n",
    "    pd.DataFrame(cog_depth)\n",
    "    .stack()\n",
    "    .rename_axis([\"cog\", \"sample\", \"species\"])\n",
    "    .to_xarray()\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "# Normalize sample names and swap the mislabeled samples.\n",
    "cog_depth[\"sample\"] = (\n",
    "    cog_depth.sample.to_series()\n",
    "    .map(lambda x: \"CF_\" + str(int(x.split(\"_\")[1])))\n",
    "    .replace({\"CF_15\": \"CF_11\", \"CF_11\": \"CF_15\"})\n",
    "    .values\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "schg_cog_list = [\n",
    "    \"COG0012\",\n",
    "    \"COG0016\",\n",
    "    \"COG0048\",\n",
    "    \"COG0049\",\n",
    "    \"COG0052\",\n",
    "    \"COG0080\",\n",
    "    \"COG0081\",\n",
    "    \"COG0085\",\n",
    "    \"COG0087\",\n",
    "    \"COG0088\",\n",
    "    \"COG0090\",\n",
    "    \"COG0091\",\n",
    "    \"COG0092\",\n",
    "    \"COG0093\",\n",
    "    \"COG0094\",\n",
    "    \"COG0096\",\n",
    "    \"COG0097\",\n",
    "    \"COG0098\",\n",
    "    \"COG0099\",\n",
    "    \"COG0100\",\n",
    "    \"COG0102\",\n",
    "    \"COG0103\",\n",
    "    \"COG0124\",\n",
    "    \"COG0184\",\n",
    "    \"COG0185\",\n",
    "    \"COG0186\",\n",
    "    \"COG0197\",\n",
    "    \"COG0200\",\n",
    "    \"COG0201\",\n",
    "    \"COG0256\",\n",
    "    \"COG0495\",\n",
    "    \"COG0522\",\n",
    "    \"COG0525\",\n",
    "    \"COG0533\",\n",
    "    # \"COG0542\",  # This one is a depth outlier...\n",
    "]\n",
    "\n",
    "len(schg_cog_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.clustermap(cog_depth.sum(\"species\").sel(cog=schg_cog_list).to_pandas(), metric='cosine', yticklabels=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_depth2 = cog_depth.sel(cog=schg_cog_list).mean(\"cog\")\n",
    "motu_rabund2 = motu_depth2 / motu_depth2.sum(\"species\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_list = [\"100003\", \"102506\", \"100022\"]\n",
    "_sample_list = [\"CF_94\", \"CF_93\"]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    len(_species_list),\n",
    "    len(_sample_list),\n",
    "    figsize=(5 * len(_sample_list), 3 * len(_species_list)),\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "axs = np.asanyarray(axs).reshape((len(_species_list), len(_sample_list)))\n",
    "\n",
    "bins = np.logspace(-3, 4, num=100)\n",
    "\n",
    "for (_species, _sample), ax in zip(product(_species_list, _sample_list), axs.flatten()):\n",
    "    ax.hist(cog_depth.sel(sample=_sample, species=_species).to_pandas(), bins=bins)\n",
    "    ax.axvline(motu_depth.loc[_sample, _species], color=\"tab:orange\")\n",
    "    ax.axvline(motu_depth2.loc[_sample, _species], color=\"black\")\n",
    "    ax.set_title((_species, _sample))\n",
    "\n",
    "ax.set_xscale(\"log\")\n",
    "# plt.xlim(0, 200)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_detection_limit = cog_depth.where(lambda x: x != 0, np.inf).min(\n",
    "    (\"sample\", \"species\")\n",
    ")\n",
    "undetected_cogs_list = idxwhere((cog_detection_limit == np.inf).to_series())\n",
    "\n",
    "cog_depth_or_detection_limit = cog_depth.where(\n",
    "    lambda x: x != 0, cog_detection_limit\n",
    ").drop_sel(cog=undetected_cogs_list)\n",
    "cog_depth_or_detection_limit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_genome_depth = (\n",
    "    cog_depth_or_detection_limit.sel(cog=schg_cog_list).median(\"cog\").sum(\"species\")  # NOTE: Mean or Median? Does it matter?\n",
    ")\n",
    "normalized_cog_depth_by_sample = (\n",
    "    cog_depth_or_detection_limit.sum(\"species\") / total_genome_depth\n",
    ")\n",
    "normalized_cog_depth_by_subject_and_type = (\n",
    "    normalized_cog_depth_by_sample.to_pandas()\n",
    "    .T.join(sample[[\"subject_id\", \"diet_or_media\"]])[\n",
    "        lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "    ]\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])\n",
    "    .median()  # NOTE: Mean or Median?\n",
    "    .unstack(\"diet_or_media\")\n",
    "    .dropna()\n",
    "    .stack(\"diet_or_media\")\n",
    ")\n",
    "normalized_cog_depth_by_subject_and_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results = {}\n",
    "for cog in tqdm(normalized_cog_depth_by_subject_and_type.columns):\n",
    "    d = normalized_cog_depth_by_subject_and_type[cog].unstack()\n",
    "    mean_een = d.EEN.mean()\n",
    "    mean_post = d.PostEEN.mean()\n",
    "    mean_log_ratio = np.log2(d.PostEEN / d.EEN).mean()\n",
    "    median_log_ratio = np.log2(d.PostEEN / d.EEN).median()\n",
    "    try:\n",
    "        result = sp.stats.wilcoxon(\n",
    "            d.EEN,\n",
    "            d.PostEEN,\n",
    "        )\n",
    "        pval = result.pvalue\n",
    "    except ValueError:\n",
    "        pval = np.nan\n",
    "    pairwise_test_results[cog] = (\n",
    "        mean_een,\n",
    "        mean_post,\n",
    "        mean_log_ratio,\n",
    "        median_log_ratio,\n",
    "        pval,\n",
    "    )\n",
    "\n",
    "pairwise_test_results = pd.DataFrame(\n",
    "    pairwise_test_results,\n",
    "    index=(\"mean_een\", \"mean_post\", \"mean_log_ratio\", \"median_log_ratio\", \"pval\"),\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Here is where I define filters on COGs:\n",
    "#    They must have a mean depth during one of the two time-periods of > 0.01\n",
    "\n",
    "pairwise_test_results_filt_with_fdr = (\n",
    "    pairwise_test_results[lambda x: (x.mean_een > 0.01) | (x.mean_post > 0.01)]\n",
    "    .assign(\n",
    "        fdr=lambda x: fdrcorrection(x.pval)[1],\n",
    "        hit=lambda x: (x.fdr < 0.1) & (np.abs(x.mean_log_ratio) > 0.2),\n",
    "    )\n",
    "    .sort_values(\"fdr\")\n",
    ")\n",
    "pairwise_test_results_filt_with_fdr[lambda x: x.hit]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pairwise_test_results_filt_with_fdr.sort_values(\"pval\").join(cog_meta)\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.scatter(\"mean_log_ratio\", \"pval\", data=d[d.hit], color=\"r\", s=5)\n",
    "ax.scatter(\"mean_log_ratio\", \"pval\", data=d[~d.hit], color=\"grey\", s=5)\n",
    "ax.invert_yaxis()\n",
    "ax.set_yscale(\"log\")\n",
    "ax.axvline(0.2, color=\"black\", lw=1, linestyle=\"--\")\n",
    "ax.axvline(-0.2, color=\"black\", lw=1, linestyle=\"--\")\n",
    "ax.axhline(0.05, color=\"black\", lw=1, linestyle=\"--\")\n",
    "ax.set_xlabel(\"Mean Log2(OR)\")\n",
    "ax.set_ylabel(\"P-value\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_x_cog_category_matrix = (\n",
    "    cog_meta.categories.map(tuple)\n",
    "    .explode()\n",
    "    .rename(\"category\")\n",
    "    .reset_index()\n",
    "    .assign(in_category=True)\n",
    "    .set_index([\"cog\", \"category\"])\n",
    "    .in_category.unstack(fill_value=False)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pairwise_test_results_filt_with_fdr[[\"hit\"]].join(cog_x_cog_category_matrix)\n",
    "\n",
    "enrichment_test = []\n",
    "for category in cog_x_cog_category_matrix.columns:\n",
    "    contingency = (\n",
    "        d[[\"hit\", category]]\n",
    "        .value_counts()\n",
    "        .reindex(\n",
    "            [(True, True), (True, False), (False, True), (False, False)], fill_value=0\n",
    "        )\n",
    "        .unstack()\n",
    "    )\n",
    "    contingency_pc = contingency + 1\n",
    "    log2_odds_ratio_pc = np.log2(\n",
    "        (contingency_pc.loc[True, True] / contingency_pc.loc[True, False])\n",
    "        / (contingency_pc.loc[False, True] / contingency_pc.loc[False, False])\n",
    "    )\n",
    "    num_hit = contingency_pc.loc[True, True]\n",
    "    enrichment_test.append(\n",
    "        [category, num_hit, log2_odds_ratio_pc, *sp.stats.fisher_exact(contingency)]\n",
    "    )\n",
    "enrichment_test = pd.DataFrame(\n",
    "    enrichment_test,\n",
    "    columns=[\"category\", \"num_hit\", \"log2_odds_ratio_pc\", \"stat\", \"pvalue\"],\n",
    ").set_index(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_test.sort_values(\"log2_odds_ratio_pc\", ascending=False).join(cog_category_meta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_x_cog_category_matrix.reindex(\n",
    "    idxwhere(\n",
    "        pairwise_test_results_filt_with_fdr.hit\n",
    "        & (pairwise_test_results_filt_with_fdr.mean_log_ratio > 0)\n",
    "    )\n",
    ").sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def _species_specificity_statistics(x):\n",
    "#     x = x.sort_values(ascending=False) / x.sum()\n",
    "#     ratio_to_top = x / x[0]\n",
    "#     ratio_to_second = x / x[1]\n",
    "#     return pd.DataFrame(dict(fraction=x, ratio_to_top=ratio_to_top, ratio_to_second=ratio_to_second)).stack()\n",
    "#     # return pd.DataFrame([[1, 2, 3], [4, 5, 6]])\n",
    "\n",
    "# d = cog_depth.mean(\"sample\").to_pandas().T\n",
    "\n",
    "cog_species_fraction = (cog_depth / cog_depth.sum(\"species\")).mean(\"sample\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select by Function"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### COG Category I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.join(cog_meta).fillna({\"categories\": \"\"})[\n",
    "    lambda x: x.hit & x.categories.str.contains(\"I\")\n",
    "].sort_values(\"mean_log_ratio\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.join(cog_meta).fillna({\"categories\": \"\"})[\n",
    "    lambda x: x.hit & x.categories.str.contains(\"I\")\n",
    "].sort_values(\"mean_log_ratio\", ascending=False).description.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3255\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG4981\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3675\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG4667\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# COGs that are *DE-enriched* in PostEEN are frequently dominated by sequences from E. coli.\n",
    "cog_species_fraction.sel(\n",
    "    cog=idxwhere(pairwise_test_results_filt_with_fdr.mean_log_ratio < -1)\n",
    ").to_pandas().idxmax(1).value_counts().head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# COGs that are *DE-enriched* in PostEEN are frequently dominated by sequences from E. coli.\n",
    "cog_species_fraction.sel(\n",
    "    cog=idxwhere(pairwise_test_results_filt_with_fdr.mean_log_ratio > 0.2)\n",
    ").to_pandas().idxmax(1).value_counts().head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### COG Category G"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.join(cog_meta).fillna({\"categories\": \"\"})[\n",
    "    lambda x: x.hit & x.categories.str.contains(\"G\")\n",
    "].sort_values(\"mean_log_ratio\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3429\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG4813\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3867\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3957\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG1621\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.join(cog_meta).fillna({\"categories\": \"\"})[\n",
    "    lambda x: x.hit & x.categories.str.contains(\"X\")\n",
    "].sort_values(\"mean_log_ratio\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_species_fraction.sel(cog=\"COG3636\").to_series().sort_values(ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Select by Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_meta.loc[\"COG4466\"].description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_specific_normalized_cog_depth_by_sample = (\n",
    "    cog_depth_or_detection_limit\n",
    "    / cog_depth_or_detection_limit.sel(cog=schg_cog_list).mean(\"cog\")\n",
    ")\n",
    "assert np.isnan(species_specific_normalized_cog_depth_by_sample.values).sum() == 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-102506 (E. coli)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"102506\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=1000).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_meta.loc[\"COG2747\"].description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.05)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ").description.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG2888"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG2888\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG2747"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG2747\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG0630"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG0630\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4667"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4667\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG1857"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG1857\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-102544 (E. lenta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"102544\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=1000).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][\n",
    "    lambda x: x.top_species == species_id\n",
    "].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(\n",
    "    cog_meta\n",
    ").sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")[lambda x: (x.mean_log_ratio > 0.5) & (x.prevalence < 0.8)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ").description.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4981"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4981\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG3675"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG3675\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG5617"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG5617\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-100099 (F. plautii)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"100099\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=1000).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta).loc[\"COG3957\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG5585"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG5585\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG3957"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG3957\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-101303 (D. scindens)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101303\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=1000).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta).loc[\"COG3255\"]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][\n",
    "    lambda x: x.top_species == species_id\n",
    "].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(\n",
    "    cog_meta\n",
    ").sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")[lambda x: (x.mean_log_ratio > 0.5) & (x.prevalence < 0.8)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-101493 (s__Clostridium_M bolteae)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101493\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=1000).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta).loc[\"COG3255\"]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][\n",
    "    lambda x: x.top_species == species_id\n",
    "].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(\n",
    "    cog_meta\n",
    ").sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")[lambda x: (x.mean_log_ratio > 0.5) & (x.prevalence < 0.8)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG3255"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG3255\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG5585"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG5585\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-101338 (s__Blautia_A wexlerae)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101338\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][\n",
    "    lambda x: x.top_species == species_id\n",
    "].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(\n",
    "    cog_meta\n",
    ").sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")[lambda x: (x.mean_log_ratio > 0.5) & (x.prevalence < 0.8)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4878"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4878\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4267"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4267\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG5585"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG5585\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-101367 (s__Clostridium_Q symbiosum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101367\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & (x.pval < 0.05)\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ").description.values"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG1353"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG1353\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-101386 s__Clostridium_M clostridioforme)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101386\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][lambda x: x.top_species == species_id].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(cog_meta).sort_values('mean_log_ratio', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_species_specificity.loc[idxwhere(pairwise_test_results_filt_with_fdr.hit)][\n",
    "    lambda x: x.top_species == species_id\n",
    "].join(pairwise_test_results_filt_with_fdr).join(cog_prevalence_in_refs).join(\n",
    "    cog_meta\n",
    ").sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")[lambda x: (x.mean_log_ratio > 0.5) & (x.prevalence < 0.8)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### sp-100032 (s__Hungatella effluvii)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"100032\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (x.mean_log_ratio > 0.3)\n",
    "    # & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"species_frac\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 100179 (g__Clostridium_M)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"100179\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.05)\n",
    "    & (x.mean_log_ratio > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta).loc[\"COG3255\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG3255"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG3255\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4938"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4938\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG0146"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG0146\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 102549"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"102549\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 103682"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"103682\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4396"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4396\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 101400"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101400\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG5599"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG5599\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 102478"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"102478\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.95)\n",
    "    & (x.pval < 0.1)\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4938"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4938\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 101292"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101292\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.clean-m10-e20-c10.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.9)\n",
    "    & x.hit\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG2920"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG2920\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG4981"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG4981\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 101346"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"101346\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "world = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    ")\n",
    "\n",
    "\n",
    "subject_abh_sample_list = list(\n",
    "    set(idxwhere(sample.subject_id.isin([\"A\", \"B\", \"H\"]))) & set(world.sample.values)\n",
    ")\n",
    "world_subject_abh = world.sel(\n",
    "    sample=subject_abh_sample_list\n",
    ").drop_low_abundance_strains(0.05)\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "sample_linkage = world.unifrac_linkage()\n",
    "strain_linkage = world.genotype.linkage()\n",
    "subject_abh_sample_linkage = world_subject_abh.unifrac_linkage()\n",
    "subject_abh_strain_linkage = world_subject_abh.genotype.linkage()\n",
    "subject_abh_position_ss_linkage = world_subject_abh.sel(\n",
    "    position=position_ss\n",
    ").genotype.linkage(\"position\")\n",
    "\n",
    "_col_colors = sample.set_index(\"fuller_label\").subject_id.map(subject_palette)\n",
    "\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    world_subject_abh.rename_coords(sample=sample.fuller_label),\n",
    "    scalex=0.3,\n",
    "    scaley=0.7,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_strain_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")\n",
    "\n",
    "sf.plot.plot_metagenotype(\n",
    "    world_subject_abh.sel(position=position_ss).rename_coords(\n",
    "        sample=sample.fuller_label\n",
    "    ),\n",
    "    scalex=0.3,\n",
    "    col_linkage_func=lambda w: subject_abh_sample_linkage,\n",
    "    row_linkage_func=lambda w: subject_abh_position_ss_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_meta = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.strain_meta.tsv\",\n",
    "    index_col=\"genome_id\",\n",
    ").rename(str)\n",
    "spgc = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10_thresh-corr450-depth200.uhgg-strain_gene.tsv\",\n",
    "    index_col=\"gene_id\",\n",
    ")\n",
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "gene_meta = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.d/proteins.emapper.annotations\",\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "gene_x_cog = (\n",
    "    pd.read_table(\n",
    "        f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_cog.tsv\"\n",
    "    )\n",
    "    .drop_duplicates()\n",
    "    .set_index(\"gene_id\")\n",
    "    .squeeze()\n",
    ")\n",
    "spgc_depth_ratio = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_depth_ratio.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").depth.unstack()\n",
    "spgc_corr = pd.read_table(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-10.strain_correlation.tsv\",\n",
    "    index_col=[\"gene_id\", \"strain\"],\n",
    ").correlation.unstack()\n",
    "gene_depth = (\n",
    "    xr.load_dataarray(\n",
    "        f\"data/group/een/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "    )\n",
    "    .to_pandas()\n",
    "    .rename(lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    ")\n",
    "eggnog_prevalence_in_refs = pd.read_table(\n",
    "    f\"data/species/sp-{species_id}/midasdb.gene75_new.eggnog-strain_gene.prevalence.tsv\",\n",
    "    names=[\"eggnog_id\", \"prevalence\"],\n",
    ").assign(cog_id=lambda x: x.eggnog_id.str.split(\"@\").str[0])\n",
    "\n",
    "high_quality_strain_list = idxwhere(\n",
    "    (spgc_strain_meta.sum_depth > 1) & (spgc_strain_meta.species_gene_frac > 0.9)\n",
    ")\n",
    "\n",
    "spgc_strain_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_prevalence_in_refs = (\n",
    "    eggnog_prevalence_in_refs[lambda x: x.cog_id.str.startswith(\"COG\")][\n",
    "        [\"cog_id\", \"prevalence\"]\n",
    "    ]\n",
    "    .drop_duplicates()\n",
    "    .groupby(\"cog_id\")\n",
    "    .prevalence.max()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pairwise_test_results_filt_with_fdr.assign(\n",
    "    species_frac=cog_species_fraction.sel(species=species_id).to_series()\n",
    ").join(cog_prevalence_in_refs).join(cog_meta)[\n",
    "    lambda x: True\n",
    "    & (x.species_frac > 0.1)\n",
    "    & (np.abs(x.mean_log_ratio) > 0.5)\n",
    "    & (x.prevalence < 0.95)\n",
    "    & (x.pval < 0.1)\n",
    "].sort_values(\n",
    "    \"mean_log_ratio\", ascending=False\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### COG1231"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_id = \"COG1231\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    species_specific_normalized_cog_depth_by_sample.sel(species=species_id)\n",
    "    .to_pandas()\n",
    "    .T\n",
    ")\n",
    "\n",
    "bins = np.logspace(-3, 4)\n",
    "plt.hist(d.mean(), bins=bins)\n",
    "plt.hist(d[cog_id], bins=bins)\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_cog_depth = cog_depth_or_detection_limit.sel(\n",
    "    species=species_id, cog=cog_id\n",
    ") / motu_depth2.sel(species=species_id)\n",
    "\n",
    "plt.scatter(\n",
    "    motu_depth2.sel(species=species_id),\n",
    "    cog_depth_or_detection_limit.sel(species=species_id, cog=cog_id),\n",
    "    c=normalized_cog_depth,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_with_gene_list = idxwhere(\n",
    "    (\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "strain_without_gene_list = idxwhere(\n",
    "    ~(\n",
    "        spgc[high_quality_strain_list].reindex(gene_x_cog[lambda x: x == cog_id].index)\n",
    "        == 1\n",
    "    ).any()\n",
    ")\n",
    "\n",
    "frac = world.community.to_pandas().rename(columns=str)\n",
    "strain_gene_unknown_list = [\"-1\"] + list(\n",
    "    set(frac.columns)\n",
    "    - set(strain_with_gene_list)\n",
    "    - set(strain_without_gene_list)\n",
    "    - set([\"-1\"])\n",
    ")\n",
    "\n",
    "strain_list = (\n",
    "    strain_with_gene_list + strain_without_gene_list + strain_gene_unknown_list\n",
    ")\n",
    "\n",
    "strain_with_gene_list, strain_without_gene_list, len(strain_gene_unknown_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_with_gene_list, cm=\"autumn\"\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_without_gene_list, cm=\"winter\", extend=strain_by_cog_palette\n",
    ")\n",
    "strain_by_cog_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_gene_unknown_list, cm=\"Greys\", extend=strain_by_cog_palette\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order[:3]\n",
    "fig, axs = lib.plot.subplots_grid(1, len(_subject_list), ax_width=20, ax_height=10)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.04))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_meta = sample.loc[motu_depth2.sample][\n",
    "    lambda x: x.diet_or_media.isin([\"EEN\", \"PostEEN\"])\n",
    "]\n",
    "_frac = frac.reindex(_meta.index, fill_value=0)\n",
    "_frac[\"-1\"] = 1 - _frac.drop(columns=[\"-1\"]).sum(1)\n",
    "d0 = (\n",
    "    _meta.join(_frac)\n",
    "    .assign(\n",
    "        species_rabund=motu_rabund2.sel(species=species_id).to_series(),\n",
    "        norm_cog_depth=normalized_cog_depth.to_series(),\n",
    "    )\n",
    "    .set_index(\"fuller_label\")\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    ")\n",
    "\n",
    "_subject_list = subject_order\n",
    "fig, axs = lib.plot.subplots_grid(6, len(_subject_list), ax_width=6, ax_height=7)\n",
    "\n",
    "for subject_id, ax in zip(_subject_list, axs.flatten()):\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id]\n",
    "    if d1.empty:\n",
    "        continue\n",
    "    d1[strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_by_cog_palette,\n",
    "        edgecolor=\"k\",\n",
    "        ax=ax,\n",
    "        lw=0.5,\n",
    "    )\n",
    "    ax.set_title(subject_id)\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax2 = plt.twinx(ax)\n",
    "    ax2.plot(d1[\"norm_cog_depth\"], color=\"k\", marker=\"o\", lw=1, markersize=5)\n",
    "    ax2.set_yscale(\"symlog\", linthresh=1e-2, linscale=0.1)\n",
    "    ax3 = plt.twinx(ax)\n",
    "    ax3.plot(d1[\"species_rabund\"], color=\"k\", marker=\"o\")\n",
    "    ax3.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "    ax3.spines.right.set_position((\"axes\", 1.15))\n",
    "\n",
    "fig.tight_layout()\n",
    "\n",
    "# frac.rename(index=sample.fuller_label).sort_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_meta.reindex(gene_x_cog[lambda x: x == cog_id].index).drop_duplicates(\n",
    "    subset=[\"seed_ortholog\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_depth_ratio.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_corr.reindex(gene_x_cog[lambda x: x == cog_id].index).dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "normalized_gene_depth = gene_depth.stack().to_xarray() / motu_depth2.sel(\n",
    "    species=species_id\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(\n",
    "        with_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_with_gene_list]\n",
    "        .sum(1),\n",
    "        without_gene_fraction=world.community.to_pandas()\n",
    "        .rename(columns=str)[strain_without_gene_list]\n",
    "        .sum(1),\n",
    "        gene_depth=gene_depth.stack()\n",
    "        .to_xarray()\n",
    "        .sel(\n",
    "            gene_id=list(\n",
    "                set(gene_x_cog[lambda x: x == cog_id].index)\n",
    "                & set(normalized_gene_depth.gene_id.values)\n",
    "            )\n",
    "        )\n",
    "        .to_pandas()\n",
    "        .sum(1),\n",
    "        species_depth=motu_depth2.sel(species=species_id).to_series(),\n",
    "    )\n",
    ").fillna({\"with_gene_fraction\": 0, \"without_gene_fraction\": 0})\n",
    "\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"with_gene_fraction\",\n",
    "    data=d.sort_values(\"with_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain With Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")\n",
    "\n",
    "fig = plt.figure()\n",
    "plt.scatter(\n",
    "    \"species_depth\",\n",
    "    \"gene_depth\",\n",
    "    c=\"without_gene_fraction\",\n",
    "    data=d.sort_values(\"without_gene_fraction\"),\n",
    "    norm=mpl.colors.PowerNorm(1, vmin=0, vmax=1),\n",
    ")\n",
    "plt.plot([1, 1000], [1, 1000])\n",
    "plt.colorbar(label=\"Strain Without Gene Fraction\")\n",
    "plt.yscale(\"symlog\", linthresh=1e-1)\n",
    "plt.xscale(\"symlog\", linthresh=1e-1)\n",
    "plt.ylabel(\"gene depth\")\n",
    "plt.xlabel(\"species depth\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz2",
   "language": "python",
   "name": "toolz2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}