{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_meta = (\n",
    "    pd.read_csv(\n",
    "        \"raw/een-mgen/2023-07-28_Kolja.Siebert@med.uni-muenchen.de/1.0_Metadata_sharedwithByron_Shotgunmetagenomics_CRC1371Juli2023_datesfixed_sheet2.csv\",\n",
    "        parse_dates=False,\n",
    "    )\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"Patient_ID\": \"subject_id\",\n",
    "            \"Start_date_EEN\": \"een_start_date\",\n",
    "            \"End_date_EEN\": \"een_end_date\",\n",
    "            \"Relapse1year\": \"endpoint_patient\",\n",
    "            \"Date_Relapse_Start\": \"relapse_start_date\",\n",
    "        }\n",
    "    )\n",
    "    .set_index(\"subject_id\")\n",
    "    # .assign(\n",
    "    #     relapse_start_date=lambda s: s.relapse_start_date.str.replace(\n",
    "    #         \".\", \"/\"\n",
    "    #     ),  # One relapse date was written with '.' separators instead of '/'\n",
    "    #     een_start_date=lambda s: s.een_start_date.str.replace(\n",
    "    #         \".\", \"/\"\n",
    "    #     ),  # One start date was written with '.' separators instead of '/'\n",
    "    # )\n",
    "    .assign(\n",
    "        een_start_date=lambda x: pd.to_datetime(x.een_start_date, format=\"%m/%d/%y\"),\n",
    "        een_end_date=lambda x: pd.to_datetime(x.een_end_date, format=\"%m/%d/%y\"),\n",
    "        relapse_start_date=lambda x: pd.to_datetime(\n",
    "            x.relapse_start_date, format=\"%m/%d/%y\"\n",
    "        ),\n",
    "        endpoint_patient_relapse=lambda x: x.endpoint_patient.fillna(\"no\"),\n",
    "    )\n",
    "    .assign(\n",
    "        een_start_date_relative_een_end=lambda x: (\n",
    "            x.een_start_date - x.een_end_date\n",
    "        ).dt.days,\n",
    "        relapse_start_date_relative_een_end=lambda x: (\n",
    "            x.relapse_start_date - x.een_end_date\n",
    "        ).dt.days,\n",
    "    )\n",
    ")\n",
    "\n",
    "assert subject_meta.index.is_unique\n",
    "\n",
    "subject_meta  # .apply(lambda x: x.value_counts().index.values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_mgen_data = (\n",
    "    pd.read_csv(\n",
    "        \"raw/een-mgen/2023-08-30_deborah.haecker@tum-create.edu.sg/mapping_file_Byron.csv\"\n",
    "    )\n",
    "    .dropna(subset=[\"ID\"])\n",
    "    .astype({\"ID\": int})\n",
    "    .assign(\n",
    "        collection_date=lambda x: pd.to_datetime(x.sampleDate, format=\"%d.%m.%Y\"),\n",
    "    )[\n",
    "        lambda x: x.Patient_ID != \"X\"\n",
    "    ]  # Patient X seems highly atypical and data is not consistent.\n",
    "    .join(subject_meta, on=\"Patient_ID\")\n",
    "    .assign(\n",
    "        collection_date_relative_een_end=lambda x: (\n",
    "            x.collection_date - x.een_end_date\n",
    "        ).dt.days,\n",
    "        has_metabolome=lambda x: x[\"matched metabolome avaliable?\"] == \"yes\",\n",
    "        status_medication_antibiotics=lambda x: x[lambda x: x.organism == 'human'][\"Antibiotics\"]\n",
    "        .fillna(\"\")\n",
    "        .str.lower()\n",
    "        .str.startswith(\"yes\"),\n",
    "        status_medication_mtx=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('MTX'),\n",
    "        status_medication_antitnf=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('AntiTNF'),\n",
    "        status_medication_other=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('OtherMedication'),\n",
    "        sample_comments=\"\",\n",
    "    )\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"Seq-Name\": \"sample_id\",\n",
    "            \"Patient_ID\": \"subject_id\",\n",
    "            \"organism\": \"sample_type\",\n",
    "            \"active_inactive\": \"status_disease_activity\",\n",
    "            # \"which_medication\": \"status_medication\",\n",
    "            # \"on_antibiotics\": \"status_antibiotics\",  # Watch out! Antibiotics is a different column name.\n",
    "            \"Therapy_weeks\": \"timepoint\",\n",
    "            \"Therapy_cluster_more\": \"timepoint_type\",\n",
    "            \"HealingGroup\": \"patient_outcome\",\n",
    "            \"Transfer_used_Inoculum\": \"source_samples\",\n",
    "            \"mouse_Histoscoring_Colon\": \"status_mouse_colon_histoscore\",\n",
    "            \"mouse_inflamed_not_inflamed\": \"status_mouse_inflamed\",\n",
    "            \"human_transfer_Diet_Media\": \"diet_or_media\",\n",
    "        }\n",
    "    )\n",
    ")\n",
    "\n",
    "new_mgen_data#.sort_values([\"subject_id\", \"sample_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = new_mgen_data[\n",
    "    [\n",
    "        \"sample_id\",\n",
    "        \"subject_id\",\n",
    "        \"collection_date_relative_een_end\",\n",
    "        \"diet_or_media\",\n",
    "        \"sample_type\",\n",
    "        \"status_disease_activity\",\n",
    "        \"status_medication_mtx\",\n",
    "        \"status_medication_antitnf\",\n",
    "        \"status_medication_antibiotics\",\n",
    "        \"status_medication_other\",\n",
    "        \"timepoint\",\n",
    "        \"timepoint_type\",\n",
    "        \"source_samples\",\n",
    "        \"mouse_genotype\",\n",
    "        \"status_mouse_colon_histoscore\",\n",
    "        \"status_mouse_inflamed\",\n",
    "        \"has_metabolome\",\n",
    "        \"sample_comments\",\n",
    "    ]\n",
    "].set_index('sample_id')\n",
    "\n",
    "# d.to_csv('meta/een-mgen/sample.tsv', sep='\\t')\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = subject_meta[['een_start_date_relative_een_end', 'endpoint_patient_relapse', 'relapse_start_date_relative_een_end']]\n",
    "# d.to_csv(\"meta/een-mgen/subject.tsv\", sep=\"\\t\")\n",
    "d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_new_mgen_data = (\n",
    "    pd.read_csv(\n",
    "        \"raw/een-mgen/2023-09-25_deborah.haecker@tum-create.edu.sg/mapping_file_Byron_final2.csv\"\n",
    "    )\n",
    "    .dropna(subset=[\"ID\"])\n",
    "    .astype({\"ID\": int})\n",
    "    .assign(\n",
    "        collection_date=lambda x: pd.to_datetime(x.sampleDate, format=\"%d.%m.%Y\"),\n",
    "    )\n",
    "    # [\n",
    "    #     lambda x: x.Patient_ID != \"X\"\n",
    "    # ]  # Patient X seems highly atypical and data is not consistent.\n",
    "    .join(subject_meta, on=\"Patient_ID\")\n",
    "    .assign(\n",
    "        collection_date_relative_een_end=lambda x: (\n",
    "            x.collection_date - x.een_end_date\n",
    "        ).dt.days,\n",
    "        has_metabolome=lambda x: x[\"matched metabolome avaliable?\"] == \"yes\",\n",
    "        status_medication_antibiotics=lambda x: x[lambda x: x.organism == 'human'][\"Antibiotics\"]\n",
    "        .fillna(\"\")\n",
    "        .str.lower()\n",
    "        .str.startswith(\"yes\"),\n",
    "        status_medication_mtx=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('MTX'),\n",
    "        status_medication_antitnf=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('AntiTNF'),\n",
    "        status_medication_other=lambda x: x[lambda x: x.organism == 'human'].which_medication.fillna('').str.contains('OtherMedication'),\n",
    "        sample_comments=\"\",\n",
    "    )\n",
    "    .rename(\n",
    "        columns={\n",
    "            \"Seq-Name\": \"sample_id\",\n",
    "            \"Patient_ID\": \"subject_id\",\n",
    "            \"organism\": \"sample_type\",\n",
    "            \"active_inactive\": \"status_disease_activity\",\n",
    "            # \"which_medication\": \"status_medication\",\n",
    "            # \"on_antibiotics\": \"status_antibiotics\",  # Watch out! Antibiotics is a different column name.\n",
    "            \"Therapy_weeks\": \"timepoint\",\n",
    "            \"Therapy_cluster_more\": \"timepoint_type\",\n",
    "            \"HealingGroup\": \"patient_outcome\",\n",
    "            \"Transfer_used_Inoculum\": \"source_samples\",\n",
    "            \"mouse_Histoscoring_Colon\": \"status_mouse_colon_histoscore\",\n",
    "            \"mouse_inflamed_not_inflamed\": \"status_mouse_inflamed\",\n",
    "            \"human_transfer_Diet_Media\": \"diet_or_media\",\n",
    "        }\n",
    "    )\n",
    ")\n",
    "\n",
    "new_new_mgen_data#.sort_values([\"subject_id\", \"sample_type\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = new_new_mgen_data[\n",
    "    [\n",
    "        \"sample_id\",\n",
    "        \"subject_id\",\n",
    "        \"collection_date_relative_een_end\",\n",
    "        \"diet_or_media\",\n",
    "        \"sample_type\",\n",
    "        \"status_disease_activity\",\n",
    "        \"status_medication_mtx\",\n",
    "        \"status_medication_antitnf\",\n",
    "        \"status_medication_antibiotics\",\n",
    "        \"status_medication_other\",\n",
    "        \"timepoint\",\n",
    "        \"timepoint_type\",\n",
    "        \"source_samples\",\n",
    "        \"mouse_genotype\",\n",
    "        \"status_mouse_colon_histoscore\",\n",
    "        \"status_mouse_inflamed\",\n",
    "        \"has_metabolome\",\n",
    "        \"sample_comments\",\n",
    "    ]\n",
    "].set_index('sample_id').sort_index()\n",
    "\n",
    "# d.to_csv('meta/een-mgen/sample.tsv', sep='\\t')\n",
    "d"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz",
   "language": "python",
   "name": "toolz"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}