{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "from warnings import filterwarnings\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "import matplotlib.ticker as mtick\n",
    "\n",
    "# from fastcluster import linkage\n",
    "from scipy.cluster.hierarchy import linkage\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from statsmodels.graphics.regressionplots import influence_plot\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "import lib.thisproject.data\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def linkage_order(linkage, labels):\n",
    "    return labels[sp.cluster.hierarchy.to_tree(linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "\n",
    "def is_prime(n):\n",
    "    if n <= 1:\n",
    "        return False\n",
    "    for i in range(2, int(n**0.5) + 1):\n",
    "        if n % i == 0:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "def iterate_primes_up_to(n, return_index=False):\n",
    "    n = int(np.ceil(n))\n",
    "    idx = 0\n",
    "    for i in range(n):\n",
    "        if is_prime(i):\n",
    "            if return_index:\n",
    "                yield (idx, i)\n",
    "            else:\n",
    "                yield i\n",
    "            idx += 1\n",
    "\n",
    "\n",
    "def maximally_shuffled_order(sorted_order):\n",
    "    n = len(sorted_order)\n",
    "    primes_list = list(iterate_primes_up_to(np.sqrt(n)))\n",
    "    table = pd.DataFrame(np.arange(n), index=sorted_order, columns=[\"original_order\"])\n",
    "    for prime in primes_list:\n",
    "        table[prime] = table.original_order % prime\n",
    "    table.sort_values(primes_list).original_order.values\n",
    "    table = table.assign(new_order=table.sort_values(primes_list).original_order.values)\n",
    "    z = table.sort_values(\"new_order\").original_order.values\n",
    "    table[\"delta\"] = [np.nan] + list(z[1:] - z[:-1])\n",
    "    return table.sort_values(\"new_order\").index.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _label_experiment_sample(x):\n",
    "    if x.sample_type == \"human\":\n",
    "        label = f\"[{x.name}] {x.collection_date_relative_een_end} {x.diet_or_media}\"\n",
    "    elif x.sample_type in [\"Fermenter_inoculum\"]:\n",
    "        label = f\"[{x.name}] {x.source_samples} inoc {x.diet_or_media}\"\n",
    "    elif x.sample_type in [\"Fermenter\"]:\n",
    "        # label = f\"[{x.name}] {x.source_samples} frmnt {x.diet_or_media}\"\n",
    "        label = f\"[{x.name}] {x.source_samples} {x.diet_or_media}\"\n",
    "    elif x.sample_type in [\"mouse\"]:\n",
    "        if x.status_mouse_inflamed == \"Inflamed\":\n",
    "            # label = f\"[{x.name}] {x.source_samples} \ud83d\udc2d {x.mouse_genotype} {x.diet_or_media} inflam\"\n",
    "            label = f\"[{x.name}] {x.source_samples} {x.diet_or_media} inflam\"\n",
    "        elif x.status_mouse_inflamed == \"not_Inflamed\":\n",
    "            # label = f\"[{x.name}] {x.source_samples} \ud83d\udc2d {x.mouse_genotype} {x.diet_or_media} not_inf\"\n",
    "            label = f\"[{x.name}] {x.source_samples} {x.diet_or_media} not_inf\"\n",
    "        else:\n",
    "            raise ValueError(f\"sample type {x.status_mouse_inflamed} not understood\")\n",
    "    else:\n",
    "        raise ValueError(f\"sample type {x.sample_type} not understood\")\n",
    "    return label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_stacked_barplot(data, x_var, order, palette=None, ax=None, **kwargs):\n",
    "    if ax is None:\n",
    "        ax = plt.subplot()\n",
    "    if palette is None:\n",
    "        palette = lib.plot.construct_ordered_palette(order)\n",
    "\n",
    "    # Bar styles\n",
    "    bar_kwargs = dict(\n",
    "        width=1.0,\n",
    "        alpha=1.0,\n",
    "        edgecolor=\"k\",\n",
    "        lw=1,\n",
    "    )\n",
    "    bar_kwargs.update(kwargs)\n",
    "\n",
    "    # Plot each bar segment\n",
    "    _last_top = 0\n",
    "    for y_var in order:\n",
    "        ax.bar(\n",
    "            x=data[x_var],\n",
    "            height=data[y_var],\n",
    "            bottom=_last_top,\n",
    "            label=y_var,\n",
    "            color=palette[y_var],\n",
    "            **bar_kwargs,\n",
    "        )\n",
    "        _last_top += data[y_var]\n",
    "    ax.set_xticks(data[x_var].values)\n",
    "    return ax\n",
    "\n",
    "\n",
    "plot_stacked_barplot(\n",
    "    pd.DataFrame(dict(t=[0, 1, 2], y1=[0.0, 0.5, 1.0], y2=[1.0, 0.5, 0.0])),\n",
    "    x_var=\"t\",\n",
    "    order=[\"y1\", \"y2\"],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pair_type_palette = {\"Transition\": \"plum\", \"EEN\": \"pink\", \"PostEEN\": \"lightblue\"}\n",
    "\n",
    "diet_palette = {\n",
    "    \"EEN\": \"lightgreen\",\n",
    "    \"PostEEN\": \"lightblue\",\n",
    "    \"InVitro\": \"plum\",\n",
    "    \"PreEEN\": \"lightpink\",\n",
    "}\n",
    "\n",
    "subject_order = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "\n",
    "# NOTE: Requires a dummy value because I want exactly 20 items.\n",
    "subject_palette = lib.plot.construct_ordered_palette(\n",
    "    subject_order + [f\"dummy{i}\" for i in range(20 - len(subject_order))], cm=\"tab20\"\n",
    ")\n",
    "subject_palette[\"X\"] = \"black\"\n",
    "pair_type_order = [\"EEN\", \"Transition\", \"PostEEN\"]\n",
    "pair_type_marker_palette = {\"EEN\": \"s\", \"Transition\": \">\", \"PostEEN\": \"o\"}\n",
    "pair_type_linestyle_palette = {\"EEN\": \":\", \"Transition\": \"-.\", \"PostEEN\": \"-\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = (\n",
    "    pd.read_table(\"meta/een-mgen/sample.tsv\")\n",
    "    .assign(\n",
    "        label=lambda x: x[\n",
    "            [\"collection_date_relative_een_end\", \"diet_or_media\", \"sample_id\"]\n",
    "        ].apply(tuple, axis=1)\n",
    "    )\n",
    "    .set_index(\"sample_id\")\n",
    "    .assign(full_label=lambda d: d.apply(_label_experiment_sample, axis=1))\n",
    ")\n",
    "subject = pd.read_table(\"meta/een-mgen/subject.tsv\", index_col=\"subject_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample.full_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_taxonomy_string(taxonomy_string):\n",
    "    values = taxonomy_string.split(\";\")\n",
    "    return pd.Series(values, index=[\"d__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_taxonomy_inpath = \"ref/uhgg_genomes_all_v2.tsv\"\n",
    "\n",
    "_motu_taxonomy = (\n",
    "    pd.read_table(motu_taxonomy_inpath)[lambda x: x.Genome == x.Species_rep]\n",
    "    .assign(species_id=lambda x: \"1\" + x.MGnify_accession.str.split(\"-\").str[2])\n",
    "    .set_index(\"species_id\")\n",
    ")\n",
    "\n",
    "# motu_lineage_string = _motu_taxonomy.Lineage\n",
    "\n",
    "motu_taxonomy = _motu_taxonomy.Lineage.apply(\n",
    "    parse_taxonomy_string\n",
    ")  # .assign(taxonomy_string=motu_lineage_string)\n",
    "motu_taxonomy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_depth = (\n",
    "    pd.read_table(\n",
    "        \"data/group/een/r.proc.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv\",\n",
    "        names=[\"sample\", \"species_id\", \"depth\"],\n",
    "        index_col=[\"sample\", \"species_id\"],\n",
    "    )\n",
    "    .depth.unstack(fill_value=0)\n",
    "    .rename(columns=str, index=lambda x: \"CF_\" + str(int(x.split(\"_\")[1])))\n",
    "    .rename({\"CF_15\": \"CF_11\", \"CF_11\": \"CF_15\"})  # Sample swap\n",
    ")\n",
    "motu_rabund = motu_depth.divide(motu_depth.sum(1), axis=0)\n",
    "\n",
    "motu_rabund"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rotu_counts = pd.read_table(\n",
    "    \"data/group/een/a.proc.zotu_counts.tsv\", index_col=\"#OTU ID\"\n",
    ").rename_axis(index=\"zotu\", columns=\"sample_id\")\n",
    "rotu_taxonomy = rotu_counts.taxonomy\n",
    "rotu_counts = rotu_counts.drop(columns=[\"taxonomy\"]).T\n",
    "rotu_rabund = rotu_counts.divide(rotu_counts.sum(1), axis=0)\n",
    "\n",
    "sample_rotu_bc_linkage = sp.cluster.hierarchy.linkage(\n",
    "    rotu_rabund, method=\"average\", metric=\"braycurtis\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Species Enrichment Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def enrichment_test(d):\n",
    "    try:\n",
    "        res = sp.stats.wilcoxon(d[\"EEN\"], d[\"PostEEN\"])\n",
    "    except ValueError:\n",
    "        res = (np.nan, np.nan)\n",
    "    log2_ratio = np.log2(d[\"PostEEN\"] / d[\"EEN\"])\n",
    "    return pd.Series(\n",
    "        [log2_ratio.mean(), d[\"EEN\"].mean(), d[\"PostEEN\"].mean(), res[1]],\n",
    "        index=[\"log2_ratio\", \"mean_EEN\", \"mean_PostEEN\", \"pvalue\"],\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample[[\"subject_id\", \"diet_or_media\"]])\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])\n",
    "    .mean()\n",
    "    .stack()\n",
    "    .unstack(\"diet_or_media\")[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    "    .rename_axis(index=[\"subject_id\", \"rotu_id\"])\n",
    "    .groupby(level=\"rotu_id\")\n",
    "    .apply(enrichment_test)\n",
    ")\n",
    "d.sort_values(\"mean_PostEEN\", ascending=False).head(20)\n",
    "# fig, ax = plt.subplots()\n",
    "# print(d.log2_ratio.mean())\n",
    "# print(sp.stats.wilcoxon(d['PostEEN'], d['EEN']))\n",
    "# ax.hist(d.log2_ratio, bins=20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Strain Time-series"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## E. coli (Zotu4 / 102506)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"102506\"\n",
    "rotu_id = \"Zotu4\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "plt.hist(d.log2_ratio, bins=20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)\n",
    "ax.set_ylim(1e-5, 1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d0 = (\n",
    "    sample.loc[\n",
    "        lambda x: (\n",
    "            True\n",
    "            # x.index.isin(sf_fit.sample.values)\n",
    "            & x.sample_type.isin([\"human\", \"Fermenter\", \"mouse\"])\n",
    "            & x.subject_id.isin([\"A\", \"B\", \"H\"])\n",
    "            # & (x.sample_type == \"Fermenter\")\n",
    "        ),\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"source_samples\",\n",
    "            \"status_mouse_inflamed\",\n",
    "            \"full_label\",\n",
    "        ],\n",
    "    ]\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"source_samples\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        rotu_rabund=rotu_rabund[rotu_id],\n",
    "    )\n",
    ")\n",
    "\n",
    "sample_type_order = [\"human\", \"Fermenter\", \"mouse\"]\n",
    "subject_order = [\"A\", \"B\", \"H\"]\n",
    "\n",
    "\n",
    "_grid_sample_counts = (\n",
    "    d0[[\"subject_id\", \"sample_type\"]]\n",
    "    .value_counts()\n",
    "    .unstack()\n",
    "    .reindex(columns=sample_type_order)\n",
    ")\n",
    "fig, axs = plt.subplots(\n",
    "    *_grid_sample_counts.shape,\n",
    "    figsize=(90, 15),\n",
    "    width_ratios=_grid_sample_counts.max().values,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(wspace=0.1, hspace=3),\n",
    ")\n",
    "\n",
    "for subject_id, ax_row in zip(subject_order, axs):\n",
    "    for sample_type, ax in zip(sample_type_order, ax_row):\n",
    "        d1 = d0[\n",
    "            lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)\n",
    "        ].assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "        ax.scatter(\n",
    "            \"xpos\", \"rotu_rabund\", data=d1, color=\"k\", s=10, label=\"__nolegend__\"\n",
    "        )\n",
    "        # ax.set_aspect(700, adjustable=\"datalim\", anchor=\"NW\")\n",
    "        ax.set_ylim(-1e-5, 1)\n",
    "        ax.set_yscale(\"symlog\", linthresh=1e-4, linscale=0.1)\n",
    "        ax.set_xlim(-0.5, _grid_sample_counts[sample_type].max())\n",
    "        ax.set_xticks(d1.xpos)\n",
    "        ax.set_xticklabels(d1.full_label)\n",
    "\n",
    "        strain_frac_sample_list = list(set(d1.index) & set(sf_fit.sample.values))\n",
    "        if len(strain_frac_sample_list) == 0:\n",
    "            print(f\"No strain analysis for {subject_id}.\")\n",
    "            comm = []\n",
    "            _strain_order = []\n",
    "        else:\n",
    "            w = (\n",
    "                sf_fit.sel(sample=strain_frac_sample_list)\n",
    "                .drop_low_abundance_strains(drop_strains_thresh)\n",
    "                .rename_coords(strain=str)\n",
    "            )\n",
    "            _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "            comm = w.community.to_pandas()\n",
    "        d2 = d1.join(comm)\n",
    "        # Plot stacked barplot\n",
    "        ax1 = ax.twinx()\n",
    "        top_last = 0\n",
    "        for strain in _strain_order:\n",
    "            ax1.bar(\n",
    "                x=\"xpos\",\n",
    "                height=strain,\n",
    "                data=d2,\n",
    "                bottom=top_last,\n",
    "                width=bar_width,\n",
    "                alpha=1.0,\n",
    "                color=strain_palette[strain],\n",
    "                edgecolor=\"k\",\n",
    "                lw=1,\n",
    "                label=\"__nolegend__\",\n",
    "            )\n",
    "            top_last += d2[strain]\n",
    "            ax.scatter(\n",
    "                [], [], color=strain_palette[strain], label=strain, marker=\"s\", s=80\n",
    "            )\n",
    "        ax1.set_yticks([])\n",
    "        # Put strains behind points:\n",
    "        ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "        ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "        ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "        ax1.set_ylim(0, 1)\n",
    "        lib.plot.rotate_xticklabels(ax=ax)\n",
    "        if sample_type == \"human\":\n",
    "            ax.legend(loc=\"upper right\")\n",
    "\n",
    "\n",
    "# fig.tight_layout()\n",
    "\n",
    "# axs[0, 0].plot([0, 100], [0, 1])\n",
    "\n",
    "# axs[0, 1].plot([0, 10], [0, 1])\n",
    "# axs[0, 1].set_aspect(10)\n",
    "\n",
    "# axs[0, 2].plot([0, 20], [0, 1])\n",
    "# axs[0, 2].set_aspect(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_grid_sample_counts.max().values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "                \"full_label\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "    ax.set_xticklabels(d0.full_label)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "(\n",
    "    sample.loc[\n",
    "        lambda x: (\n",
    "            x.index.isin(sf_fit.sample.values)\n",
    "            & x.sample_type.isin([\"mouse\"])\n",
    "            & x.subject_id.isin([\"A\", \"B\", \"H\"])\n",
    "            # & (x.sample_type == \"Fermenter\")\n",
    "        ),\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"source_samples\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ],\n",
    "    ]\n",
    "    .sort_values(\n",
    "        [\n",
    "            \"subject_id\",\n",
    "            \"collection_date_relative_een_end\",\n",
    "            \"sample_type\",\n",
    "            \"diet_or_media\",\n",
    "            \"mouse_genotype\",\n",
    "            \"source_samples\",\n",
    "            \"status_mouse_inflamed\",\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        rotu_rabund=rotu_rabund[rotu_id],\n",
    "    )\n",
    "    # .join(comm)\n",
    "    .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## E. lenta (Zotu172 / 102544)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"102544\"\n",
    "rotu_id = \"Zotu172\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = [\"A\", \"B\", \"H\", \"N\", \"M\", \"S\"]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list) // 2,\n",
    "    ncols=2,\n",
    "    figsize=(10 * 2, 4 * len(subject_list) // 2),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    # ax1.legend()\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "ax.hist(d.log2_ratio, bins=20)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## F. plautii (Zotu49 / 100099)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"100099\"\n",
    "rotu_id = \"Zotu49\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "ax.hist(d.log2_ratio, bins=20)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bacteroides dorei (Zotu1 / 102478)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"102478\"\n",
    "rotu_id = \"Zotu1\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "ax.hist(d.log2_ratio, bins=20)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Bacteroides uniformis (Zotu6 / 101346)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"101346\"\n",
    "rotu_id = \"Zotu6\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order[:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "ax.hist(d.log2_ratio, bins=20)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## s__Bacteroides fragilis (Zotu12 / 101337)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "motu_id = \"101337\"\n",
    "rotu_id = \"Zotu12\"\n",
    "drop_strains_thresh = 0.5\n",
    "ylinthresh = 1e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{motu_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.01)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "mgtp_error = sf.evaluation.metagenotype_error2(sf_fit, discretized=False)[1]\n",
    "entrp_error = sf.evaluation.metagenotype_entropy_error(\n",
    "    sf_fit, discretized=False, p=1, montecarlo_draws=10\n",
    ")[1]\n",
    "comm_entrp = sf_fit.community.entropy().to_series()\n",
    "high_mgtp_error = mgtp_error >= 0.1\n",
    "high_entrp_error = entrp_error >= 0.2\n",
    "high_comm_entrp = comm_entrp >= 1.5\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = sf_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        sf_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sf_fit,\n",
    "    scalex=0.4,\n",
    "    scaley=0.6,\n",
    "    row_linkage_func=lambda w: strain_linkage,\n",
    "    row_colors=sf_fit.strain.to_series().map(strain_palette),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trnsfm_x = lambda x: np.sign(x) * np.sqrt(np.abs(x))\n",
    "bar_width = 1.0\n",
    "\n",
    "subject_list = subject_order  # [:3]\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    nrows=len(subject_list),\n",
    "    figsize=(10, 4 * len(subject_list)),\n",
    "    squeeze=False,\n",
    "    sharex=True,\n",
    "    sharey=True,\n",
    ")\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: x.subject_id == subject_id].index.to_series().pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[lambda x: (x.subject_id == subject_id) & x.sample_type.isin([\"human\"])]\n",
    "        .sort_values([\"collection_date_relative_een_end\"])\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[subject_sample_order, [\"collection_date_relative_een_end\"]]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "            trnsfm_collection_date_relative_een_end=lambda x: x.collection_date_relative_een_end.apply(\n",
    "                trnsfm_x\n",
    "            ),\n",
    "        )\n",
    "        .join(comm)\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"trnsfm_collection_date_relative_een_end\",\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].trnsfm_collection_date_relative_een_end,\n",
    "            height=d1,\n",
    "            bottom=top_last,\n",
    "            width=bar_width,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # Start and end of EEN\n",
    "    ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    ax.axvline(\n",
    "        trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "        lw=1,\n",
    "        linestyle=\"--\",\n",
    "        color=\"k\",\n",
    "    )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=1)\n",
    "    xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax1.set_ylim(0, 1.0)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    rotu_rabund.apply(lambda x: x + x.replace({0: np.inf}).min())\n",
    "    .join(sample)\n",
    "    .groupby([\"subject_id\", \"diet_or_media\"])[rotu_id]\n",
    "    .mean()\n",
    "    .unstack()[[\"EEN\", \"PostEEN\"]]\n",
    "    .dropna()\n",
    "    .assign(log2_ratio=lambda x: np.log2(x[\"PostEEN\"] / x[\"EEN\"]))\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "print(d.log2_ratio.mean())\n",
    "print(sp.stats.wilcoxon(d[\"PostEEN\"], d[\"EEN\"]))\n",
    "ax.hist(d.log2_ratio, bins=20)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3, 8))\n",
    "max_c_value = np.abs(d.log2_ratio).max()\n",
    "for subject_id, (een_rabund, post_rabund, log2_ratio, c) in d.assign(\n",
    "    c=lambda x: ((x.log2_ratio / max_c_value) + 1) / 2\n",
    ").iterrows():\n",
    "    ax.plot([0, 1], [een_rabund, post_rabund], c=mpl.cm.coolwarm(c), lw=4)\n",
    "ax.set_yscale(\"log\")\n",
    "ax.set_xticks([0, 1])\n",
    "ax.set_xticklabels([\"EEN\", \"PostEEN\"])\n",
    "ax.set_xlim(-0.1, 1.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for subject_id, sample_type in product([\"A\", \"B\", \"H\"], [\"Fermenter\", \"mouse\"]):\n",
    "    all_sample_list = (\n",
    "        sample[lambda x: (x.sample_type == sample_type) & (x.subject_id == subject_id)]\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    subject_sample_order = (\n",
    "        sample[\n",
    "            lambda x: (x.sample_type == sample_type)\n",
    "            & (x.subject_id == subject_id)\n",
    "            & (x.index.isin(sf_fit.sample.values))\n",
    "        ]\n",
    "        .sort_values(\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ]\n",
    "        )\n",
    "        .index.to_series()\n",
    "        .pipe(list)\n",
    "    )\n",
    "    # Cull low abundance strains for each subject\n",
    "\n",
    "    strain_frac_sample_list = list(set(all_sample_list) & set(sf_fit.sample.values))\n",
    "    if len(strain_frac_sample_list) == 0:\n",
    "        print(f\"No strain analysis for {subject_id}.\")\n",
    "        comm = []\n",
    "        _strain_order = []\n",
    "    else:\n",
    "        w = (\n",
    "            sf_fit.sel(sample=strain_frac_sample_list)\n",
    "            .drop_low_abundance_strains(drop_strains_thresh)\n",
    "            .rename_coords(strain=str)\n",
    "        )\n",
    "        _strain_order = [s for s in strain_order if s in w.strain] + [\"-1\"]\n",
    "        comm = w.community.to_pandas()\n",
    "\n",
    "    d0 = (\n",
    "        sample.loc[\n",
    "            subject_sample_order,\n",
    "            [\n",
    "                \"sample_type\",\n",
    "                \"diet_or_media\",\n",
    "                \"mouse_genotype\",\n",
    "                \"source_samples\",\n",
    "                \"status_mouse_inflamed\",\n",
    "            ],\n",
    "        ]\n",
    "        .assign(\n",
    "            rotu_rabund=rotu_rabund[rotu_id],\n",
    "        )\n",
    "        .join(comm)\n",
    "        .assign(xpos=lambda x: np.arange(len(x.index)))\n",
    "    )\n",
    "\n",
    "    fig, ax = plt.subplots(\n",
    "        figsize=(0.4 * len(strain_frac_sample_list), 4),\n",
    "    )\n",
    "\n",
    "    ax.plot(\n",
    "        \"rotu_rabund\",\n",
    "        data=d0,\n",
    "        marker=\"o\",\n",
    "        linestyle=\"\",\n",
    "        label=\"__nolegend__\",\n",
    "        color=\"k\",\n",
    "        markersize=5,\n",
    "    )\n",
    "\n",
    "    # Plot stacked barplot\n",
    "    ax1 = ax.twinx()\n",
    "    top_last = 0\n",
    "    for strain, d1 in d0[_strain_order].T.iterrows():\n",
    "        ax1.bar(\n",
    "            x=d0.loc[d1.index].xpos,\n",
    "            height=d1,\n",
    "            width=1,\n",
    "            bottom=top_last,\n",
    "            alpha=1.0,\n",
    "            color=strain_palette[strain],\n",
    "            edgecolor=\"k\",\n",
    "            lw=1,\n",
    "            label=strain,\n",
    "        )\n",
    "        top_last += d1\n",
    "    ax1.set_yticks([])\n",
    "\n",
    "    # Put strains behind points:\n",
    "    ax.set_zorder(ax1.get_zorder() + 1)  # put ax in front of ax1\n",
    "    ax.patch.set_visible(False)  # hide the 'canvas'\n",
    "    ax1.patch.set_visible(True)  # show the 'canvas'\n",
    "\n",
    "    # # Start and end of EEN\n",
    "    # ax.axvline(0, lw=1, linestyle=\"--\", color=\"k\")\n",
    "    # ax.axvline(\n",
    "    #     trnsfm_x(subject.een_start_date_relative_een_end.loc[subject_id]),\n",
    "    #     lw=1,\n",
    "    #     linestyle=\"--\",\n",
    "    #     color=\"k\",\n",
    "    # )\n",
    "    ax1.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "    ax.set_title(subject_id)\n",
    "\n",
    "    # xtick_pos = np.array([-100, -50, -10, 0, 10, 50, 100, 300])\n",
    "    # ax.set_xticks(trnsfm_x(xtick_pos))\n",
    "    # ax.set_xticklabels(xtick_pos)\n",
    "    ax.set_yscale(\"symlog\", linthresh=ylinthresh)\n",
    "    ax.set_ylim(0, 1.0)\n",
    "    ax.set_xlim(-0.5, len(strain_frac_sample_list) - 0.5)\n",
    "    lib.plot.rotate_xticklabels(ax=ax)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot complex experimental design"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_transfer_sample_lists = {\n",
    "    # B \u2013 PreEEN (active disease)\n",
    "    # CF-7 with direct transfer CF 379,380,381,384,385,386\n",
    "    # CF7 with ex vivo samples CF 97,98,99,100 (first FR, then EEN)\n",
    "    # Ex vivo with post ex vivo transfer\n",
    "    # CF 97,98 with 397,406,408,409,395,402\n",
    "    # CF 99,100 with 430,426,427,428,429,431\n",
    "    \"B\": {\n",
    "        \"Subject B:\\nTime Series\": (0, [\n",
    "            \"CF_7\",\n",
    "            \"CF_8\",\n",
    "            \"CF_9\",\n",
    "            \"CF_10\",\n",
    "            \"CF_11\",\n",
    "            \"CF_12\",\n",
    "            \"CF_13\",\n",
    "            \"CF_14\",\n",
    "        ]),\n",
    "        \"Direct Transfer\": (1, [\n",
    "            \"CF_7\",\n",
    "            \"CF_379\",\n",
    "            \"CF_380\",\n",
    "            \"CF_381\",\n",
    "            \"CF_384\",\n",
    "            \"CF_385\",\n",
    "            \"CF_386\",\n",
    "        ]),\n",
    "        \"Ex Vivo\": (1, [\n",
    "            \"CF_7\",\n",
    "            \"CF_97\",\n",
    "            \"CF_98\",\n",
    "            \"CF_99\",\n",
    "            \"CF_100\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo Transfer #1\": (2, [\n",
    "            \"CF_97\",\n",
    "            \"CF_98\",\n",
    "            \"CF_397\",\n",
    "            \"CF_406\",\n",
    "            \"CF_408\",\n",
    "            \"CF_409\",\n",
    "            \"CF_395\",\n",
    "            \"CF_402\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo Transfer #2\": (2, [\n",
    "            \"CF_99\",\n",
    "            \"CF_100\",\n",
    "            \"CF_430\",\n",
    "            \"CF_426\",\n",
    "            \"CF_427\",\n",
    "            \"CF_428\",\n",
    "            \"CF_429\",\n",
    "            \"CF_431\",\n",
    "        ]),\n",
    "    },\n",
    "    # A \u2013 EEN (remission)\n",
    "    # CF-3 with direct transfer CF 140,141,142, 149,150,151\n",
    "    # CF-3 with ex vivo samples CF_103,104, 101,102 (started with EEN, then FR)\n",
    "    # Ex vivo with post ex vivo transfer\n",
    "    # cf 103,104 with 107-175\n",
    "    # cf 101,102 with 152-157\n",
    "    \"A\": {\n",
    "        \"Subject A:\\nTime Series\": (0, [\n",
    "            \"CF_1\",\n",
    "            \"CF_2\",\n",
    "            \"CF_3\",\n",
    "            \"CF_6\",\n",
    "        ]),\n",
    "        \"Direct Transfer\": (1, [\n",
    "            \"CF_3\",\n",
    "            \"CF_140\",\n",
    "            \"CF_141\",\n",
    "            \"CF_142\",\n",
    "            \"CF_149\",\n",
    "            \"CF_150\",\n",
    "            \"CF_151\",\n",
    "        ]),\n",
    "        \"Ex Vivo\": (1, [\n",
    "            \"CF_3\",\n",
    "            \"CF_103\",\n",
    "            \"CF_104\",\n",
    "            \"CF_101\",\n",
    "            \"CF_102\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo Transfer #1\": (2, [\n",
    "            \"CF_103\",\n",
    "            \"CF_104\",\n",
    "            \"CF_170\",\n",
    "            \"CF_171\",\n",
    "            \"CF_172\",\n",
    "            \"CF_173\",\n",
    "            \"CF_174\",\n",
    "            \"CF_175\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo Transfer #2\": (2, [\n",
    "            \"CF_101\",\n",
    "            \"CF_102\",\n",
    "            \"CF_152\",\n",
    "            \"CF_153\",\n",
    "            \"CF_154\",\n",
    "            \"CF_155\",\n",
    "            \"CF_156\",\n",
    "            \"CF_157\",\n",
    "        ]),\n",
    "    },\n",
    "    # H \u2013 EEN (persistent inflammation)\n",
    "    # CF48 with direct transfer CF 115-120\n",
    "    # CF48 with ex vivo samples 107,108 (109,110) (first EEN, then FR (no FR transfer in mice)\n",
    "    # Ex vivo with post ex vivo transfer\n",
    "    # 107,108 with 127-133 (chow) and 667-672 (EEN-like mouse diet; PD)\n",
    "    \"H\": {\n",
    "        \"Subject H:\\nTime Series\": (0, [\n",
    "            \"CF_46\",\n",
    "            \"CF_47\",\n",
    "            \"CF_48\",\n",
    "            \"CF_49\",\n",
    "            \"CF_50\",\n",
    "            \"CF_51\",\n",
    "        ]),\n",
    "        \"Direct Transfer\": (1, [\n",
    "            \"CF_48\",\n",
    "            \"CF_115\",\n",
    "            \"CF_116\",\n",
    "            \"CF_117\",\n",
    "            \"CF_118\",\n",
    "            \"CF_119\",\n",
    "            \"CF_120\",\n",
    "        ]),\n",
    "        \"Ex Vivo\": (1, [\n",
    "            \"CF_48\",\n",
    "            \"CF_107\",\n",
    "            \"CF_108\",\n",
    "            \"CF_109\",\n",
    "            \"CF_110\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo - Chow Diet\": (2, [\n",
    "            \"CF_107\",\n",
    "            \"CF_108\",\n",
    "            \"CF_127\",\n",
    "            \"CF_128\",\n",
    "            # \"CF_129\",  # ???  # In Debbie's email, but seems to be wrong.\n",
    "            \"CF_130\",\n",
    "            \"CF_131\",\n",
    "            \"CF_132\",\n",
    "            \"CF_133\",\n",
    "        ]),\n",
    "        \"Post Ex Vivo - EEN-Like Diet\": (2, [\n",
    "            \"CF_107\",\n",
    "            \"CF_108\",\n",
    "            \"CF_667\",\n",
    "            \"CF_668\",\n",
    "            \"CF_669\",\n",
    "            \"CF_670\",\n",
    "            \"CF_671\",\n",
    "            \"CF_672\",\n",
    "        ]),\n",
    "    },\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = \"101493\"\n",
    "subject_order = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "    \"M\",\n",
    "    \"P\",\n",
    "]\n",
    "\n",
    "print(motu_taxonomy.loc[species])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_fit = (\n",
    "    sf.data.World.load(\n",
    "        f\"data/group/een/species/sp-{species}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    "    )\n",
    "    .rename_coords(sample=lambda s: \"CF_{}\".format(int(s.split(\"_\")[1])))\n",
    "    .rename_coords(sample={\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "    .drop_low_abundance_strains(0.2)\n",
    "    .rename_coords(strain=str)\n",
    ")\n",
    "print(strain_fit.sizes)\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_linkage = strain_fit.genotype.linkage(optimal_ordering=True)\n",
    "strain_order = list(\n",
    "    linkage_order(\n",
    "        strain_linkage,\n",
    "        strain_fit.strain.values,\n",
    "    )\n",
    ")\n",
    "strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "strain_order.append(\"-1\")  # Add to end of list\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    "    extend={\"-1\": \"lightgrey\"},\n",
    ")\n",
    "\n",
    "sf.evaluation.metagenotype_error2(strain_fit, discretized=True)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "sample_linkage = strain_fit.unifrac_linkage(optimal_ordering=True)\n",
    "position_ss = strain_fit.random_sample(position=min(strain_fit.sizes['position'], 1000)).position\n",
    "\n",
    "sf.plot.plot_metagenotype(strain_fit.sel(position=position_ss), col_linkage_func=lambda w: sample_linkage)\n",
    "sf.plot.plot_community(strain_fit, col_linkage_func=lambda w: sample_linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ncols = 4\n",
    "\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=ncols,\n",
    "    naxes=len(subject_order),\n",
    "    ax_width=4,\n",
    "    gridspec_kw=dict(hspace=1.2, wspace=0.0),\n",
    "    sharey=True,\n",
    ")\n",
    "\n",
    "for subject, ax in zip(subject_order, axs.flatten()):\n",
    "    subject_sample_order = sample.sort_values(\"collection_date_relative_een_end\")[\n",
    "        lambda x: (x.subject_id == subject)\n",
    "        & (x.sample_type == \"human\")\n",
    "        & (x.index.isin(motu_rabund.index))\n",
    "    ].index\n",
    "    try:\n",
    "        subject_comm_sample_list = list(set(subject_sample_order) & set(strain_fit.sample.values))\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    ax.set_title(subject)\n",
    "    d = (\n",
    "        sample.reindex(subject_sample_order)\n",
    "        # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "        # .sort_values(\"collection_date_relative_een_end\")\n",
    "        .assign(\n",
    "            t=lambda x: range(len(x)),\n",
    "        )\n",
    "    ).join(subject_comm)\n",
    "\n",
    "    plot_stacked_barplot(\n",
    "        data=d,\n",
    "        x_var=\"t\",\n",
    "        order=[s for s in strain_order if s in subject_comm.columns],\n",
    "        palette=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    # ax.legend(bbox_to_anchor=(1, 1))\n",
    "    ax.set_xticklabels(d.timepoint)\n",
    "    ax.set_aspect(3, anchor=\"NW\")\n",
    "    lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "    ax.set_yticks([0, 0.5, 1.0])\n",
    "    ax.set_xlim(-0.5, d.shape[0] - 0.5)\n",
    "\n",
    "fig.savefig(f\"fig/{species}.een_strain_time_series_plot.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(\n",
    "    3,\n",
    "    5,\n",
    "    figsize=(5 * 5, 3 * 3.5),\n",
    "    squeeze=False,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(hspace=1.5, wspace=0),\n",
    ")\n",
    "\n",
    "for subject, axs_row in zip(subject_transfer_sample_lists, axs):\n",
    "    subject_comm_sample_list = list(\n",
    "        set(idxwhere(sample.subject_id == subject)) & set(strain_fit.sample.values)\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    for (sample_list_label, (num_offset_samples, sample_list)), ax in zip(\n",
    "        subject_transfer_sample_lists[subject].items(), axs_row\n",
    "    ):\n",
    "        d = (\n",
    "            sample.reindex(sample_list)\n",
    "            # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "            # .sort_values(\"collection_date_relative_een_end\")\n",
    "            .assign(\n",
    "                t=lambda x: range(len(x)),\n",
    "                simple_label=lambda x: np.where(\n",
    "                    x.sample_type == \"human\",\n",
    "                    x.timepoint,\n",
    "                    np.where(\n",
    "                        x.sample_type == \"Fermenter\",\n",
    "                        x.diet_or_media,\n",
    "                        np.where(\n",
    "                            x.status_mouse_inflamed == \"Inflamed\",\n",
    "                            x.diet_or_media + \" / inflam\",\n",
    "                            x.diet_or_media + \" / not\",\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            )\n",
    "        ).join(subject_comm)\n",
    "        d.loc[d.index[:num_offset_samples], 't'] -= 0.2  # Offset width\n",
    "        \n",
    "\n",
    "        plot_stacked_barplot(\n",
    "            data=d,\n",
    "            x_var=\"t\",\n",
    "            order=[s for s in strain_order if s in subject_comm.columns],\n",
    "            palette=strain_palette,\n",
    "            ax=ax,\n",
    "            width=0.8,\n",
    "            lw=0.5,\n",
    "        )\n",
    "        \n",
    "        ax.set_title(sample_list_label)\n",
    "        ax.set_xticklabels(d.simple_label)\n",
    "        ax.set_aspect(4, anchor=\"NW\")\n",
    "        lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "        ax.set_yticks([0, 0.5, 1.0])\n",
    "        ax.set_xlim(d.t.min() - 0.5, d.t.max() + 0.5)\n",
    "    ax.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "fig.savefig(f\"fig/{species}.een_strain_transfer_expt_plot.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(\n",
    "    3,\n",
    "    5,\n",
    "    figsize=(5 * 5, 3 * 3.5),\n",
    "    squeeze=False,\n",
    "    sharey=True,\n",
    "    gridspec_kw=dict(hspace=1.5, wspace=0),\n",
    ")\n",
    "\n",
    "for subject, axs_row in zip(subject_transfer_sample_lists, axs):\n",
    "    subject_comm_sample_list = list(\n",
    "        set(idxwhere(sample.subject_id == subject)) & set(strain_fit.sample.values)\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    for (sample_list_label, (num_offset_samples, sample_list)), ax in zip(\n",
    "        subject_transfer_sample_lists[subject].items(), axs_row\n",
    "    ):\n",
    "        d = (\n",
    "            sample.reindex(sample_list)\n",
    "            # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "            # .sort_values(\"collection_date_relative_een_end\")\n",
    "            .assign(\n",
    "                t=lambda x: range(len(x)),\n",
    "                simple_label=lambda x: np.where(\n",
    "                    x.sample_type == \"human\",\n",
    "                    x.timepoint,\n",
    "                    np.where(\n",
    "                        x.sample_type == \"Fermenter\",\n",
    "                        x.diet_or_media,\n",
    "                        np.where(\n",
    "                            x.status_mouse_inflamed == \"Inflamed\",\n",
    "                            x.diet_or_media + \" / inflam\",\n",
    "                            x.diet_or_media + \" / not\",\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            )\n",
    "        ).join(subject_comm)\n",
    "        d.loc[d.index[:num_offset_samples], 't'] -= 0.2  # Offset width\n",
    "        \n",
    "\n",
    "        plot_stacked_barplot(\n",
    "            data=d,\n",
    "            x_var=\"t\",\n",
    "            order=[s for s in strain_order if s in subject_comm.columns],\n",
    "            palette=strain_palette,\n",
    "            ax=ax,\n",
    "            width=0.8,\n",
    "            lw=0.5,\n",
    "        )\n",
    "        \n",
    "        ax.set_title(sample_list_label)\n",
    "        ax.set_xticklabels(d.index)\n",
    "        ax.set_aspect(4, anchor=\"NW\")\n",
    "        lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "        ax.set_yticks([0, 0.5, 1.0])\n",
    "        ax.set_xlim(d.t.min() - 0.5, d.t.max() + 0.5)\n",
    "    ax.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "\n",
    "fig.savefig(f\"fig/{species}.een_strain_transfer_expt_plot.sample_id.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf.plot.plot_metagenotype_frequency_spectrum_compare_samples(strain_fit, ['CF_6', 'CF_3'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(\n",
    "    3,\n",
    "    5,\n",
    "    figsize=(5 * 5, 3 * 3.5),\n",
    "    squeeze=False,\n",
    "    sharey=False,\n",
    "    gridspec_kw=dict(hspace=1.5, wspace=0),\n",
    ")\n",
    "\n",
    "for subject, axs_row in zip(subject_transfer_sample_lists, axs):\n",
    "    subject_comm_sample_list = list(\n",
    "        set(idxwhere(sample.subject_id == subject)) & set(strain_fit.sample.values)\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    for (sample_list_label, (num_offset_samples, sample_list)), ax in zip(\n",
    "        subject_transfer_sample_lists[subject].items(), axs_row\n",
    "    ):\n",
    "        d = (\n",
    "            sample.reindex(sample_list)\n",
    "            # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "            # .sort_values(\"collection_date_relative_een_end\")\n",
    "            .assign(\n",
    "                t=lambda x: range(len(x)),\n",
    "                simple_label=lambda x: np.where(\n",
    "                    x.sample_type == \"human\",\n",
    "                    x.timepoint,\n",
    "                    np.where(\n",
    "                        x.sample_type == \"Fermenter\",\n",
    "                        x.diet_or_media,\n",
    "                        np.where(\n",
    "                            x.status_mouse_inflamed == \"Inflamed\",\n",
    "                            x.diet_or_media + \" / inflam\",\n",
    "                            x.diet_or_media + \" / not\",\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            )\n",
    "        ).join(subject_comm)\n",
    "        # d.loc[d.index[:num_offset_samples], 't'] -= 0.7  # Offset width\n",
    "\n",
    "        plot_stacked_barplot(\n",
    "            data=d,\n",
    "            x_var=\"t\",\n",
    "            order=[s for s in strain_order if s in subject_comm.columns],\n",
    "            palette=strain_palette,\n",
    "            ax=ax,\n",
    "            width=0.8,\n",
    "            lw=0,\n",
    "        )\n",
    "\n",
    "        # Mark inflammation\n",
    "        for (_, x) in d.iterrows():\n",
    "            if x.status_mouse_inflamed == 'Inflamed':\n",
    "                ax.annotate('*', xy=(x.t, 1.0))\n",
    "\n",
    "        ax.set_title(sample_list_label, pad=25)\n",
    "        ax.set_xticklabels(d.simple_label)\n",
    "        ax.set_aspect(9, anchor=\"NW\")\n",
    "        ax.set_ylim(0, 1.0)\n",
    "        lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "        ax.set_yticks(np.linspace(0, 1.0, num=6))\n",
    "        ax.yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, symbol=''))\n",
    "        ax.set_xlim(d.t.min() - 0.5, d.t.max() + 0.5)\n",
    "        ax.spines[['right', 'top']].set_visible(False)\n",
    "    ax.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "            \n",
    "fig.savefig(f\"fig/{species}.een_strain_transfer_expt_plot.paper_style.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf.plot.plot_genotype(strain_fit.random_sample(position=1000), row_linkage_func=lambda w: strain_linkage, scaley=0.4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_subject_coabundance_linkage = sp.cluster.hierarchy.linkage(strain_fit.community.to_pandas().groupby(sample.subject_id).mean().T, metric='cosine', optimal_ordering=True)\n",
    "shuffled_strain_order = maximally_shuffled_order(linkage_order(strain_subject_coabundance_linkage, labels=strain_fit.strain.values))\n",
    "shuffled_strain_order.remove(\"-1\")  # Drop \"other\" strain.\n",
    "shuffled_strain_order.append(\"-1\")  # Add to end of list\n",
    "shuffled_strain_palette = lib.plot.construct_ordered_palette(\n",
    "    shuffled_strain_order,\n",
    "    cm=\"rainbow\",\n",
    "    extend={\"-1\": \"lightgrey\"},\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    3,\n",
    "    5,\n",
    "    figsize=(5 * 5, 3 * 3.5),\n",
    "    squeeze=False,\n",
    "    sharey=False,\n",
    "    gridspec_kw=dict(hspace=1.5, wspace=0),\n",
    ")\n",
    "\n",
    "for subject, axs_row in zip(subject_transfer_sample_lists, axs):\n",
    "    subject_comm_sample_list = list(\n",
    "        set(idxwhere(sample.subject_id == subject)) & set(strain_fit.sample.values)\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    for (sample_list_label, (num_offset_samples, sample_list)), ax in zip(\n",
    "        subject_transfer_sample_lists[subject].items(), axs_row\n",
    "    ):\n",
    "        d = (\n",
    "            sample.reindex(sample_list)\n",
    "            # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "            # .sort_values(\"collection_date_relative_een_end\")\n",
    "            .assign(\n",
    "                t=lambda x: range(len(x)),\n",
    "                simple_label=lambda x: np.where(\n",
    "                    x.sample_type == \"human\",\n",
    "                    x.timepoint,\n",
    "                    np.where(\n",
    "                        x.sample_type == \"Fermenter\",\n",
    "                        x.diet_or_media,\n",
    "                        np.where(\n",
    "                            x.status_mouse_inflamed == \"Inflamed\",\n",
    "                            x.diet_or_media + \" / inflam\",\n",
    "                            x.diet_or_media + \" / not\",\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            )\n",
    "        ).join(subject_comm)\n",
    "        # d.loc[d.index[:num_offset_samples], 't'] -= 0.7  # Offset width\n",
    "\n",
    "        plot_stacked_barplot(\n",
    "            data=d,\n",
    "            x_var=\"t\",\n",
    "            order=[s for s in shuffled_strain_order if s in subject_comm.columns],\n",
    "            palette=shuffled_strain_palette,\n",
    "            ax=ax,\n",
    "            width=0.8,\n",
    "            lw=0,\n",
    "        )\n",
    "\n",
    "        # Mark inflammation\n",
    "        for (_, x) in d.iterrows():\n",
    "            if x.status_mouse_inflamed == 'Inflamed':\n",
    "                ax.annotate('*', xy=(x.t, 1.0))\n",
    "\n",
    "        ax.set_title(sample_list_label, pad=25)\n",
    "        ax.set_xticklabels(d.simple_label)\n",
    "        ax.set_aspect(9, anchor=\"NW\")\n",
    "        ax.set_ylim(0, 1.0)\n",
    "        lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "        ax.set_yticks(np.linspace(0, 1.0, num=6))\n",
    "        ax.yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, symbol=''))\n",
    "        ax.set_xlim(d.t.min() - 0.5, d.t.max() + 0.5)\n",
    "        ax.spines[['right', 'top']].set_visible(False)\n",
    "    ax.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "            \n",
    "fig.savefig(f\"fig/{species}.een_strain_transfer_expt_plot.paper_style.pdf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "more_colors_strain_palette = lib.plot.construct_ordered_palette(\n",
    "    strain_order,\n",
    "    cm=\"rainbow\",\n",
    "    extend={\"-1\": \"lightgrey\"},\n",
    "    desaturate_levels=[1.0, 0.7, 0.4],\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    3,\n",
    "    5,\n",
    "    figsize=(5 * 5, 3 * 3.5),\n",
    "    squeeze=False,\n",
    "    sharey=False,\n",
    "    gridspec_kw=dict(hspace=1.5, wspace=0),\n",
    ")\n",
    "\n",
    "for subject, axs_row in zip(subject_transfer_sample_lists, axs):\n",
    "    subject_comm_sample_list = list(\n",
    "        set(idxwhere(sample.subject_id == subject)) & set(strain_fit.sample.values)\n",
    "    )\n",
    "\n",
    "    try:\n",
    "        subject_comm = (\n",
    "            strain_fit.sel(sample=subject_comm_sample_list)\n",
    "            .drop_low_abundance_strains(0.2, agg_strain_coord=\"-1\")\n",
    "            .community.to_pandas()\n",
    "        )\n",
    "    except ValueError:\n",
    "        subject_comm = pd.DataFrame([], columns=[\"-1\"])\n",
    "\n",
    "    for (sample_list_label, (num_offset_samples, sample_list)), ax in zip(\n",
    "        subject_transfer_sample_lists[subject].items(), axs_row\n",
    "    ):\n",
    "        d = (\n",
    "            sample.reindex(sample_list)\n",
    "            # .dropna(subset=[\"collection_date_relative_een_end\"])\n",
    "            # .sort_values(\"collection_date_relative_een_end\")\n",
    "            .assign(\n",
    "                t=lambda x: range(len(x)),\n",
    "                simple_label=lambda x: np.where(\n",
    "                    x.sample_type == \"human\",\n",
    "                    x.timepoint,\n",
    "                    np.where(\n",
    "                        x.sample_type == \"Fermenter\",\n",
    "                        x.diet_or_media,\n",
    "                        np.where(\n",
    "                            x.status_mouse_inflamed == \"Inflamed\",\n",
    "                            x.diet_or_media + \" / inflam\",\n",
    "                            x.diet_or_media + \" / not\",\n",
    "                        ),\n",
    "                    ),\n",
    "                ),\n",
    "            )\n",
    "        ).join(subject_comm)\n",
    "        # d.loc[d.index[:num_offset_samples], 't'] -= 0.7  # Offset width\n",
    "\n",
    "        plot_stacked_barplot(\n",
    "            data=d,\n",
    "            x_var=\"t\",\n",
    "            order=[s for s in strain_order if s in subject_comm.columns],\n",
    "            palette=more_colors_strain_palette,\n",
    "            ax=ax,\n",
    "            width=0.8,\n",
    "            lw=0,\n",
    "        )\n",
    "\n",
    "        # Mark inflammation\n",
    "        for (_, x) in d.iterrows():\n",
    "            if x.status_mouse_inflamed == 'Inflamed':\n",
    "                ax.annotate('*', xy=(x.t, 1.0))\n",
    "\n",
    "        ax.set_title(sample_list_label, pad=25)\n",
    "        ax.set_xticklabels(d.simple_label)\n",
    "        ax.set_aspect(9, anchor=\"NW\")\n",
    "        ax.set_ylim(0, 1.0)\n",
    "        lib.plot.rotate_xticklabels(rotation=45, ax=ax)\n",
    "        ax.set_yticks(np.linspace(0, 1.0, num=6))\n",
    "        ax.yaxis.set_major_formatter(mtick.PercentFormatter(xmax=1, symbol=''))\n",
    "        ax.set_xlim(d.t.min() - 0.5, d.t.max() + 0.5)\n",
    "        ax.spines[['right', 'top']].set_visible(False)\n",
    "    ax.legend(bbox_to_anchor=(1, 1), ncols=2)\n",
    "            \n",
    "fig.savefig(f\"fig/{species}.een_strain_transfer_expt_plot.paper_style.pdf\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "toolz2",
   "language": "python",
   "name": "toolz2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}