{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "from warnings import filterwarnings\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from statsmodels.graphics.regressionplots import influence_plot\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "import lib.thisproject.data\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _calculate_2tailed_pvalue_from_perm(obs, perms):\n",
    "    hypoth_left = perms > obs\n",
    "    hypoth_right = perms < obs\n",
    "    null_p_left = (hypoth_left.sum() + 1) / (len(hypoth_left) + 1)\n",
    "    null_p_right = (hypoth_right.sum() + 1) / (len(hypoth_right) + 1)\n",
    "    return np.minimum(null_p_left, null_p_right) * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def linkage_order(linkage, labels):\n",
    "    return labels[sp.cluster.hierarchy.to_tree(linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "\n",
    "\n",
    "def is_prime(n):\n",
    "    if n <= 1:\n",
    "        return False\n",
    "    for i in range(2, int(n**0.5) + 1):\n",
    "        if n % i == 0:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "def iterate_primes_up_to(n, return_index=False):\n",
    "    n = int(np.ceil(n))\n",
    "    idx = 0\n",
    "    for i in range(n):\n",
    "        if is_prime(i):\n",
    "            if return_index:\n",
    "                yield (idx, i)\n",
    "            else:\n",
    "                yield i\n",
    "            idx += 1\n",
    "\n",
    "\n",
    "def maximally_shuffled_order(sorted_order):\n",
    "    n = len(sorted_order)\n",
    "    primes_list = list(iterate_primes_up_to(np.sqrt(n)))\n",
    "    table = pd.DataFrame(np.arange(n), index=sorted_order, columns=[\"original_order\"])\n",
    "    for prime in primes_list:\n",
    "        table[prime] = table.original_order % prime\n",
    "    table.sort_values(primes_list).original_order.values\n",
    "    table = table.assign(new_order=table.sort_values(primes_list).original_order.values)\n",
    "    z = table.sort_values(\"new_order\").original_order.values\n",
    "    table[\"delta\"] = [np.nan] + list(z[1:] - z[:-1])\n",
    "    return table.sort_values(\"new_order\").index.to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = (\n",
    "    pd.read_table(\n",
    "        \"data/group/een/r.proc.gtpro.species_depth.tsv\",\n",
    "        index_col=[\"sample\", \"species_id\"],\n",
    "    )\n",
    "    .depth.unstack(fill_value=0)\n",
    "    .rename(columns=str)\n",
    ")\n",
    "\n",
    "bins = np.linspace(0, 30_000, num=200)\n",
    "\n",
    "fig, axs = plt.subplots(2, sharex=True)\n",
    "\n",
    "for (title, x), ax in zip(\n",
    "    dict(\n",
    "        total_depth_by_sample=species_depth.sum(1),\n",
    "        total_depth_by_species=species_depth.sum(0),\n",
    "    ).items(),\n",
    "    axs.flatten(),\n",
    "):\n",
    "    ax.hist(x, bins=np.logspace(-1, 5, num=100))\n",
    "    ax.set_title(title)\n",
    "    ax.set_xscale(\"log\")\n",
    "fig.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_rabund = species_depth.divide(species_depth.sum(1), axis=0)\n",
    "species_rabund.mean().sort_values(ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_rabund = species_depth.divide(species_depth.sum(1), axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_species = 10\n",
    "top_species = (\n",
    "    (species_rabund > 1e-5).sum().sort_values(ascending=False).head(n_species).index\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    n_species, figsize=(5, 0.3 * n_species), sharex=True, sharey=True\n",
    ")\n",
    "\n",
    "bins = np.logspace(-8, 1, num=51)\n",
    "\n",
    "for species_id, ax in zip(top_species, axs):\n",
    "    # ax.hist(rabund_subset[species_id], bins=bins, alpha=0.7)\n",
    "    ax.hist(species_rabund[species_id], bins=bins, alpha=0.7)\n",
    "    ax.set_xscale(\"log\")\n",
    "    prevalence = (species_rabund[species_id] > 1e-5).mean()\n",
    "    ax.set_title(\"\")\n",
    "    # ax.set_xticks()\n",
    "    # ax.set_yticks()\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.patch.set_alpha(0.0)\n",
    "    for spine in [\"left\", \"right\", \"top\", \"bottom\"]:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    ax.annotate(\n",
    "        f\"{species_id} ({prevalence:0.0%})\",\n",
    "        xy=(0.05, 0.1),\n",
    "        ha=\"left\",\n",
    "        xycoords=\"axes fraction\",\n",
    "    )\n",
    "    ax.set_xlim(left=1e-9)\n",
    "    ax.set_ylim(top=20)\n",
    "    ax.axvline(1e-5, lw=1, linestyle=\":\", color=\"k\")\n",
    "\n",
    "ax.xaxis.set_visible(True)\n",
    "ax.spines[\"bottom\"].set_visible(True)\n",
    "ax.set_xticks([1e-4, 1e-2, 1e-0])\n",
    "ax.set_xticklabels([\"0.01%\", \"1%\", \"100%\"])\n",
    "ax.set_xlabel(\"Relative Abundance\")\n",
    "\n",
    "# fig.subplots_adjust(hspace=-0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(\n",
    "    \"ref/gtpro/species_taxonomy_ext.tsv\"\n",
    ")\n",
    "species_taxonomy.loc[\"102506\", \"s__\"] = \"s__Escherichia coli\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _species_id in top_species.astype(str):\n",
    "    print(_species_id, \":\", species_taxonomy.taxonomy_string.loc[_species_id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy[lambda x: x.s__.str.endswith(\"hansenii\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _species_id in [\"102544\", \"102506\", \"101303\", \"100150\", \"102330\", \"101704\"]:\n",
    "    print(\n",
    "        _species_id,\n",
    "        (species_rabund[_species_id] > 0.0001).mean().round(2),\n",
    "        (species_rabund[_species_id] > 0.001).mean().round(2),\n",
    "        species_taxonomy.loc[_species_id].s__,\n",
    "        sep=\"\\t\\t\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n_species = 20\n",
    "top_species = (\n",
    "    (species_rabund > 1e-3).sum().sort_values(ascending=False).head(n_species).index\n",
    ")\n",
    "\n",
    "fig, axs = plt.subplots(\n",
    "    n_species, figsize=(5, 0.3 * n_species), sharex=True, sharey=True\n",
    ")\n",
    "\n",
    "bins = np.logspace(-8, 1, num=51)\n",
    "\n",
    "for species_id, ax in zip(top_species, axs):\n",
    "    # ax.hist(rabund_subset[species_id], bins=bins, alpha=0.7)\n",
    "    ax.hist(species_rabund[species_id], bins=bins, alpha=0.7)\n",
    "    ax.set_xscale(\"log\")\n",
    "    prevalence = (species_rabund[species_id] > 1e-3).mean()\n",
    "    ax.set_title(\"\")\n",
    "    # ax.set_xticks()\n",
    "    # ax.set_yticks()\n",
    "    ax.yaxis.set_visible(False)\n",
    "    ax.xaxis.set_visible(False)\n",
    "    ax.patch.set_alpha(0.0)\n",
    "    for spine in [\"left\", \"right\", \"top\", \"bottom\"]:\n",
    "        ax.spines[spine].set_visible(False)\n",
    "    ax.annotate(\n",
    "        f\"{species_id} ({prevalence:0.0%})\",\n",
    "        xy=(0.05, 0.1),\n",
    "        ha=\"left\",\n",
    "        xycoords=\"axes fraction\",\n",
    "    )\n",
    "    ax.set_xlim(left=1e-9)\n",
    "    ax.set_ylim(top=20)\n",
    "    ax.axvline(1e-5, lw=1, linestyle=\":\", color=\"k\")\n",
    "\n",
    "ax.xaxis.set_visible(True)\n",
    "ax.spines[\"bottom\"].set_visible(True)\n",
    "ax.set_xticks([1e-4, 1e-2, 1e-0])\n",
    "ax.set_xticklabels([\"0.01%\", \"1%\", \"100%\"])\n",
    "ax.set_xlabel(\"Relative Abundance\")\n",
    "\n",
    "# fig.subplots_adjust(hspace=-0.75)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_depth = []\n",
    "missing_files = []\n",
    "for species_id in species_depth.columns:\n",
    "    path = f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.comm.tsv\"\n",
    "    try:\n",
    "        d = pd.read_table(path, index_col=[\"sample\", \"strain\"]).squeeze().unstack()\n",
    "    except FileNotFoundError:\n",
    "        missing_files.append(path)\n",
    "        d = pd.DataFrame([])\n",
    "    _keep_strains = idxwhere(d.sum() > 0.05)\n",
    "    d = d.reindex(index=species_depth.index, columns=_keep_strains, fill_value=0)\n",
    "    d = d.assign(__other=lambda x: 1 - x.sum(1)).rename(columns={\"__other\": -1})\n",
    "    d[d < 0] = 0\n",
    "    d = d.divide(d.sum(1), axis=0)\n",
    "    d = d.multiply(species_depth[species_id], axis=0)\n",
    "    d = d.rename(columns=lambda s: f\"{species_id}_{s}\")\n",
    "    strain_depth.append(d)\n",
    "strain_depth = pd.concat(strain_depth, axis=1)\n",
    "strain_rabund = strain_depth.divide(strain_depth.sum(1), axis=0)\n",
    "len(species_depth.columns), len(missing_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(strain_rabund.values.flatten() + 1e-10, bins=np.logspace(-10, 0))\n",
    "plt.yscale(\"log\")\n",
    "plt.xscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _species_id in top_species.astype(str):\n",
    "    print(_species_id, \":\", species_taxonomy.taxonomy_string.loc[_species_id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta = (\n",
    "    pd.read_table(\"meta/een-mgen/stool.tsv\")\n",
    "    # .rename(columns={'Seq-Name': 'sample', 'CED/Patient-recoded': 'subject_id', 'sampleDate': 'date', 'Diet (=PreEEN, EEN, PostEEN)': 'sample_type'})\n",
    "    # .assign(\n",
    "    #     date=lambda x: pd.to_datetime(x.date),\n",
    "    #     sample_type=lambda x: x.sample_type.fillna('???')\n",
    "    # )\n",
    "    .set_index(\"mgen_id\")\n",
    "    # .sort_values(['subject_id', 'date', 'sample_type'])\n",
    ")\n",
    "\n",
    "# FIXME: Metadata seems to include a swap in the metagenomic data of CF_11 and CF_15.\n",
    "meta = meta.rename({\"CF_11\": \"CF_15\", \"CF_15\": \"CF_11\"})\n",
    "meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "micro_meta = (\n",
    "    pd.read_table(\"meta/een-mgen/microcosm.tsv\")\n",
    "    .set_index(\"mgen_id\")\n",
    "    .rename(columns={\"inoculum_subject_id\": \"subject_id\"})\n",
    ")\n",
    "micro_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert meta.index.isin(species_rabund.index).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta[[\"subject_id\", \"sample_type\"]].value_counts().unstack(fill_value=0).sort_values(\n",
    "    \"EEN\", ascending=False\n",
    ").head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Species Metagenotype / Strain Deconvolution"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_id = \"100099\"\n",
    "species_taxonomy.loc[species_id]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(0)\n",
    "\n",
    "mgtp_all = sf.data.Metagenotype.load(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.mgtp.nc\"\n",
    ")\n",
    "world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "position_ss = world.random_sample(position=min(1000, world.sizes[\"position\"])).position\n",
    "\n",
    "strain_frac = world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "abundant_strain_list = list(strain_frac.columns)\n",
    "abundant_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(abundant_strain_list, cm=\"rainbow\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sf.plot.plot_metagenotype(\n",
    "    world.sel(position=position_ss),\n",
    "    col_linkage_func=lambda w: w.metagenotype.linkage(),\n",
    "    dwidth=0.01,\n",
    "    scalex=0.147,\n",
    "    col_colors_func=None,\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    world.sel(position=position_ss),\n",
    "    col_linkage_func=lambda w: w.metagenotype.linkage(),\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(),\n",
    "    dwidth=0.01,\n",
    "    row_colors_func=None,\n",
    ")\n",
    "sf.plot.plot_genotype(\n",
    "    world.sel(position=position_ss),\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(),\n",
    "    row_colors_func=lambda w: w.genotype.entropy(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert (world.sample.isin(micro_meta.index) | world.sample.isin(meta.index)).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Focal Species Plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### g__Escherichia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102506\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    # f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "    # f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts49-s85-seed0.world.nc\"\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "\n",
    "# # Arbitrarily ordered palette:\n",
    "# _strain_list = list(_frac.columns)\n",
    "# _strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "_world_drop_low_abund = _world.drop_low_abundance_strains(0.05)\n",
    "_strain_list = list(linkage_order(\n",
    "    _world_drop_low_abund.genotype.linkage(optimal_ordering=True), _world_drop_low_abund.strain.values\n",
    "))\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    _strain_list, cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains (e.g. found in the same subject)\n",
    "# # showing up with similar colors.\n",
    "# _strain_list = linkage_order(\n",
    "#     sp.cluster.hierarchy.linkage(\n",
    "#         _frac.groupby(meta.subject_id).mean().T,\n",
    "#         method=\"average\",\n",
    "#         metric=\"cosine\",\n",
    "#         optimal_ordering=True,\n",
    "#     ),\n",
    "#     index=_frac.columns,\n",
    "# )\n",
    "# _strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         _strain_list\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(_species_id)\n",
    "_world0 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "_world1 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts49-s85-seed0.world.nc\"\n",
    ")\n",
    "w0 = (\n",
    "    _world0.random_sample(position=1000)\n",
    "    .sel(sample=d0.dropna(subset=[-1]).index)\n",
    "    .drop_low_abundance_strains(0.05)\n",
    ")\n",
    "w1 = _world1.sel(position=w0.position, sample=w0.sample).drop_low_abundance_strains(\n",
    "    0.05\n",
    ")\n",
    "\n",
    "plt.scatter(w0.community.entropy(), w1.community.entropy())\n",
    "plt.plot([0, 1], [0, 1])\n",
    "\n",
    "sf.plot.plot_metagenotype(w0, col_cluster=False)\n",
    "sf.plot.plot_community(\n",
    "    w0,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    w1,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")\n",
    "sf.plot.plot_genotype(\n",
    "    w0,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### s__Eggerthella lenta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102544\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         linkage_order(\n",
    "#             sp.cluster.hierarchy.linkage(\n",
    "#                 _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "#                 method=\"average\",\n",
    "#                 metric=\"cosine\",\n",
    "#                 optimal_ordering=True,\n",
    "#             ),\n",
    "#             index=_strain_list,\n",
    "#         )\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(_species_id)\n",
    "_world0 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "_world1 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    ")\n",
    "w0 = (\n",
    "    _world0.random_sample(position=1000)\n",
    "    .sel(sample=d0.dropna(subset=[-1]).index)\n",
    "    .drop_low_abundance_strains(0.05)\n",
    ")\n",
    "w1 = _world1.sel(position=w0.position, sample=w0.sample).drop_low_abundance_strains(\n",
    "    0.05\n",
    ")\n",
    "\n",
    "plt.scatter(w0.community.entropy(), w1.community.entropy())\n",
    "plt.plot([0, 1], [0, 1])\n",
    "\n",
    "sf.plot.plot_metagenotype(w0, col_cluster=False)\n",
    "sf.plot.plot_community(\n",
    "    w0,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    w1,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample='CF_035'\n",
    "print(_world0.metagenotype.mean_depth().sel(sample=sample).values)\n",
    "sf.plot.plot_metagenotype_frequency_spectrum(_world0, sample)\n",
    "# plt.yscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### g__Dorea scindens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"101303\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    # f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         linkage_order(\n",
    "#             sp.cluster.hierarchy.linkage(\n",
    "#                 _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "#                 method=\"average\",\n",
    "#                 metric=\"cosine\",\n",
    "#                 optimal_ordering=True,\n",
    "#             ),\n",
    "#             index=_strain_list,\n",
    "#         )\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_world0 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "_world1 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    ")\n",
    "w0 = (\n",
    "    _world0.random_sample(position=1000)\n",
    "    .sel(sample=d0.dropna(subset=[-1]).index)\n",
    "    .drop_low_abundance_strains(0.05)\n",
    ")\n",
    "w1 = _world1.sel(position=w0.position, sample=w0.sample).drop_low_abundance_strains(\n",
    "    0.05\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "ax.scatter(w0.community.entropy(), w1.community.entropy())\n",
    "ax.plot([0, 1], [0, 1])\n",
    "ax.set_aspect(1)\n",
    "\n",
    "\n",
    "sf.plot.plot_metagenotype(w0, col_cluster=False)\n",
    "sf.plot.plot_community(\n",
    "    w0,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    w1,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### s__Hungatella hathewayi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"101303\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         linkage_order(\n",
    "#             sp.cluster.hierarchy.linkage(\n",
    "#                 _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "#                 method=\"average\",\n",
    "#                 metric=\"cosine\",\n",
    "#                 optimal_ordering=True,\n",
    "#             ),\n",
    "#             index=_strain_list,\n",
    "#         )\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### s__Eisenbergiella tayi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102330\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         linkage_order(\n",
    "#             sp.cluster.hierarchy.linkage(\n",
    "#                 _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "#                 method=\"average\",\n",
    "#                 metric=\"cosine\",\n",
    "#                 optimal_ordering=True,\n",
    "#             ),\n",
    "#             index=_strain_list,\n",
    "#         )\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### s__Blautia hansenii"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"101704\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# Genotype similarity ordered palette:\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "# # Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     maximally_shuffled_order(\n",
    "#         linkage_order(\n",
    "#             sp.cluster.hierarchy.linkage(\n",
    "#                 _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "#                 method=\"average\",\n",
    "#                 metric=\"cosine\",\n",
    "#                 optimal_ordering=True,\n",
    "#             ),\n",
    "#             index=_strain_list,\n",
    "#         )\n",
    "#     ),\n",
    "#     cm=\"rainbow\",\n",
    "# )\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Other Species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### s__Flavonifractor plautii"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"100099\"\n",
    "print(species_taxonomy.loc[_species_id])\n",
    "_species_name = species_taxonomy.loc[_species_id].s__[len(\"s__\") :].replace(\" \", \"_\")\n",
    "_species_rabund = species_rabund[_species_id]\n",
    "_world = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "    # f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    ")\n",
    "_frac = _world.drop_low_abundance_strains(0.05).community.to_pandas()\n",
    "_strain_list = list(_frac.columns)\n",
    "_strain_list.remove(-1)  # Drop \"other\" strain.\n",
    "# # Arbitrarily ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(_strain_list, cm=\"rainbow\")\n",
    "\n",
    "# # Genotype similarity ordered palette:\n",
    "# strain_palette = lib.plot.construct_ordered_palette(\n",
    "#     linkage_order(_world.genotype.linkage(optimal_ordering=True), _world.strain), cm=\"rainbow\"\n",
    "# )\n",
    "\n",
    "# Construct an ordered palette, but use the order to AVOID closely correlated strains showing up with similar colors.\n",
    "strain_palette = lib.plot.construct_ordered_palette(\n",
    "    maximally_shuffled_order(\n",
    "        linkage_order(\n",
    "            sp.cluster.hierarchy.linkage(\n",
    "                _frac[_strain_list].groupby(meta.subject_id).mean().T,\n",
    "                method=\"average\",\n",
    "                metric=\"cosine\",\n",
    "                optimal_ordering=True,\n",
    "            ),\n",
    "            index=_strain_list,\n",
    "        )\n",
    "    ),\n",
    "    cm=\"rainbow\",\n",
    ")\n",
    "\n",
    "d0 = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            meta.assign(\n",
    "                label=lambda x: x.assign(idx=x.index)[\n",
    "                    [\"idx\", \"collection_date_relative_een_end\", \"sample_type\"]\n",
    "                ].apply(tuple, axis=1)\n",
    "            ),\n",
    "            micro_meta.assign(\n",
    "                collection_date_relative_een_end=np.inf,\n",
    "                sample_type=lambda x: x.inoculum_mgen_id.map(lambda s: f\"invitro[{s}]\"),\n",
    "                label=lambda x: x.assign(idx=x.index)[[\"idx\", \"sample_type\"]].apply(\n",
    "                    tuple, axis=1\n",
    "                ),\n",
    "            ),\n",
    "        ]\n",
    "    )\n",
    "    .assign(\n",
    "        # has_strain_deconv=lambda x: x.index.isin(_comm.index),\n",
    "        species_rabund=_species_rabund,\n",
    "    )\n",
    "    .join(_frac)\n",
    "    .sort_values([\"subject_id\", \"collection_date_relative_een_end\", \"sample_type\"])\n",
    ")\n",
    "\n",
    "subject_list = [\n",
    "    \"A\",\n",
    "    \"B\",\n",
    "    \"H\",\n",
    "    \"C\",\n",
    "    \"D\",\n",
    "    \"E\",\n",
    "    \"F\",\n",
    "    \"G\",\n",
    "    \"K\",\n",
    "    \"L\",\n",
    "    \"M\",\n",
    "    \"N\",\n",
    "    \"O\",\n",
    "    \"P\",\n",
    "    \"Q\",\n",
    "    \"R\",\n",
    "    \"S\",\n",
    "    \"T\",\n",
    "    \"U\",\n",
    "]\n",
    "fig, axs = lib.plot.subplots_grid(\n",
    "    ncols=4,\n",
    "    naxes=len(subject_list),\n",
    "    ax_width=5,\n",
    "    ax_height=4,\n",
    ")\n",
    "fig.suptitle(_species_name)\n",
    "for subject_id, ax in zip(subject_list, axs.flatten()):\n",
    "    ax.set_title(subject_id)\n",
    "    twin_ax = ax.twinx()\n",
    "    d1 = d0[lambda x: x.subject_id == subject_id].set_index(\"label\")\n",
    "    d1[[-1] + _strain_list].plot(\n",
    "        kind=\"bar\",\n",
    "        width=0.95,\n",
    "        stacked=True,\n",
    "        color=strain_palette,\n",
    "        ax=ax,\n",
    "    )\n",
    "    d1.species_rabund.plot(kind=\"line\", ax=twin_ax, color=\"k\")\n",
    "    ax.legend_.set_visible(False)\n",
    "    ax.set_ylim(0, 1)\n",
    "    ax.set_ylabel(\"strain fraction\")\n",
    "    ax.set_xlabel(\"\")\n",
    "    twin_ax.set_ylabel(\"species relative abundance\")\n",
    "    twin_ax.set_ylim(0)\n",
    "    lib.plot.rotate_xticklabels(ax)\n",
    "    ax.set_xlim(-0.5, 14)\n",
    "fig.tight_layout()\n",
    "\n",
    "_outpath = f\"fig/een_{_species_name}_strain_tracking.pdf\"\n",
    "print(_outpath)\n",
    "fig.savefig(_outpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(_species_id)\n",
    "_world0 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts37-s85-seed0.world.nc\"\n",
    ")\n",
    "_world1 = sf.data.World.load(\n",
    "    f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.filt-poly05-cvrg05.ss-g10000-block0-seed0.fit-sfacts48-s85-seed0.world.nc\"\n",
    ")\n",
    "w0 = (\n",
    "    _world0.random_sample(position=1000)\n",
    "    .sel(sample=d0.dropna(subset=[-1]).index)\n",
    "    .drop_low_abundance_strains(0.05)\n",
    ")\n",
    "w1 = _world1.sel(position=w0.position, sample=w0.sample).drop_low_abundance_strains(\n",
    "    0.05\n",
    ")\n",
    "\n",
    "plt.scatter(w0.community.entropy(), w1.community.entropy())\n",
    "plt.plot([0, 2.5], [0, 2.5])\n",
    "\n",
    "sf.plot.plot_metagenotype(w0, col_cluster=False)\n",
    "sf.plot.plot_community(\n",
    "    w0,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")\n",
    "sf.plot.plot_community(\n",
    "    w1,\n",
    "    col_cluster=False,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(optimal_ordering=True),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples=['CF_022']\n",
    "print(_world0.metagenotype.mean_depth().sel(sample=samples).values)\n",
    "sf.plot.plot_metagenotype_frequency_spectrum_compare_samples(_world0, samples)\n",
    "plt.yscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Permutation Test"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Species Turnover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    species_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_rabund = _rabund.rename(_perm).loc[_rabund.index]\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_cdist = pdist(_perm_rabund, metric=\"braycurtis\")\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc=perm_bc_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"np.log(time_delta)\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Turnover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_rabund = _rabund.rename(_perm).loc[_rabund.index]\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_cdist = pdist(_perm_rabund, metric=\"braycurtis\")\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc=perm_bc_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"np.log(time_delta)\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Turnover beyond Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select data\n",
    "_strain_rabund, _species_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    strain_rabund, species_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_strain_cdist = pdist(_strain_rabund, metric=\"braycurtis\")\n",
    "bc_species_cdist = pdist(_species_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc_strain=bc_strain_cdist,\n",
    "        bc_species=bc_species_cdist,\n",
    "        bc_diff=bc_strain_cdist - bc_species_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc_diff\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc_diff\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc_diff ~ 0 + C(pair_type) + C(subject_id, Sum) + time_delta\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_strain_rabund = _strain_rabund.rename(_perm).loc[_rabund.index]\n",
    "    _perm_species_rabund = _species_rabund.rename(_perm).loc[_rabund.index]\n",
    "\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_strain_cdist = pdist(_perm_strain_rabund, metric=\"braycurtis\")\n",
    "    perm_bc_species_cdist = pdist(_perm_species_rabund, metric=\"braycurtis\")\n",
    "\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc_strain=perm_bc_strain_cdist,\n",
    "            bc_species=perm_bc_species_cdist,\n",
    "            bc_diff=perm_bc_strain_cdist - perm_bc_species_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"time_delta\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc_diff\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    plt.scatter(x=\"bc_species\", y=\"bc_strain\", label=pair_type, data=d3)\n",
    "plt.legend()\n",
    "plt.xlabel(\"Species BC\")\n",
    "plt.ylabel(\"Strain BC\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Within Species Turnover Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### E. coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102506\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        sampleA_depth=_world.metagenotype.mean_depth()\n",
    "        .to_series()\n",
    "        .loc[pairs.str[0]]\n",
    "        .values,\n",
    "        sampleB_depth=_world.metagenotype.mean_depth()\n",
    "        .to_series()\n",
    "        .loc[pairs.str[1]]\n",
    "        .values,\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    "    min_sample_depth=lambda x: np.minimum(x.sampleA_depth, x.sampleB_depth),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_rabund = _rabund.rename(_perm).loc[_rabund.index]\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_cdist = pdist(_perm_rabund, metric=\"braycurtis\")\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc=perm_bc_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"np.log(time_delta)\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### E. lenta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102544\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        sampleA_depth=_world.metagenotype.mean_depth()\n",
    "        .to_series()\n",
    "        .loc[pairs.str[0]]\n",
    "        .values,\n",
    "        sampleB_depth=_world.metagenotype.mean_depth()\n",
    "        .to_series()\n",
    "        .loc[pairs.str[1]]\n",
    "        .values,\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    "    min_sample_depth=lambda x: np.minimum(x.sampleA_depth, x.sampleB_depth),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_rabund = _rabund.rename(_perm).loc[_rabund.index]\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_cdist = pdist(_perm_rabund, metric=\"braycurtis\")\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc=perm_bc_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"np.log(time_delta)\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### C. scindens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"101303\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = (\n",
    "    _world.sel(sample=_world.metagenotype.mean_depth() > 1.0)\n",
    "    .drop_low_abundance_strains(0.05)\n",
    "    .community.to_pandas()\n",
    ")\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute labels within subjects\n",
    "    _perm = (\n",
    "        _meta.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .mgen_id.transform(np.random.permutation)\n",
    "    )\n",
    "    _perm_rabund = _rabund.rename(_perm).loc[_rabund.index]\n",
    "    # Calculate pairwise comparisons\n",
    "    perm_bc_cdist = pdist(_perm_rabund, metric=\"braycurtis\")\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            bc=perm_bc_cdist,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(4)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[3]\n",
    "param_name = \"np.log(time_delta)\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Permutation Test Prototype #2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Species Turnover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    species_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Turnover"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    xx = np.linspace(10, 500)\n",
    "    logxx = np.log(xx)\n",
    "    yy = fit.params[f'C(pair_type)[{pair_type}]'] + logxx * fit.params[f'np.log(time_delta)']\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "    ax.plot(xx, yy)\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Within Species Turnover Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### E. coli"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102506\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### E. lenta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"102544\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### C. scindens"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"101303\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### F. plautii"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_species_id = \"100099\"\n",
    "path = f\"data/group/een/species/sp-{_species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "_world = sf.data.World.load(path)\n",
    "single_species_strain_rabund = _world.sel(\n",
    "    sample=_world.metagenotype.mean_depth() > 0.5\n",
    ").community.to_pandas()\n",
    "\n",
    "# Select data\n",
    "_rabund, _meta = lib.pandas_util.align_indexes(\n",
    "    single_species_strain_rabund, meta[meta.sample_type.isin([\"EEN\", \"PostEEN\"])]\n",
    ")\n",
    "\n",
    "# Calculate pairwise comparisons\n",
    "bc_cdist = pdist(_rabund, metric=\"braycurtis\")\n",
    "time_cdist = pdist(\n",
    "    _meta[[\"collection_date_relative_een_end\"]], metric=lambda x, y: np.abs(x - y)\n",
    ")\n",
    "diff_subject_cdist = pdist(_meta[[\"subject_id\"]], metric=lambda x, y: x != y)\n",
    "same_subject_cdist = (1 - diff_subject_cdist).astype(bool)\n",
    "type_transition_indicator = pdist(\n",
    "    _meta[[\"sample_type\"]],\n",
    "    metric=lambda x, y: (x == \"PostEEN\").astype(float) + (y == \"PostEEN\").astype(float),\n",
    ")\n",
    "# same_type_cdist = (1 - diff_type_cdist).astype(bool)\n",
    "pairs = (\n",
    "    pd.DataFrame(\n",
    "        squareform(diff_subject_cdist),\n",
    "        index=_meta.index,\n",
    "        columns=_meta.index,\n",
    "    )\n",
    "    .rename_axis(index=\"sampleA\", columns=\"sampleB\")\n",
    "    .unstack()\n",
    "    .index.to_frame()\n",
    "    .assign(pair=lambda x: x[[\"sampleA\", \"sampleB\"]].apply(tuple, axis=1))\n",
    "    .pair.unstack()\n",
    ")\n",
    "pairs = pd.Series(pairs.values[np.triu_indices_from(pairs.values, k=1)])\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        sampleA=pairs.str[0],\n",
    "        sampleB=pairs.str[1],\n",
    "        bc=bc_cdist,\n",
    "        same_subject=same_subject_cdist,\n",
    "        type_transition_indicator=type_transition_indicator,\n",
    "        diff_type=type_transition_indicator == 1,\n",
    "        time_delta=time_cdist,\n",
    "    )\n",
    ").assign(\n",
    "    pair_type=lambda x: x.type_transition_indicator.map(\n",
    "        {0: \"EEN\", 1: \"Transition\", 2: \"PostEEN\"}\n",
    "    ),\n",
    "    subject_id=lambda x: x.sampleA.map(_meta.subject_id),\n",
    "    subject_id_other=lambda x: x.sampleB.map(_meta.subject_id),\n",
    ")\n",
    "d1 = d0[lambda x: x.same_subject]\n",
    "\n",
    "\n",
    "# Plot and observed relationship\n",
    "fig, axs = plt.subplots(1, 2)\n",
    "ax = axs[0]\n",
    "sns.boxenplot(y=\"bc\", x=\"same_subject\", hue=\"pair_type\", data=d0, ax=ax)\n",
    "# ax.legend()\n",
    "ax = axs[1]\n",
    "sns.stripplot(y=\"bc\", x=\"pair_type\", hue=\"subject_id\", data=d1, ax=ax)\n",
    "ax.legend_.set_visible(False)\n",
    "\n",
    "\n",
    "# Fit observed relationship\n",
    "formula = \"bc ~ 0 + C(pair_type) + C(subject_id, Sum) + np.log(time_delta)\"\n",
    "fit = smf.ols(formula, data=d1).fit()\n",
    "observed_stat = fit.params\n",
    "observed_stat[\"EEN - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "observed_stat[\"EEN - Transition\"] = (\n",
    "    observed_stat[\"C(pair_type)[EEN]\"] - observed_stat[\"C(pair_type)[Transition]\"]\n",
    ")\n",
    "observed_stat[\"Transition - PostEEN\"] = (\n",
    "    observed_stat[\"C(pair_type)[Transition]\"] - observed_stat[\"C(pair_type)[PostEEN]\"]\n",
    ")\n",
    "\n",
    "# Calculate permutations\n",
    "np.random.seed(1)\n",
    "perm_stat = {}\n",
    "for i in tqdm(range(999)):\n",
    "    # Permute sample-pair labels within subjects\n",
    "    _perm_type = (\n",
    "        d0.assign(mgen_id=lambda x: x.index)\n",
    "        .groupby(\"subject_id\")\n",
    "        .pair_type.transform(np.random.permutation)\n",
    "    )\n",
    "    perm_fit = smf.ols(\n",
    "        formula,\n",
    "        data=d0.assign(\n",
    "            pair_type=_perm_type,\n",
    "        )[lambda x: x.same_subject],\n",
    "    ).fit()\n",
    "    _stat = perm_fit.params\n",
    "    _stat[\"EEN - PostEEN\"] = _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    _stat[\"EEN - Transition\"] = (\n",
    "        _stat[\"C(pair_type)[EEN]\"] - _stat[\"C(pair_type)[Transition]\"]\n",
    "    )\n",
    "    _stat[\"Transition - PostEEN\"] = (\n",
    "        _stat[\"C(pair_type)[Transition]\"] - _stat[\"C(pair_type)[PostEEN]\"]\n",
    "    )\n",
    "    perm_stat[i] = _stat\n",
    "perm_stat = pd.DataFrame(perm_stat).T\n",
    "\n",
    "perm_pvalues = _calculate_2tailed_pvalue_from_perm(observed_stat, perm_stat)\n",
    "\n",
    "fig, axs = plt.subplots(3)\n",
    "# Plot permutation tests\n",
    "ax = axs[0]\n",
    "param_name = \"EEN - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[1]\n",
    "param_name = \"EEN - Transition\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "ax = axs[2]\n",
    "param_name = \"Transition - PostEEN\"\n",
    "ax.set_title(param_name)\n",
    "ax.hist(perm_stat[param_name], bins=20)\n",
    "ax.axvline(observed_stat[param_name])\n",
    "hypoth = perm_stat[param_name] > observed_stat[param_name]\n",
    "null_p = (hypoth.sum() + 1) / (len(hypoth) + 1)\n",
    "ax.annotate(\n",
    "    perm_pvalues[param_name],\n",
    "    xy=(0.95, 0.95),\n",
    "    xycoords=\"axes fraction\",\n",
    "    ha=\"right\",\n",
    "    va=\"top\",\n",
    ")\n",
    "fig.tight_layout()\n",
    "\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fit.params"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d2 = d1.assign(\n",
    "    predict=fit.predict(),\n",
    "    resid=fit.resid_pearson,\n",
    "    influence=fit.get_influence().summary_frame().cooks_d,\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "for pair_type, d3 in d2.groupby(\"pair_type\"):\n",
    "    xx = np.linspace(10, 500)\n",
    "    logxx = np.log(xx)\n",
    "    yy = fit.params[f'C(pair_type)[{pair_type}]'] + logxx * fit.params[f'np.log(time_delta)']\n",
    "    ax.scatter(\n",
    "        \"time_delta\",\n",
    "        \"bc\",\n",
    "        label=pair_type,\n",
    "        data=d3,\n",
    "    )\n",
    "    ax.plot(xx, yy)\n",
    "ax.legend()\n",
    "ax.set_ylabel(\"Within-Subjects\\nBray-Curtis Distance\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "art = ax.scatter(\n",
    "    \"time_delta\",\n",
    "    \"resid\",\n",
    "    c=\"influence\",\n",
    "    data=d2,\n",
    ")\n",
    "fig.colorbar(art, label=\"Cook's D\")\n",
    "ax.set_ylabel(\"Residual BC\\n(standardized)\")\n",
    "ax.set_xlabel(\"Within-Subjects Days between Samples\")\n",
    "ax.set_xscale(\"symlog\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}