{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Template Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels as sm\n",
    "import statsmodels.formula.api as smf\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from sklearn.metrics import adjusted_mutual_info_score, adjusted_rand_score\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "import lib.thisproject.data\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Papermill parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell is tagged \"parameters\" for papermill.\n",
    "# See <https://papermill.readthedocs.io/en/latest/usage-parameterize.html#how-parameters-work> for some gotchas.\n",
    "# NOTE: *ALL* parameters should be passed to papermill. Values set here are only for prototyping.\n",
    "species_id = \"102506\"\n",
    "show_unimportant_figures = \"True\"  # Since this is coming in through papermill, I'll need to treat it as a boolean below, and the only False string is \"\".\n",
    "\n",
    "species_taxonomy_inpath = \"ref/gtpro/species_taxonomy_ext.tsv\"\n",
    "sample_to_spgc_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.spgc_ss-all.strain_samples.tsv\"\n",
    "sfacts_fit_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    "# ref_geno_inpath = f\"data/species/sp-{species_id}/midasdb.geno.nc\"  # \"De-novo GT-Pro genotype\"\n",
    "ref_geno_inpath = (\n",
    "    f\"data/species/sp-{species_id}/gtpro_ref.mgtp.nc\"  # \"Reference GT-Pro genotype\"\n",
    ")\n",
    "spgc_meta_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_meta.tsv\"\n",
    "ref_gene_copy_number_uhgg_inpath = (\n",
    "    f\"data/species/sp-{species_id}/gene75_new.reference_copy_number.nc\"\n",
    ")\n",
    "spgc_gene_uhgg_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_gene.tsv\"\n",
    "spgc_gene_uhgg_depth_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_depth_ratio.tsv\"\n",
    "spgc_gene_uhgg_corr_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_correlation.tsv\"\n",
    "uhgg_x_eggnog_inpath = (\n",
    "    f\"data/species/sp-{species_id}/pangenome_new.centroids.emapper.gene_x_eggnog.tsv\"\n",
    ")\n",
    "uhgg_x_top_eggnog_inpath = f\"data/species/sp-{species_id}/pangenome_new.centroids.emapper.gene_x_top_eggnog.tsv\"\n",
    "uhgg_gene_length_inpath = (\n",
    "    f\"ref/midasdb_uhgg_new/pangenomes/{species_id}/cluster_info.txt\"\n",
    ")\n",
    "gene_annotations_inpath = f\"data/species/sp-{species_id}/pangenome_new.centroids.emapper.d/proteins.emapper.annotations\"\n",
    "uhgg_depth_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    "species_depth_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv\"\n",
    "mgen_inpath = \"meta/hmp2/mgen.tsv\"\n",
    "preparation_inpath = \"meta/hmp2/preparation.tsv\"\n",
    "stool_inpath = \"meta/hmp2/stool.tsv\"\n",
    "subject_inpath = \"meta/hmp2/subject.tsv\"\n",
    "\n",
    "# TODO: Export SPGC gene info: prevalence, passes filtering against references,\n",
    "#    concordance with SNP dissimilarity in each genome set (both using SPGC-only, SPGC+refs, refs-only),\n",
    "#    cluster membership for each clustering analysis (SPGC-only, SPGC+refs, refs-only),\n",
    "#    compiled metadata (length, COG category table), etc.\n",
    "gene_stats_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.gene_stats.tsv\"\n",
    "# TODO: Export info for each SPGC strain: passes filtering? relationship between\n",
    "#   minimum genotype distance to a reference and the associated minimum gene\n",
    "#   distance (both weighted and unweighted)\n",
    "spgc_strain_stats_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_strain_stats.tsv\"\n",
    "# TODO: Export info for each reference strain: relationship between\n",
    "#   minimum genotype distance to a reference and the associated minimum gene\n",
    "#   distance (both weighted and unweighted)\n",
    "ref_strain_stats_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.ref_strain_stats.tsv\"\n",
    "\n",
    "# TODO: Export anything else I need for cross-species comparisons. What might that be?\n",
    "# strain_mwas_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_mwas.tsv\"\n",
    "html_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_ref_comparison.html\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Loading / Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(species_taxonomy_inpath)\n",
    "species_taxonomy.loc[species_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_x_eggnog = pd.read_table(uhgg_x_eggnog_inpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_x_top_eggnog = pd.read_table(uhgg_x_top_eggnog_inpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_gene_length = (\n",
    "    pd.read_table(uhgg_gene_length_inpath)\n",
    "    .groupby(\"centroid_75\")\n",
    "    .centroid_99_length.mean()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "_gene_annotations = (\n",
    "    pd.read_table(\n",
    "        gene_annotations_inpath,\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "_gene_annotations = uhgg_gene_length.to_frame().join(_gene_annotations)\n",
    "\n",
    "_gene_annotations.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category1 = (\n",
    "    _gene_annotations.COG_category.fillna(\"-\").apply(list).explode()[lambda x: x != \"-\"]\n",
    ")\n",
    "gene_x_cog_category1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_x_category = pd.read_table(\n",
    "    \"ref/cog-20.meta.tsv\",\n",
    "    names=[\"cog\", \"cog_category\", \"description\", \"short_name\", \"_4\", \"_5\", \"_6\"],\n",
    "    index_col=\"cog\",\n",
    ").cog_category\n",
    "cog_x_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_annotations.eggNOG_OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog = (\n",
    "    _gene_annotations.eggNOG_OGs.fillna(\"\")\n",
    "    .str.split(\",\")\n",
    "    .explode()[lambda x: x.str.startswith(\"COG\")]\n",
    "    .str.split(\"@\")\n",
    "    .str[0]\n",
    ")\n",
    "gene_x_cog.value_counts().head()\n",
    "gene_x_cog_category2 = gene_x_cog.map(cog_x_category).dropna().apply(list).explode()\n",
    "gene_x_cog_category2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            gene_x_cog_category1,\n",
    "            gene_x_cog_category2,  # FIXME: Which metadata table do I want?\n",
    "        ]\n",
    "    )\n",
    "    .reset_index()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "gene_x_cog_category.columns = [\"centroid_75\", \"cog_category\"]\n",
    "gene_x_cog_category = gene_x_cog_category.set_index(\"centroid_75\").cog_category\n",
    "gene_x_cog_category.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotations = _gene_annotations.assign(\n",
    "    COG_category=gene_x_cog_category.sort_values()\n",
    "    .reset_index()\n",
    "    .groupby(\"centroid_75\")\n",
    "    .apply(lambda x: \"\".join(x.cog_category.values))\n",
    ").assign(COG_category=lambda x: x.COG_category.fillna(\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category_matrix = (\n",
    "    gene_x_cog_category.reset_index()\n",
    "    .assign(tally=True)\n",
    "    .set_index([\"centroid_75\", \"cog_category\"])\n",
    "    .tally.unstack(\"cog_category\")\n",
    "    .fillna(False)\n",
    "    .reindex(gene_annotations.index, fill_value=False)\n",
    "    .assign(no_category=lambda x: x.sum(1) == 0)\n",
    ")\n",
    "gene_x_cog_category_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Raw Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = pd.read_table(\n",
    "    species_depth_inpath, names=[\"sample\", \"depth\"], index_col=[\"sample\"]\n",
    ").depth\n",
    "species_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen = pd.read_table(mgen_inpath, index_col=\"library_id\")\n",
    "preparation = pd.read_table(preparation_inpath, index_col=\"preparation_id\")\n",
    "stool = pd.read_table(stool_inpath, index_col=\"stool_id\")\n",
    "subject = pd.read_table(subject_inpath, index_col=\"subject_id\")\n",
    "\n",
    "mgen_meta = (\n",
    "    mgen.join(preparation, on=\"preparation_id\", lsuffix=\"_mgen\", rsuffix=\"_preparation\")\n",
    "    .join(stool, on=\"stool_id\")\n",
    "    .join(subject, on=\"subject_id\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_depth = xr.load_dataarray(uhgg_depth_inpath)\n",
    "subject_uhgg_depth = (\n",
    "    mgen_meta[[\"subject_id\"]]\n",
    "    .join(uhgg_depth.to_pandas())\n",
    "    .groupby(\"subject_id\")\n",
    "    .mean()\n",
    "    .dropna()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_depth_ratio = uhgg_depth / species_depth.to_xarray().sel(sample=uhgg_depth.sample)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_mean_species_depth = (\n",
    "    mgen_meta[[\"subject_id\"]]\n",
    "    .join(species_depth)\n",
    "    .groupby(\"subject_id\")\n",
    "    .depth.mean()\n",
    "    .dropna()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_subject_uhgg_depth = (\n",
    "    mgen_meta[[\"subject_id\"]]\n",
    "    .join(uhgg_depth.to_pandas())\n",
    "    .groupby(\"subject_id\")\n",
    "    .sum()\n",
    "    .dropna()\n",
    ")\n",
    "total_subject_species_depth = (\n",
    "    mgen_meta[[\"subject_id\"]]\n",
    "    .join(species_depth)\n",
    "    .groupby(\"subject_id\")\n",
    "    .depth.sum()\n",
    "    .dropna()\n",
    ")\n",
    "subject_uhgg_depth_ratio = total_subject_uhgg_depth.divide(\n",
    "    total_subject_species_depth, axis=0\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SPGC Strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_to_spgc = pd.read_table(sample_to_spgc_inpath, index_col=\"sample\").strain.astype(\n",
    "    str\n",
    ")\n",
    "\n",
    "spgc_palette = lib.plot.construct_ordered_palette(\n",
    "    sample_to_spgc, other=(0.8, 0.8, 0.8, 1.0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sfacts_fit = sf.World.load(sfacts_fit_inpath).drop_low_abundance_strains(0.5)\n",
    "# Strains should be str not int.\n",
    "sfacts_fit.data[\"strain\"] = sfacts_fit.strain.values.astype(str)\n",
    "print(dict(sfacts_fit.sizes))\n",
    "\n",
    "spgc_est_mgtp = sf.Metagenotype(\n",
    "    sfacts_fit.metagenotype.data.sel(sample=sample_to_spgc.index)\n",
    "    .groupby(sample_to_spgc.to_xarray())\n",
    "    .sum()\n",
    "    .rename(strain=\"sample\")\n",
    ")\n",
    "spgc_est_geno = spgc_est_mgtp.to_estimated_genotype(pseudo=0)\n",
    "\n",
    "# Pre-calculate shared heatmap decorations\n",
    "position_ss = sfacts_fit.random_sample(\n",
    "    position=min(500, sfacts_fit.sizes[\"position\"])\n",
    ").position\n",
    "w = sfacts_fit.sel(position=position_ss)\n",
    "sample_linkage = w.unifrac_linkage()\n",
    "try:\n",
    "    position_linkage = spgc_est_geno.linkage(\"position\")\n",
    "except ValueError as err:\n",
    "    print(err)\n",
    "    position_linkage = None\n",
    "sample_colors = w.sample.to_series().map(sample_to_spgc).map(spgc_palette)\n",
    "spgc_linkage = w.genotype.linkage()\n",
    "spgc_colors = w.strain.to_series().map(spgc_palette)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if show_unimportant_figures:\n",
    "    sf.plot.plot_community(sfacts_fit.sel(position=position_ss))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if show_unimportant_figures:\n",
    "    sf.plot.plot_metagenotype(\n",
    "        sfacts_fit.sel(position=position_ss),\n",
    "        col_linkage_func=lambda w: w.community.linkage(\"sample\"),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sf.data.Genotype.concat(\n",
    "    dict(mgen=spgc_est_geno, fit=sfacts_fit.genotype), dim=\"strain\"\n",
    ")\n",
    "g_pdist = g.pdist()\n",
    "g_linkage = g.linkage()\n",
    "\n",
    "sf.plot.plot_genotype(\n",
    "    g.sel(position=position_ss), transpose=True, col_linkage=g_linkage\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ref Strains"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# \"De-novo GT-Pro genotype\"\n",
    "ref_geno = xr.load_dataarray(ref_geno_inpath)\n",
    "ref_geno['genome_id'] = ref_geno.genome_id.to_series().map(lambda s: \"UHGG\" + s[len(\"GUT_GENOME\"):])\n",
    "ref_geno = ref_geno.fillna(0.5)\n",
    "ref_geno = sf.data.Genotype(ref_geno.rename({'genome_id': 'strain'}).T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Reference GT-Pro genotype\"\n",
    "ref_geno = sf.Metagenotype.load(ref_geno_inpath).to_estimated_genotype()\n",
    "ref_geno.data[\"strain\"] = ref_geno.strain.to_series().map(\n",
    "    lambda s: \"UHGG\" + s[len(\"GUT_GENOME\") :]\n",
    ")\n",
    "ref_geno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_meta = (\n",
    "    pd.read_table(\"ref/uhgg_genomes_all_4644.tsv\", index_col=\"Genome\")\n",
    "    .rename_axis(index=\"genome_id\")[\n",
    "        lambda x: x.MGnify_accession == \"MGYG-HGUT-\" + species_id[1:]\n",
    "    ]\n",
    "    .rename(lambda s: \"UHGG\" + s[len(\"GUT_GENOME\") :])\n",
    "    .loc[ref_geno.strain]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 100, num=101)\n",
    "plt.hist(reference_meta.Completeness, bins=bins)\n",
    "plt.hist(reference_meta.Contamination, bins=bins)\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Select any ref genotype that is within the top-10 closest distances from an SPGC strain.\n",
    "spgc_to_ref_geno_cdist = ref_geno.sel(\n",
    "    position=spgc_est_geno.position,\n",
    "    strain=idxwhere(\n",
    "        (reference_meta.Completeness > 97) & (reference_meta.Contamination < 2)\n",
    "    ),\n",
    ").cdist(spgc_est_geno)\n",
    "ref_list = list(\n",
    "    spgc_to_ref_geno_cdist.apply(lambda x: x.sort_values().head(10).index)\n",
    "    .stack()\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_meta = pd.read_table(spgc_meta_inpath, index_col=\"strain\").rename(str)\n",
    "print(spgc_meta.shape)\n",
    "spgc_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_copy_number_uhgg = xr.load_dataarray(ref_gene_copy_number_uhgg_inpath)\n",
    "ref_gene_uhgg = (ref_gene_copy_number_uhgg > 0).astype(int).to_pandas().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg = pd.read_table(spgc_gene_uhgg_inpath, index_col=\"gene_id\").rename_axis(\n",
    "    columns=\"strain\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_num_genes_uhgg = ref_gene_uhgg.sum()\n",
    "spgc_num_genes_uhgg = spgc_gene_uhgg.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ref_gene_uhgg_prevalence = ref_gene_uhgg[ref_list].mean(1)\n",
    "ref_gene_uhgg_entropy = (\n",
    "    -_ref_gene_uhgg_prevalence * np.log2(_ref_gene_uhgg_prevalence)\n",
    ").fillna(0)\n",
    "plt.hist(ref_gene_uhgg_entropy)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Selection / Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = spgc_meta[lambda x: x.species_gene_frac > 0.9].num_genes\n",
    "y = ref_num_genes_uhgg\n",
    "_df, _loc, _scale = sp.stats.t.fit(x.values, fix_df=2)\n",
    "_dist0 = sp.stats.t(_df, _loc, _scale)\n",
    "_dist1 = sp.stats.norm(_loc, _scale)\n",
    "\n",
    "thresh_max_num_uhgg_genes = _dist1.ppf(0.999)\n",
    "thresh_min_num_uhgg_genes = _dist1.ppf(0.001)\n",
    "\n",
    "\n",
    "bins = np.linspace(0, x.max() * 1.5, num=50)\n",
    "xx = np.linspace(0, x.max() * 1.5, num=1000)\n",
    "\n",
    "plt.hist(x, density=True, bins=bins, alpha=0.2)\n",
    "plt.hist(y, density=True, bins=bins, alpha=0.2)\n",
    "\n",
    "plt.plot(xx, _dist0.pdf(xx), color=\"k\")\n",
    "plt.plot(xx, _dist1.pdf(xx), color=\"k\", linestyle=\"--\")\n",
    "plt.axvline(thresh_max_num_uhgg_genes, lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.axvline(thresh_min_num_uhgg_genes, lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh_min_num_uhgg_genes, thresh_max_num_uhgg_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"sum_depth\",\n",
    "    \"species_gene_frac\",\n",
    "    c=(spgc_meta.num_genes - _loc) / _scale,\n",
    "    data=spgc_meta,\n",
    "    norm=mpl.colors.PowerNorm(1 / 1, vmin=-4, vmax=4),\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"logit\")\n",
    "plt.axvline(1, lw=1, color=\"k\", linestyle=\"--\")\n",
    "plt.axhline(0.9, lw=1, color=\"k\", linestyle=\"--\")\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Select SPGC strains that pass various filters\n",
    "spgc_list = idxwhere(\n",
    "    (spgc_meta.sum_depth > 1)\n",
    "    & (spgc_meta.species_gene_frac > 0.9)\n",
    "    & (spgc_num_genes_uhgg <= thresh_max_num_uhgg_genes)\n",
    "    & (spgc_num_genes_uhgg >= thresh_min_num_uhgg_genes)\n",
    ")\n",
    "\n",
    "print(len(ref_list), len(spgc_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(spgc_list) > 2"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "strain_geno_all = sf.Genotype.concat(dict(\n",
    "    ref=ref_geno.sel(position=spgc_est_geno.position),\n",
    "    spgc=spgc_est_geno.sel(strain=spgc_list),\n",
    "), dim='strain', rename=False).mlift('fillna', 0.5)\n",
    "\n",
    "print(ref_geno.sizes['strain'], len(spgc_list))\n",
    "strain_geno_all_pdist = strain_geno_all.pdist(q=1)\n",
    "strain_geno_all_linkage = strain_geno_all.linkage(pdist_kwargs=dict(q=1), optimal_ordering=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d0 = strain_geno_all_pdist.copy() + np.eye(strain_geno_all_pdist.shape[0])\n",
    "d1 = d0.min().to_frame(name='min_dist_to_ref').assign(is_ref=lambda x: x.index.isin(ref_geno.strain.values))\n",
    "\n",
    "sns.stripplot(data=d1, x='is_ref', y='min_dist_to_ref')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_geno = sf.Genotype.concat(\n",
    "    dict(\n",
    "        ref=ref_geno.sel(strain=ref_list, position=spgc_est_geno.position),\n",
    "        spgc=spgc_est_geno.sel(strain=spgc_list),\n",
    "    ),\n",
    "    dim=\"strain\",\n",
    "    rename=False,\n",
    ").mlift(\"fillna\", 0.5)\n",
    "\n",
    "strain_geno_pdist = strain_geno.pdist(q=1)\n",
    "strain_geno_linkage = strain_geno.linkage(pdist_kwargs=dict(q=1), optimal_ordering=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sf.plot.plot_community(w, scaley=0.5, col_linkage_func=lambda w: w.metagenotype.linkage())"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sf.plot.plot_metagenotype(w, col_linkage_func=lambda w: w.metagenotype.linkage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type = reference_meta.Genome_type.reindex(strain_geno.strain)\n",
    "genome_type.loc[genome_type.index.isin(spgc_list)] = \"SPGC\"\n",
    "genome_type_order = [\"Isolate\", \"MAG\", \"SPGC\"]\n",
    "genome_type_palette = lib.plot.construct_ordered_palette(genome_type_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_list = idxwhere(genome_type == \"Isolate\")\n",
    "mag_list = idxwhere(genome_type == \"MAG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(2, 1))\n",
    "for _genome_type in genome_type_order:\n",
    "    ax.scatter([], [], color=genome_type_palette[_genome_type], label=_genome_type)\n",
    "ax.legend()\n",
    "\n",
    "lib.plot.hide_axes_and_spines(ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(2, 1))\n",
    "for _genome_type in genome_type_order:\n",
    "    ax.scatter([], [], color=genome_type_palette[_genome_type], label=_genome_type)\n",
    "ax.legend(ncols=3)\n",
    "\n",
    "lib.plot.hide_axes_and_spines(ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_colors = genome_type.map(genome_type_palette)\n",
    "sf.plot.plot_genotype(\n",
    "    strain_geno.sel(position=position_ss),\n",
    "    transpose=True,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    col_colors=_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_colors = genome_type.map(genome_type_palette)\n",
    "sns.clustermap(\n",
    "    strain_geno_pdist,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_geno_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    # figsize=(40, 40),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 2))\n",
    "sns.heatmap([[]], vmin=0, vmax=1, ax=ax, cbar_kws=dict(location=\"bottom\"))\n",
    "lib.plot.hide_axes_and_spines(ax=ax)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import squareform\n",
    "\n",
    "bins = np.linspace(0, 1)\n",
    "plt.hist(\n",
    "    squareform(strain_geno_pdist.loc[ref_list, ref_list]),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    strain_geno_pdist.loc[ref_list, spgc_list].values.flatten(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    squareform(strain_geno_pdist.loc[spgc_list, spgc_list]),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=200)\n",
    "\n",
    "_pdist = strain_geno_pdist + np.eye(len(strain_geno_pdist))\n",
    "\n",
    "plt.hist(\n",
    "    _pdist.loc[ref_list, ref_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[ref_list, spgc_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[spgc_list, ref_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"ref-to-spgc\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[spgc_list, spgc_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "\n",
    "plt.legend()\n",
    "# plt.axvline(0.43, lw=1, linestyle='--', color='k')\n",
    "plt.xlabel(\"minimum distance\")\n",
    "plt.ylabel(\"cumulative fraction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg.shape, ref_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg = pd.concat(\n",
    "    [ref_gene_uhgg[ref_list], spgc_gene_uhgg[spgc_list]], axis=1\n",
    ").fillna(0)[lambda x: x.sum(1) > 0]\n",
    "strain_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 10000)\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"ref\",\n",
    "    density=True,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"spgc\",\n",
    "    density=True,\n",
    ")\n",
    "plt.legend()\n",
    "\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Geno/Gene Spaces"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# NOTE: This will take ~2 minutes to run for 40,000 genes.\n",
    "# TODO: Drop this calculation.\n",
    "gene_uhgg_cdmat = sp.spatial.distance.pdist(strain_gene_uhgg, metric=\"cosine\")\n",
    "gene_uhgg_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(gene_uhgg_cdmat),\n",
    "    index=strain_gene_uhgg.index,\n",
    "    columns=strain_gene_uhgg.index,\n",
    ")\n",
    "gene_uhgg_linkage = sp.cluster.hierarchy.linkage(gene_uhgg_cdmat, method=\"average\")\n",
    "gene_uhgg_linkage.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_gene_uhgg.T,\n",
    "    metric=\"cosine\",\n",
    "    w=ref_gene_uhgg_entropy.loc[strain_gene_uhgg.index],\n",
    ")\n",
    "strain_gene_uhgg_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_gene_uhgg_cdmat),\n",
    "    index=strain_gene_uhgg.columns,\n",
    "    columns=strain_gene_uhgg.columns,\n",
    ")\n",
    "\n",
    "assert (strain_geno_pdist.index == strain_gene_uhgg_pdist.index).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_gene_uhgg_cdmat, method=\"average\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# FIXME: This requires gene_uhgg_linkage, which takes too long to calculate.\n",
    "\n",
    "x = strain_gene_uhgg\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# FIXME: This requires gene_uhgg_linkage, which takes too long to calculate.\n",
    "\n",
    "x = strain_gene_uhgg\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg_depth = (\n",
    "    pd.read_table(spgc_gene_uhgg_depth_inpath, index_col=[\"gene_id\", \"strain\"])\n",
    "    .depth.unstack(\"strain\")\n",
    "    .rename(columns=str)\n",
    ")\n",
    "spgc_gene_uhgg_corr = (\n",
    "    pd.read_table(spgc_gene_uhgg_corr_inpath, index_col=[\"gene_id\", \"strain\"])\n",
    "    .correlation.unstack(\"strain\")\n",
    "    .rename(columns=str)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ref = ref_gene_uhgg[ref_list]\n",
    "_spgc = spgc_gene_uhgg_depth[spgc_list]\n",
    "strain_gene_uhgg_depth = (\n",
    "    pd.concat([_ref, _spgc], axis=1).fillna(0).loc[strain_gene_uhgg.index]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ref = ref_gene_uhgg[ref_list]\n",
    "_spgc = spgc_gene_uhgg_corr[spgc_list]\n",
    "strain_gene_uhgg_corr = (\n",
    "    pd.concat([_ref, _spgc], axis=1).fillna(0).loc[strain_gene_uhgg.index]\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# FIXME: This requires gene_uhgg_linkage, which takes too long to calculate.\n",
    "\n",
    "x = strain_gene_uhgg_depth\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x,\n",
    "        row_cluster=False,\n",
    "        col_linkage=_col_linkage,\n",
    "        col_colors=_col_colors,\n",
    "        norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "x = strain_gene_uhgg_corr\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "print(\n",
    "    \"ref-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "        squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "    squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-spgc\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-spgc\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "        squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "    _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-ref\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "        _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.legend(markerscale=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=200)\n",
    "\n",
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, ref_list].min(),\n",
    "    _pdistB.loc[ref_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, spgc_list].min(),\n",
    "    _pdistB.loc[spgc_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].min(),\n",
    "    _pdistB.loc[spgc_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-spgc\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, spgc_list].min(),\n",
    "    _pdistB.loc[ref_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.xlabel(\"minimum_genotype_diss\")\n",
    "plt.ylabel(\"minimum_gene_diss\")\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    color=\"k\",\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "\n",
    "for spgc_strain_id in spgc_list:\n",
    "    plt.scatter(\n",
    "        _pdistA.loc[spgc_strain_id, ref_list],\n",
    "        _pdistB.loc[spgc_strain_id, ref_list],\n",
    "        s=1,\n",
    "        alpha=0.5,\n",
    "        color=spgc_palette[spgc_strain_id],\n",
    "        # label='spgc-ref',\n",
    "    )\n",
    "    print(\n",
    "        spgc_strain_id,\n",
    "        sp.stats.pearsonr(\n",
    "            _pdistA.loc[spgc_strain_id, ref_list], _pdistB.loc[spgc_strain_id, ref_list]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Enriched/Depleted in SPGC Strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_uhgg_prevalence = strain_gene_uhgg[ref_list].mean(1)  # NOTE: Set above\n",
    "spgc_gene_uhgg_prevalence = strain_gene_uhgg[spgc_list].mean(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ref_gene_uhgg_prevalence\n",
    "y = spgc_gene_uhgg_prevalence\n",
    "\n",
    "print(sp.stats.pearsonr(x, y))\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(5, 10))\n",
    "\n",
    "bins0 = np.linspace(0.0, 1.0, num=50)\n",
    "axs[0].hist2d(x, y, bins=bins0, norm=mpl.colors.PowerNorm(1 / 3, vmin=0, vmax=1e3))\n",
    "\n",
    "bins1 = np.linspace(0.1, 0.9, num=40)\n",
    "axs[1].hist2d(x, y, bins=bins1, norm=mpl.colors.PowerNorm(1 / 3))\n",
    "axs[1].set_xlabel(\"reference prevalence\")\n",
    "axs[1].set_ylabel(\"inferred prevalence\")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-4, 4)\n",
    "plt.hist(\n",
    "    (spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence).replace(\n",
    "        {0: 1e-3, np.inf: 1e3}\n",
    "    ),\n",
    "    bins=bins,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.hist(\n",
    "    (ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence).replace(\n",
    "        {0: 1e-3, np.inf: 1e3}\n",
    "    ),\n",
    "    bins=bins,\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "spgc_extremely_enriched = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.9)\n",
    "spgc_extremely_depleted = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.9)\n",
    "spgc_very_enriched = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.5)\n",
    "spgc_very_depleted = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.5)\n",
    "spgc_highly_enriched = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.35)\n",
    "spgc_highly_depleted = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.35)\n",
    "spgc_enriched = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.25)\n",
    "spgc_depleted = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.25)\n",
    "spgc_similar = idxwhere(((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > -0.25) & ((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < 0.25))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_extremely_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence) > 1000\n",
    ")\n",
    "spgc_extremely_depleted = idxwhere(\n",
    "    (ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence) > 1000\n",
    ")\n",
    "spgc_very_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence) > 100\n",
    ")\n",
    "spgc_very_depleted = idxwhere(\n",
    "    (ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence) > 100\n",
    ")\n",
    "spgc_highly_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence) > 10\n",
    ")\n",
    "spgc_highly_depleted = idxwhere(\n",
    "    (ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence) > 10\n",
    ")\n",
    "spgc_enriched = idxwhere((spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence) > 5)\n",
    "spgc_depleted = idxwhere((ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence) > 5)\n",
    "spgc_similar = idxwhere(\n",
    "    ((spgc_gene_uhgg_prevalence / ref_gene_uhgg_prevalence) < 5)\n",
    "    & ((ref_gene_uhgg_prevalence / spgc_gene_uhgg_prevalence) < 5)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Length / Singletons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for spgc_enrichment_class_list in [\n",
    "    spgc_extremely_enriched,\n",
    "    spgc_very_enriched,\n",
    "    spgc_highly_enriched,\n",
    "    spgc_enriched,\n",
    "    spgc_similar,\n",
    "    spgc_depleted,\n",
    "    spgc_highly_depleted,\n",
    "    spgc_very_depleted,\n",
    "    spgc_extremely_depleted,\n",
    "]:\n",
    "    print(\n",
    "        len(spgc_enrichment_class_list),\n",
    "        uhgg_gene_length.loc[spgc_enrichment_class_list].mean(),\n",
    "        uhgg_gene_length.loc[spgc_enrichment_class_list].std(),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "short_genes = idxwhere(uhgg_gene_length < 300)\n",
    "singleton_genes = idxwhere(strain_gene_uhgg.sum(1) <= 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Final Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_list = spgc_highly_enriched + spgc_highly_depleted + short_genes + singleton_genes\n",
    "print(\n",
    "    len(spgc_highly_enriched),\n",
    "    len(spgc_highly_depleted),\n",
    "    len(short_genes),\n",
    "    len(singleton_genes),\n",
    ")\n",
    "\n",
    "strain_uhgg_filt = strain_gene_uhgg.drop(index=drop_list, errors=\"ignore\")\n",
    "print(strain_gene_uhgg.shape[0], strain_uhgg_filt.shape[0])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "strain_gene_eggnog = (\n",
    "    uhgg_x_eggnog\n",
    "    .join(strain_gene_uhgg, on='gene_id')\n",
    "    .drop(columns=['gene_id'])\n",
    "    .groupby('eggnog')\n",
    "    .sum()\n",
    "    .gt(0)\n",
    "    [lambda x: x.sum(1) > 0]\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "strain_gene_filt_eggnog = (\n",
    "    uhgg_x_eggnog\n",
    "    .join(strain_uhgg_filt, on='gene_id')\n",
    "    .drop(columns=['gene_id'])\n",
    "    .groupby('eggnog')\n",
    "    .sum()\n",
    "    .gt(0)\n",
    "    [lambda x: x.sum(1) > 0]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, strain_gene_uhgg[spgc_list + ref_list].sum().max(), num=50)\n",
    "\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    label=\"ref (uhgg)\",\n",
    "    density=True,\n",
    "    color=\"tab:blue\",\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    label=\"spgc (uhgg)\",\n",
    "    density=True,\n",
    "    color=\"tab:orange\",\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "plt.hist(\n",
    "    strain_uhgg_filt[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"ref (filtered uhgg)\",\n",
    "    density=True,\n",
    "    linestyle=\"-\",\n",
    "    color=\"tab:blue\",\n",
    "    lw=2,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_uhgg_filt[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"spgc (filtered uhgg)\",\n",
    "    density=True,\n",
    "    linestyle=\"-\",\n",
    "    color=\"tab:orange\",\n",
    "    lw=2,\n",
    ")\n",
    "\n",
    "# plt.hist(strain_gene_eggnog[ref_list].sum(), bins=bins, histtype='stepfilled', label='ref (eggnog)', density=True, color='tab:blue', alpha=0.3)\n",
    "# plt.hist(strain_gene_eggnog[spgc_list].sum(), bins=bins, histtype='stepfilled', label='spgc (eggnog)', density=True, color='tab:orange', alpha=0.3)\n",
    "\n",
    "# plt.hist(strain_gene_filt_eggnog[ref_list].sum(), bins=bins, histtype='step', label='ref (filtered eggnog)', density=True, linestyle='--', color='tab:blue', lw=1.5)\n",
    "# plt.hist(strain_gene_filt_eggnog[spgc_list].sum(), bins=bins, histtype='step', label='spgc (filtered eggnog)', density=True, linestyle='--', color='tab:orange', lw=1.5)\n",
    "\n",
    "plt.legend()\n",
    "\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_filt_cdmat = sp.spatial.distance.pdist(strain_uhgg_filt, metric=\"cosine\")\n",
    "uhgg_filt_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(uhgg_filt_cdmat),\n",
    "    index=strain_uhgg_filt.index,\n",
    "    columns=strain_uhgg_filt.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_filt_linkage = sp.cluster.hierarchy.linkage(uhgg_filt_cdmat, method=\"average\")\n",
    "uhgg_filt_linkage.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_uhgg_filt.T,\n",
    "    metric=\"cosine\",\n",
    "    w=ref_gene_uhgg_entropy.loc[strain_uhgg_filt.index],\n",
    ")\n",
    "strain_uhgg_filt_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_uhgg_filt_cdmat),\n",
    "    index=strain_uhgg_filt.columns,\n",
    "    columns=strain_uhgg_filt.columns,\n",
    ")\n",
    "strain_uhgg_filt_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_uhgg_filt_cdmat, method=\"average\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt_unweighted_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_uhgg_filt.T, metric=\"cosine\"\n",
    ")\n",
    "strain_uhgg_filt_unweighted_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_uhgg_filt_unweighted_cdmat),\n",
    "    index=strain_uhgg_filt.columns,\n",
    "    columns=strain_uhgg_filt.columns,\n",
    ")\n",
    "strain_uhgg_filt_unweighted_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_uhgg_filt_unweighted_cdmat, method=\"average\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt_unweighted_jacc_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_uhgg_filt.T, metric=\"jaccard\"\n",
    ")\n",
    "strain_uhgg_filt_unweighted_jacc_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_uhgg_filt_unweighted_jacc_cdmat),\n",
    "    index=strain_uhgg_filt.columns,\n",
    "    columns=strain_uhgg_filt.columns,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_unweighted_jacc_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_gene_uhgg.T, metric=\"jaccard\"\n",
    ")\n",
    "strain_uhgg_unweighted_jacc_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_uhgg_unweighted_jacc_cdmat),\n",
    "    index=strain_uhgg_filt.columns,\n",
    "    columns=strain_uhgg_filt.columns,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = uhgg_filt_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt\n",
    "_col_linkage = strain_uhgg_filt_linkage\n",
    "_row_linkage = uhgg_filt_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt\n",
    "_col_linkage = strain_uhgg_filt_unweighted_linkage\n",
    "_row_linkage = uhgg_filt_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Content Divergence from Refs\n",
    "\n",
    "Are SPGC Genomes Believably Distinct?\n",
    "\n",
    "Test if SPGC Strains are surprisingly distinct from refs in gene content relative to their distinction in genotype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Min-genotype / Min-gene content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "print(\n",
    "    \"ref-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "        squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "    squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-spgc\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-spgc\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "        squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "    _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-ref\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "        _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.legend(markerscale=5)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, ref_list].min(),\n",
    "    _pdistB.loc[ref_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label='ref-to-ref',\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, spgc_list].min(),\n",
    "    _pdistB.loc[ref_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label='spgc-to-ref',\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, spgc_list].min(),\n",
    "    _pdistB.loc[spgc_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label='spgc-to-spgc',\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].min(),\n",
    "    _pdistB.loc[spgc_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label='ref-to-spgc',\n",
    ")\n",
    "plt.xlabel('minimum_genotype_diss')\n",
    "plt.ylabel('minimum_gene_diss')\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import statsmodels.formula.api as smf\n",
    "\n",
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "geno_pdist_adjust = (1 / sfacts_fit.sizes['position'])\n",
    "# gene_pdist_adjust = (1 / strain_gene_filt_eggnog.shape[0])\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "ref_to_ref_min_pdistA = _pdistA.loc[ref_list, ref_list].min()\n",
    "ref_to_ref_min_pdistB = _pdistB.loc[ref_list, ref_list].min()\n",
    "spgc_to_ref_min_pdistA = _pdistA.loc[ref_list, spgc_list].min()\n",
    "spgc_to_ref_min_pdistB = _pdistB.loc[ref_list, spgc_list].min()\n",
    "\n",
    "d0 = pd.DataFrame(dict(\n",
    "    geno_dist=pd.concat([ref_to_ref_min_pdistA, spgc_to_ref_min_pdistA]),\n",
    "    gene_dist=pd.concat([ref_to_ref_min_pdistB, spgc_to_ref_min_pdistB]),\n",
    "    spgc=np.concatenate([np.zeros_like(ref_to_ref_min_pdistA), np.ones_like(spgc_to_ref_min_pdistA)]).astype(bool),\n",
    "))#.assign(\n",
    "#     geno_dist_pc=lambda x: x.geno_dist + geno_pdist_adjust,\n",
    "#     gene_dist_pc=lambda x: x.gene_dist + gene_pdist_adjust,\n",
    "# )\n",
    "fit = smf.ols(f'gene_dist ~ np.log2(geno_dist + {geno_pdist_adjust}) * spgc', data=d0).fit()\n",
    "d1 = d0.assign(gene_dist_predict=lambda x: fit.predict(), gene_dist_resid_pearson=fit.resid_pearson).sort_values('geno_dist')\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "ax = axs[0]\n",
    "for _spgc, d2 in d1.groupby('spgc'):\n",
    "    ax.scatter('geno_dist', 'gene_dist', data=d2, label=_spgc)\n",
    "    ax.plot('geno_dist', 'gene_dist_predict', data=d2, label='__nolegend__')\n",
    "\n",
    "ax = axs[1]\n",
    "for _spgc, d2 in d1.groupby('spgc'):\n",
    "    ax.scatter('geno_dist', 'gene_dist', data=d2, label=_spgc)\n",
    "    ax.plot('geno_dist', 'gene_dist_predict', data=d2, label='__nolegend__')\n",
    "ax.legend(title='SPGC')\n",
    "ax.set_xscale('symlog', linthresh=1e-4)\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "import statsmodels.formula.api as smf\n",
    "\n",
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "geno_pdist_adjust = (1 / sfacts_fit.sizes['position'])\n",
    "# gene_pdist_adjust = (1 / strain_gene_filt_eggnog.shape[0])\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "isolate_to_ref_min_pdistA = _pdistA.loc[ref_list, isolate_list].min()\n",
    "isolate_to_ref_min_pdistB = _pdistB.loc[ref_list, isolate_list].min()\n",
    "mag_to_ref_min_pdistA = _pdistA.loc[ref_list, mag_list].min()\n",
    "mag_to_ref_min_pdistB = _pdistB.loc[ref_list, mag_list].min()\n",
    "spgc_to_ref_min_pdistA = _pdistA.loc[ref_list, spgc_list].min()\n",
    "spgc_to_ref_min_pdistB = _pdistB.loc[ref_list, spgc_list].min()\n",
    "\n",
    "d0 = pd.DataFrame(dict(\n",
    "    geno_dist=pd.concat([isolate_to_ref_min_pdistA, mag_to_ref_min_pdistA, spgc_to_ref_min_pdistA]) + geno_pdist_adjust,\n",
    "    gene_dist=pd.concat([isolate_to_ref_min_pdistB, mag_to_ref_min_pdistB, spgc_to_ref_min_pdistB]),\n",
    "    genome_type=['Isolate'] * len(isolate_to_ref_min_pdistA) + ['MAG'] * len(mag_to_ref_min_pdistA) + ['SPGC'] * len(spgc_to_ref_min_pdistA),\n",
    "))#.assign(\n",
    "#     geno_dist_pc=lambda x: x.geno_dist + geno_pdist_adjust,\n",
    "#     gene_dist_pc=lambda x: x.gene_dist + gene_pdist_adjust,\n",
    "# )\n",
    "fit = smf.ols(f'gene_dist ~ np.log2(geno_dist) * genome_type', data=d0).fit()\n",
    "d1 = d0.assign(gene_dist_predict=lambda x: fit.predict(), gene_dist_resid_pearson=fit.resid_pearson).sort_values('geno_dist')\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "ax = axs[0]\n",
    "for _genome_type, d2 in d1.groupby('genome_type'):\n",
    "    ax.scatter('geno_dist', 'gene_dist', data=d2, label=_genome_type)\n",
    "    ax.plot('geno_dist', 'gene_dist_predict', data=d2, label='__nolegend__')\n",
    "\n",
    "ax = axs[1]\n",
    "for _genome_type, d2 in d1.groupby('genome_type'):\n",
    "    ax.scatter('geno_dist', 'gene_dist', data=d2, label=_genome_type)\n",
    "    ax.plot('geno_dist', 'gene_dist_predict', data=d2, label='__nolegend__')\n",
    "ax.legend()\n",
    "ax.set_xscale('symlog', linthresh=1e-4)\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "genome_type_to_label = {'isolate': 'Isolate', 'mag': 'MAG', 'spgc': 'SPGC'}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "for _genome_type, d2 in d1.groupby('genome_type'):\n",
    "    ax.scatter('geno_dist', 'gene_dist', data=d2, color=genome_type_palette[_genome_type], label=_genome_type, s=50, alpha=0.7, lw=0)\n",
    "    ax.plot('geno_dist', 'gene_dist_predict', data=d2, color=genome_type_palette[_genome_type], label='__nolegend__')\n",
    "# ax.legend(markerscale=3)\n",
    "ax.set_xscale('symlog', linthresh=1e-4)\n",
    "ax.set_xlabel('SNP Dissimilarity')\n",
    "ax.set_ylabel('Gene Content Dissimilarity')\n",
    "ax.set_yticks([0, 0.1, 0.2, 0.3, 0.4])"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "for _genome_type, d2 in d1.groupby('genome_type'):\n",
    "    plt.scatter('gene_dist_predict', 'gene_dist_resid_pearson', data=d2, label=_genome_type)\n",
    "plt.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    color=\"k\",\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "\n",
    "for spgc_strain_id in spgc_list:\n",
    "    plt.scatter(\n",
    "        _pdistA.loc[spgc_strain_id, ref_list],\n",
    "        _pdistB.loc[spgc_strain_id, ref_list],\n",
    "        s=1,\n",
    "        alpha=0.5,\n",
    "        color=spgc_palette[spgc_strain_id],\n",
    "        # label='spgc-ref',\n",
    "    )\n",
    "    print(\n",
    "        spgc_strain_id,\n",
    "        sp.stats.pearsonr(\n",
    "            _pdistA.loc[spgc_strain_id, ref_list], _pdistB.loc[spgc_strain_id, ref_list]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    color=\"k\",\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "\n",
    "for spgc_strain_id in spgc_list:\n",
    "    plt.scatter(\n",
    "        _pdistA.loc[spgc_strain_id, ref_list],\n",
    "        _pdistB.loc[spgc_strain_id, ref_list],\n",
    "        s=1,\n",
    "        alpha=0.5,\n",
    "        color=spgc_palette[spgc_strain_id],\n",
    "        # label='spgc-ref',\n",
    "    )\n",
    "    print(\n",
    "        spgc_strain_id,\n",
    "        sp.stats.pearsonr(\n",
    "            _pdistA.loc[spgc_strain_id, ref_list], _pdistB.loc[spgc_strain_id, ref_list]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Min-genotype / Matched-gene content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "geno_pdist_adjust = 1 / sfacts_fit.sizes[\"position\"]\n",
    "# gene_pdist_adjust = (1 / strain_gene_filt_eggnog.shape[0])\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "isolate_to_ref_min = _pdistA.loc[ref_list, isolate_list].idxmin()\n",
    "isolate_to_ref_min_pdistA = _pdistA.stack()[\n",
    "    isolate_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "isolate_to_ref_min_pdistB = _pdistB.stack()[\n",
    "    isolate_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "\n",
    "mag_to_ref_min = _pdistA.loc[ref_list, mag_list].idxmin()\n",
    "mag_to_ref_min_pdistA = _pdistA.stack()[\n",
    "    mag_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "mag_to_ref_min_pdistB = _pdistB.stack()[\n",
    "    mag_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "\n",
    "spgc_to_ref_min = _pdistA.loc[ref_list, spgc_list].idxmin()\n",
    "spgc_to_ref_min_pdistA = _pdistA.stack()[\n",
    "    spgc_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "spgc_to_ref_min_pdistB = _pdistB.stack()[\n",
    "    spgc_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        min_geno_ref=pd.concat(\n",
    "            [\n",
    "                isolate_to_ref_min,\n",
    "                mag_to_ref_min,\n",
    "                spgc_to_ref_min,\n",
    "            ]\n",
    "        ),\n",
    "        geno_dist=(\n",
    "            pd.concat(\n",
    "                [\n",
    "                    isolate_to_ref_min_pdistA,\n",
    "                    mag_to_ref_min_pdistA,\n",
    "                    spgc_to_ref_min_pdistA,\n",
    "                ]\n",
    "            )\n",
    "            + geno_pdist_adjust\n",
    "        ),\n",
    "        gene_dist=pd.concat(\n",
    "            [isolate_to_ref_min_pdistB, mag_to_ref_min_pdistB, spgc_to_ref_min_pdistB]\n",
    "        ),\n",
    "        genome_type=[\"Isolate\"] * len(isolate_to_ref_min_pdistA)\n",
    "        + [\"MAG\"] * len(mag_to_ref_min_pdistA)\n",
    "        + [\"SPGC\"] * len(spgc_to_ref_min_pdistA),\n",
    "    )\n",
    ")  # .assign(\n",
    "#     geno_dist_pc=lambda x: x.geno_dist + geno_pdist_adjust,\n",
    "#     gene_dist_pc=lambda x: x.gene_dist + gene_pdist_adjust,\n",
    "# )\n",
    "fit = smf.ols(f\"gene_dist ~ np.log2(geno_dist) * genome_type\", data=d0).fit()\n",
    "genome_distance_comparison = d0.assign(\n",
    "    gene_dist_predict=lambda x: fit.predict(), gene_dist_resid_pearson=fit.resid_pearson\n",
    ").sort_values(\"geno_dist\")\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "ax = axs[0]\n",
    "for _genome_type, d2 in genome_distance_comparison.groupby(\"genome_type\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_genome_type)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "\n",
    "ax = axs[1]\n",
    "for _genome_type, d2 in genome_distance_comparison.groupby(\"genome_type\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_genome_type)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "ax.legend()\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type_to_label = {\"isolate\": \"Isolate\", \"mag\": \"MAG\", \"spgc\": \"SPGC\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "for _genome_type, d2 in genome_distance_comparison.groupby(\"genome_type\"):\n",
    "    ax.scatter(\n",
    "        \"geno_dist\",\n",
    "        \"gene_dist\",\n",
    "        data=d2,\n",
    "        color=genome_type_palette[_genome_type],\n",
    "        label=_genome_type,\n",
    "        s=50,\n",
    "        alpha=0.7,\n",
    "        lw=0,\n",
    "    )\n",
    "    ax.plot(\n",
    "        \"geno_dist\",\n",
    "        \"gene_dist_predict\",\n",
    "        data=d2,\n",
    "        color=genome_type_palette[_genome_type],\n",
    "        label=\"__nolegend__\",\n",
    "    )\n",
    "# ax.legend(markerscale=3)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\")\n",
    "# ax.set_yticks([0, 0.1, 0.2, 0.3, 0.4])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistB = strain_uhgg_filt_unweighted_jacc_pdist\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "isolate_to_ref_min_pdistB_raw = _pdistB.stack()[\n",
    "    isolate_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "mag_to_ref_min_pdistB_raw = _pdistB.stack()[\n",
    "    mag_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "spgc_to_ref_min_pdistB_raw = _pdistB.stack()[\n",
    "    spgc_to_ref_min.reset_index().set_index([\"index\", 0]).index\n",
    "].reset_index(level=1, drop=True)\n",
    "\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        min_geno_ref=pd.concat(\n",
    "            [\n",
    "                isolate_to_ref_min,\n",
    "                mag_to_ref_min,\n",
    "                spgc_to_ref_min,\n",
    "            ]\n",
    "        ),\n",
    "        geno_dist=(\n",
    "            pd.concat(\n",
    "                [\n",
    "                    isolate_to_ref_min_pdistA,\n",
    "                    mag_to_ref_min_pdistA,\n",
    "                    spgc_to_ref_min_pdistA,\n",
    "                ]\n",
    "            )\n",
    "            + geno_pdist_adjust\n",
    "        ),\n",
    "        gene_dist=pd.concat(\n",
    "            [\n",
    "                isolate_to_ref_min_pdistB_raw,\n",
    "                mag_to_ref_min_pdistB_raw,\n",
    "                spgc_to_ref_min_pdistB_raw,\n",
    "            ]\n",
    "        ),\n",
    "        genome_type=[\"Isolate\"] * len(isolate_to_ref_min_pdistA)\n",
    "        + [\"MAG\"] * len(mag_to_ref_min_pdistA)\n",
    "        + [\"SPGC\"] * len(spgc_to_ref_min_pdistA),\n",
    "    )\n",
    ")\n",
    "fit = smf.ols(f\"gene_dist ~ np.log2(geno_dist) * genome_type\", data=d0).fit()\n",
    "genome_distance_comparison_raw = d0.assign(\n",
    "    gene_dist_predict=lambda x: fit.predict(), gene_dist_resid_pearson=fit.resid_pearson\n",
    ").sort_values(\"geno_dist\")\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "ax = axs[0]\n",
    "for _genome_type, d2 in genome_distance_comparison_raw.groupby(\"genome_type\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_genome_type)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "\n",
    "ax = axs[1]\n",
    "for _genome_type, d2 in genome_distance_comparison_raw.groupby(\"genome_type\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_genome_type)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "ax.legend()\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type_to_label = {\"isolate\": \"Isolate\", \"mag\": \"MAG\", \"spgc\": \"SPGC\"}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 5))\n",
    "for _genome_type, d2 in genome_distance_comparison_raw.groupby(\"genome_type\"):\n",
    "    ax.scatter(\n",
    "        \"geno_dist\",\n",
    "        \"gene_dist\",\n",
    "        data=d2,\n",
    "        color=genome_type_palette[_genome_type],\n",
    "        label=_genome_type,\n",
    "        s=50,\n",
    "        alpha=0.7,\n",
    "        lw=0,\n",
    "    )\n",
    "    ax.plot(\n",
    "        \"geno_dist\",\n",
    "        \"gene_dist_predict\",\n",
    "        data=d2,\n",
    "        color=genome_type_palette[_genome_type],\n",
    "        label=\"__nolegend__\",\n",
    "    )\n",
    "# ax.legend(markerscale=3)\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "ax.set_xlabel(\"SNP Profile Dissimilarity\")\n",
    "ax.set_ylabel(\"Gene Content Dissimilarity\")\n",
    "# ax.set_yticks([0, 0.1, 0.2, 0.3, 0.4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Diversity Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_geno_pdist.copy()\n",
    "dii, djj = np.diag_indices_from(x)\n",
    "x.values[dii, djj] = np.nan\n",
    "\n",
    "_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "# g.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_geno_pdist.copy()\n",
    "dii, djj = np.diag_indices_from(x)\n",
    "x.values[dii, djj] = np.nan\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        g=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    dendrogram_ratio=0.1,\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    tree_kws=dict(lw=1.0),\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "g.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt_pdist.copy()\n",
    "dii, djj = np.diag_indices_from(x)\n",
    "x.values[dii, djj] = np.nan\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        g=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    dendrogram_ratio=0.1,\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    tree_kws=dict(lw=1.0),\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "g.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "x = strain_uhgg_filt_pdist.replace({0: np.nan})\n",
    "\n",
    "_colors = pd.DataFrame(dict(\n",
    "    g=genome_type.map(genome_type_palette),\n",
    "))\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    dendrogram_ratio=0.1,\n",
    "    vmin=0, vmax=1,\n",
    "    tree_kws=dict(lw=1.5),\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor('aqua')\n",
    "g.cax.set_visible(False)\n",
    "g.ax_col_dendrogram.set_ylim(0.1, 1)\n",
    "g.ax_row_dendrogram.set_xlim(1, 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prevalence Comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_uhgg_prevalence = strain_gene_uhgg[ref_list].mean(1)\n",
    "spgc_uhgg_prevalence = strain_gene_uhgg[spgc_list].mean(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ref_uhgg_prevalence\n",
    "y = spgc_uhgg_prevalence\n",
    "\n",
    "print(sp.stats.pearsonr(x, y))\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(5, 10))\n",
    "\n",
    "bins0 = np.linspace(0.0, 1.0, num=50)\n",
    "axs[0].hist2d(x, y, bins=bins0, norm=mpl.colors.PowerNorm(1 / 3, vmin=0, vmax=1e3))\n",
    "\n",
    "bins1 = np.linspace(0.1, 0.9, num=40)\n",
    "axs[1].hist2d(x, y, bins=bins1, norm=mpl.colors.PowerNorm(1 / 3))\n",
    "axs[1].set_xlabel(\"reference prevalence\")\n",
    "axs[1].set_ylabel(\"inferred prevalence\")\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Core / Shell / Cloud Pangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=51)\n",
    "\n",
    "\n",
    "def _assign_prevalence_class(p):\n",
    "    if p > 0.95:\n",
    "        return \"core\"\n",
    "    elif p > 0.1:\n",
    "        return \"shell\"\n",
    "    elif p < 0.1:\n",
    "        return \"cloud\"\n",
    "\n",
    "\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]\n",
    "prevalence_class_palette = {\n",
    "    \"core\": \"tab:blue\",\n",
    "    \"shell\": \"tab:orange\",\n",
    "    \"cloud\": \"tab:green\",\n",
    "}\n",
    "\n",
    "spgc_uhgg_class = spgc_uhgg_prevalence.map(_assign_prevalence_class)\n",
    "\n",
    "for prevalence_class in prevalence_class_order:\n",
    "    plt.hist(\n",
    "        spgc_uhgg_prevalence[spgc_uhgg_class == prevalence_class],\n",
    "        bins=bins,\n",
    "        label=prevalence_class,\n",
    "        color=prevalence_class_palette[prevalence_class],\n",
    "    )\n",
    "plt.legend()\n",
    "plt.yscale(\"log\")\n",
    "# core_genes = idxwhere(strain_uhgg_filt_prevalence > 0.9)\n",
    "# shell_genes = idxwhere((strain_uhgg_filt_prevalence < 0.9) & (strain_uhgg_filt_prevalence > 0.1))\n",
    "# cloud_genes = idxwhere(strain_uhgg_filt_prevalence < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=51)\n",
    "\n",
    "\n",
    "def _assign_prevalence_class(p):\n",
    "    if p > 0.95:\n",
    "        return \"core\"\n",
    "    elif p > 0.1:\n",
    "        return \"shell\"\n",
    "    elif p < 0.1:\n",
    "        return \"cloud\"\n",
    "\n",
    "\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]\n",
    "prevalence_class_palette = {\n",
    "    \"core\": \"tab:blue\",\n",
    "    \"shell\": \"tab:orange\",\n",
    "    \"cloud\": \"tab:green\",\n",
    "}\n",
    "\n",
    "ref_uhgg_class = ref_uhgg_prevalence.map(_assign_prevalence_class)\n",
    "\n",
    "for prevalence_class in prevalence_class_order:\n",
    "    plt.hist(\n",
    "        ref_uhgg_prevalence[ref_uhgg_class == prevalence_class],\n",
    "        bins=bins,\n",
    "        label=prevalence_class,\n",
    "        color=prevalence_class_palette[prevalence_class],\n",
    "    )\n",
    "plt.legend()\n",
    "plt.yscale(\"log\")\n",
    "# core_genes = idxwhere(strain_uhgg_filt_prevalence > 0.9)\n",
    "# shell_genes = idxwhere((strain_uhgg_filt_prevalence < 0.9) & (strain_uhgg_filt_prevalence > 0.1))\n",
    "# cloud_genes = idxwhere(strain_uhgg_filt_prevalence < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "strain_genome_spgc_prevalence_class_tally = (\n",
    "    strain_gene_uhgg.groupby(spgc_uhgg_class)\n",
    "    .sum()\n",
    "    .reindex(prevalence_class_order, fill_value=0)\n",
    "    .T.astype(int)\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(10, 5), tight_layout=True)\n",
    "\n",
    "gs = gridspec.GridSpec(\n",
    "    1, 2, width_ratios=[len(ref_list) / len(spgc_list), 1], figure=fig\n",
    ")\n",
    "ax0 = fig.add_subplot(gs[0, 0])\n",
    "ax1 = fig.add_subplot(gs[0, 1], sharey=ax0)\n",
    "\n",
    "# , axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "for ax, strain_list, title in zip(\n",
    "    [ax0, ax1], [ref_list, spgc_list], [\"references\", \"spgc\"]\n",
    "):\n",
    "    # TODO: Fix the colors to always match the previous palette.\n",
    "    strain_genome_spgc_prevalence_class_tally.loc[strain_list].sort_values(\n",
    "        \"shell\"\n",
    "    ).plot.bar(stacked=True, ax=ax, width=1)\n",
    "    ax.set_title(title)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_aspect(0.08)\n",
    "    ax.legend_.set_visible(False)\n",
    "ax0.legend_.set_visible(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "strain_genome_ref_prevalence_class_tally = (\n",
    "    strain_gene_uhgg.groupby(ref_uhgg_class)\n",
    "    .sum()\n",
    "    .reindex(prevalence_class_order, fill_value=0)\n",
    "    .T.astype(int)\n",
    ")\n",
    "\n",
    "fig = plt.figure(figsize=(10, 5), tight_layout=True)\n",
    "\n",
    "gs = gridspec.GridSpec(\n",
    "    1, 2, width_ratios=[len(ref_list) / len(spgc_list), 1], figure=fig\n",
    ")\n",
    "ax0 = fig.add_subplot(gs[0, 0])\n",
    "ax1 = fig.add_subplot(gs[0, 1], sharey=ax0)\n",
    "\n",
    "# , axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "for ax, strain_list, title in zip(\n",
    "    [ax0, ax1], [ref_list, spgc_list], [\"references\", \"spgc\"]\n",
    "):\n",
    "    # TODO: Fix the colors to always match the previous palette.\n",
    "    strain_genome_ref_prevalence_class_tally.loc[strain_list].sort_values(\n",
    "        \"shell\"\n",
    "    ).plot.bar(stacked=True, ax=ax, width=1)\n",
    "    ax.set_title(title)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_aspect(0.08)\n",
    "    ax.legend_.set_visible(False)\n",
    "ax0.legend_.set_visible(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotations.groupby(spgc_uhgg_class).centroid_99_length.quantile(\n",
    "    [0.1, 0.25, 0.5, 0.75, 0.9]\n",
    ").unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_order = gene_x_cog_category_matrix.columns\n",
    "cog_category_palette = lib.plot.construct_ordered_palette(\n",
    "    cog_category_order, cm=\"rainbow\", extend=dict(no_category=\"grey\")\n",
    ")\n",
    "for cog_category in cog_category_order:\n",
    "    plt.scatter([], [], c=cog_category_palette[cog_category], label=cog_category)\n",
    "plt.legend(ncols=4)\n",
    "\n",
    "lib.plot.hide_axes_and_spines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"cog_category\", \"description\"],\n",
    "    index_col=\"cog_category\",\n",
    ").assign(description=lambda x: x.index + \": \" + x.description)\n",
    "cog_category_description.loc[\"no_category\", \"description\"] = \"-: No Annotation\"\n",
    "cog_category_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = spgc_uhgg_class\n",
    "y = gene_x_cog_category_matrix.reindex(\n",
    "    x.index, fill_value=False\n",
    ")  # .assign(no_category=lambda x: x.sum(1) == 0)\n",
    "gene_list = spgc_uhgg_class.index\n",
    "\n",
    "cog_category_gene_class_enrichment_test = []\n",
    "\n",
    "for _prevalence_class, _cog_category in product(\n",
    "    prevalence_class_order, list(cog_category_order)\n",
    "):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_prev_class=(x == _prevalence_class),\n",
    "                is_cog_category=y[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_gene_class_enrichment_test.append(\n",
    "        (\n",
    "            _prevalence_class,\n",
    "            _cog_category,\n",
    "            _test[0],\n",
    "            _test[1],\n",
    "            contingency_table.loc[True, True],\n",
    "        )\n",
    "    )\n",
    "\n",
    "cog_category_gene_class_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_gene_class_enrichment_test,\n",
    "        columns=[\n",
    "            \"prevalence_class\",\n",
    "            \"cog_category\",\n",
    "            \"statistic\",\n",
    "            \"pvalue\",\n",
    "            \"gene_count\",\n",
    "        ],\n",
    "    )\n",
    "    .set_index([\"prevalence_class\", \"cog_category\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio=lambda x: np.log2(x.statistic),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _assign_significance_marker(pvalue):\n",
    "    if pvalue < 1e-5:\n",
    "        return \"***\"\n",
    "    elif pvalue < 1e-3:\n",
    "        return \"**\"\n",
    "    elif pvalue < 0.05:\n",
    "        return \"*\"\n",
    "    else:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    cog_category_gene_class_enrichment_test.log2_odds_ratio.unstack(\"prevalence_class\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"prevalence_class\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = x[\"core\"].sort_values(ascending=False).index\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-5,\n",
    "    vmax=5,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "for prevalence_class in prevalence_class_order:\n",
    "    d = cog_category_gene_class_enrichment_test.xs(\n",
    "        prevalence_class, level=\"prevalence_class\"\n",
    "    )\n",
    "    ax.scatter(\n",
    "        \"log2_odds_ratio\",\n",
    "        \"negative_log10_pvalue\",\n",
    "        data=d,\n",
    "        c=prevalence_class_palette[prevalence_class],\n",
    "        label=prevalence_class,\n",
    "        s=5,\n",
    "    )\n",
    "\n",
    "# Annotate top-10\n",
    "for idx in (\n",
    "    cog_category_gene_class_enrichment_test.sort_values(\n",
    "        \"negative_log10_pvalue\", ascending=False\n",
    "    )\n",
    "    .head(10)\n",
    "    .index\n",
    "):\n",
    "    ax.annotate(\n",
    "        idx[1],\n",
    "        xy=cog_category_gene_class_enrichment_test.loc[idx][\n",
    "            [\"log2_odds_ratio\", \"negative_log10_pvalue\"]\n",
    "        ],\n",
    "        xytext=(2, 1),\n",
    "        textcoords=\"offset pixels\",\n",
    "        color=prevalence_class_palette[idx[0]],\n",
    "        # ha='center',\n",
    "        # va='center',\n",
    "        fontweight=\"bold\",\n",
    "    )\n",
    "ax.axvline(0, linestyle=\"--\", lw=1, color=\"k\")\n",
    "ax.axhline(\n",
    "    -np.log10(0.05 / cog_category_gene_class_enrichment_test.shape[0]),\n",
    "    linestyle=\"--\",\n",
    "    lw=1,\n",
    "    color=\"k\",\n",
    ")\n",
    "ax.legend()\n",
    "# cog_category_gene_class_enrichment_test.sort_values('negative_log10_pvalue', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strength of Phylogenetic Signal in Shell Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shell_gene_list = idxwhere(spgc_uhgg_class == \"shell\")\n",
    "\n",
    "x = strain_gene_uhgg.loc[shell_gene_list]\n",
    "\n",
    "strain_shell_gene_cdist = sp.spatial.distance.pdist(\n",
    "    x.T, metric=\"cosine\", w=ref_gene_uhgg_entropy.loc[x.index]\n",
    ")\n",
    "strain_shell_gene_pdist = pd.DataFrame(\n",
    "    squareform(strain_shell_gene_cdist), index=x.columns, columns=x.columns\n",
    ")\n",
    "strain_shell_gene_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_shell_gene_cdist, method=\"average\", optimal_ordering=True\n",
    ")\n",
    "\n",
    "# _col_linkage = strain_shell_gene_linkage\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "_col_linkage = strain_shell_gene_linkage\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(x, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import pdist, squareform\n",
    "\n",
    "# FIXME: x is ambiguous here.\n",
    "x = strain_shell_gene_pdist.replace({0: np.nan})\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = strain_shell_gene_linkage\n",
    "_colors = genome_type.map(genome_type_palette)\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    col_linkage=_col_linkage,\n",
    "    row_linkage=_row_linkage,\n",
    "    col_colors=_colors,\n",
    "    row_colors=_colors,\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geno_pdist_to_weights(pdist, rate=2):\n",
    "    pdist = np.array(pdist)\n",
    "    w = np.exp(-rate * pdist)\n",
    "    w[np.diag_indices_from(w)] = 0\n",
    "    return w"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_genome_list = strain_geno_pdist.index\n",
    "_results = {}\n",
    "w = geno_pdist_to_weights(strain_geno_pdist.loc[_genome_list, _genome_list], rate=1)\n",
    "\n",
    "d = pd.DataFrame(w, index=_genome_list, columns=_genome_list)\n",
    "sns.clustermap(d, col_linkage=strain_geno_linkage, row_linkage=strain_geno_linkage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def morans_i(x, w):\n",
    "    \"\"\"TODO\n",
    "\n",
    "    w should be a weight matrix, e.g.\n",
    "    \n",
    "        w = np.exp(-(strain_geno_pdist.loc[_genome_list, _genome_list].values))\n",
    "        w[np.diag_indices_from(w)] = 0\n",
    "\n",
    "    \"\"\"\n",
    "    # Moran's I.\n",
    "    x_centered = x - x.mean()\n",
    "    denomenator = (x_centered**2).sum()\n",
    "    if denomenator == 0:\n",
    "        return np.nan\n",
    "\n",
    "    n = len(x)\n",
    "    w_sum = w.sum()\n",
    "    \n",
    "    numerator = (np.outer(x_centered, x_centered) * w).sum()\n",
    "    normalizer = n / w_sum\n",
    "    observed = numerator * normalizer / denomenator\n",
    "\n",
    "    # Expected value of Moran's I under the null hypothesis:\n",
    "    expected = -1 / (len(x) - 1)\n",
    "\n",
    "    # Expected variance under the null\n",
    "    s1 = 0.5 * ((w + w.T)**2).sum()\n",
    "    s2 = ((w.sum(0) + w.sum(1))**2).sum()\n",
    "    s3 = ((1 / n) * (x_centered**4).sum()) / ((1 / n) * x_centered**2).sum()**2\n",
    "    s4 = (n**2 - 3*n + 3) * s1 - n * s2 + 3 * w_sum**2\n",
    "    s5 = (n**2 - n) * s1 - 2*n*s2 + 6*w_sum**2\n",
    "\n",
    "    var = (n*s4 - s3*s5) / ((n - 1) * (n - 2) * (n - 3) * w_sum**2) - expected**2\n",
    "    \n",
    "    zscore = (observed - expected) / np.sqrt(var)\n",
    "    pvalue = 1 - sp.stats.norm(loc=0, scale=1).cdf(zscore)\n",
    "    return observed, zscore, pvalue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_id = 'UHGG000026_00043'\n",
    "_genome_list = ref_list\n",
    "w = geno_pdist_to_weights(strain_geno_pdist.loc[_genome_list, _genome_list], rate=5)\n",
    "morans_i(\n",
    "    strain_gene_uhgg.loc[gene_id, _genome_list].values,\n",
    "    w,\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "def morans_i2(x, w):\n",
    "    n = len(x)\n",
    "    w = np.array(w)\n",
    "    x = np.array(x)\n",
    "\n",
    "    # Calculate residuals\n",
    "    dev_from_mean = x - np.mean(x)\n",
    "    # import pdb; pdb.set_trace()\n",
    "    \n",
    "    # Calculate Moran's I\n",
    "    numerator = np.sum(np.outer(dev_from_mean, dev_from_mean) * w)\n",
    "    denominator = np.sum(dev_from_mean**2)\n",
    "    I = (n / np.sum(w)) * (numerator / denominator)\n",
    "\n",
    "    # Calculate expectation under the null\n",
    "    E_I = -1 / (n - 1)\n",
    "\n",
    "    # Calculate components for variance under the null\n",
    "    S1 = 0.5 * np.sum((w + w.T)**2)\n",
    "    S2 = np.sum((np.sum(w, axis=1) + np.sum(w, axis=0))**2)\n",
    "    # S3 = ((1 / n) * (dev_from_mean**4).sum()) / ((1 / n) * dev_from_mean**2).sum()**2\n",
    "    # S3 = (n**-1) * np.sum(dev_from_mean**4) / ((n**-1) * np.sum(dev_from_mean**2)**2)\n",
    "    S3 = ((n**-1) * np.sum(dev_from_mean**4)) / ((n**-1) * np.sum(dev_from_mean**2))**2\n",
    "    S4 = (n**2 - 3*n + 3) * S1 - n * S2 + 3 * np.sum(w)**2\n",
    "    S5 = (n**2 - n) * S1 - 2 * n * S2 + 6 * np.sum(w)**2\n",
    "\n",
    "    # Calculate variance under the null\n",
    "    var_I = (n * S4 - S3 * S5) / ((n - 1) * (n - 2) * (n - 3) * np.sum(w)**2) - E_I**2\n",
    "    zscore = (I - E_I) / np.sqrt(var_I)\n",
    "    pvalue = 1 - sp.stats.norm(loc=0, scale=1).cdf(zscore)\n",
    "    return I, zscore, pvalue"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_id = 'UHGG000026_00043'\n",
    "_genome_list = ref_list\n",
    "w = np.exp(-2 * (strain_geno_pdist.loc[_genome_list, _genome_list].values))\n",
    "w[np.diag_indices_from(w)] = 0\n",
    "\n",
    "\n",
    "list(zip(\n",
    "    morans_i(\n",
    "        strain_gene_uhgg.loc[gene_id, _genome_list].values,\n",
    "        w,\n",
    "    ),\n",
    "    morans_i2(\n",
    "        strain_gene_uhgg.loc[gene_id, _genome_list].values,\n",
    "        w,\n",
    "    )\n",
    "))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_genome_list = ref_list\n",
    "_results = {}\n",
    "w = geno_pdist_to_weights(strain_geno_pdist.loc[_genome_list, _genome_list], rate=2)\n",
    "for gene_id in tqdm(strain_gene_uhgg.index):\n",
    "    _results[gene_id] = morans_i(\n",
    "        strain_gene_uhgg.loc[gene_id, _genome_list].values,\n",
    "        w,\n",
    "    )\n",
    "phylogenetic_signal_ref = pd.DataFrame(\n",
    "    _results, index=[\"morans_i\", \"zscore\", \"pvalue\"]\n",
    ").T\n",
    "\n",
    "\n",
    "_genome_list = spgc_list\n",
    "_results = {}\n",
    "w = geno_pdist_to_weights(strain_geno_pdist.loc[_genome_list, _genome_list], rate=2)\n",
    "for gene_id in tqdm(strain_gene_uhgg.index):\n",
    "    _results[gene_id] = morans_i(\n",
    "        strain_gene_uhgg.loc[gene_id, _genome_list].values,\n",
    "        w,\n",
    "    )\n",
    "phylogenetic_signal_spgc = pd.DataFrame(\n",
    "    _results, index=[\"morans_i\", \"zscore\", \"pvalue\"]\n",
    ").T"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "d = pd.DataFrame(dict(ref=morans_ref.morans_i, spgc=morans_spgc.morans_i))\n",
    "\n",
    "plt.hist2d('ref', 'spgc', data=d.dropna())"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# FIXME: Move geno_pdist_* out of the loop.\n",
    "\n",
    "_gene_list = idxwhere(\n",
    "    ((spgc_uhgg_prevalence > 0.1) & (spgc_uhgg_prevalence < 0.9))\n",
    "    | ((ref_uhgg_prevalence > 0.1) & (ref_uhgg_prevalence < 0.9))\n",
    ")\n",
    "\n",
    "phylogenetic_signal_spgc = {}\n",
    "geno_pdist_spgc = squareform(strain_geno_pdist.loc[spgc_list, spgc_list])\n",
    "for gene_id in tqdm(_gene_list):\n",
    "    gene_pdist_spgc = pdist(strain_gene_uhgg.loc[[gene_id], spgc_list].T)\n",
    "    if (gene_pdist_spgc == 0).all():\n",
    "        continue\n",
    "    phylogenetic_signal_spgc[gene_id] = sp.stats.pearsonr(\n",
    "        np.sqrt(geno_pdist_spgc), gene_pdist_spgc\n",
    "    )\n",
    "phylogenetic_signal_spgc = pd.DataFrame(\n",
    "    phylogenetic_signal_spgc, index=[\"r_statistic\", \"pvalue\"]\n",
    ").T\n",
    "\n",
    "phylogenetic_signal_ref = {}\n",
    "geno_pdist_ref = squareform(strain_geno_pdist.loc[ref_list, ref_list])\n",
    "for gene_id in tqdm(_gene_list):\n",
    "    gene_pdist_ref = pdist(strain_gene_uhgg.loc[[gene_id], ref_list].T)\n",
    "    if (gene_pdist_ref == 0).all():\n",
    "        continue\n",
    "    phylogenetic_signal_ref[gene_id] = sp.stats.pearsonr(\n",
    "        np.sqrt(geno_pdist_ref), gene_pdist_ref\n",
    "    )\n",
    "phylogenetic_signal_ref = pd.DataFrame(\n",
    "    phylogenetic_signal_ref, index=[\"r_statistic\", \"pvalue\"]\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(-0.2, 0.2)\n",
    "plt.hist(phylogenetic_signal_spgc.morans_i, alpha=0.5, bins=bins)\n",
    "plt.hist(phylogenetic_signal_ref.morans_i, alpha=0.5, bins=bins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = pd.DataFrame(\n",
    "    dict(x=phylogenetic_signal_ref.morans_i, y=phylogenetic_signal_spgc.morans_i)\n",
    ").dropna()\n",
    "\n",
    "print(sp.stats.pearsonr(d.x, d.y))\n",
    "\n",
    "plt.hist2d(\n",
    "    x=\"x\",\n",
    "    y=\"y\",\n",
    "    data=d,\n",
    "    # kind=\"reg\",\n",
    "    bins=20,\n",
    "    norm=mpl.colors.PowerNorm(1/3),\n",
    ")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = phylogenetic_signal_ref.assign(prevalence=ref_uhgg_prevalence)[lambda x: (x.prevalence > 0.1) & (x.prevalence < 0.9)]\n",
    "plt.scatter('prevalence', 'morans_i', data=d, c='zscore', norm=mpl.colors.SymLogNorm(linthresh=0.1), s=5, alpha=0.5)\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg.loc[shell_gene_list]  # , spgc_list]\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "_row_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        phylo=phylogenetic_signal_spgc.morans_i.map(\n",
    "            lambda x: mpl.cm.coolwarm((x * 5 + 1) / 2)\n",
    "        ),\n",
    "        entrp=ref_gene_uhgg_entropy.map(mpl.cm.viridis),\n",
    "    )\n",
    ")\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x,\n",
    "        row_colors=_row_colors,\n",
    "        col_linkage=strain_geno_linkage,\n",
    "        col_colors=_col_colors,\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_phylogenetic_signal_test = {}\n",
    "\n",
    "for _cog_category in tqdm(list(cog_category_order)):\n",
    "    d = pd.DataFrame(\n",
    "        dict(\n",
    "            phylogenetic_signal=phylogenetic_signal_spgc.morans_i,\n",
    "            is_cog_category=gene_x_cog_category_matrix.reindex(\n",
    "                phylogenetic_signal_spgc.index\n",
    "            ).fillna(False)[_cog_category],\n",
    "        )\n",
    "    ).loc[shell_gene_list]\n",
    "    x = d[d.is_cog_category].phylogenetic_signal\n",
    "    y = d[~d.is_cog_category].phylogenetic_signal\n",
    "    if (len(x) > 5) and (len(y) > 5):\n",
    "        _test = sp.stats.mannwhitneyu(x, y)\n",
    "    else:\n",
    "        _test = (np.nan, np.nan)\n",
    "    cog_category_phylogenetic_signal_test[_cog_category] = (\n",
    "        len(x),\n",
    "        x.median(),\n",
    "        y.median(),\n",
    "        x.mean(),\n",
    "        y.mean(),\n",
    "        *_test,\n",
    "    )\n",
    "\n",
    "cog_category_phylogenetic_signal_test = pd.DataFrame(\n",
    "    cog_category_phylogenetic_signal_test,\n",
    "    index=[\n",
    "        \"num_genes_in_category\",\n",
    "        \"cog_median_i\",\n",
    "        \"not_cog_median_i\",\n",
    "        \"cog_mean_i\",\n",
    "        \"not_cog_mean_i\",\n",
    "        \"mwu_statistic\",\n",
    "        \"pvalue\",\n",
    "    ],\n",
    ").T.assign(\n",
    "    negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "    median_diff=lambda x: x.cog_median_i - x.not_cog_median_i,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = cog_category_phylogenetic_signal_test.join(cog_category_description).dropna(\n",
    "    subset=[\"median_diff\"]\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "plt.scatter(\n",
    "    \"median_diff\",\n",
    "    \"pvalue\",\n",
    "    data=d,\n",
    ")\n",
    "# lib.plot.rotate_xticklabels()\n",
    "ax.axvline(0, linestyle=\"--\", lw=1, color=\"k\")\n",
    "# ax.axhline(-np.log10(0.05 / d.shape[0]), linestyle='--', lw=1, color='k')\n",
    "\n",
    "\n",
    "for cog_category in idxwhere((np.abs(d.median_diff) > 0) & (d.pvalue < 1e-4)):\n",
    "    plt.annotate(\n",
    "        cog_category,\n",
    "        xy=d[[\"median_diff\", \"pvalue\"]].loc[cog_category],\n",
    "        xytext=(4, 2),\n",
    "        textcoords=\"offset points\",\n",
    "        # color=prevalence_class_palette[prevalence_class],\n",
    "        # ha='center',\n",
    "        # va='center',\n",
    "        # fontweight='bold',\n",
    "    )\n",
    "\n",
    "plt.yscale(\"log\")\n",
    "plt.xlim(-0.04, 0.04)\n",
    "plt.ylim(1e-15, 1)\n",
    "plt.gca().invert_yaxis()\n",
    "plt.xlabel(\"Difference in Median Moran's I\\n(in COG category vs. not)\")\n",
    "plt.ylabel(\"P-value\")\n",
    "\n",
    "d.sort_values(\"median_diff\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    cog_category_phylogenetic_signal_test.join(cog_category_description)[\n",
    "        [\"median_diff\"]\n",
    "    ]\n",
    "    .dropna()\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")\n",
    ")\n",
    "\n",
    "# Order rows based on the COG-category enrichment analysis\n",
    "_row_order = (\n",
    "    cog_category_gene_class_enrichment_test.log2_odds_ratio.unstack(\"prevalence_class\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    "    .fillna(0)[\"core\"]\n",
    "    .sort_values(ascending=False)\n",
    "    .index\n",
    ")\n",
    "annot = (\n",
    "    cog_category_phylogenetic_signal_test.pvalue.map(_assign_significance_marker)\n",
    "    .to_frame()\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")\n",
    ")\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(3.5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"PuOr\",\n",
    "    center=0,  # phylogenetic_signal_spgc.morans_i.mean(),\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True,\n",
    "        location=\"left\",\n",
    "        label=\"Difference in Median Moran's I\",\n",
    "        extend=\"both\",\n",
    "        fraction=0.5,\n",
    "    ),\n",
    "    # ax=ax,\n",
    "    # yticklabels=1,\n",
    "    # xticklabels=1,\n",
    "    # annot_kws=dict(va='center'),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# d = phylogenetic_signal_spgc.loc[shell_gene_list].r_statistic.dropna().to_frame().join(gene_x_cog_category).fillna(\"-\")\n",
    "d = (\n",
    "    phylogenetic_signal_spgc.morans_i.dropna()\n",
    "    .to_frame()\n",
    "    .join(gene_x_cog_category)\n",
    "    .fillna(\"-\")\n",
    ")\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 5))\n",
    "sns.violinplot(\n",
    "    data=d,\n",
    "    x=\"cog_category\",\n",
    "    y=\"morans_i\",\n",
    "    ax=ax,\n",
    "    order=cog_category_phylogenetic_signal_test.cog_median_i.sort_values().index,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Co-occurence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Found using SPGC only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_prevalence = strain_uhgg_filt[spgc_list].mean(1)\n",
    "plt.hist(gene_prevalence)\n",
    "plt.yscale(\"log\")\n",
    "variable_genes = idxwhere((gene_prevalence > 0.05) & (gene_prevalence < 0.95))\n",
    "len(variable_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg.astype(bool)\n",
    "y = x[spgc_list]\n",
    "\n",
    "drop_nohit_genes_list = idxwhere(y.sum(1) == 0)\n",
    "drop_ubiq_genes_list = idxwhere((~y).sum(1) == 0)\n",
    "drop_single_hit_genes_list = idxwhere(y.sum(1) == 1)\n",
    "drop_only_one_missing_genes_list = idxwhere((~y).sum(1) == 1)\n",
    "z = y.drop(\n",
    "    drop_nohit_genes_list\n",
    "    + drop_ubiq_genes_list\n",
    "    + drop_single_hit_genes_list\n",
    "    + drop_only_one_missing_genes_list\n",
    ")\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "clust1 = pd.Series(\n",
    "    AgglomerativeClustering(\n",
    "        n_clusters=None,\n",
    "        metric=\"correlation\",\n",
    "        linkage=\"average\",\n",
    "        distance_threshold=0.035,\n",
    "    ).fit_predict(z),\n",
    "    index=z.index,\n",
    ")\n",
    "# Add back ubiquitous and nohit genes as clusters.\n",
    "clust1 = pd.concat(\n",
    "    [\n",
    "        clust1,\n",
    "        pd.Series(-1, index=drop_nohit_genes_list),\n",
    "        pd.Series(-2, index=drop_ubiq_genes_list),\n",
    "        pd.Series(-3, index=drop_single_hit_genes_list),\n",
    "        pd.Series(-4, index=drop_only_one_missing_genes_list),\n",
    "    ]\n",
    ")\n",
    "clust1_palette = lib.plot.construct_ordered_palette(clust1.unique())\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x.loc[variable_genes],\n",
    "        row_colors=x.loc[variable_genes]\n",
    "        .index.to_series()\n",
    "        .map(clust1)\n",
    "        .map(clust1_palette),\n",
    "        col_linkage=strain_geno_linkage,\n",
    "        col_colors=_col_colors,\n",
    "        metric=\"cosine\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust1_sizes = clust1.value_counts()\n",
    "\n",
    "clust1_sizes_meta = (\n",
    "    clust1.value_counts().to_frame(\"tally\")\n",
    "    # .groupby('clust')\n",
    "    # .count()\n",
    "    # .sort_values(ascending=False)\n",
    "    .assign(\n",
    "        top5_cc=gene_x_cog_category_matrix.groupby(clust1)\n",
    "        .sum()\n",
    "        .rename(int)\n",
    "        .astype(int)\n",
    "        .apply(lambda x: x.sort_values(ascending=False).head(5).index.to_list(), axis=1)\n",
    "    )\n",
    ")\n",
    "\n",
    "clust1_sizes_meta.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust1_sizes_meta.drop([-1, -2, -3, -4]).index[0]\n",
    "\n",
    "_gene_list = idxwhere(clust1 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(phylogenetic_signal_spgc.loc[_gene_list].median())\n",
    "print(phylogenetic_signal_ref.loc[_gene_list].median())\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Found using SPGC and Ref"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_prevalence = strain_uhgg_filt[spgc_list + list(ref_list)].mean(1)\n",
    "plt.hist(gene_prevalence)\n",
    "plt.yscale('log')\n",
    "variable_genes = idxwhere((gene_prevalence > 0.05) & (gene_prevalence < 0.95))\n",
    "len(variable_genes)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "x = strain_gene_uhgg.astype(bool)\n",
    "y = x[spgc_list + ref_list]\n",
    "\n",
    "drop_nohit_genes_list = idxwhere(y.sum(1) == 0)\n",
    "drop_ubiq_genes_list = idxwhere((~y).sum(1) == 0)\n",
    "drop_single_hit_genes_list = idxwhere(y.sum(1) == 1)\n",
    "drop_only_one_missing_genes_list = idxwhere((~y).sum(1) == 1)\n",
    "z = y.drop(drop_nohit_genes_list + drop_ubiq_genes_list + drop_single_hit_genes_list + drop_only_one_missing_genes_list)\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "clust2 = pd.Series(\n",
    "    AgglomerativeClustering(\n",
    "        n_clusters=None,\n",
    "        metric=\"correlation\",\n",
    "        linkage=\"average\",\n",
    "        distance_threshold=0.035,\n",
    "    ).fit_predict(z),\n",
    "    index=z.index,\n",
    ")\n",
    "# Add back ubiquitous and nohit genes as clusters.\n",
    "clust2 = pd.concat(\n",
    "    [\n",
    "        clust2,\n",
    "        pd.Series(-1, index=drop_nohit_genes_list),\n",
    "        pd.Series(-2, index=drop_ubiq_genes_list),\n",
    "        pd.Series(-3, index=drop_single_hit_genes_list),\n",
    "        pd.Series(-4, index=drop_only_one_missing_genes_list),\n",
    "    ]\n",
    ")\n",
    "clust2_palette = lib.plot.construct_ordered_palette(clust2.unique())\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x.loc[variable_genes],\n",
    "        row_colors=x.loc[variable_genes].index.to_series().map(clust2).map(clust2_palette),\n",
    "        col_linkage=strain_geno_linkage,\n",
    "        col_colors=_col_colors,\n",
    "        metric=\"cosine\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "clust2_sizes = clust2.value_counts()\n",
    "\n",
    "clust2_sizes_meta = (\n",
    "    clust2.value_counts().to_frame(\"tally\")\n",
    "    # .groupby('clust')\n",
    "    # .count()\n",
    "    # .sort_values(ascending=False)\n",
    "    .assign(\n",
    "        top5_cc=gene_x_cog_category_matrix.groupby(clust2)\n",
    "        .sum()\n",
    "        .rename(int)\n",
    "        .astype(int)\n",
    "        .apply(lambda x: x.sort_values(ascending=False).head(5).index.to_list(), axis=1)\n",
    "    )\n",
    ")\n",
    "\n",
    "clust2_sizes_meta.head(10)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_clust = clust2_sizes_meta.index[9]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Found using Ref only"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "gene_prevalence = strain_uhgg_filt[list(ref_list)].mean(1)\n",
    "plt.hist(gene_prevalence)\n",
    "plt.yscale('log')\n",
    "variable_genes = idxwhere((gene_prevalence > 0.05) & (gene_prevalence < 0.95))\n",
    "len(variable_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg.astype(bool)\n",
    "y = x[ref_list]\n",
    "\n",
    "drop_nohit_genes_list = idxwhere(y.sum(1) == 0)\n",
    "drop_ubiq_genes_list = idxwhere((~y).sum(1) == 0)\n",
    "drop_single_hit_genes_list = idxwhere(y.sum(1) == 1)\n",
    "drop_only_one_missing_genes_list = idxwhere((~y).sum(1) == 1)\n",
    "z = y.drop(\n",
    "    drop_nohit_genes_list\n",
    "    + drop_ubiq_genes_list\n",
    "    + drop_single_hit_genes_list\n",
    "    + drop_only_one_missing_genes_list\n",
    ")\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "clust3 = pd.Series(\n",
    "    AgglomerativeClustering(\n",
    "        n_clusters=None,\n",
    "        metric=\"correlation\",\n",
    "        linkage=\"average\",\n",
    "        distance_threshold=0.035,\n",
    "    ).fit_predict(z),\n",
    "    index=z.index,\n",
    ")\n",
    "# Add back ubiquitous and nohit genes as clusters.\n",
    "clust3 = pd.concat(\n",
    "    [\n",
    "        clust3,\n",
    "        pd.Series(-1, index=drop_nohit_genes_list),\n",
    "        pd.Series(-2, index=drop_ubiq_genes_list),\n",
    "        pd.Series(-3, index=drop_single_hit_genes_list),\n",
    "        pd.Series(-4, index=drop_only_one_missing_genes_list),\n",
    "    ]\n",
    ")\n",
    "clust3_palette = lib.plot.construct_ordered_palette(clust3.unique())\n",
    "\n",
    "if show_unimportant_figures:\n",
    "    sns.clustermap(\n",
    "        x.loc[variable_genes],\n",
    "        row_colors=x.loc[variable_genes]\n",
    "        .index.to_series()\n",
    "        .map(clust3)\n",
    "        .map(clust3_palette),\n",
    "        col_linkage=strain_geno_linkage,\n",
    "        col_colors=_col_colors,\n",
    "        metric=\"cosine\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust3_sizes = clust3.value_counts()\n",
    "\n",
    "clust3_sizes_meta = (\n",
    "    clust3.value_counts().to_frame(\"tally\")\n",
    "    # .groupby('clust')\n",
    "    # .count()\n",
    "    # .sort_values(ascending=False)\n",
    "    .assign(\n",
    "        top5_cc=gene_x_cog_category_matrix.groupby(clust3)\n",
    "        .sum()\n",
    "        .rename(int)\n",
    "        .astype(int)\n",
    "        .apply(lambda x: x.sort_values(ascending=False).head(5).index.to_list(), axis=1)\n",
    "    )\n",
    ")\n",
    "clust3_sizes_meta.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust3_sizes_meta.head(10).index[-1]\n",
    "\n",
    "_gene_list = idxwhere(clust3 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(phylogenetic_signal_spgc.loc[_gene_list].median())\n",
    "print(phylogenetic_signal_ref.loc[_gene_list].median())\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Comparison Among Clusterings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_list = idxwhere((gene_prevalence >= 0) & (gene_prevalence <= 1))\n",
    "\n",
    "print(adjusted_mutual_info_score(clust1[_gene_list], clust3[_gene_list]))\n",
    "# print(adjusted_mutual_info_score(clust1[_gene_list], clust2[_gene_list]))\n",
    "# print(adjusted_mutual_info_score(clust2[_gene_list], clust3[_gene_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_list = idxwhere((gene_prevalence >= 0.05) & (gene_prevalence <= 0.95))\n",
    "\n",
    "print(adjusted_mutual_info_score(clust1[_gene_list], clust3[_gene_list]))\n",
    "# print(adjusted_mutual_info_score(clust1[_gene_list], clust2[_gene_list]))\n",
    "# print(adjusted_mutual_info_score(clust2[_gene_list], clust3[_gene_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_list = idxwhere((gene_prevalence >= 0) & (gene_prevalence <= 1))\n",
    "\n",
    "print(adjusted_rand_score(clust1[_gene_list], clust3[_gene_list]))\n",
    "# print(adjusted_rand_score(clust1[_gene_list], clust2[_gene_list]))\n",
    "# print(adjusted_rand_score(clust2[_gene_list], clust3[_gene_list]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_list = idxwhere((gene_prevalence >= 0.05) & (gene_prevalence <= 0.95))\n",
    "\n",
    "print(adjusted_rand_score(clust1[_gene_list], clust3[_gene_list]))\n",
    "# print(adjusted_rand_score(clust1[_gene_list], clust2[_gene_list]))\n",
    "# print(adjusted_rand_score(clust2[_gene_list], clust3[_gene_list]))"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "from sklearn.metrics.cluster import contingency_matrix\n",
    "\n",
    "_inspect_clust = clust2\n",
    "_against_clust = clust3\n",
    "\n",
    "contingency = pd.DataFrame(\n",
    "    contingency_matrix(_against_clust, _inspect_clust),\n",
    "    index=sorted(_against_clust.unique()),\n",
    "    columns=sorted(_inspect_clust.unique()),\n",
    ")\n",
    "spgc_gene_clust_meta = (\n",
    "        pd.DataFrame(\n",
    "        dict(\n",
    "            clust_size=contingency.sum(),\n",
    "            largest_overlap_size=contingency.max(),\n",
    "            largest_overlap_clust=contingency.idxmax(),\n",
    "            prevalence_spgc=(\n",
    "                strain_gene_uhgg[spgc_list].groupby(_inspect_clust).mean()\n",
    "            ).mean(1),\n",
    "            prevalence_ref=(\n",
    "                strain_gene_uhgg[ref_list].groupby(_inspect_clust).mean()\n",
    "            ).mean(1),\n",
    "        )\n",
    "    )\n",
    "    .assign(\n",
    "        interest_score=lambda x: x.prevalence_ref * np.log(x.prevalence_ref) * x.prevalence_spgc * np.log(x.prevalence_spgc) * np.log(x.clust_size),\n",
    "    )\n",
    "    .sort_values('interest_score', ascending=False)\n",
    ")\n",
    "spgc_gene_clust_meta.head(10)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "_clust = spgc_gene_clust_meta.index[0]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SPGC-MWAS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# _row_linkage = gene_uhgg_linkage\n",
    "_gene_data = strain_gene_uhgg\n",
    "\n",
    "u = (\n",
    "    sfacts_fit.community.sel(strain=spgc_list)\n",
    "    .to_pandas()\n",
    "    .groupby(mgen_meta.subject_id)\n",
    "    .mean()\n",
    "    > 0.2\n",
    ")[lambda x: x.any(1)]\n",
    "v = _gene_data[u.columns]\n",
    "subject_strain_gene_content = (u @ v.T).T > 0\n",
    "subject_strain_gene_prevalence = subject_strain_gene_content.mean(1)\n",
    "num_strain_subjects = subject_strain_gene_content.shape[1]\n",
    "\n",
    "subject_ibd_diagnosis = subject.loc[subject_strain_gene_content.columns].ibd_diagnosis\n",
    "\n",
    "_col_colors = subject_ibd_diagnosis.replace(\n",
    "    {\"CD\": \"tab:green\", \"UC\": \"tab:blue\", \"nonIBD\": \"lightgrey\"}\n",
    ")\n",
    "# row_order = sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)\n",
    "# sns.clustermap(subject_strain_gene_content.iloc[row_order], col_colors=_col_colors, row_cluster=False)\n",
    "if not u.empty:\n",
    "    sns.clustermap(\n",
    "        subject_strain_gene_content, col_colors=_col_colors, row_cluster=False\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwas_results = []\n",
    "for _gene in tqdm(strain_gene_uhgg.index):\n",
    "    contingency_table0 = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                diagnosis=subject_ibd_diagnosis,\n",
    "                gene=subject_strain_gene_content.loc[_gene],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack(\"diagnosis\")\n",
    "        .reindex(columns=[\"CD\", \"UC\", \"nonIBD\"])\n",
    "        .assign(IBD=lambda x: x.CD + x.UC)\n",
    "    )\n",
    "    for ibd_comparison, (groupA, groupB) in dict(\n",
    "        cd=(\"nonIBD\", \"CD\"), uc=(\"nonIBD\", \"UC\"), ibd=(\"nonIBD\", \"IBD\")\n",
    "    ).items():\n",
    "        contingency_table1 = contingency_table0.reindex(\n",
    "            index=[True, False], columns=[groupA, groupB]\n",
    "        ).fillna(0)\n",
    "        _test = sp.stats.fisher_exact(contingency_table1)\n",
    "        contingency_table_pc = contingency_table1 + 1\n",
    "        oddsratio_pc = (\n",
    "            contingency_table_pc.loc[True, groupB]\n",
    "            / contingency_table_pc.loc[False, groupB]\n",
    "        ) / (\n",
    "            contingency_table_pc.loc[True, groupA]\n",
    "            / contingency_table_pc.loc[False, groupA]\n",
    "        )\n",
    "        mwas_results.append((ibd_comparison, _gene, *_test, oddsratio_pc))\n",
    "\n",
    "mwas_results = (\n",
    "    pd.DataFrame(\n",
    "        mwas_results,\n",
    "        columns=[\"ibd_comparison\", \"gene\", \"statistic\", \"pvalue\", \"oddsratio_pc\"],\n",
    "    )\n",
    "    # .join(\n",
    "    #     gene_annotations[\n",
    "    #         [\n",
    "    #             \"Description\",\n",
    "    #             \"COG_category\",\n",
    "    #             \"eggNOG_OGs\",\n",
    "    #             \"centroid_99_length\",\n",
    "    #             \"score\",\n",
    "    #             \"Preferred_name\",\n",
    "    #             \"KEGG_ko\",\n",
    "    #             \"PFAMs\",\n",
    "    #         ]\n",
    "    #     ],\n",
    "    #     on=\"gene\",\n",
    "    # )\n",
    "    .sort_values(\"pvalue\")\n",
    ")\n",
    "\n",
    "mwas_results.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.logspace(-5, 0, num=20)\n",
    "# bins = np.linspace(0, 1, num=51)\n",
    "for ibd_comparison in mwas_results.ibd_comparison.unique():\n",
    "    plt.hist(\n",
    "        mwas_results[lambda x: x.ibd_comparison == ibd_comparison].pvalue,\n",
    "        label=ibd_comparison,\n",
    "        alpha=0.5,\n",
    "        bins=bins,\n",
    "    )\n",
    "\n",
    "plt.plot(bins[1:], (bins[1:] - bins[:-1]) * mwas_results.shape[0] / 3)\n",
    "plt.legend()\n",
    "\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"log\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene = mwas_results.gene.iloc[0]\n",
    "print(_gene)\n",
    "pd.DataFrame(\n",
    "    dict(diagnosis=subject_ibd_diagnosis, gene=subject_strain_gene_content.loc[_gene])\n",
    ").value_counts().unstack().fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_list = mwas_results.head(10).gene\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = genome_type.map(genome_type_palette)\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_data = strain_gene_uhgg.loc[_gene_list]\n",
    "\n",
    "_subject_x_strain = (\n",
    "    sfacts_fit.community.sel(strain=spgc_list)\n",
    "    .to_pandas()\n",
    "    .groupby(mgen_meta.subject_id)\n",
    "    .mean()\n",
    "    > 0.2\n",
    ")[lambda x: x.any(1)]\n",
    "_strain_x_gene = _gene_data[u.columns]\n",
    "_subject_x_gene = (_subject_x_strain @ _strain_x_gene.T).T > 0\n",
    "\n",
    "_col_colors = subject.loc[_subject_x_gene.columns].ibd_diagnosis.replace(\n",
    "    {\"CD\": \"tab:green\", \"UC\": \"tab:blue\", \"nonIBD\": \"lightgrey\"}\n",
    ")\n",
    "if not _subject_x_gene.empty:\n",
    "    sns.clustermap(_subject_x_gene, col_colors=_col_colors, row_cluster=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_data = subject_uhgg_depth_ratio.T.loc[_gene_list].applymap(\n",
    "    np.nan_to_num, nan=0, posinf=100, neginf=-100\n",
    ")\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        ibd_diagnosis=subject.ibd_diagnosis.replace(\n",
    "            {\"CD\": \"tab:green\", \"UC\": \"tab:blue\", \"nonIBD\": \"lightgrey\"}\n",
    "        ),\n",
    "        species_depth=subject_mean_species_depth.map(\n",
    "            lambda x: mpl.cm.viridis(x / subject_mean_species_depth.max())\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "sns.clustermap(\n",
    "    _gene_data + 1e-5,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.SymLogNorm(1e-3, vmin=1e-2, vmax=1e2),\n",
    "    metric=\"cosine\",\n",
    "    row_cluster=False,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mean_subject_depth = subject_uhgg_depth_ratio[_gene]\n",
    "\n",
    "d = subject[[\"ibd_diagnosis\"]].join(mean_subject_depth).dropna()\n",
    "for ibd_diagnosis in d.ibd_diagnosis.unique():\n",
    "    plt.hist(\n",
    "        d.loc[d.ibd_diagnosis == ibd_diagnosis, _gene] + 1e-4,\n",
    "        bins=np.logspace(-4, 3),\n",
    "        alpha=0.5,\n",
    "        label=ibd_diagnosis,\n",
    "    )\n",
    "\n",
    "thresh = 1e-1\n",
    "plt.axvline(thresh, lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.xscale(\"log\")\n",
    "# plt.yscale('log')\n",
    "plt.legend()\n",
    "\n",
    "d.assign(gene=lambda x: x[_gene] > thresh)[\n",
    "    [\"ibd_diagnosis\", \"gene\"]\n",
    "].value_counts().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Write Statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_stats = (\n",
    "    pd.DataFrame(\n",
    "        dict(\n",
    "            nlength=gene_annotations[\"centroid_99_length\"],\n",
    "            eggnog=gene_annotations[\"eggNOG_OGs\"],\n",
    "            cog_categories=gene_annotations[\"COG_category\"],\n",
    "            description=gene_annotations[\"Description\"],\n",
    "            gene_name=gene_annotations[\"Preferred_name\"],\n",
    "            ko=gene_annotations[\"KEGG_ko\"],\n",
    "            pfam=gene_annotations[\"PFAMs\"],\n",
    "            prevalence_spgc=spgc_gene_uhgg_prevalence,\n",
    "            prevalence_ref=ref_gene_uhgg_prevalence,\n",
    "            prevalence_mwas_subject=subject_strain_gene_prevalence,\n",
    "            num_mwas_subject=num_strain_subjects,\n",
    "            phylogenetic_i_spgc=phylogenetic_signal_spgc.morans_i,\n",
    "            phylogenetic_i_ref=phylogenetic_signal_ref.morans_i,\n",
    "            coclust_label_spgc=clust1,\n",
    "            # coclust_label_both=clust2,\n",
    "            coclust_label_ref=clust3,\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        mwas_results.set_index([\"gene\", \"ibd_comparison\"])\n",
    "        .pvalue.unstack()\n",
    "        .rename(columns=lambda x: x + \"_mwas_pvalue\")\n",
    "    )\n",
    "    .join(\n",
    "        mwas_results.set_index([\"gene\", \"ibd_comparison\"])\n",
    "        .oddsratio_pc.unstack()\n",
    "        .rename(columns=lambda x: x + \"_mwas_oddsratio_pc\")\n",
    "    )\n",
    "    .join(gene_x_cog_category_matrix.rename(columns=lambda x: \"cog_category_\" + x))\n",
    "    .rename_axis(index=\"gene_id\")\n",
    ")\n",
    "\n",
    "gene_stats.to_csv(gene_stats_outpath, sep=\"\\t\")\n",
    "gene_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_strain_stats = (\n",
    "    pd.DataFrame(\n",
    "        dict(\n",
    "            nearest_ref_geno=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "            ][\"min_geno_ref\"],\n",
    "            nearest_ref_geno_diss=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "            ][\"geno_dist\"],\n",
    "            nearest_ref_gene_diss=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "            ][\"gene_dist\"],\n",
    "            nearest_ref_gene_raw_diss=genome_distance_comparison_raw[\n",
    "                lambda x: x.genome_type.isin([\"SPGC\"])\n",
    "            ][\"gene_dist\"],\n",
    "        )\n",
    "    )\n",
    "    .join(spgc_meta.loc[spgc_list])\n",
    "    .join(\n",
    "        strain_genome_spgc_prevalence_class_tally.rename(\n",
    "            columns=lambda x: f\"spgc_{x}_gene_tally\"\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        strain_genome_ref_prevalence_class_tally.rename(\n",
    "            columns=lambda x: f\"ref_{x}_gene_tally\"\n",
    "        )\n",
    "    )\n",
    "    .rename_axis(index=\"strain\")\n",
    ")\n",
    "\n",
    "spgc_strain_stats.to_csv(spgc_strain_stats_outpath, sep=\"\\t\")\n",
    "spgc_strain_stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_strain_stats = (\n",
    "    pd.DataFrame(\n",
    "        dict(\n",
    "            nearest_ref_geno=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "            ][\"min_geno_ref\"],\n",
    "            nearest_ref_geno_diss=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "            ][\"geno_dist\"],\n",
    "            nearest_ref_gene_diss=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "            ][\"gene_dist\"],\n",
    "            nearest_ref_gene_raw_diss=genome_distance_comparison_raw[\n",
    "                lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "            ][\"gene_dist\"],\n",
    "            genome_type=genome_distance_comparison[\n",
    "                lambda x: x.genome_type.isin([\"Isolate\", \"MAG\"])\n",
    "            ][\"genome_type\"],\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        reference_meta[\n",
    "            [\n",
    "                \"Length\",\n",
    "                \"N_contigs\",\n",
    "                \"N50\",\n",
    "                \"GC_content\",\n",
    "                \"Completeness\",\n",
    "                \"Contamination\",\n",
    "            ]\n",
    "        ]\n",
    "    )\n",
    "    .join(\n",
    "        strain_genome_spgc_prevalence_class_tally.rename(\n",
    "            columns=lambda x: f\"spgc_{x}_gene_tally\"\n",
    "        )\n",
    "    )\n",
    "    .join(\n",
    "        strain_genome_ref_prevalence_class_tally.rename(\n",
    "            columns=lambda x: f\"ref_{x}_gene_tally\"\n",
    "        )\n",
    "    )\n",
    "    .rename_axis(index=\"strain\")\n",
    ")\n",
    "\n",
    "ref_strain_stats.to_csv(ref_strain_stats_outpath, sep=\"\\t\")\n",
    "ref_strain_stats"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1847.611523,
   "end_time": "2023-06-24T02:46:52.897060",
   "environment_variables": {},
   "exception": null,
   "input_path": "nb/prototype_compare_spgc_and_refs.ipynb",
   "output_path": "nb/papermill/data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_ref_comparison.ipynb",
   "parameters": {
    "gene_annotations_path": "data/species/sp-101346/pangenome.centroids.emapper.d/proteins.emapper.annotations",
    "mgen_path": "meta/hmp2/mgen.tsv",
    "nb_path": "nb/prototype_compare_spgc_and_refs.ipynb",
    "preparation_path": "meta/hmp2/preparation.tsv",
    "ref_gene_copy_number_uhgg_path": "ref/midasdb_uhgg_pangenomes/101346/gene75.reference_copy_number.nc",
    "ref_geno_path": "data/species/sp-101346/gtpro_ref.mgtp.nc",
    "sample_to_spgc_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.spgc_ss-all.strain_samples.tsv",
    "sfacts_fit_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.world.nc",
    "species_id": "101346",
    "species_taxonomy_path": "ref/gtpro/species_taxonomy_ext.tsv",
    "spgc_gene_uhgg_corr_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_correlation.tsv",
    "spgc_gene_uhgg_depth_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_depth_ratio.tsv",
    "spgc_gene_uhgg_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_gene.tsv",
    "spgc_meta_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_meta.tsv",
    "stool_path": "meta/hmp2/stool.tsv",
    "subject_path": "meta/hmp2/subject.tsv",
    "uhgg_depth_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gene99-v22-agg75.depth2.nc",
    "uhgg_gene_length_path": "ref/midasdb_uhgg/pangenomes/101346/cluster_info.txt",
    "uhgg_x_eggnog_path": "data/species/sp-101346/pangenome.centroids.emapper.gene_x_eggnog.tsv",
    "uhgg_x_top_eggnog_path": "data/species/sp-101346/pangenome.centroids.emapper.gene_x_top_eggnog.tsv"
   },
   "start_time": "2023-06-24T02:16:05.285537",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}