{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Template Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os as _os\n",
    "\n",
    "_os.chdir(_os.environ[\"PROJECT_ROOT\"])\n",
    "_os.path.realpath(_os.path.curdir)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import time\n",
    "import warnings\n",
    "from itertools import chain, product\n",
    "from tempfile import mkstemp\n",
    "\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy as sp\n",
    "import seaborn as sns\n",
    "import sfacts as sf\n",
    "import statsmodels as sm\n",
    "import xarray as xr\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from scipy.spatial.distance import pdist, squareform\n",
    "from sklearn.cluster import AgglomerativeClustering\n",
    "from statsmodels.stats.multitest import fdrcorrection\n",
    "from tqdm import tqdm\n",
    "\n",
    "import lib.plot\n",
    "import lib.thisproject.data\n",
    "from lib.pandas_util import align_indexes, aligned_index, idxwhere, invert_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set_context(\"talk\")\n",
    "plt.rcParams[\"figure.dpi\"] = 50"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Papermill parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This cell is tagged \"parameters\" for papermill.\n",
    "# See <https://papermill.readthedocs.io/en/latest/usage-parameterize.html#how-parameters-work> for some gotchas.\n",
    "# NOTE: *ALL* parameters should be passed to papermill. Values set here are only for prototyping.\n",
    "species_id = \"102506\"\n",
    "\n",
    "species_taxonomy_inpath = f\"ref/gtpro/species_taxonomy_ext.tsv\"\n",
    "sample_to_spgc_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.spgc_ss-all.strain_samples.tsv\"\n",
    "sfacts_fit_inpath = (\n",
    "    f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.world.nc\"\n",
    ")\n",
    "ref_geno_inpath = f\"data/species/sp-{species_id}/gtpro_ref.mgtp.nc\"\n",
    "spgc_meta_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_meta.tsv\"\n",
    "ref_gene_copy_number_uhgg_inpath = (\n",
    "    f\"data/species/sp-{species_id}/gene75_new.reference_copy_number.nc\"\n",
    ")\n",
    "spgc_gene_uhgg_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_gene.tsv\"\n",
    "spgc_gene_uhgg_depth_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_depth_ratio.tsv\"\n",
    "spgc_gene_uhgg_corr_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_correlation.tsv\"\n",
    "uhgg_x_eggnog_inpath = (\n",
    "    f\"data/species/sp-{species_id}/pangenome.centroids.emapper.gene_x_eggnog.tsv\"\n",
    ")\n",
    "uhgg_x_top_eggnog_inpath = (\n",
    "    f\"data/species/sp-{species_id}/pangenome_new.centroids.emapper.gene_x_top_eggnog.tsv\"\n",
    ")\n",
    "uhgg_gene_length_inpath = f\"ref/midasdb_uhgg_new/pangenomes/{species_id}/cluster_info.txt\"\n",
    "gene_annotations_inpath = f\"data/species/sp-{species_id}/pangenome_new.centroids.emapper.d/proteins.emapper.annotations\"\n",
    "uhgg_depth_inpath = (\n",
    "    f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.depth2.nc\"\n",
    ")\n",
    "species_depth_inpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95.species_depth.tsv\"\n",
    "\n",
    "mgen_inpath = f\"meta/ucfmt/mgen.tsv\"\n",
    "sample_inpath = f\"meta/ucfmt/sample.tsv\"\n",
    "subject_inpath = f\"meta/ucfmt/subject.tsv\"\n",
    "\n",
    "html_outpath = f\"data/group/xjin_ucfmt_hmp2/species/sp-{species_id}/r.proc.gtpro.sfacts-fit.gene99_new-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.donor_comparison.html\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Loading / Validation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_taxonomy = lib.thisproject.data.load_species_taxonomy(species_taxonomy_inpath)\n",
    "species_taxonomy.loc[species_id]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_x_eggnog = pd.read_table(uhgg_x_eggnog_inpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_x_top_eggnog = pd.read_table(uhgg_x_top_eggnog_inpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_gene_length = (\n",
    "    pd.read_table(uhgg_gene_length_inpath)\n",
    "    .groupby(\"centroid_75\")\n",
    "    .centroid_99_length.mean()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eggnog_column_names = \"query seed_ortholog evalue score eggNOG_OGs max_annot_lvl COG_category Description Preferred_name GOs EC KEGG_ko KEGG_inpathway KEGG_Module KEGG_Reaction KEGG_rclass BRITE KEGG_TC CAZy BiGG_Reaction PFAMs\".split(\n",
    "    \" \"\n",
    ")\n",
    "_gene_annotations = (\n",
    "    pd.read_table(\n",
    "        gene_annotations_inpath,\n",
    "        comment=\"#\",\n",
    "        names=eggnog_column_names,\n",
    "        index_col=\"query\",\n",
    "    )\n",
    "    .rename_axis(index=\"gene_id\")\n",
    "    .replace({\"-\": np.nan})\n",
    ")\n",
    "_gene_annotations = uhgg_gene_length.to_frame().join(_gene_annotations)\n",
    "\n",
    "_gene_annotations.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category1 = (\n",
    "    _gene_annotations.COG_category.fillna(\"-\").apply(list).explode()[lambda x: x != \"-\"]\n",
    ")\n",
    "gene_x_cog_category1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_x_category = pd.read_table(\n",
    "    \"ref/cog-20.meta.tsv\",\n",
    "    names=[\"cog\", \"cog_category\", \"description\", \"short_name\", \"_4\", \"_5\", \"_6\"],\n",
    "    index_col=\"cog\",\n",
    ").cog_category\n",
    "cog_x_category"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_gene_annotations.eggNOG_OGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog = (\n",
    "    _gene_annotations.eggNOG_OGs.fillna(\"\")\n",
    "    .str.split(\",\")\n",
    "    .explode()[lambda x: x.str.startswith(\"COG\")]\n",
    "    .str.split(\"@\")\n",
    "    .str[0]\n",
    ")\n",
    "gene_x_cog.value_counts().head()\n",
    "gene_x_cog_category2 = gene_x_cog.map(cog_x_category).dropna().apply(list).explode()\n",
    "gene_x_cog_category2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            gene_x_cog_category1,\n",
    "            gene_x_cog_category2,  # FIXME: Which metadata table do I want?\n",
    "        ]\n",
    "    )\n",
    "    .reset_index()\n",
    "    .drop_duplicates()\n",
    ")\n",
    "\n",
    "gene_x_cog_category.columns = [\"centroid_75\", \"cog_category\"]\n",
    "gene_x_cog_category = gene_x_cog_category.set_index(\"centroid_75\").cog_category\n",
    "gene_x_cog_category.shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotations = _gene_annotations.assign(\n",
    "    COG_category=gene_x_cog_category.sort_values()\n",
    "    .reset_index()\n",
    "    .groupby(\"centroid_75\")\n",
    "    .apply(lambda x: \"\".join(x.cog_category.values))\n",
    ").assign(COG_category=lambda x: x.COG_category.fillna(\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_x_cog_category_matrix = (\n",
    "    gene_x_cog_category.reset_index()\n",
    "    .assign(tally=True)\n",
    "    .set_index([\"centroid_75\", \"cog_category\"])\n",
    "    .tally.unstack(\"cog_category\")\n",
    "    .fillna(False)\n",
    "    .reindex(gene_annotations.index, fill_value=False)\n",
    "    .assign(no_category=lambda x: x.sum(1) == 0)\n",
    ")\n",
    "gene_x_cog_category_matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Raw Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_depth = pd.read_table(\n",
    "    species_depth_inpath, names=[\"sample\", \"depth\"], index_col=[\"sample\"]\n",
    ").depth\n",
    "species_depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_depth = xr.load_dataarray(uhgg_depth_inpath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_depth_ratio = uhgg_depth / species_depth.to_xarray().sel(sample=uhgg_depth.sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SPGC Strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_to_spgc = pd.read_table(sample_to_spgc_inpath, index_col=\"sample\").strain.astype(\n",
    "    str\n",
    ")\n",
    "\n",
    "spgc_palette = lib.plot.construct_ordered_palette(\n",
    "    sample_to_spgc, other=(0.8, 0.8, 0.8, 1.0)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sfacts_fit = sf.World.load(sfacts_fit_inpath).drop_low_abundance_strains(0.5)\n",
    "# Strains should be str not int.\n",
    "sfacts_fit.data[\"strain\"] = sfacts_fit.strain.values.astype(str)\n",
    "print(dict(sfacts_fit.sizes))\n",
    "\n",
    "spgc_est_geno = sf.Metagenotype(\n",
    "    sfacts_fit.metagenotype.data.sel(sample=sample_to_spgc.index)\n",
    "    .groupby(sample_to_spgc.to_xarray())\n",
    "    .sum()\n",
    "    .rename(strain=\"sample\")\n",
    ").to_estimated_genotype(pseudo=0)\n",
    "\n",
    "# Pre-calculate shared heatmap decorations\n",
    "position_ss = sfacts_fit.random_sample(\n",
    "    position=min(500, sfacts_fit.sizes[\"position\"])\n",
    ").position\n",
    "w = sfacts_fit.sel(position=position_ss)\n",
    "sample_linkage = w.unifrac_linkage()\n",
    "try:\n",
    "    position_linkage = spgc_est_geno.linkage(\"position\")\n",
    "except ValueError as err:\n",
    "    print(err)\n",
    "    position_linkage = None\n",
    "sample_colors = w.sample.to_series().map(sample_to_spgc).map(spgc_palette)\n",
    "spgc_linkage = w.genotype.linkage()\n",
    "spgc_colors = w.strain.to_series().map(spgc_palette)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "g = sf.data.Genotype.concat(\n",
    "    dict(mgen=spgc_est_geno, fit=sfacts_fit.genotype), dim=\"strain\"\n",
    ")\n",
    "g_pdist = g.pdist()\n",
    "g_linkage = g.linkage()\n",
    "sf.plot.plot_genotype(\n",
    "    g.sel(position=position_ss), transpose=True, col_linkage=g_linkage\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Ref Strains"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "# \"De-novo GT-Pro genotype\"\n",
    "ref_geno = xr.load_dataarray(ref_geno_inpath)\n",
    "ref_geno['genome_id'] = ref_geno.genome_id.to_series().map(lambda s: \"UHGG\" + s[len(\"GUT_GENOME\"):])\n",
    "ref_geno = ref_geno.fillna(0.5)\n",
    "ref_geno = sf.data.Genotype(ref_geno.rename({'genome_id': 'strain'}).T)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"Reference GT-Pro genotype\"\n",
    "ref_geno = sf.Metagenotype.load(ref_geno_inpath).to_estimated_genotype()\n",
    "ref_geno.data[\"strain\"] = ref_geno.strain.to_series().map(\n",
    "    lambda s: \"UHGG\" + s[len(\"GUT_GENOME\") :]\n",
    ")\n",
    "ref_geno.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_meta = (\n",
    "    pd.read_table(\"ref/uhgg_genomes_all_4644.tsv\", index_col=\"Genome\")\n",
    "    .rename_axis(index=\"genome_id\")[\n",
    "        lambda x: x.MGnify_accession == \"MGYG-HGUT-\" + species_id[1:]\n",
    "    ]\n",
    "    .rename(lambda s: \"UHGG\" + s[len(\"GUT_GENOME\") :])\n",
    "    .loc[ref_geno.strain]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 100, num=101)\n",
    "plt.hist(reference_meta.Completeness, bins=bins)\n",
    "plt.hist(reference_meta.Contamination, bins=bins)\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Select any ref genotype that is within the top-10 closest distances from an SPGC strain.\n",
    "spgc_to_ref_geno_cdist = ref_geno.sel(\n",
    "    position=spgc_est_geno.position,\n",
    "    strain=idxwhere(\n",
    "        (reference_meta.Completeness > 97) & (reference_meta.Contamination < 2)\n",
    "    ),\n",
    ").cdist(spgc_est_geno)\n",
    "ref_list = list(\n",
    "    spgc_to_ref_geno_cdist.apply(lambda x: x.sort_values().head(10).index)\n",
    "    .stack()\n",
    "    .unique()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_meta = pd.read_table(spgc_meta_inpath, index_col=\"strain\").rename(str)\n",
    "print(spgc_meta.shape)\n",
    "spgc_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_copy_number_uhgg = xr.load_dataarray(ref_gene_copy_number_uhgg_inpath)\n",
    "ref_gene_uhgg = (ref_gene_copy_number_uhgg > 0).astype(int).to_pandas().T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg = pd.read_table(spgc_gene_uhgg_inpath, index_col=\"gene_id\").rename_axis(\n",
    "    columns=\"strain\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_num_genes_uhgg = ref_gene_uhgg.sum()\n",
    "spgc_num_genes_uhgg = spgc_gene_uhgg.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_uhgg_prevalence = ref_gene_uhgg[ref_list].mean(1)\n",
    "ref_gene_uhgg_entropy = (\n",
    "    -ref_gene_uhgg_prevalence * np.log2(ref_gene_uhgg_prevalence)\n",
    ").fillna(0)\n",
    "plt.hist(ref_gene_uhgg_entropy)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strain Selection / Filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = spgc_meta[lambda x: x.species_gene_frac > 0.9].num_genes\n",
    "y = ref_num_genes_uhgg\n",
    "_df, _loc, _scale = sp.stats.t.fit(x.values, fix_df=2)\n",
    "_dist0 = sp.stats.t(_df, _loc, _scale)\n",
    "_dist1 = sp.stats.norm(_loc, _scale)\n",
    "\n",
    "thresh_max_num_uhgg_genes = _dist1.ppf(0.999)\n",
    "thresh_min_num_uhgg_genes = _dist1.ppf(0.001)\n",
    "\n",
    "\n",
    "bins = np.linspace(0, x.max() * 1.5, num=50)\n",
    "xx = np.linspace(0, x.max() * 1.5, num=1000)\n",
    "\n",
    "plt.hist(x, density=True, bins=bins, alpha=0.2)\n",
    "plt.hist(y, density=True, bins=bins, alpha=0.2)\n",
    "\n",
    "plt.plot(xx, _dist0.pdf(xx), color=\"k\")\n",
    "plt.plot(xx, _dist1.pdf(xx), color=\"k\", linestyle=\"--\")\n",
    "plt.axvline(thresh_max_num_uhgg_genes, lw=1, linestyle=\"--\", color=\"k\")\n",
    "plt.axvline(thresh_min_num_uhgg_genes, lw=1, linestyle=\"--\", color=\"k\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresh_min_num_uhgg_genes, thresh_max_num_uhgg_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(\n",
    "    \"sum_depth\",\n",
    "    \"species_gene_frac\",\n",
    "    c=(spgc_meta.num_genes - _loc) / _scale,\n",
    "    data=spgc_meta,\n",
    "    norm=mpl.colors.PowerNorm(1 / 1, vmin=-4, vmax=4),\n",
    ")\n",
    "plt.xscale(\"log\")\n",
    "plt.yscale(\"logit\")\n",
    "plt.axvline(1, lw=1, color=\"k\", linestyle=\"--\")\n",
    "plt.axhline(0.9, lw=1, color=\"k\", linestyle=\"--\")\n",
    "plt.colorbar()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: Select SPGC strains that pass various filters\n",
    "spgc_list = idxwhere(\n",
    "    (spgc_meta.sum_depth > 1)\n",
    "    & (spgc_meta.species_gene_frac > 0.9)\n",
    "    & (spgc_num_genes_uhgg <= thresh_max_num_uhgg_genes)\n",
    "    & (spgc_num_genes_uhgg >= thresh_min_num_uhgg_genes)\n",
    ")\n",
    "\n",
    "print(len(ref_list), len(spgc_list))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(spgc_list) >= 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_geno = sf.Genotype.concat(\n",
    "    dict(\n",
    "        ref=ref_geno.sel(strain=ref_list, position=spgc_est_geno.position),\n",
    "        spgc=spgc_est_geno.sel(strain=spgc_list),\n",
    "    ),\n",
    "    dim=\"strain\",\n",
    "    rename=False,\n",
    ").mlift(\"fillna\", 0.5)\n",
    "\n",
    "strain_geno_pdist = strain_geno.pdist(q=1)\n",
    "strain_geno_linkage = strain_geno.linkage(pdist_kwargs=dict(q=1), optimal_ordering=True)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sf.plot.plot_community(w, scaley=0.5, col_linkage_func=lambda w: w.metagenotype.linkage())"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "sf.plot.plot_metagenotype(w, col_linkage_func=lambda w: w.metagenotype.linkage())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genome_type = reference_meta.Genome_type.reindex(strain_geno.strain)\n",
    "genome_type.loc[genome_type.index.isin(spgc_list)] = \"SPGC\"\n",
    "genome_type_order = [\"Isolate\", \"MAG\", \"SPGC\"]\n",
    "genome_type_palette = lib.plot.construct_ordered_palette(genome_type_order)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(2, 1))\n",
    "for _genome_type in genome_type_order:\n",
    "    ax.scatter([], [], color=genome_type_palette[_genome_type], label=_genome_type)\n",
    "ax.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_colors = genome_type.map(genome_type_palette)\n",
    "sf.plot.plot_genotype(\n",
    "    strain_geno.sel(position=position_ss),\n",
    "    transpose=True,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    col_colors=_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_colors = genome_type.map(genome_type_palette)\n",
    "sns.clustermap(\n",
    "    strain_geno_pdist,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_geno_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.spatial.distance import squareform\n",
    "\n",
    "bins = np.linspace(0, 1)\n",
    "plt.hist(\n",
    "    squareform(strain_geno_pdist.loc[ref_list, ref_list]),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    strain_geno_pdist.loc[ref_list, spgc_list].values.flatten(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    squareform(strain_geno_pdist.loc[spgc_list, spgc_list]),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    density=True,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=200)\n",
    "\n",
    "_pdist = strain_geno_pdist + np.eye(len(strain_geno_pdist))\n",
    "\n",
    "plt.hist(\n",
    "    _pdist.loc[ref_list, ref_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[ref_list, spgc_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[spgc_list, ref_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"ref-to-spgc\",\n",
    ")\n",
    "plt.hist(\n",
    "    _pdist.loc[spgc_list, spgc_list].min(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    cumulative=True,\n",
    "    density=True,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "\n",
    "plt.legend()\n",
    "# plt.axvline(0.43, lw=1, linestyle='--', color='k')\n",
    "plt.xlabel(\"minimum distance\")\n",
    "plt.ylabel(\"cumulative fraction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg.shape, ref_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg = pd.concat(\n",
    "    [ref_gene_uhgg[ref_list], spgc_gene_uhgg[spgc_list]], axis=1\n",
    ").fillna(0)[lambda x: x.sum(1) > 0]\n",
    "strain_gene_uhgg.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 10000)\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"ref\",\n",
    "    density=True,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"spgc\",\n",
    "    density=True,\n",
    ")\n",
    "plt.legend()\n",
    "\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Donor Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen = pd.read_table(mgen_inpath, index_col=\"mgen_id\")\n",
    "sample = pd.read_table(sample_inpath, index_col=\"sample_id\")\n",
    "subject = pd.read_table(subject_inpath, index_col=\"subject_id\")\n",
    "\n",
    "mgen_meta = mgen.join(sample, on=\"sample_id\").join(subject, on=\"subject_id\")\n",
    "\n",
    "# subject_uhgg_depth = mgen_meta[['subject_id']].join(uhgg_depth.to_pandas()).groupby('subject_id').mean().dropna()\n",
    "# subject_mean_species_depth = mgen_meta[['subject_id']].join(species_depth).groupby('subject_id').depth.mean().dropna()\n",
    "\n",
    "# total_subject_uhgg_depth = mgen_meta[['subject_id']].join(uhgg_depth.to_pandas()).groupby('subject_id').sum().dropna()\n",
    "# total_subject_species_depth = mgen_meta[['subject_id']].join(species_depth).groupby('subject_id').depth.sum().dropna()\n",
    "# subject_uhgg_depth_ratio = total_subject_uhgg_depth.divide(total_subject_species_depth, axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample.sample_type.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_meta.donor_subject_id.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_type_order = [\"donor\", \"baseline\", \"maintenance\", \"followup\", \"post_antibiotic\"]\n",
    "donor_order = [\"D0044\", \"D0097\", \"D0485\", \"D0065\"]\n",
    "\n",
    "subject_palette = lib.plot.construct_ordered_palette(\n",
    "    subject.sort_values(\"donor_subject_id\").index, cm=\"rainbow\"\n",
    ")\n",
    "sample_type_palette = lib.plot.construct_ordered_palette(sample_type_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(2, 2))\n",
    "for s in donor_order:\n",
    "    ax.scatter([], [], c=subject_palette[s], label=s)\n",
    "ax.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_list = list(set(sfacts_fit.sample.values) & set(mgen_meta.index))\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        is_donor=(mgen_meta.sample_type == \"donor\").map(\n",
    "            {True: \"black\", False: \"darkgrey\"}\n",
    "        ),\n",
    "        subject=mgen_meta.loc[mgen_list].subject_id.map(subject_palette),\n",
    "        donor=mgen_meta.loc[mgen_list].donor_subject_id.map(subject_palette),\n",
    "        spgc=sample_to_spgc.reindex(mgen_list).dropna().map(spgc_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "_col_linkage = sfacts_fit.sel(sample=mgen_list).unifrac_linkage()\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sfacts_fit.sel(sample=mgen_list).drop_low_abundance_strains(0.05),\n",
    "    col_colors=_col_colors,\n",
    "    col_linkage=_col_linkage,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(),\n",
    ")\n",
    "sf.plot.plot_metagenotype(\n",
    "    sfacts_fit.sel(sample=mgen_list, position=position_ss).drop_low_abundance_strains(0.05),\n",
    "    col_colors=_col_colors,\n",
    "    col_linkage=_col_linkage,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Most abundant strains in donor samples:\n",
    "\n",
    "x = sfacts_fit.community.to_pandas().groupby(mgen_meta.subject_id).mean().T\n",
    "print(x[\"D0044\"].sort_values(ascending=False).head(3))\n",
    "print(x[\"D0097\"].sort_values(ascending=False).head(3))\n",
    "spgc_meta.loc[x.idxmax()[[\"D0044\", \"D0097\"]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_pairs = idxwhere(\n",
    "    sfacts_fit.community.to_pandas()\n",
    "    .groupby(mgen_meta.subject_id)\n",
    "    .mean()\n",
    "    .groupby(subject.donor_subject_id)\n",
    "    .mean()\n",
    "    .stack()\n",
    "    > 0.1\n",
    ")\n",
    "donor_strain_list = list(map(lambda x: x[1], donor_strain_pairs))\n",
    "donor_strain_list\n",
    "print(donor_strain_pairs)\n",
    "spgc_meta.reindex(donor_strain_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Most abundant strains in subjects with each donor:\n",
    "\n",
    "x = sfacts_fit.community.to_pandas().groupby(mgen_meta.donor_subject_id).mean().T\n",
    "print(x[\"D0044\"].sort_values(ascending=False).head(3))\n",
    "print(x[\"D0097\"].sort_values(ascending=False).head(3))\n",
    "d44_strain, d97_strain = x.idxmax()[[\"D0044\", \"D0097\"]]\n",
    "spgc_meta.loc[x.idxmax()[[\"D0044\", \"D0097\"]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Most frequently \"pure\" strains in subjects with each donor:\n",
    "\n",
    "x = (\n",
    "    sample_to_spgc.groupby(mgen_meta.donor_subject_id)\n",
    "    .value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .T\n",
    ")\n",
    "\n",
    "print(x[\"D0044\"].sort_values(ascending=False).head(3))\n",
    "print(x[\"D0097\"].sort_values(ascending=False).head(3))\n",
    "spgc_meta.loc[x.idxmax()[[\"D0044\", \"D0097\"]]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_geno_pdist\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.index.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_geno_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    # figsize=(40, 40),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(squareform(strain_geno_pdist))\n",
    "print(strain_geno_pdist.loc[d44_strain, d97_strain])\n",
    "plt.axvline(\n",
    "    strain_geno_pdist.loc[d44_strain, d97_strain], lw=1, linestyle=\"--\", color=\"k\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        is_donor=(mgen_meta.sample_type == \"donor\").map(\n",
    "            {True: \"black\", False: \"darkgrey\"}\n",
    "        ),\n",
    "        subject=mgen_meta.loc[sfacts_fit.sample].subject_id.map(subject_palette),\n",
    "        donor=mgen_meta.loc[sfacts_fit.sample].donor_subject_id.map(subject_palette),\n",
    "        spgc=sample_to_spgc.map(spgc_palette),\n",
    "    )\n",
    ")\n",
    "\n",
    "_col_linkage = sfacts_fit.unifrac_linkage()\n",
    "\n",
    "sf.plot.plot_community(\n",
    "    sfacts_fit,\n",
    "    col_colors=_col_colors,\n",
    "    col_linkage=_col_linkage,\n",
    "    row_linkage_func=lambda w: w.genotype.linkage(),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_meta.subject_id.value_counts().head(4).to_frame(name=\"num_mgen\").join(subject)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgen_meta[lambda x: x.donor_subject_id.isin([\"D0097\"])].subject_id.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subject_lists = {\n",
    "    \"D0044\": [\n",
    "        \"S0041\",\n",
    "        \"S0053\",\n",
    "        \"S0055\",\n",
    "        # 'S0056', 'S0061',\n",
    "        # 'S0060', 'S0059'\n",
    "    ],  # D0044\n",
    "    \"D0097\": [\n",
    "        # 'S0004',\n",
    "        \"S0001\",\n",
    "        \"S0024\",\n",
    "        # 'S0027',\n",
    "        \"S0021\",\n",
    "    ],\n",
    "}\n",
    "all_subjects_list = subject_lists[\"D0044\"] + subject_lists[\"D0097\"]\n",
    "drop_mgen_list = [\"SS01117\", \"SS01120\", \"SS01126\", \"SS01185\", \"SS01008\"]\n",
    "sample_type_specific_x = {\n",
    "    \"baseline\": 0,\n",
    "    \"pre_maintenance_1\": 1,\n",
    "    \"pre_maintenance_2\": 2,\n",
    "    \"pre_maintenance_3\": 3,\n",
    "    \"pre_maintenance_4\": 4,\n",
    "    \"pre_maintenance_5\": 5,\n",
    "    \"pre_maintenance_6\": 6,\n",
    "    \"followup_1\": 7,\n",
    "    \"followup_2\": 8,\n",
    "}\n",
    "sample_type_specific_order = list(sample_type_specific_x.keys())\n",
    "\n",
    "top_strains_list = (\n",
    "    sfacts_fit.community.to_pandas()\n",
    "    .reindex(\n",
    "        mgen_meta.drop(drop_mgen_list)[\n",
    "            lambda x: x.subject_id.isin(all_subjects_list + [\"D0044\", \"D0097\"])\n",
    "        ].index.to_list()\n",
    "    )\n",
    "    .mean()\n",
    "    .sort_values(ascending=False)\n",
    "    .head(20)\n",
    "    .sample(frac=1.0, random_state=3)\n",
    "    .index\n",
    ")\n",
    "_palette = lib.plot.construct_ordered_palette(\n",
    "    top_strains_list, cm=\"Spectral\", extend={d44_strain: \"blue\", d97_strain: \"green\"}\n",
    ")\n",
    "\n",
    "ncols = len(subject_lists[\"D0044\"])\n",
    "nrows = 2\n",
    "fig, axs = plt.subplots(\n",
    "    nrows, ncols, figsize=(4 * ncols, 3 * nrows), sharex=True, sharey=True\n",
    ")\n",
    "\n",
    "for donor_id, ax_row in zip([\"D0044\", \"D0097\"], axs):\n",
    "    for subject_id, ax in zip(subject_lists[donor_id], ax_row):\n",
    "        ax.set_title(subject_id)\n",
    "        # Select data\n",
    "        # subject_id = 'S0004'  # S0041 has D0044 and 12 samples\n",
    "        _subject_baseline_sample_list = mgen_meta[\n",
    "            lambda x: (x.subject_id == subject_id)\n",
    "            & (x.sample_type_specific == \"baseline\")\n",
    "            & (~x.index.isin(drop_mgen_list))\n",
    "        ].index\n",
    "        assert (\n",
    "            len(_subject_baseline_sample_list) == 1\n",
    "        ), f\"No baseline sample for {subject_id}\"\n",
    "        subject_baseline_strain_list = idxwhere(\n",
    "            sfacts_fit.community.data.to_pandas()\n",
    "            .reindex([_subject_baseline_sample_list[0]], fill_value=0)\n",
    "            .squeeze()\n",
    "            > 1e-1\n",
    "        )\n",
    "\n",
    "        # Reshape data\n",
    "        _strain_list = set(donor_strain_list) | set(subject_baseline_strain_list)\n",
    "        _meta = (\n",
    "            mgen_meta[\n",
    "                lambda x: (x.subject_id == subject_id) & (~x.index.isin(drop_mgen_list))\n",
    "            ]\n",
    "            # .reset_index()\n",
    "        )\n",
    "        _strain_frac = sfacts_fit.community.to_pandas().reindex(_meta.index).dropna()\n",
    "        # .set_index('sample_type_specific').reindex(sample_type_specific_order)\n",
    "        # .dropna(subset='mgen_id')\n",
    "        _meta, _strain_frac = lib.pandas_util.align_indexes(_meta, _strain_frac)\n",
    "        _strain_frac[-1] = 1 - _strain_frac[list(_strain_list)].sum(1)\n",
    "        _strain_frac = _strain_frac[list(_strain_list | {-1})]\n",
    "        _strain_list |= {-1}\n",
    "        _strain_depth = _strain_frac.multiply(species_depth, axis=0)\n",
    "\n",
    "        _strain_frac = (\n",
    "            _strain_frac.rename(_meta.sample_type_specific)\n",
    "            .reindex(sample_type_specific_order)\n",
    "            .dropna()\n",
    "        )\n",
    "        _strain_depth = (\n",
    "            _strain_depth.rename(_meta.sample_type_specific)\n",
    "            .reindex(sample_type_specific_order)\n",
    "            .dropna()\n",
    "        )\n",
    "        _meta = (\n",
    "            _meta.rename(_meta.sample_type_specific)\n",
    "            .reindex(sample_type_specific_order)\n",
    "            .dropna(subset=[\"sample_id\"])\n",
    "        )\n",
    "\n",
    "        for _strain in _strain_list:\n",
    "            lw = {True: 4, False: 2}[_strain in donor_strain_list]\n",
    "            s = {True: 10, False: 5}[_strain in donor_strain_list]\n",
    "            ax.plot(\n",
    "                _meta.index.map(sample_type_specific_x),\n",
    "                _strain_frac[_strain],\n",
    "                color=_palette[_strain],\n",
    "                lw=lw,\n",
    "                alpha=0.9,\n",
    "                marker=\".\",\n",
    "                markersize=s,\n",
    "            )\n",
    "\n",
    "    # TODO: Strain legend?\n",
    "    # label=_strain, color=spgc_palette[_strain]\n",
    "    # ax.set_yscale('log')\n",
    "    ax.set_yscale(\"symlog\", linthresh=1e-2)\n",
    "    ax.set_ylim(5e-3, 2e0)\n",
    "    # ax.legend(bbox_to_anchor=(1, 1))\n",
    "axs[0, 0].set_xticks(np.arange(9))\n",
    "fig.tight_layout()\n",
    "\n",
    "# fig, ax = plt.subplots()\n",
    "# for _strain_id in top_strains_list:\n",
    "#     ax.scatter([], [], color=_palette[_strain_id], label=_strain_id)\n",
    "# ax.legend()"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "(sfacts_fit\n",
    "    .community\n",
    "    .to_pandas().reindex(mgen_meta[lambda x: x.subject_id == 'S0061'].index)\n",
    ")"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "subject_id = 'S0004'  # S0041 has D0044 and 12 samples\n",
    "sample_type_specific_x = {'baseline': 0, 'pre_maintenance_1': 1, 'pre_maintenance_2': 2, 'pre_maintenance_3': 3, 'pre_maintenance_4': 4, 'pre_maintenance_5': 5, 'pre_maintenance_6': 6, 'followup_1': 7, 'followup_2': 8}\n",
    "sample_type_specific_order = list(sample_type_specific_x.keys())\n",
    "drop_mgen_list = ['SS01117', 'SS01120', 'SS01126']\n",
    "\n",
    "_meta = (\n",
    "    mgen_meta\n",
    "    [lambda x: (x.subject_id == subject_id) & (~x.index.isin(drop_mgen_list))]\n",
    "    .reset_index().set_index('sample_type_specific')\n",
    "    .reindex(sample_type_specific_order)\n",
    ")\n",
    "_strain_frac = sfacts_fit.community.to_pandas()\n",
    "x = _meta.index.to_series().map(sample_type_specific_x)\n",
    "_strain_list = list({d44_strain, d97_strain} | set(idxwhere(_strain_frac.reindex(_meta.mgen_id).dropna().max() > 0.1)))\n",
    "_strain_frac[-1] = 1 - _strain_frac[_strain_list].sum(1)\n",
    "_strain_frac = _strain_frac[_strain_list + [-1]]\n",
    "_strain_depth = _strain_frac.multiply(species_depth, axis=0).loc[_meta.mgen_id]\n",
    "\n",
    "for _strain in _strain_list + [-1]:\n",
    "    plt.plot(x, _strain_depth[_strain], label=_strain, color=spgc_palette[_strain])\n",
    "plt.yscale('symlog', linthresh=1e-3)\n",
    "plt.ylim(1e-5)\n",
    "plt.legend(bbox_to_anchor=(1, 1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Geno/Gene Spaces"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: This will take ~2 minutes to run for 40,000 genes.\n",
    "gene_uhgg_cdmat = sp.spatial.distance.pdist(strain_gene_uhgg, metric=\"cosine\")\n",
    "gene_uhgg_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(gene_uhgg_cdmat),\n",
    "    index=strain_gene_uhgg.index,\n",
    "    columns=strain_gene_uhgg.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_uhgg_linkage = sp.cluster.hierarchy.linkage(gene_uhgg_cdmat, method=\"average\")\n",
    "gene_uhgg_linkage.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_gene_uhgg.T,\n",
    "    metric=\"cosine\",\n",
    "    w=ref_gene_uhgg_entropy.loc[strain_gene_uhgg.index],\n",
    ")\n",
    "strain_gene_uhgg_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_gene_uhgg_cdmat),\n",
    "    index=strain_gene_uhgg.columns,\n",
    "    columns=strain_gene_uhgg.columns,\n",
    ")\n",
    "\n",
    "assert (strain_geno_pdist.index == strain_gene_uhgg_pdist.index).all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_gene_uhgg_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_gene_uhgg_cdmat, method=\"average\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_cluster=False,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_cluster=False,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_uhgg_depth = (\n",
    "    pd.read_table(spgc_gene_uhgg_depth_inpath, index_col=[\"gene_id\", \"strain\"])\n",
    "    .depth.unstack(\"strain\")\n",
    "    .rename(columns=str)\n",
    ")\n",
    "spgc_gene_uhgg_corr = (\n",
    "    pd.read_table(spgc_gene_uhgg_corr_inpath, index_col=[\"gene_id\", \"strain\"])\n",
    "    .correlation.unstack(\"strain\")\n",
    "    .rename(columns=str)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ref = ref_gene_uhgg[ref_list]\n",
    "_spgc = spgc_gene_uhgg_depth[spgc_list]\n",
    "strain_gene_uhgg_depth = (\n",
    "    pd.concat([_ref, _spgc], axis=1).fillna(0).loc[strain_gene_uhgg.index]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_ref = ref_gene_uhgg[ref_list]\n",
    "_spgc = spgc_gene_uhgg_corr[spgc_list]\n",
    "strain_gene_uhgg_corr = (\n",
    "    pd.concat([_ref, _spgc], axis=1).fillna(0).loc[strain_gene_uhgg.index]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg_depth\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_cluster=False,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg_corr\n",
    "_col_linkage = strain_gene_uhgg_linkage\n",
    "_row_linkage = gene_uhgg_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "print(\n",
    "    \"ref-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "        squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "    squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-spgc\",\n",
    ")\n",
    "if len(spgc_list) > 2:\n",
    "    print(\n",
    "        \"spgc-spgc\",\n",
    "        sp.stats.pearsonr(\n",
    "            squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "            squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "    _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-ref\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "        _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.legend(markerscale=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=200)\n",
    "\n",
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, ref_list].min(),\n",
    "    _pdistB.loc[ref_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, spgc_list].min(),\n",
    "    _pdistB.loc[spgc_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].min(),\n",
    "    _pdistB.loc[spgc_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-spgc\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, spgc_list].min(),\n",
    "    _pdistB.loc[ref_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.xlabel(\"minimum_genotype_diss\")\n",
    "plt.ylabel(\"minimum_gene_diss\")\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_gene_uhgg_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    color=\"k\",\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "\n",
    "for spgc_strain_id in spgc_list:\n",
    "    plt.scatter(\n",
    "        _pdistA.loc[spgc_strain_id, ref_list],\n",
    "        _pdistB.loc[spgc_strain_id, ref_list],\n",
    "        s=1,\n",
    "        alpha=0.5,\n",
    "        color=spgc_palette[spgc_strain_id],\n",
    "        # label='spgc-ref',\n",
    "    )\n",
    "    print(\n",
    "        spgc_strain_id,\n",
    "        sp.stats.pearsonr(\n",
    "            _pdistA.loc[spgc_strain_id, ref_list], _pdistB.loc[spgc_strain_id, ref_list]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Enriched/Depleted in SPGC Strains"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_gene_uhgg_prevalence = strain_gene_uhgg[ref_list].mean(1)\n",
    "spgc_gene_uhgg_prevalence = strain_gene_uhgg[spgc_list].mean(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ref_gene_uhgg_prevalence\n",
    "y = spgc_gene_uhgg_prevalence\n",
    "\n",
    "print(sp.stats.pearsonr(x, y))\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(5, 10))\n",
    "\n",
    "bins0 = np.linspace(0.0, 1.0, num=50)\n",
    "axs[0].hist2d(x, y, bins=bins0, norm=mpl.colors.PowerNorm(1 / 3, vmin=0, vmax=1e3))\n",
    "\n",
    "bins1 = np.linspace(0.1, 0.9, num=40)\n",
    "axs[1].hist2d(x, y, bins=bins1, norm=mpl.colors.PowerNorm(1 / 3))\n",
    "axs[1].set_xlabel(\"reference prevalence\")\n",
    "axs[1].set_ylabel(\"inferred prevalence\")\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_extremely_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.9\n",
    ")\n",
    "spgc_extremely_depleted = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.9\n",
    ")\n",
    "spgc_very_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.5\n",
    ")\n",
    "spgc_very_depleted = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.5\n",
    ")\n",
    "spgc_highly_enriched = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.35\n",
    ")\n",
    "spgc_highly_depleted = idxwhere(\n",
    "    (spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.35\n",
    ")\n",
    "spgc_enriched = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > 0.25)\n",
    "spgc_depleted = idxwhere((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < -0.25)\n",
    "spgc_similar = idxwhere(\n",
    "    ((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) > -0.25)\n",
    "    & ((spgc_gene_uhgg_prevalence - ref_gene_uhgg_prevalence) < 0.25)\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Length / Singletons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for spgc_enrichment_class_list in [\n",
    "    spgc_extremely_enriched,\n",
    "    spgc_very_enriched,\n",
    "    spgc_highly_enriched,\n",
    "    spgc_enriched,\n",
    "    spgc_similar,\n",
    "    spgc_depleted,\n",
    "    spgc_highly_depleted,\n",
    "    spgc_very_depleted,\n",
    "    spgc_extremely_depleted,\n",
    "]:\n",
    "    print(\n",
    "        len(spgc_enrichment_class_list),\n",
    "        uhgg_gene_length.loc[spgc_enrichment_class_list].mean(),\n",
    "        uhgg_gene_length.loc[spgc_enrichment_class_list].std(),\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "short_genes = idxwhere(uhgg_gene_length < 300)\n",
    "singleton_genes = idxwhere(strain_gene_uhgg.sum(1) <= 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Final Filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "drop_list = spgc_highly_enriched + spgc_highly_depleted + short_genes + singleton_genes\n",
    "print(\n",
    "    len(spgc_highly_enriched),\n",
    "    len(spgc_highly_depleted),\n",
    "    len(short_genes),\n",
    "    len(singleton_genes),\n",
    ")\n",
    "\n",
    "strain_uhgg_filt = strain_gene_uhgg.drop(index=drop_list, errors=\"ignore\")\n",
    "print(strain_gene_uhgg.shape[0], strain_uhgg_filt.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, strain_gene_uhgg[spgc_list + ref_list].sum().max(), num=50)\n",
    "\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    label=\"ref (uhgg)\",\n",
    "    density=True,\n",
    "    color=\"tab:blue\",\n",
    "    alpha=0.5,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_gene_uhgg[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"stepfilled\",\n",
    "    label=\"spgc (uhgg)\",\n",
    "    density=True,\n",
    "    color=\"tab:orange\",\n",
    "    alpha=0.5,\n",
    ")\n",
    "\n",
    "plt.hist(\n",
    "    strain_uhgg_filt[ref_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"ref (filtered uhgg)\",\n",
    "    density=True,\n",
    "    linestyle=\"-\",\n",
    "    color=\"tab:blue\",\n",
    "    lw=2,\n",
    ")\n",
    "plt.hist(\n",
    "    strain_uhgg_filt[spgc_list].sum(),\n",
    "    bins=bins,\n",
    "    histtype=\"step\",\n",
    "    label=\"spgc (filtered uhgg)\",\n",
    "    density=True,\n",
    "    linestyle=\"-\",\n",
    "    color=\"tab:orange\",\n",
    "    lw=2,\n",
    ")\n",
    "\n",
    "# plt.hist(strain_gene_eggnog[ref_list].sum(), bins=bins, histtype='stepfilled', label='ref (eggnog)', density=True, color='tab:blue', alpha=0.3)\n",
    "# plt.hist(strain_gene_eggnog[spgc_list].sum(), bins=bins, histtype='stepfilled', label='spgc (eggnog)', density=True, color='tab:orange', alpha=0.3)\n",
    "\n",
    "# plt.hist(strain_gene_filt_eggnog[ref_list].sum(), bins=bins, histtype='step', label='ref (filtered eggnog)', density=True, linestyle='--', color='tab:blue', lw=1.5)\n",
    "# plt.hist(strain_gene_filt_eggnog[spgc_list].sum(), bins=bins, histtype='step', label='spgc (filtered eggnog)', density=True, linestyle='--', color='tab:orange', lw=1.5)\n",
    "\n",
    "plt.legend()\n",
    "\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_filt_cdmat = sp.spatial.distance.pdist(strain_uhgg_filt, metric=\"cosine\")\n",
    "uhgg_filt_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(uhgg_filt_cdmat),\n",
    "    index=strain_uhgg_filt.index,\n",
    "    columns=strain_uhgg_filt.index,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "uhgg_filt_linkage = sp.cluster.hierarchy.linkage(uhgg_filt_cdmat, method=\"average\")\n",
    "uhgg_filt_linkage.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt_cdmat = sp.spatial.distance.pdist(\n",
    "    strain_uhgg_filt.T,\n",
    "    metric=\"cosine\",\n",
    "    w=ref_gene_uhgg_entropy.loc[strain_uhgg_filt.index],\n",
    ")\n",
    "strain_uhgg_filt_pdist = pd.DataFrame(\n",
    "    sp.spatial.distance.squareform(strain_uhgg_filt_cdmat),\n",
    "    index=strain_uhgg_filt.columns,\n",
    "    columns=strain_uhgg_filt.columns,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_uhgg_filt_cdmat, method=\"average\", optimal_ordering=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = uhgg_filt_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt\n",
    "_col_linkage = strain_uhgg_filt_linkage\n",
    "_row_linkage = uhgg_filt_linkage\n",
    "# Order x by leaf order.\n",
    "# See <https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.ClusterNode.pre_order.html#scipy.cluster.hierarchy.ClusterNode.pre_order>\n",
    "x = x.iloc[sp.cluster.hierarchy.to_tree(_row_linkage).pre_order(lambda x: x.id)]\n",
    "\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(x, row_cluster=False, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene Content Divergence from Refs\n",
    "\n",
    "Are SPGC Genomes Believably Distinct?\n",
    "\n",
    "Test if SPGC Strains are surprisingly distinct from refs in gene content relative to their distinction in genotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "print(\n",
    "    \"ref-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "        squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "    squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-spgc\",\n",
    ")\n",
    "if len(spgc_list) > 2:\n",
    "    print(\n",
    "        \"spgc-spgc\",\n",
    "        sp.stats.pearsonr(\n",
    "            squareform(_pdistA.loc[spgc_list, spgc_list]),\n",
    "            squareform(_pdistB.loc[spgc_list, spgc_list]),\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "    _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-ref\",\n",
    ")\n",
    "print(\n",
    "    \"spgc-ref\",\n",
    "    sp.stats.pearsonr(\n",
    "        _pdistA.loc[spgc_list, ref_list].values.flatten(),\n",
    "        _pdistB.loc[spgc_list, ref_list].values.flatten(),\n",
    "    ),\n",
    ")\n",
    "\n",
    "plt.legend(markerscale=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, ref_list].min(),\n",
    "    _pdistB.loc[ref_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-ref\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[ref_list, spgc_list].min(),\n",
    "    _pdistB.loc[ref_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-ref\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, spgc_list].min(),\n",
    "    _pdistB.loc[spgc_list, spgc_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"spgc-to-spgc\",\n",
    ")\n",
    "plt.scatter(\n",
    "    _pdistA.loc[spgc_list, ref_list].min(),\n",
    "    _pdistB.loc[spgc_list, ref_list].min(),\n",
    "    s=10,\n",
    "    alpha=0.5,\n",
    "    label=\"ref-to-spgc\",\n",
    ")\n",
    "plt.xlabel(\"minimum_genotype_diss\")\n",
    "plt.ylabel(\"minimum_gene_diss\")\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statsmodels.formula.api as smf\n",
    "\n",
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "geno_pdist_adjust = 1 / sfacts_fit.sizes[\"position\"]\n",
    "# gene_pdist_adjust = (1 / strain_gene_filt_eggnog.shape[0])\n",
    "\n",
    "\n",
    "# Remove the diagonal from \"minimum distance\".\n",
    "_pdistA = _pdistA + np.eye(len(_pdistA))\n",
    "_pdistB = _pdistB + np.eye(len(_pdistB))\n",
    "\n",
    "ref_to_ref_min_pdistA = _pdistA.loc[ref_list, ref_list].min()\n",
    "ref_to_ref_min_pdistB = _pdistB.loc[ref_list, ref_list].min()\n",
    "spgc_to_ref_min_pdistA = _pdistA.loc[ref_list, spgc_list].min()\n",
    "spgc_to_ref_min_pdistB = _pdistB.loc[ref_list, spgc_list].min()\n",
    "\n",
    "d0 = pd.DataFrame(\n",
    "    dict(\n",
    "        geno_dist=pd.concat([ref_to_ref_min_pdistA, spgc_to_ref_min_pdistA]),\n",
    "        gene_dist=pd.concat([ref_to_ref_min_pdistB, spgc_to_ref_min_pdistB]),\n",
    "        spgc=np.concatenate(\n",
    "            [np.zeros_like(ref_to_ref_min_pdistA), np.ones_like(spgc_to_ref_min_pdistA)]\n",
    "        ).astype(bool),\n",
    "    )\n",
    ")  # .assign(\n",
    "#     geno_dist_pc=lambda x: x.geno_dist + geno_pdist_adjust,\n",
    "#     gene_dist_pc=lambda x: x.gene_dist + gene_pdist_adjust,\n",
    "# )\n",
    "fit = smf.ols(\n",
    "    f\"gene_dist ~ np.log2(geno_dist + {geno_pdist_adjust}) * spgc\", data=d0\n",
    ").fit()\n",
    "d1 = d0.assign(\n",
    "    gene_dist_predict=lambda x: fit.predict(), gene_dist_resid_pearson=fit.resid_pearson\n",
    ").sort_values(\"geno_dist\")\n",
    "\n",
    "fig, axs = plt.subplots(1, 2, figsize=(10, 4))\n",
    "\n",
    "ax = axs[0]\n",
    "for _spgc, d2 in d1.groupby(\"spgc\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_spgc)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "\n",
    "ax = axs[1]\n",
    "for _spgc, d2 in d1.groupby(\"spgc\"):\n",
    "    ax.scatter(\"geno_dist\", \"gene_dist\", data=d2, label=_spgc)\n",
    "    ax.plot(\"geno_dist\", \"gene_dist_predict\", data=d2, label=\"__nolegend__\")\n",
    "ax.legend(title=\"SPGC\")\n",
    "ax.set_xscale(\"symlog\", linthresh=1e-4)\n",
    "fit.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for _spgc, d2 in d1.groupby(\"spgc\"):\n",
    "    plt.scatter(\"gene_dist_predict\", \"gene_dist_resid_pearson\", data=d2, label=_spgc)\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_pdistA = strain_geno_pdist\n",
    "_pdistB = strain_uhgg_filt_pdist\n",
    "\n",
    "plt.scatter(\n",
    "    squareform(_pdistA.loc[ref_list, ref_list]),\n",
    "    squareform(_pdistB.loc[ref_list, ref_list]),\n",
    "    s=1,\n",
    "    alpha=0.5,\n",
    "    color=\"k\",\n",
    "    label=\"ref-ref\",\n",
    ")\n",
    "\n",
    "for spgc_strain_id in spgc_list:\n",
    "    plt.scatter(\n",
    "        _pdistA.loc[spgc_strain_id, ref_list],\n",
    "        _pdistB.loc[spgc_strain_id, ref_list],\n",
    "        s=1,\n",
    "        alpha=0.5,\n",
    "        color=spgc_palette[spgc_strain_id],\n",
    "        # label='spgc-ref',\n",
    "    )\n",
    "    print(\n",
    "        spgc_strain_id,\n",
    "        sp.stats.pearsonr(\n",
    "            _pdistA.loc[spgc_strain_id, ref_list], _pdistB.loc[spgc_strain_id, ref_list]\n",
    "        ),\n",
    "    )\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Strain Diversity Analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_geno_pdist.replace({0: np.nan})\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt_pdist.replace({0: np.nan})\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        g=genome_type.map(genome_type_palette),\n",
    "        d=x.columns.to_series()\n",
    "        .map(\n",
    "            {d44_strain: spgc_palette[d44_strain], d97_strain: spgc_palette[d97_strain]}\n",
    "        )\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "# g.cax.set_visible(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt_pdist.replace({0: np.nan})\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        g=genome_type.map(genome_type_palette),\n",
    "        # d=x.columns.to_series().map({d44_strain: spgc_palette[d44_strain], d97_strain: spgc_palette[d97_strain]}).fillna('grey'),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    "    vmin=0,\n",
    "    vmax=1,\n",
    "    tree_kws=dict(lw=1.5),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "# g.cax.set_visible(False)\n",
    "\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")\n",
    "g.cax.set_visible(False)\n",
    "g.ax_col_dendrogram.set_ylim(0.1, 1)\n",
    "g.ax_row_dendrogram.set_xlim(1, 0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Prevalence Comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_uhgg_prevalence = strain_uhgg_filt[ref_list].mean(1)\n",
    "spgc_uhgg_prevalence = strain_uhgg_filt[spgc_list].mean(1)\n",
    "gene_prevalence = strain_uhgg_filt.mean(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = ref_uhgg_prevalence\n",
    "y = spgc_uhgg_prevalence\n",
    "\n",
    "print(sp.stats.pearsonr(x, y))\n",
    "\n",
    "fig, axs = plt.subplots(2, figsize=(5, 10))\n",
    "\n",
    "bins0 = np.linspace(0.0, 1.0, num=50)\n",
    "axs[0].hist2d(x, y, bins=bins0, norm=mpl.colors.PowerNorm(1 / 3, vmin=0, vmax=1e3))\n",
    "\n",
    "bins1 = np.linspace(0.1, 0.9, num=40)\n",
    "axs[1].hist2d(x, y, bins=bins1, norm=mpl.colors.PowerNorm(1 / 3))\n",
    "axs[1].set_xlabel(\"reference prevalence\")\n",
    "axs[1].set_ylabel(\"inferred prevalence\")\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Core / Shell / Cloud Pangenome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins = np.linspace(0, 1, num=51)\n",
    "\n",
    "\n",
    "def _assign_prevalence_class(p):\n",
    "    if p > 0.95:\n",
    "        return \"core\"\n",
    "    elif p > 0.1:\n",
    "        return \"shell\"\n",
    "    elif p < 0.1:\n",
    "        return \"cloud\"\n",
    "\n",
    "\n",
    "prevalence_class_order = [\"core\", \"shell\", \"cloud\"]\n",
    "prevalence_class_palette = {\n",
    "    \"core\": \"tab:blue\",\n",
    "    \"shell\": \"tab:orange\",\n",
    "    \"cloud\": \"tab:green\",\n",
    "}\n",
    "\n",
    "spgc_uhgg_class = spgc_uhgg_prevalence.map(_assign_prevalence_class)\n",
    "\n",
    "for prevalence_class in prevalence_class_order:\n",
    "    plt.hist(\n",
    "        spgc_uhgg_prevalence[spgc_uhgg_class == prevalence_class],\n",
    "        bins=bins,\n",
    "        label=prevalence_class,\n",
    "        color=prevalence_class_palette[prevalence_class],\n",
    "    )\n",
    "plt.legend()\n",
    "plt.yscale(\"log\")\n",
    "# core_genes = idxwhere(strain_uhgg_filt_prevalence > 0.9)\n",
    "# shell_genes = idxwhere((strain_uhgg_filt_prevalence < 0.9) & (strain_uhgg_filt_prevalence > 0.1))\n",
    "# cloud_genes = idxwhere(strain_uhgg_filt_prevalence < 0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.gridspec as gridspec\n",
    "\n",
    "d = strain_gene_uhgg.groupby(spgc_uhgg_class).sum().T[prevalence_class_order]\n",
    "\n",
    "fig = plt.figure(figsize=(10, 5), tight_layout=True)\n",
    "\n",
    "gs = gridspec.GridSpec(\n",
    "    1, 2, width_ratios=[len(ref_list) / len(spgc_list), 1], figure=fig\n",
    ")\n",
    "ax0 = fig.add_subplot(gs[0, 0])\n",
    "ax1 = fig.add_subplot(gs[0, 1], sharey=ax0)\n",
    "\n",
    "# , axs = plt.subplots(1, 2, figsize=(14, 5), sharey=True)\n",
    "for ax, strain_list, title in zip(\n",
    "    [ax0, ax1], [ref_list, spgc_list], [\"references\", \"spgc\"]\n",
    "):\n",
    "    # TODO: Fix the colors to always match the previous palette.\n",
    "    d.loc[strain_list].sort_values(\"shell\").plot.bar(stacked=True, ax=ax, width=1)\n",
    "    ax.set_title(title)\n",
    "    ax.set_xticks([])\n",
    "    ax.set_aspect(0.08)\n",
    "    ax.legend_.set_visible(False)\n",
    "ax0.legend_.set_visible(True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_annotations.groupby(spgc_uhgg_class).centroid_99_length.quantile(\n",
    "    [0.1, 0.25, 0.5, 0.75, 0.9]\n",
    ").unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_order = gene_x_cog_category_matrix.columns\n",
    "cog_category_palette = lib.plot.construct_ordered_palette(\n",
    "    cog_category_order, cm=\"rainbow\", extend=dict(no_category=\"grey\")\n",
    ")\n",
    "for cog_category in cog_category_order:\n",
    "    plt.scatter([], [], c=cog_category_palette[cog_category], label=cog_category)\n",
    "plt.legend(ncols=4)\n",
    "\n",
    "lib.plot.hide_axes_and_spines()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description = pd.read_table(\n",
    "    \"ref/cog-20.categories.tsv\",\n",
    "    names=[\"cog_category\", \"description\"],\n",
    "    index_col=\"cog_category\",\n",
    ").assign(description=lambda x: x.index + \": \" + x.description)\n",
    "cog_category_description.loc[\"no_category\", \"description\"] = \"-: No Annotation\"\n",
    "cog_category_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = spgc_uhgg_class\n",
    "y = gene_x_cog_category_matrix\n",
    "gene_list = spgc_uhgg_class.index\n",
    "\n",
    "cog_category_gene_class_enrichment_test = []\n",
    "\n",
    "for _prevalence_class, _cog_category in product(\n",
    "    prevalence_class_order, cog_category_order\n",
    "):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_prev_class=(x == _prevalence_class),\n",
    "                is_cog_category=y[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_gene_class_enrichment_test.append(\n",
    "        (\n",
    "            _prevalence_class,\n",
    "            _cog_category,\n",
    "            _test[0],\n",
    "            _test[1],\n",
    "            contingency_table.loc[True, True],\n",
    "        )\n",
    "    )\n",
    "\n",
    "cog_category_gene_class_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_gene_class_enrichment_test,\n",
    "        columns=[\n",
    "            \"prevalence_class\",\n",
    "            \"cog_category\",\n",
    "            \"statistic\",\n",
    "            \"pvalue\",\n",
    "            \"gene_count\",\n",
    "        ],\n",
    "    )\n",
    "    .set_index([\"prevalence_class\", \"cog_category\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio=lambda x: np.log2(x.statistic),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def _assign_significance_marker(pvalue):\n",
    "    if pvalue < 1e-3:\n",
    "        return \"***\"\n",
    "    elif pvalue < 1e-2:\n",
    "        return \"**\"\n",
    "    elif pvalue < 0.05:\n",
    "        return \"*\"\n",
    "    else:\n",
    "        return \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = (\n",
    "    cog_category_gene_class_enrichment_test.log2_odds_ratio.unstack(\"prevalence_class\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"prevalence_class\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[prevalence_class_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = x[\"core\"].sort_values(ascending=False).index\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-5,\n",
    "    vmax=5,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "for prevalence_class in prevalence_class_order:\n",
    "    d = cog_category_gene_class_enrichment_test.xs(\n",
    "        prevalence_class, level=\"prevalence_class\"\n",
    "    )\n",
    "    ax.scatter(\n",
    "        \"log2_odds_ratio\",\n",
    "        \"negative_log10_pvalue\",\n",
    "        data=d,\n",
    "        c=prevalence_class_palette[prevalence_class],\n",
    "        label=prevalence_class,\n",
    "        s=5,\n",
    "    )\n",
    "\n",
    "# Annotate top-10\n",
    "for idx in (\n",
    "    cog_category_gene_class_enrichment_test.sort_values(\n",
    "        \"negative_log10_pvalue\", ascending=False\n",
    "    )\n",
    "    .head(10)\n",
    "    .index\n",
    "):\n",
    "    ax.annotate(\n",
    "        idx[1],\n",
    "        xy=cog_category_gene_class_enrichment_test.loc[idx][\n",
    "            [\"log2_odds_ratio\", \"negative_log10_pvalue\"]\n",
    "        ],\n",
    "        xytext=(2, 1),\n",
    "        textcoords=\"offset pixels\",\n",
    "        color=prevalence_class_palette[idx[0]],\n",
    "        # ha='center',\n",
    "        # va='center',\n",
    "        fontweight=\"bold\",\n",
    "    )\n",
    "ax.axvline(0, linestyle=\"--\", lw=1, color=\"k\")\n",
    "ax.axhline(\n",
    "    -np.log10(0.05 / cog_category_gene_class_enrichment_test.shape[0]),\n",
    "    linestyle=\"--\",\n",
    "    lw=1,\n",
    "    color=\"k\",\n",
    ")\n",
    "ax.legend()\n",
    "# cog_category_gene_class_enrichment_test.sort_values('negative_log10_pvalue', ascending=False).head(20)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Strength of Phylogenetic Signal in Shell Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shell_gene_list = idxwhere(spgc_uhgg_class == \"shell\")\n",
    "\n",
    "x = strain_gene_uhgg.loc[shell_gene_list]\n",
    "\n",
    "strain_shell_gene_cdist = sp.spatial.distance.pdist(\n",
    "    x.T, metric=\"cosine\", w=ref_gene_uhgg_entropy.loc[x.index]\n",
    ")\n",
    "strain_shell_gene_pdist = pd.DataFrame(\n",
    "    squareform(strain_shell_gene_cdist), index=x.columns, columns=x.columns\n",
    ")\n",
    "strain_shell_gene_linkage = sp.cluster.hierarchy.linkage(\n",
    "    strain_shell_gene_cdist, method=\"average\", optimal_ordering=True\n",
    ")\n",
    "\n",
    "# _col_linkage = strain_shell_gene_linkage\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_col_linkage = strain_shell_gene_linkage\n",
    "\n",
    "sns.clustermap(x, col_linkage=_col_linkage, col_colors=_col_colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FIXME: x is ambiguous here.\n",
    "x = strain_shell_gene_pdist.replace({0: np.nan})\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = strain_shell_gene_linkage\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    col_linkage=_col_linkage,\n",
    "    row_linkage=_row_linkage,\n",
    "    col_colors=_colors,\n",
    "    row_colors=_colors,\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using all genomes, not just SPGC\n",
    "phylogenetic_signal = {}\n",
    "for gene_id in tqdm(idxwhere((gene_prevalence > 0.02) & (gene_prevalence < 0.98))):\n",
    "    gene_pdist = pdist(strain_gene_uhgg.loc[[gene_id]].T)\n",
    "    x = squareform(strain_geno_pdist)\n",
    "    y = gene_pdist\n",
    "    if len(x) > 1:\n",
    "        phylogenetic_signal[gene_id] = sp.stats.pearsonr(x, y)\n",
    "\n",
    "phylogenetic_signal = pd.DataFrame(\n",
    "    phylogenetic_signal, index=[\"r_statistic\", \"pvalue\"]\n",
    ").T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "phylogenetic_signal.sort_values(\"r_statistic\")\n",
    "plt.plot(phylogenetic_signal.r_statistic.sort_values().values)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_gene_uhgg.loc[shell_gene_list]  # , spgc_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        phylo=phylogenetic_signal.r_statistic.map(\n",
    "            lambda x: mpl.cm.coolwarm((x + 1) / 2)\n",
    "        ),\n",
    "        entrp=ref_gene_uhgg_entropy.map(mpl.cm.viridis),\n",
    "    )\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x, row_colors=_row_colors, col_linkage=strain_geno_linkage, col_colors=_col_colors\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "is_phylogenetic = pd.concat(\n",
    "    [\n",
    "        pd.Series(\n",
    "            True,\n",
    "            index=phylogenetic_signal[\n",
    "                lambda x: (x.r_statistic > 0.1) & (x.pvalue < 0.001)\n",
    "            ].index,\n",
    "        ),\n",
    "        pd.Series(\n",
    "            False,\n",
    "            index=phylogenetic_signal[\n",
    "                lambda x: (x.r_statistic < 0.1) & (x.pvalue > 0.05)\n",
    "            ].index,\n",
    "        ),\n",
    "    ]\n",
    ")[lambda x: x.index.isin(shell_gene_list)]\n",
    "\n",
    "cog_category_phylogenetic_enrichment_test = []\n",
    "\n",
    "for _cog_category in tqdm(list(cog_category_order)):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_phylogenetic=is_phylogenetic,\n",
    "                is_cog_category=gene_x_cog_category_matrix[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .dropna()\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_phylogenetic_enrichment_test.append(\n",
    "        (_cog_category, *_test, *contingency_table.stack())\n",
    "    )\n",
    "\n",
    "cog_category_phylogenetic_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_phylogenetic_enrichment_test,\n",
    "        columns=[\"cog_category\", \"statistic\", \"pvalue\", \"np_nc\", \"np_c\", \"p-nc\", \"p-c\"],\n",
    "    )\n",
    "    .set_index([\"cog_category\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio=lambda x: np.log2(x.statistic),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_phylogenetic_enrichment_test.sort_values(\n",
    "    \"log2_odds_ratio\", ascending=False\n",
    ").join(cog_category_description)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_phylogenetic_signal_test = {}\n",
    "\n",
    "for _cog_category in tqdm(list(cog_category_order)):\n",
    "    d = pd.DataFrame(\n",
    "        dict(\n",
    "            phylogenetic_signal=phylogenetic_signal.r_statistic,\n",
    "            is_cog_category=gene_x_cog_category_matrix.reindex(\n",
    "                phylogenetic_signal.index\n",
    "            ).fillna(False)[_cog_category],\n",
    "        )\n",
    "    )[lambda x: x.index.isin(shell_gene_list)]\n",
    "    x = d[d.is_cog_category].phylogenetic_signal\n",
    "    y = d[~d.is_cog_category].phylogenetic_signal\n",
    "    if (len(x) > 0) and (len(y) > 0):\n",
    "        _test = sp.stats.mannwhitneyu(x, y)\n",
    "    else:\n",
    "        _test = (np.nan, np.nan)\n",
    "    cog_category_phylogenetic_signal_test[_cog_category] = (\n",
    "        len(x),\n",
    "        x.median(),\n",
    "        y.median(),\n",
    "        x.mean(),\n",
    "        y.mean(),\n",
    "        *_test,\n",
    "    )\n",
    "\n",
    "cog_category_phylogenetic_signal_test = pd.DataFrame(\n",
    "    cog_category_phylogenetic_signal_test,\n",
    "    index=[\n",
    "        \"num_genes_in_category\",\n",
    "        \"cog_median_r\",\n",
    "        \"not_cog_median_r\",\n",
    "        \"cog_mean_r\",\n",
    "        \"not_cog_mean_r\",\n",
    "        \"mwu_statistic\",\n",
    "        \"pvalue\",\n",
    "    ],\n",
    ").T.assign(\n",
    "    negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "    median_diff=lambda x: x.cog_median_r - x.not_cog_median_r,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = cog_category_phylogenetic_signal_test.join(cog_category_description)\n",
    "\n",
    "fig, ax = plt.subplots()\n",
    "plt.scatter(\n",
    "    \"median_diff\",\n",
    "    \"negative_log10_pvalue\",\n",
    "    data=d[[\"description\", \"median_diff\", \"negative_log10_pvalue\"]]\n",
    "    .fillna(0)\n",
    "    .sort_values(\"median_diff\", ascending=False),\n",
    ")\n",
    "# lib.plot.rotate_xticklabels()\n",
    "ax.axvline(0, linestyle=\"--\", lw=1, color=\"k\")\n",
    "ax.axhline(-np.log10(0.05 / d.shape[0]), linestyle=\"--\", lw=1, color=\"k\")\n",
    "\n",
    "\n",
    "for cog_category in idxwhere((np.abs(d.median_diff) > 0) & (d.pvalue < 1e-4)):\n",
    "    ax.annotate(\n",
    "        cog_category,\n",
    "        xy=d[[\"median_diff\", \"negative_log10_pvalue\"]].loc[cog_category],\n",
    "        xytext=(2, 1),\n",
    "        textcoords=\"offset pixels\",\n",
    "        # color=prevalence_class_palette[prevalence_class],\n",
    "        # ha='center',\n",
    "        # va='center',\n",
    "        fontweight=\"bold\",\n",
    "    )\n",
    "\n",
    "ax.set_ylabel(\"-log10(p-value)\")\n",
    "ax.set_xlabel(\"Difference in Median $r$\\n(Genes in COG category vs. Others)\")\n",
    "ax.set_xlim(-0.35, 0.35)\n",
    "ax.set_ylim(0, 20)\n",
    "\n",
    "d.sort_values(\"median_diff\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = phylogenetic_signal.r_statistic.to_frame().join(gene_x_cog_category).fillna(\"-\")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(20, 5))\n",
    "sns.violinplot(\n",
    "    data=d,\n",
    "    x=\"cog_category\",\n",
    "    y=\"r_statistic\",\n",
    "    ax=ax,\n",
    "    order=cog_category_phylogenetic_signal_test.cog_median_r.sort_values().index,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Donor Comparison"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt_pdist.replace({0: np.nan})\n",
    "\n",
    "_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.index.to_series()\n",
    "        .map(\n",
    "            {d44_strain: spgc_palette[d44_strain], d97_strain: spgc_palette[d97_strain]}\n",
    "        )\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "g = sns.clustermap(\n",
    "    x,\n",
    "    row_colors=_colors,\n",
    "    col_colors=_colors,\n",
    "    row_linkage=strain_uhgg_filt_linkage,\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    # figsize=(40, 40),\n",
    ")\n",
    "g.ax_heatmap.set_facecolor(\"aqua\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d44_strain, d97_strain"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_uhgg_filt[[d44_strain, d97_strain]].astype(bool).value_counts().unstack()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_specificity = (\n",
    "    strain_uhgg_filt[[d44_strain, d97_strain]]\n",
    "    .astype(bool)\n",
    "    .apply(tuple, axis=1)\n",
    "    .map(\n",
    "        {\n",
    "            (True, True): \"both\",\n",
    "            (True, False): \"d44\",\n",
    "            (False, True): \"d97\",\n",
    "            (False, False): \"neither\",\n",
    "        }\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d44_only_genes = idxwhere(donor_strain_specificity == \"d44\")\n",
    "d97_only_genes = idxwhere(donor_strain_specificity == \"d97\")\n",
    "both_donor_strain_genes = idxwhere(donor_strain_specificity == \"both\")\n",
    "\n",
    "len(d44_only_genes), len(d97_only_genes), len(both_donor_strain_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = donor_strain_specificity[donor_strain_specificity != \"neither\"]\n",
    "y = gene_x_cog_category_matrix\n",
    "gene_list = x.index\n",
    "\n",
    "cog_category_donor_strain_enrichment_test = []\n",
    "\n",
    "for _donor_strain_specificity, _cog_category in product(x.unique(), cog_category_order):\n",
    "    contingency_table = (\n",
    "        pd.DataFrame(\n",
    "            dict(\n",
    "                is_donor_strain=(x == _donor_strain_specificity),\n",
    "                is_cog_category=y[_cog_category],\n",
    "            )\n",
    "        )\n",
    "        .value_counts()\n",
    "        .unstack()\n",
    "        .reindex(index=[False, True], columns=[False, True])\n",
    "        .fillna(0)\n",
    "    )\n",
    "    _test = sp.stats.fisher_exact(contingency_table)\n",
    "    cog_category_donor_strain_enrichment_test.append(\n",
    "        (\n",
    "            _donor_strain_specificity,\n",
    "            _cog_category,\n",
    "            _test[0],\n",
    "            _test[1],\n",
    "            contingency_table.loc[True, True],\n",
    "        )\n",
    "    )\n",
    "\n",
    "cog_category_donor_strain_enrichment_test = (\n",
    "    pd.DataFrame(\n",
    "        cog_category_donor_strain_enrichment_test,\n",
    "        columns=[\"donor_strain\", \"cog_category\", \"statistic\", \"pvalue\", \"gene_count\"],\n",
    "    )\n",
    "    .set_index([\"donor_strain\", \"cog_category\"])\n",
    "    .assign(\n",
    "        negative_log10_pvalue=lambda x: -np.log10(x.pvalue),\n",
    "        log2_odds_ratio=lambda x: np.log2(x.statistic),\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_order = [\"d44\", \"both\", \"d97\"]\n",
    "x = (\n",
    "    cog_category_donor_strain_enrichment_test.log2_odds_ratio.unstack(\"donor_strain\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_donor_strain_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"donor_strain\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = x[\"both\"].sort_values(ascending=False).index\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-5,\n",
    "    vmax=5,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_order = [\"d44\", \"both\", \"d97\"]\n",
    "x = (\n",
    "    cog_category_donor_strain_enrichment_test.log2_odds_ratio.unstack(\"donor_strain\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    cog_category_donor_strain_enrichment_test.pvalue.map(_assign_significance_marker)\n",
    "    .unstack(\"donor_strain\")\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = [\n",
    "    \"F: Nucleotide transport and metabolism\",\n",
    "    \"A: RNA processing and modification\",\n",
    "    \"J: Translation, ribosomal structure and biogenesis\",\n",
    "    \"C: Energy production and conversion\",\n",
    "    \"E: Amino acid transport and metabolism\",\n",
    "    \"P: Inorganic ion transport and metabolism\",\n",
    "    \"Z: Cytoskeleton\",\n",
    "    \"H: Coenzyme transport and metabolism\",\n",
    "    \"O: Post-translational modification, protein turnover, and chaperones\",\n",
    "    \"I: Lipid transport and metabolism\",\n",
    "    \"T: Signal transduction mechanisms\",\n",
    "    \"Q: Secondary metabolites biosynthesis, transport, and catabolism\",\n",
    "    \"G: Carbohydrate transport and metabolism\",\n",
    "    \"R: General function prediction only\",\n",
    "    \"K: Transcription\",\n",
    "    \"M: Cell wall/membrane/envelope biogenesis\",\n",
    "    \"B: Chromatin structure and dynamics\",\n",
    "    \"D: Cell cycle control, cell division, chromosome partitioning\",\n",
    "    \"V: Defense mechanisms\",\n",
    "    \"S: Function unknown\",\n",
    "    \"N: Cell motility\",\n",
    "    \"L: Replication, recombination and repair\",\n",
    "    \"U: Intracellular trafficking, secretion, and vesicular transport\",\n",
    "    \"W: Extracellular structures\",\n",
    "    \"-: No Annotation\",\n",
    "    \"X: Mobilome: prophages, transposons\",\n",
    "]\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-5,\n",
    "    vmax=5,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_order = [\"d44\", \"both\", \"d97\"]\n",
    "x = (\n",
    "    cog_category_donor_strain_enrichment_test.log2_odds_ratio.unstack(\"donor_strain\")\n",
    "    .replace({np.inf: np.nan, -np.inf: np.nan})\n",
    "    .join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = (cog_category_gene_class_enrichment_test.pvalue.map(_assign_significance_marker) + '|' + cog_category_gene_class_enrichment_test.gene_count.astype(int).astype(str)).unstack('prevalence_class')[prevalence_class_order]\n",
    "annot = (\n",
    "    donor_strain_specificity.to_frame(name=\"donor_strain\")\n",
    "    .join(gene_x_cog_category_matrix)\n",
    "    .groupby(\"donor_strain\")\n",
    "    .sum()\n",
    "    .T.join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "# annot = cog_category_gene_class_enrichment_test.gene_count.unstack('prevalence_class')[prevalence_class_order].astype(int)\n",
    "\n",
    "_row_order = [\n",
    "    \"F: Nucleotide transport and metabolism\",\n",
    "    \"A: RNA processing and modification\",\n",
    "    \"J: Translation, ribosomal structure and biogenesis\",\n",
    "    \"C: Energy production and conversion\",\n",
    "    \"E: Amino acid transport and metabolism\",\n",
    "    \"P: Inorganic ion transport and metabolism\",\n",
    "    \"Z: Cytoskeleton\",\n",
    "    \"H: Coenzyme transport and metabolism\",\n",
    "    \"O: Post-translational modification, protein turnover, and chaperones\",\n",
    "    \"I: Lipid transport and metabolism\",\n",
    "    \"T: Signal transduction mechanisms\",\n",
    "    \"Q: Secondary metabolites biosynthesis, transport, and catabolism\",\n",
    "    \"G: Carbohydrate transport and metabolism\",\n",
    "    \"R: General function prediction only\",\n",
    "    \"K: Transcription\",\n",
    "    \"M: Cell wall/membrane/envelope biogenesis\",\n",
    "    \"B: Chromatin structure and dynamics\",\n",
    "    \"D: Cell cycle control, cell division, chromosome partitioning\",\n",
    "    \"V: Defense mechanisms\",\n",
    "    \"S: Function unknown\",\n",
    "    \"N: Cell motility\",\n",
    "    \"L: Replication, recombination and repair\",\n",
    "    \"U: Intracellular trafficking, secretion, and vesicular transport\",\n",
    "    \"W: Extracellular structures\",\n",
    "    \"-: No Annotation\",\n",
    "    \"X: Mobilome: prophages, transposons\",\n",
    "]\n",
    "# x, annot = lib.pandas_util.align_indexes(x, annot)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    x.reindex(_row_order),\n",
    "    annot=annot.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"coolwarm\",\n",
    "    center=0,\n",
    "    vmin=-5,\n",
    "    vmax=5,\n",
    "    cbar_kws=dict(\n",
    "        use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\", extend=\"both\"\n",
    "    ),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(fontsize=12, va=\"center\"),\n",
    "    # norm=mpl.colors.SymLogNorm(linthresh=1e1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "d = (\n",
    "    donor_strain_specificity.to_frame(name=\"donor_strain\")\n",
    "    .join(gene_x_cog_category_matrix)\n",
    "    .groupby(\"donor_strain\")\n",
    "    .sum()\n",
    "    .T.join(cog_category_description)\n",
    "    .set_index(\"description\")[donor_strain_order]\n",
    "    .fillna(0)\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(5, 12))\n",
    "ax = sns.heatmap(\n",
    "    d.reindex(_row_order),\n",
    "    fmt=\"\",\n",
    "    cmap=\"Oranges\",\n",
    "    # center=0, vmin=-5, vmax=5,\n",
    "    cbar_kws=dict(use_gridspec=True, location=\"left\", label=\"log2(odds ratio)\"),\n",
    "    ax=ax,\n",
    "    yticklabels=1,\n",
    "    xticklabels=1,\n",
    "    annot_kws=dict(va=\"center\"),\n",
    "    norm=mpl.colors.PowerNorm(1),\n",
    "    # center=0,\n",
    ")\n",
    "\n",
    "ax.yaxis.set_ticks_position(\"right\")\n",
    "ax.set_ylabel(\"\")\n",
    "lib.plot.rotate_yticklabels(rotation=-0, va=\"center\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_donor_strain_enrichment_test.xs(\"both\", level=\"donor_strain\").sort_values(\n",
    "    \"log2_odds_ratio\", ascending=False\n",
    ").head(50).join(cog_category_description)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Co-occurence"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Found using SPGC and Ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_prevalence = strain_uhgg_filt[spgc_list + list(ref_list)].mean(1)\n",
    "plt.hist(gene_prevalence)\n",
    "plt.yscale(\"log\")\n",
    "variable_genes = idxwhere((gene_prevalence > 0.05) & (gene_prevalence < 0.95))\n",
    "len(variable_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt.loc[variable_genes]  # , spgc_list]\n",
    "y = x[spgc_list + list(ref_list)]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "        # donor_strain=x.index.to_series().map({d44_strain: 'black', d97_strain: 'white'}).fillna('grey'),\n",
    "    )\n",
    ")\n",
    "\n",
    "clust2 = pd.Series(\n",
    "    AgglomerativeClustering(\n",
    "        n_clusters=None,\n",
    "        metric=\"correlation\",\n",
    "        linkage=\"average\",\n",
    "        distance_threshold=0.035,\n",
    "    ).fit_predict(y),\n",
    "    index=x.index,\n",
    ")\n",
    "clust2_palette = lib.plot.construct_ordered_palette(\n",
    "    clust2.value_counts().sample(frac=1.0).index, cm=\"rainbow\"\n",
    ")\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_colors=x.index.to_series().map(clust2).map(clust2_palette),\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust2_sizes = clust2.value_counts()\n",
    "\n",
    "clust2_sizes_meta = (\n",
    "    clust2.to_frame(\"clust\")\n",
    "    .assign(donor=donor_strain_specificity)\n",
    "    .groupby(\"clust\")\n",
    "    .donor.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .assign(sum_of_d44_d97=lambda x: x.d44 + x.d97)\n",
    "    .sort_values(\"sum_of_d44_d97\", ascending=False)\n",
    "    .assign(\n",
    "        top5_cc=gene_x_cog_category_matrix.groupby(clust2)\n",
    "        .sum()\n",
    "        .rename(int)\n",
    "        .astype(int)\n",
    "        .apply(lambda x: x.sort_values(ascending=False).head(5).index.to_list(), axis=1)\n",
    "    )\n",
    ")\n",
    "\n",
    "clust2_sizes_meta.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[0]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "        # donor_strain=x.index.to_series().map({d44_strain: 'black', d97_strain: 'white'}).fillna('grey'),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[1]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[2]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[3]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[4]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[5]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.index.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust2_sizes_meta.index[6]\n",
    "\n",
    "_gene_list = idxwhere(clust2 == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Detailed Donor Comparison of Clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cog_category_description"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_specificity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Choose clusters that are only found in one of d44 or d97\n",
    "_clust_list = []\n",
    "_clust_list += (\n",
    "    clust2_sizes_meta.d44.sort_values(ascending=False).head(10).index.to_list()\n",
    ")\n",
    "_clust_list += (\n",
    "    clust2_sizes_meta.d97.sort_values(ascending=False).head(10).index.to_list()\n",
    ")\n",
    "\n",
    "_gene_list = idxwhere(clust2.isin(_clust_list))\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "        # donor_strain=x.index.to_series().map({d44_strain: 'black', d97_strain: 'white'}).fillna('grey'),\n",
    "    )\n",
    ")\n",
    "_row_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        # length=gene_annotations.reindex(_gene_list).centroid_99_length.map(np.log10).map(lambda x: x / 4).map(mpl.cm.viridis),\n",
    "        phylo=phylogenetic_signal.r_statistic.pipe(\n",
    "            lambda x: x - x.min() / (x.max() - x.min())\n",
    "        ).map(mpl.cm.viridis),\n",
    "        x=gene_x_cog_category_matrix[\"X\"].map(\n",
    "            {True: \"black\", False: \"white\"}\n",
    "        ),  # Mobilome\n",
    "        m=gene_x_cog_category_matrix[\"M\"].map(\n",
    "            {True: \"black\", False: \"white\"}\n",
    "        ),  # Cell wall\n",
    "        n=gene_x_cog_category_matrix[\"N\"].map(\n",
    "            {True: \"black\", False: \"white\"}\n",
    "        ),  # Motility\n",
    "        w=gene_x_cog_category_matrix[\"W\"].map(\n",
    "            {True: \"black\", False: \"white\"}\n",
    "        ),  # Extracellular struct.\n",
    "        u=gene_x_cog_category_matrix[\"U\"].map(\n",
    "            {True: \"black\", False: \"white\"}\n",
    "        ),  # Trafficing and secretion\n",
    "        clust=clust2.map(clust2_palette),\n",
    "        donor_strain=donor_strain_specificity.map(\n",
    "            {\n",
    "                \"d44\": \"tab:blue\",\n",
    "                \"d97\": \"tab:green\",\n",
    "                \"both\": \"tab:purple\",\n",
    "                \"neither\": \"grey\",\n",
    "            }\n",
    "        ),\n",
    "    )\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 10),\n",
    "    row_linkage=_row_linkage,\n",
    "    row_colors=_row_colors,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "# sns.clustermap(y, figsize=(10, 5), row_linkage=_row_linkage, col_linkage=_col_linkage, col_colors=_col_colors, norm=mpl.colors.PowerNorm(1/2, vmin=0, vmax=2), xticklabels=0, yticklabels=0)\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "# gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "donor_strain_specificity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Found using SPGC Only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "spgc_gene_prevalence = strain_uhgg_filt[spgc_list].mean(1)\n",
    "plt.hist(spgc_gene_prevalence)\n",
    "plt.yscale(\"log\")\n",
    "spgc_variable_genes = idxwhere(\n",
    "    (spgc_gene_prevalence > 0.1) & (spgc_gene_prevalence < 0.9)\n",
    ")\n",
    "len(spgc_variable_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "x = strain_uhgg_filt.loc[spgc_variable_genes]  # , spgc_list]\n",
    "y = x[spgc_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "\n",
    "clust = pd.Series(\n",
    "    AgglomerativeClustering(\n",
    "        n_clusters=None,\n",
    "        metric=\"correlation\",\n",
    "        linkage=\"average\",\n",
    "        distance_threshold=0.035,\n",
    "    ).fit_predict(y),\n",
    "    index=x.index,\n",
    ")\n",
    "clust_palette = lib.plot.construct_ordered_palette(clust.unique())\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    row_colors=x.index.to_series().map(clust).map(clust_palette),\n",
    "    col_linkage=strain_geno_linkage,\n",
    "    col_colors=_col_colors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clust_sizes = clust.value_counts()\n",
    "\n",
    "clust_sizes_meta = (\n",
    "    clust.to_frame(\"clust\")\n",
    "    .assign(donor=donor_strain_specificity)\n",
    "    .groupby(\"clust\")\n",
    "    .donor.value_counts()\n",
    "    .unstack(fill_value=0)\n",
    "    .assign(sum_of_d44_d97=lambda x: x.d44 + x.d97)\n",
    "    .sort_values(\"sum_of_d44_d97\", ascending=False)\n",
    "    .assign(\n",
    "        top5_cc=gene_x_cog_category_matrix.groupby(clust)\n",
    "        .sum()\n",
    "        .rename(int)\n",
    "        .astype(int)\n",
    "        .apply(lambda x: x.sort_values(ascending=False).head(5).index.to_list(), axis=1)\n",
    "    )\n",
    ")\n",
    "\n",
    "clust_sizes_meta.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[0]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[1]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[2]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[3]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[4]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[5]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "_clust = clust_sizes_meta.index[6]\n",
    "\n",
    "_gene_list = idxwhere(clust == _clust)\n",
    "x = strain_gene_uhgg.loc[_gene_list]\n",
    "y = strain_gene_uhgg_depth.loc[_gene_list]\n",
    "_col_colors = pd.DataFrame(\n",
    "    dict(\n",
    "        genome_type=genome_type.map(genome_type_palette),\n",
    "        donor_strain=x.columns.to_series()\n",
    "        .map({d44_strain: \"black\", d97_strain: \"white\"})\n",
    "        .fillna(\"grey\"),\n",
    "    )\n",
    ")\n",
    "_row_colors = (\n",
    "    gene_annotations.reindex(_gene_list)\n",
    "    .centroid_99_length.map(np.log10)\n",
    "    .map(lambda x: x / 4)\n",
    "    .map(mpl.cm.viridis)\n",
    ")\n",
    "_col_linkage = strain_geno_linkage\n",
    "_row_linkage = sp.cluster.hierarchy.linkage(\n",
    "    x, method=\"average\", metric=\"cosine\", optimal_ordering=True\n",
    ")  # TODO\n",
    "\n",
    "sns.clustermap(\n",
    "    x,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "sns.clustermap(\n",
    "    y,\n",
    "    figsize=(10, 5),\n",
    "    row_linkage=_row_linkage,\n",
    "    col_linkage=_col_linkage,\n",
    "    col_colors=_col_colors,\n",
    "    norm=mpl.colors.PowerNorm(1 / 2, vmin=0, vmax=2),\n",
    "    xticklabels=0,\n",
    "    yticklabels=0,\n",
    ")\n",
    "\n",
    "print(\n",
    "    gene_x_cog_category_matrix.reindex(_gene_list)\n",
    "    .sum()\n",
    "    .sort_values(ascending=False)[lambda x: x > 0]\n",
    "    .to_frame(\"tally\")\n",
    "    .join(cog_category_description)\n",
    ")\n",
    "gene_annotations.reindex(_gene_list)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sfacts",
   "language": "python",
   "name": "sfacts"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1847.611523,
   "end_time": "2023-06-24T02:46:52.897060",
   "environment_variables": {},
   "exception": null,
   "input_path": "nb/prototype_compare_spgc_and_refs.ipynb",
   "output_path": "nb/papermill/data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.spgc_ref_comparison.ipynb",
   "parameters": {
    "gene_annotations_path": "data/species/sp-101346/pangenome.centroids.emapper.d/proteins.emapper.annotations",
    "mgen_path": "meta/hmp2/mgen.tsv",
    "nb_path": "nb/prototype_compare_spgc_and_refs.ipynb",
    "preparation_path": "meta/hmp2/preparation.tsv",
    "ref_gene_copy_number_uhgg_path": "ref/midasdb_uhgg_pangenomes/101346/gene75.reference_copy_number.nc",
    "ref_geno_path": "data/species/sp-101346/gtpro_ref.mgtp.nc",
    "sample_to_spgc_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.spgc_ss-all.strain_samples.tsv",
    "sfacts_fit_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.world.nc",
    "species_id": "101346",
    "species_taxonomy_path": "ref/gtpro/species_taxonomy_ext.tsv",
    "spgc_gene_uhgg_corr_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_correlation.tsv",
    "spgc_gene_uhgg_depth_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30.strain_depth_ratio.tsv",
    "spgc_gene_uhgg_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_gene.tsv",
    "spgc_meta_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gtpro.sfacts-fit.gene99-v22-agg75.spgc_specgene-ref-t25-p95_ss-all_t-30_thresh-corr100-depth250.strain_meta.tsv",
    "stool_path": "meta/hmp2/stool.tsv",
    "subject_path": "meta/hmp2/subject.tsv",
    "uhgg_depth_path": "data/group/xjin_hmp2/species/sp-101346/r.proc.gene99-v22-agg75.depth2.nc",
    "uhgg_gene_length_path": "ref/midasdb_uhgg/pangenomes/101346/cluster_info.txt",
    "uhgg_x_eggnog_path": "data/species/sp-101346/pangenome.centroids.emapper.gene_x_eggnog.tsv",
    "uhgg_x_top_eggnog_path": "data/species/sp-101346/pangenome.centroids.emapper.gene_x_top_eggnog.tsv"
   },
   "start_time": "2023-06-24T02:16:05.285537",
   "version": "2.4.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}